// @c8y/ngx-components/esm2015/core/dynamic-forms/radio/radio.type.component.js
var FieldRadio_1;
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
let FieldRadio = FieldRadio_1 = class FieldRadio extends FieldType {
    constructor() {
        super(...arguments);
        this.defaultOptions = {
            templateOptions: {
                options: [],
                formCheck: 'custom' // 'custom' | 'custom-inline' | 'stacked' | 'inline'
            }
        };
    }
};
FieldRadio.CONFIG = {
    types: [
        {
            name: 'radio',
            component: FieldRadio_1,
            wrappers: ['c8y-form-field']
        }
    ]
};
FieldRadio = FieldRadio_1 = tslib_1.__decorate([
    Component({
        selector: 'c8y-field-radio',
        template: "<label\n  *ngFor=\"let option of to.options | formlySelectOptions: field | async; let i = index\"\n  [title]=\"option.label\"\n  class=\"c8y-radio\"\n>\n  <input\n    type=\"radio\"\n    [id]=\"id + '_' + i\"\n    [name]=\"field.name || id\"\n    [attr.value]=\"option.value\"\n    [value]=\"option.value\"\n    [formControl]=\"formControl\"\n    [formlyAttributes]=\"field\"\n    [attr.disabled]=\"option.disabled || formControl.disabled ? true : null\"\n  />\n  <span></span>\n  <span>{{ option.label }}</span>\n</label>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], FieldRadio);
export { FieldRadio };
