// @c8y/ngx-components/esm5/operations/stepper-bulk-type-firmware/stepper-bulk-type-firmware.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { RepositoryModule } from '@c8y/ngx-components/repository';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { HOOK_LIST_BULK_TYPE, baseUrl, BulkOperationType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { SelectFirmwareComponent } from './select-firmware.component';
import { VersionOrPatchComponent } from './version-or-patch.component';
import { StepperBulkTypeFirmwareComponent } from './stepper-bulk-type-firmware.component';
var ɵ0 = {
    type: BulkOperationType.FIRMWARE,
    c8yIcon: 'c8y-firmware',
    name: gettext('Firmware update'),
    path: baseUrl + "firmware",
    component: StepperBulkTypeFirmwareComponent,
    fragments: ['c8y_Firmware'],
    selected: false
};
/** Module for the 'Firmware update' operation type stepper */
var StepperBulkTypeFirmwareModule = /** @class */ (function () {
    function StepperBulkTypeFirmwareModule() {
    }
    StepperBulkTypeFirmwareModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                StepperBulkTypeFirmwareComponent,
                VersionOrPatchComponent,
                SelectFirmwareComponent
            ],
            imports: [
                CoreModule,
                FormsModule,
                ReactiveFormsModule,
                BulkOperationStepperModule,
                RepositoryModule
            ],
            providers: [
                {
                    provide: HOOK_LIST_BULK_TYPE,
                    useValue: ɵ0,
                    multi: true
                }
            ],
            entryComponents: [StepperBulkTypeFirmwareComponent],
            exports: [StepperBulkTypeFirmwareComponent]
        })
    ], StepperBulkTypeFirmwareModule);
    return StepperBulkTypeFirmwareModule;
}());
export { StepperBulkTypeFirmwareModule };
export { ɵ0 };
