// @c8y/ngx-components/esm5/core/bootstrap/bootstrap.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ActionBarModule } from '../action-bar/action-bar.module';
import { AlertModule } from '../alert/alert.module';
import { CommonModule } from '../common/common.module';
import { HeaderModule } from '../header/header.module';
import { LoginModule } from '../login/login.module';
import { ModalModule } from '../modal/modal.module';
import { NavigatorModule } from '../navigator/navigator.module';
import { TabsModule } from '../tabs/tabs.module';
import { BootstrapComponent } from './bootstrap.component';
import { CookieBannerPreferencesModalComponent } from './cookie-banner/cookie-banner-preferences-modal/cookie-banner-preferences-modal.component';
import { CookieBannerComponent } from './cookie-banner/cookie-banner.component';
import { CookieBannerService } from './cookie-banner/cookie-banner.service';
/**
 * The module which allows to bootstrap a custom application.
 * Use it as `bootstrap: []` component in your main app module
 * or as `<c8y-bootstrap>` in your main app.
 */
var BootstrapModule = /** @class */ (function () {
    function BootstrapModule() {
    }
    BootstrapModule = tslib_1.__decorate([
        NgModule({
            declarations: [BootstrapComponent, CookieBannerComponent, CookieBannerPreferencesModalComponent],
            imports: [
                FormsModule,
                CommonModule,
                HeaderModule,
                LoginModule,
                RouterModule,
                TabsModule,
                ActionBarModule,
                AlertModule,
                NavigatorModule,
                TooltipModule,
                ModalModule
            ],
            exports: [BootstrapComponent],
            entryComponents: [BootstrapComponent, CookieBannerPreferencesModalComponent],
            providers: [CookieBannerService]
        })
    ], BootstrapModule);
    return BootstrapModule;
}());
export { BootstrapModule };
