// @c8y/ngx-components/esm2015/protocol-lora/lora-set-device-protocol.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { LoraSetDeviceProtocolService } from './lora-set-device-protocol.service';
import { AlertService, gettext } from '@c8y/ngx-components';
import { IManagedObject, InventoryService, IResultList } from '@c8y/client';
import { Router } from '@angular/router';
import { pipe } from 'rxjs';
import { map } from 'rxjs/operators';
let LoraAssignDeviceProtocolComponent = class LoraAssignDeviceProtocolComponent {
    constructor(loraService, alertService, router, inventoryService) {
        this.loraService = loraService;
        this.alertService = alertService;
        this.router = router;
        this.inventoryService = inventoryService;
        this.filterCurrentProtocol = pipe(map((protocols) => protocols.filter(protocol => !this.currentProtocol || this.currentProtocol.id !== protocol.id)));
    }
    ngOnInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.reload();
        });
    }
    reload() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.loading = true;
            this.newProtocol = null;
            try {
                yield this.loadDevice();
                this.availableProtocols = yield this.loraService.getAvailableProtocols();
                this.currentProtocol = yield this.loraService.getCurrentProtocol(this.device);
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
            finally {
                this.loading = false;
            }
        });
    }
    loadDevice() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const deviceId = this.router.routerState.snapshot.url.match(/\d+/)[0];
            const { data } = yield this.inventoryService.detail(deviceId);
            this.device = data;
        });
    }
    apply(selectedProtocol) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                yield this.loraService.applyProtocol(this.device, selectedProtocol);
                yield this.reload();
                this.alertService.success(gettext('Device protocol set.'));
                this.loraSetDeviceProtocolForm.reset('dirty');
            }
            catch (ex) {
                this.alertService.danger(gettext('Could not set device protocol.'));
            }
        });
    }
};
LoraAssignDeviceProtocolComponent.ctorParameters = () => [
    { type: LoraSetDeviceProtocolService },
    { type: AlertService },
    { type: Router },
    { type: InventoryService }
];
tslib_1.__decorate([
    ViewChild('loraSetDeviceProtocolForm', { static: false })
], LoraAssignDeviceProtocolComponent.prototype, "loraSetDeviceProtocolForm", void 0);
LoraAssignDeviceProtocolComponent = tslib_1.__decorate([
    Component({
        selector: 'set-device-protocol',
        template: "<form #loraSetDeviceProtocolForm=\"ngForm\" class=\"row\">\n  <div class=\"col-md-6\">\n    <div class=\"card\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\">\n          {{ 'LoRa device protocol' | translate }}\n        </h4>\n      </div>\n      <div *ngIf=\"loading\" class=\"card-block\">\n        <div class=\"p-relative\" style=\"min-height: 40px; min-width: 55px;\">\n          <div class=\"spinner\">\n            <div class=\"rect1\"></div>\n            <div class=\"rect2\"></div>\n            <div class=\"rect3\"></div>\n            <div class=\"rect4\"></div>\n          </div>\n        </div>\n        <span translate>Loading</span>\n      </div>\n      <div *ngIf=\"!loading\">\n        <div class=\"card-block\">\n          <div class=\"form-group\">\n            <label translate>Current device protocol</label>\n            <p class=\"form-control-static\" *ngIf=\"!currentProtocol\">\n              {{ 'c8y_LoriotLora' }}\n            </p>\n            <p class=\"form-control-static\" *ngIf=\"currentProtocol\">\n              {{ currentProtocol.name }}\n            </p>\n          </div>\n          <div class=\"form-group\">\n            <label translate>New device protocol</label>\n            <div class=\"c8y-select-wrapper\">\n              <select class=\"form-control\" [(ngModel)]=\"newProtocol\" name=\"newProtocol\" required>\n                <option *ngIf=\"!newProtocol\" [ngValue]=\"null\">\n                  {{ 'Select device protocol' | translate }}\n                </option>\n                <option\n                  *c8yFor=\"let protocol of availableProtocols; pipe: filterCurrentProtocol\"\n                  [ngValue]=\"protocol\"\n                >\n                  {{ protocol.name }}\n                </option>\n              </select>\n            </div>\n          </div>\n        </div>\n        <div class=\"card-footer separator\">\n          <button\n            title=\"{{ 'Apply' | translate }}\"\n            class=\"btn btn-primary\"\n            (click)=\"apply(newProtocol)\"\n            [disabled]=\"!loraSetDeviceProtocolForm.dirty\"\n          >\n            {{ 'Apply' | translate }}\n          </button>\n        </div>\n      </div>\n    </div>\n  </div>\n</form>\n\u200C\n"
    })
], LoraAssignDeviceProtocolComponent);
export { LoraAssignDeviceProtocolComponent };
