// @c8y/ngx-components/esm2015/protocol-lora/lora-set-device-protocol.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { IManagedObject, InventoryService, IResultList } from '@c8y/client';
let LoraSetDeviceProtocolService = class LoraSetDeviceProtocolService {
    constructor(inventoryService) {
        this.inventoryService = inventoryService;
    }
    getCurrentProtocol(device) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const protocolId = device.c8y_LpwanDevice.type
                ? device.c8y_LpwanDevice.type.split('/')[2]
                : null;
            if (!protocolId) {
                return null;
            }
            const { data } = yield this.inventoryService.detail(protocolId);
            return data;
        });
    }
    applyProtocol(device, selectedProtocol) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            device.c8y_LpwanDevice.type = 'inventory/managedObjects/' + selectedProtocol.id;
            device.type = selectedProtocol.name;
            return this.inventoryService.update(device);
        });
    }
    getAvailableProtocols() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const query = {
                type: {
                    __in: ['c8y_ActilityDeviceType', 'c8y_LoraDeviceType']
                }
            };
            return this.inventoryService.listQuery(query);
        });
    }
};
LoraSetDeviceProtocolService.ctorParameters = () => [
    { type: InventoryService }
];
LoraSetDeviceProtocolService = tslib_1.__decorate([
    Injectable()
], LoraSetDeviceProtocolService);
export { LoraSetDeviceProtocolService };
