// @c8y/ngx-components/esm5/core/common/permissions.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { AppStateService } from './ui-state.service';
import { IManagedObject, InventoryService } from '@c8y/client';
import { property } from 'lodash-es';
import { memoize } from './memoize.decorator';
import { IUser, UserService } from '@c8y/client';
var Permissions = /** @class */ (function () {
    function Permissions(appState, inventory, user) {
        this.appState = appState;
        this.inventory = inventory;
        this.user = user;
    }
    /**
     * Checks if the current user has write permissions for the given mo.
     * (either through global role, individual device permissions or via inventory roles).
     *
     * @param {array} roleIds Array of roles which the current user should have.
     * @param {object} mo The managed object for which we are checking whether the user has access.
     */
    Permissions.prototype.canEdit = function (roleIds, mo) {
        return this.checkIfCanEdit(roleIds, mo);
    };
    Permissions.prototype.hasRole = function (roleId) {
        var currentUser = this.appState.currentUser.value;
        if (!currentUser) {
            throw new Error('Roles can only be requested if the user is logged in.');
        }
        return this.user.hasRole(currentUser, roleId);
    };
    Permissions.prototype.hasAllRoles = function (roleIds) {
        var currentUser = this.appState.currentUser.value;
        if (!currentUser) {
            throw new Error('Roles can only be requested if the user is logged in.');
        }
        return this.user.hasAllRoles(currentUser, roleIds);
    };
    Permissions.prototype.hasAnyRole = function (roleIds) {
        var currentUser = this.appState.currentUser.value;
        if (!currentUser) {
            throw new Error('Roles can only be requested if the user is logged in.');
        }
        return this.user.hasAnyRole(currentUser, roleIds);
    };
    Permissions.prototype.checkIfOwner = function (mo) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var currentUserName, _a, data, res;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.appState.currentUser.value.userName];
                    case 1:
                        currentUserName = _b.sent();
                        return [4 /*yield*/, this.inventory.detail(mo.id)];
                    case 2:
                        _a = _b.sent(), data = _a.data, res = _a.res;
                        return [2 /*return*/, currentUserName === data.owner];
                }
            });
        });
    };
    Permissions.prototype.checkWithRequest = function (mo) {
        var moName = mo.name;
        var moId = mo.id;
        var partialUpdateObject = {
            id: moId,
            name: moName
        };
        return this.inventory
            .update(partialUpdateObject)
            .then(function (success) {
            return true;
        })
            .catch(function (err) {
            return false;
        });
    };
    Permissions.prototype.checkIfCanEdit = function (roleIds, mo) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.hasAnyRole(roleIds)];
                    case 1:
                        if (!_a.sent()) return [3 /*break*/, 2];
                        return [2 /*return*/, true];
                    case 2: return [4 /*yield*/, this.checkIfOwner(mo)];
                    case 3:
                        if (!_a.sent()) return [3 /*break*/, 4];
                        return [2 /*return*/, true];
                    case 4: return [4 /*yield*/, this.checkWithRequest(mo)];
                    case 5:
                        if (_a.sent()) {
                            return [2 /*return*/, true];
                        }
                        _a.label = 6;
                    case 6: return [2 /*return*/, false];
                }
            });
        });
    };
    Permissions.ctorParameters = function () { return [
        { type: AppStateService },
        { type: InventoryService },
        { type: UserService }
    ]; };
    tslib_1.__decorate([
        memoize(property('id'))
    ], Permissions.prototype, "checkIfOwner", null);
    tslib_1.__decorate([
        memoize(property('id'))
    ], Permissions.prototype, "checkWithRequest", null);
    Permissions = tslib_1.__decorate([
        Injectable()
    ], Permissions);
    return Permissions;
}());
export { Permissions };
