// @c8y/ngx-components/esm5/core/common/tenant-ui.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ApplicationService, TenantService, IApplication } from '@c8y/client';
/** The helper UI service for tenant related methods built upon client services. */
var TenantUiService = /** @class */ (function () {
    function TenantUiService(tenantService, applicationService) {
        this.tenantService = tenantService;
        this.applicationService = applicationService;
    }
    /**
     * Checks whether current tenant is the management tenant.
     * @returns True if current tenant is the management tenant.
     */
    TenantUiService.prototype.isManagementTenant = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var currentTenant;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.tenantService.current()];
                    case 1:
                        currentTenant = (_a.sent()).data;
                        return [2 /*return*/, currentTenant.name === 'management'];
                }
            });
        });
    };
    /**
     * Checks whether current tenant is an enterprise tenant.
     * An enterprise tenant is a tenant which has subscribed:
     * - `branding` microservice or `feature-branding` feature app,
     * - `sslmanagement` microservice,
     * - `feature-user-hierarchy` feature app,
     * - `feature-broker` feature app.
     *
     * See https://cumulocity.com/guides/users-guide/enterprise-edition/ for details about such tenants.
     *
     * @returns True, if current tenant is an enterprise tenant.
     */
    TenantUiService.prototype.isEnterpriseTenant = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var hasBranding, _a, hasSslManagement, hasUserHierarchy, hasDataBroker;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.hasApp({ name: 'branding' })];
                    case 1:
                        _a = (_b.sent());
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.hasApp({ name: 'feature-branding' })];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        hasBranding = _a;
                        return [4 /*yield*/, this.hasApp({ name: 'sslmanagement' })];
                    case 4:
                        hasSslManagement = _b.sent();
                        return [4 /*yield*/, this.hasApp({ name: 'feature-user-hierarchy' })];
                    case 5:
                        hasUserHierarchy = _b.sent();
                        return [4 /*yield*/, this.hasApp({ name: 'feature-broker' })];
                    case 6:
                        hasDataBroker = _b.sent();
                        return [2 /*return*/, hasBranding && hasSslManagement && hasUserHierarchy && hasDataBroker];
                }
            });
        });
    };
    TenantUiService.prototype.hasApp = function (app) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.applicationService.isAvailable(app)];
                    case 1: return [2 /*return*/, (_a.sent()).data];
                }
            });
        });
    };
    TenantUiService.ctorParameters = function () { return [
        { type: TenantService },
        { type: ApplicationService }
    ]; };
    TenantUiService = tslib_1.__decorate([
        Injectable()
    ], TenantUiService);
    return TenantUiService;
}());
export { TenantUiService };
