// @c8y/ngx-components/esm5/context-dashboard/report-dashboard/report-dashboard-list.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { BsModalService } from 'ngx-bootstrap/modal';
import { IManagedObject, InventoryService, IResultList, QueriesUtil } from '@c8y/client';
import { debounceTime, distinctUntilChanged, shareReplay, switchMap, take, tap } from 'rxjs/operators';
import { FilterInputComponent, gettext, ModalService, Status, AlertService } from '@c8y/ngx-components';
import { DashboardDetailComponent } from '../dashboard-detail.component';
import { ContextDashboardService } from '../context-dashboard.service';
import { ContextDashboardType } from '../context-dashboard.model';
import { TranslateService } from '@ngx-translate/core';
import { ReportDashboardService } from './report-dashboard.service';
var ReportDashboardListComponent = /** @class */ (function () {
    function ReportDashboardListComponent(inventoryService, contextDashboardService, bsModal, translateService, modal, alertService, reportDashboardService) {
        var _this = this;
        this.inventoryService = inventoryService;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.translateService = translateService;
        this.modal = modal;
        this.alertService = alertService;
        this.reportDashboardService = reportDashboardService;
        this.textFilter$ = new BehaviorSubject('');
        this.reload$ = new BehaviorSubject(null);
        this.reloading = false;
        this.reports$ = combineLatest(this.textFilter$.pipe(debounceTime(400), distinctUntilChanged()), this.reload$).pipe(tap(function () {
            _this.reloading = true;
        }), switchMap(function (_a) {
            var _b = tslib_1.__read(_a, 1), text = _b[0];
            return _this.loadReports(text);
        }), tap(function () {
            _this.reloading = false;
        }), shareReplay(1));
        this.DELETED_SUCCESS_MSG = gettext('Report deleted.');
    }
    ReportDashboardListComponent.prototype.loadReports = function (partialName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, partialName
                        ? this.reportDashboardService.listReports({ filter: { name: "*" + partialName + "*" } })
                        : this.reportDashboardService.listReports()];
            });
        });
    };
    ReportDashboardListComponent.prototype.add = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var initialState, modal, cfg, name_1, icon, c8y_IsNavigatorNode, priority, description, dashboardCfg, report, ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        initialState = {
                            isReport: true,
                            isNamedDashboard: true
                        };
                        modal = this.bsModal.show(DashboardDetailComponent, {
                            class: 'modal-lg',
                            ignoreBackdropClick: true,
                            initialState: initialState
                        }).content;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 5, , 6]);
                        return [4 /*yield*/, modal.result];
                    case 2:
                        cfg = _a.sent();
                        name_1 = cfg.name, icon = cfg.icon, c8y_IsNavigatorNode = cfg.c8y_IsNavigatorNode, priority = cfg.priority, description = cfg.description, dashboardCfg = tslib_1.__rest(cfg, ["name", "icon", "c8y_IsNavigatorNode", "priority", "description"]);
                        return [4 /*yield*/, this.reportDashboardService.createReport({
                                name: name_1,
                                icon: icon,
                                c8y_IsNavigatorNode: c8y_IsNavigatorNode,
                                priority: priority,
                                description: description
                            })];
                    case 3:
                        report = (_a.sent()).data;
                        return [4 /*yield*/, this.contextDashboardService.create(dashboardCfg, "" + this.contextDashboardService.REPORT_PARTIAL_NAME + report.id)];
                    case 4:
                        _a.sent();
                        if (report.c8y_IsNavigatorNode) {
                            this.reportDashboardService.addReportNavigatorNode(report);
                        }
                        this.reload$.next();
                        modal.close();
                        return [3 /*break*/, 6];
                    case 5:
                        ex_1 = _a.sent();
                        if (ex_1) {
                            throw new Error("Something went wrong: " + ex_1);
                        }
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    ReportDashboardListComponent.prototype.delete = function (report) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var msg, ex_2;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        msg = gettext("You are about to delete the report \"{{ reportName }}\". Do you want to proceed?");
                        return [4 /*yield*/, this.modal.confirm(gettext('Delete report'), this.translateService.instant(msg, {
                                reportName: report.name
                            }), Status.DANGER, {
                                ok: gettext('Delete'),
                                cancel: gettext('Cancel')
                            })];
                    case 1:
                        _a.sent();
                        this.contextDashboardService
                            .getDashboard$("report_" + report.id, [ContextDashboardType.Named])
                            .pipe(take(1))
                            .subscribe(function (dashboard) { return _this.contextDashboardService.delete(dashboard, false); });
                        return [4 /*yield*/, this.inventoryService.delete(report.id)];
                    case 2:
                        _a.sent();
                        this.alertService.success(this.DELETED_SUCCESS_MSG);
                        if (report.c8y_IsNavigatorNode) {
                            this.reportDashboardService.removeNavigatorNode(report);
                        }
                        this.reload$.next();
                        return [3 /*break*/, 4];
                    case 3:
                        ex_2 = _a.sent();
                        if (ex_2) {
                            throw new Error("Something went wrong: " + ex_2);
                        }
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ReportDashboardListComponent.prototype.update = function (report) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        report.c8y_IsNavigatorNode = report.c8y_IsNavigatorNode ? {} : null;
                        return [4 /*yield*/, this.inventoryService.update(report)];
                    case 1:
                        _a.sent();
                        report.c8y_IsNavigatorNode
                            ? this.reportDashboardService.addReportNavigatorNode(report)
                            : this.reportDashboardService.removeNavigatorNode(report);
                        return [2 /*return*/];
                }
            });
        });
    };
    ReportDashboardListComponent.ctorParameters = function () { return [
        { type: InventoryService },
        { type: ContextDashboardService },
        { type: BsModalService },
        { type: TranslateService },
        { type: ModalService },
        { type: AlertService },
        { type: ReportDashboardService }
    ]; };
    tslib_1.__decorate([
        ViewChild(FilterInputComponent, { static: false })
    ], ReportDashboardListComponent.prototype, "filter", void 0);
    ReportDashboardListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-report-dashboard-list',
            template: "<c8y-title>\n  <span translate>\n    Reports\n  </span>&nbsp;\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form\">\n  <div class=\"input-group input-group-search\">\n    <input class=\"form-control\"\n      type=\"search\"\n      title=\"{{ 'Filter\u2026' | translate }}\"\n      placeholder=\"{{ 'Filter\u2026' | translate }}\"\n      [ngModel]=\"textFilter$ | async\"\n      (ngModelChange)=\"textFilter$.next($event)\"\n    />\n    <span class=\"input-group-addon\">\n      <i c8yIcon=\"search\"\n        *ngIf=\"(textFilter$ | async).length === 0\"\n      ></i>\n      <i class=\"text-muted\"\n        c8yIcon=\"times\"\n        *ngIf=\"(textFilter$ | async).length > 0\"\n        (click)=\"textFilter$.next('')\"\n      ></i>\n    </span>\n  </div>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\"\n    title=\"{{ 'Add report' | translate }}\"\n    (click)=\"add()\"\n    >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add report' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\"\n    title=\"{{ 'Reload' | translate }}\"\n    (click)=\"loadReports()\"\n  >\n    <i [ngClass]=\"{ 'icon-spin': reloading }\"\n      c8yIcon=\"refresh\"\n    ></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"p-b-32\">\n  <c8y-list-group>\n    <c8y-li class=\"page-sticky-header hidden-xs hidden-sm\"\n      *ngIf=\"(reports$ | async)?.data.length > 0\"\n    >\n      <c8y-li-icon>\n        <i class=\"fa\"></i>\n      </c8y-li-icon>\n      <c8y-li-body class=\"content-flex-80\">\n        <div class=\"col-3\">\n          {{ 'Report' | translate }}\n        </div>\n        <div class=\"col-6\">\n          {{ 'Description' | translate }}\n        </div>\n        <div class=\"col-2\">\n          {{ 'Show in navigator' | translate }}\n        </div>\n      </c8y-li-body>\n    </c8y-li>\n\n    <c8y-li *c8yFor=\"let report of reports$; let i = index; loadMore: 'auto'\">\n      <c8y-li-icon [icon]=\"report.icon\"></c8y-li-icon>\n      <c8y-li-body class=\"content-flex-70\">\n        <div class=\"col-3\">\n          <button class=\"btn-clean\"\n            title=\"{{ report.name }}\"\n            routerLink=\"/reports/{{ report.id }}\"\n          >\n            <span class=\"text-truncate\">\n              {{ report.name }}\n            </span>\n          </button>\n        </div>\n        <div class=\"col-6\">\n          <small class=\"text-truncate-wrap\">\n            <em class=\"text-muted\"\n              *ngIf=\"!report.description; else showDescription\"\n            >\n              {{ 'No description available' | translate }}\n            </em>\n            <ng-template #showDescription>\n              {{ report.description }}\n            </ng-template>\n          </small>\n        </div>\n        <div class=\"col-2\">\n          <span class=\"m-t-8 visible-xs\"></span>\n          <label class=\"c8y-switch c8y-switch--inline\"\n            title=\"{{ 'Show in navigator' | translate }}\"\n          >\n            <input\n              [(ngModel)]=\"!!report.c8y_IsNavigatorNode\"\n              type=\"checkbox\"\n              (change)=\"update(report)\"\n            />\n            <span></span>\n            <span class=\"visible-xs\">\n              {{ 'Show in navigator' | translate }}\n            </span>\n          </label>\n        </div>\n        <div class=\"col-1 text-right hidden-xs\">\n          <button class=\"btn-dot showOnHover pull-right\"\n            (click)=\"delete(report)\"\n            title=\"{{ 'Remove report' | translate }}\"\n          >\n            <i class=\"text-danger\"\n            c8yIcon=\"minus-circle\"\n            ></i>\n          </button>\n        </div>\n        <div class=\"visible-xs p-t-8 text-right\">\n          <button class=\"btn-danger btn btn-xs\"\n            (click)=\"delete(report)\"\n            title=\"{{ 'Remove report' | translate }}\"\n          >\n            <i c8yIcon=\"trash\"></i>\n            Delete\n          </button>\n        </div>\n      </c8y-li-body>\n    </c8y-li>\n  </c8y-list-group>\n</div>\n\n<div class=\"c8y-empty-state text-center m-t-40\"\n  *ngIf=\"(reports$ | async)?.data.length === 0\"\n  >\n  <h1 c8yIcon=\"c8y-reports\" class=\"c8y-icon-duocolor\"></h1>\n  <h3 translate>\n    There are no reports defined\n  </h3>\n  <p translate>\n    Add a report first.\n  </p>\n  <div>\n    <button class=\"btn btn-primary\"\n      (click)=\"add()\"\n      translate\n    >\n      Add report\n    </button>\n  </div>\n  <p c8y-guide-docs>\n    <small translate ngNonBindable\n      >Find out more in the\n      <a c8y-guide-href=\"users-guide/cockpit/#reports\">\n        User guide`KEEP_ORIGINAL`\n      </a>.\n      </small>\n  </p>\n</div>\n"
        })
    ], ReportDashboardListComponent);
    return ReportDashboardListComponent;
}());
export { ReportDashboardListComponent };
