// @c8y/ngx-components/esm5/context-dashboard/report-dashboard/report-dashboard.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { IManagedObject, InventoryService, QueriesUtil } from '@c8y/client';
import { NavigatorNode, NavigatorService } from '@c8y/ngx-components';
var ReportDashboardService = /** @class */ (function () {
    function ReportDashboardService(inventoryService, navigatorService) {
        this.inventoryService = inventoryService;
        this.navigatorService = navigatorService;
        this.queriesUtil = new QueriesUtil();
    }
    ReportDashboardService.prototype.listReports = function (options) {
        if (options === void 0) { options = {}; }
        var query = { __has: 'c8y_Report' };
        if (options.filter) {
            query = this.queriesUtil.addAndFilter(query, options.filter);
        }
        var filters = {
            query: this.queriesUtil.buildQuery(query),
            pageSize: options.pageSize || 50,
            withTotalPages: true
        };
        return this.inventoryService.list(filters);
    };
    ReportDashboardService.prototype.createReport = function (reportCfg) {
        var report = {};
        Object.assign(report, reportCfg);
        Object.assign(report, { c8y_Report: {} });
        return this.inventoryService.create(report);
    };
    ReportDashboardService.prototype.addReportNavigatorNode = function (report) {
        var node = new NavigatorNode({
            label: report.name,
            path: "reports/" + report.id,
            icon: report.icon,
            priority: report.priority
        });
        this.navigatorService.add(node);
    };
    ReportDashboardService.prototype.removeNavigatorNode = function (report) {
        var _this = this;
        this.navigatorService.state.forEach(function (node) {
            if (node.path === "reports/" + report.id) {
                _this.navigatorService.remove(node);
                return;
            }
        });
    };
    ReportDashboardService.ctorParameters = function () { return [
        { type: InventoryService },
        { type: NavigatorService }
    ]; };
    ReportDashboardService = tslib_1.__decorate([
        Injectable()
    ], ReportDashboardService);
    return ReportDashboardService;
}());
export { ReportDashboardService };
