// @c8y/ngx-components/esm5/operations/bulk-operations-list/modals/bulk-operations-modal.component.js
import * as tslib_1 from "tslib";
import { Component, ViewContainerRef, ViewChild, ComponentFactoryResolver } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
var BulkOperationsModalComponent = /** @class */ (function () {
    function BulkOperationsModalComponent(bsModalRef, resolver) {
        this.bsModalRef = bsModalRef;
        this.resolver = resolver;
        this.showComponent = false;
    }
    BulkOperationsModalComponent.prototype.cancel = function () {
        this.bsModalRef.hide();
    };
    BulkOperationsModalComponent.prototype.createComponent = function (item) {
        if (item && item.component) {
            this.showComponent = true;
            var factory = this.resolver.resolveComponentFactory(item.component);
            this.container.createComponent(factory);
        }
    };
    BulkOperationsModalComponent.ctorParameters = function () { return [
        { type: BsModalRef },
        { type: ComponentFactoryResolver }
    ]; };
    tslib_1.__decorate([
        ViewChild('container', { static: false, read: ViewContainerRef })
    ], BulkOperationsModalComponent.prototype, "container", void 0);
    BulkOperationsModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-bulk-operations-modal',
            template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <h1 class=\"c8y-icon c8y-icon-energy\"></h1>\n    <h4 translate>New bulk operation</h4>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body text-center\" [hidden]=\"showComponent\">\n      <h4 translate>Select one of the available options</h4>\n      <p c8y-guide-docs>\n        <small translate ngNonBindable>\n          Check the\n          <a c8y-guide-href=\"users-guide/device-management/#bulk-operation-wizard\"\n            >User guide`KEEP_ORIGINAL`</a\n          >\n          for more details.</small\n        >\n      </p>\n    </div>\n\n    <c8y-bulk-operation-type-list\n      (onCustomType)=\"createComponent($event)\"\n    ></c8y-bulk-operation-type-list>\n    <template #container></template>\n  </div>\n  <div class=\"modal-footer\" [hidden]=\"showComponent\">\n    <button\n      (click)=\"cancel()\"\n      type=\"button\"\n      class=\"btn btn-default\"\n      title=\"{{ 'Cancel' | translate }}\"\n    >\n      <span translate>Cancel</span>\n    </button>\n  </div>\n</div>\n"
        })
    ], BulkOperationsModalComponent);
    return BulkOperationsModalComponent;
}());
export { BulkOperationsModalComponent };
