// @c8y/ngx-components/esm5/operations/bulk-operations-list/modals/bulk-operation-modals.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { BulkOperationsModalComponent } from './bulk-operations-modal.component';
var BulkOperationModalsService = /** @class */ (function () {
    function BulkOperationModalsService(bsModalService) {
        this.bsModalService = bsModalService;
    }
    BulkOperationModalsService.prototype.showNewBulkOperationModal = function () {
        this.bsModalRef = this.bsModalService.show(BulkOperationsModalComponent, {
            backdrop: 'static',
            class: 'modal-sm'
        });
    };
    BulkOperationModalsService.prototype.hideNewBulkOperationModal = function () {
        if (this.bsModalRef) {
            this.bsModalRef.hide();
        }
    };
    BulkOperationModalsService.ctorParameters = function () { return [
        { type: BsModalService }
    ]; };
    BulkOperationModalsService = tslib_1.__decorate([
        Injectable()
    ], BulkOperationModalsService);
    return BulkOperationModalsService;
}());
export { BulkOperationModalsService };
