// @c8y/ngx-components/esm5/core/login/password-strength-validator.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { StrengthValidatorService } from './strength-validator-service';
var PasswordStrengthValidatorDirective = /** @class */ (function () {
    function PasswordStrengthValidatorDirective(passwordService) {
        this.passwordService = passwordService;
    }
    PasswordStrengthValidatorDirective_1 = PasswordStrengthValidatorDirective;
    Object.defineProperty(PasswordStrengthValidatorDirective.prototype, "passwordStrengthEnforced", {
        set: function (value) {
            this.forced = value;
        },
        enumerable: true,
        configurable: true
    });
    PasswordStrengthValidatorDirective.prototype.validate = function (control) {
        var strengthFulfilled = this.passwordService.isStrong(control.value || '');
        var enforcementForcedAndNotFulfilled = this.forced && !strengthFulfilled;
        return enforcementForcedAndNotFulfilled ? { passwordStrength: true } : null;
    };
    var PasswordStrengthValidatorDirective_1;
    PasswordStrengthValidatorDirective.ctorParameters = function () { return [
        { type: StrengthValidatorService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], PasswordStrengthValidatorDirective.prototype, "passwordStrengthEnforced", null);
    PasswordStrengthValidatorDirective = PasswordStrengthValidatorDirective_1 = tslib_1.__decorate([
        Directive({
            selector: '[passwordStrengthEnforced]',
            providers: [{ provide: NG_VALIDATORS, useExisting: PasswordStrengthValidatorDirective_1, multi: true }]
        })
    ], PasswordStrengthValidatorDirective);
    return PasswordStrengthValidatorDirective;
}());
export { PasswordStrengthValidatorDirective };
