// @c8y/ngx-components/esm5/core/search/search-outlet.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { sortBy } from 'lodash';
import { AlertService } from '../alert/alert.service';
var SearchOutletComponent = /** @class */ (function () {
    function SearchOutletComponent(alertService) {
        this.alertService = alertService;
        this.search = [];
    }
    SearchOutletComponent.prototype.getByPriority = function () {
        return sortBy(this.search, this.byPriority);
    };
    SearchOutletComponent.prototype.byPriority = function (item) {
        return -item.priority;
    };
    SearchOutletComponent.ctorParameters = function () { return [
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], SearchOutletComponent.prototype, "search", void 0);
    SearchOutletComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-search-outlet',
            template: "<ng-container *ngFor=\"let item of getByPriority()\">\n  <ng-container *c8yOutlet=\"item.template\"></ng-container>\n</ng-container>\n"
        })
    ], SearchOutletComponent);
    return SearchOutletComponent;
}());
export { SearchOutletComponent };
