// @c8y/ngx-components/esm5/context-dashboard/dashboard-actions.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Action, ActionFactory, gettext, ViewContext, Permissions } from '@c8y/ngx-components';
import { ActivatedRoute, Router } from '@angular/router';
import { ContextDashboardService } from './context-dashboard.service';
import { BsModalService } from 'ngx-bootstrap/modal';
import { DashboardDetailComponent } from './dashboard-detail.component';
var DashboardActionsFactory = /** @class */ (function () {
    function DashboardActionsFactory(router, contextDashboardService, bsModal, permissions) {
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.permissions = permissions;
    }
    DashboardActionsFactory.prototype.get = function (activatedRoute) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var context, canEditDashboard;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        context = !activatedRoute.parent || activatedRoute.snapshot.data.context
                            ? activatedRoute.snapshot.data
                            : activatedRoute.parent.snapshot.data;
                        return [4 /*yield*/, this.isEnabled(context)];
                    case 1:
                        canEditDashboard = _a.sent();
                        return [2 /*return*/, [
                                {
                                    label: gettext('Add dashboard'),
                                    priority: 2500,
                                    action: function () { return _this.addDashboard(context); },
                                    disabled: !canEditDashboard
                                }
                            ]];
                }
            });
        });
    };
    DashboardActionsFactory.prototype.addDashboard = function (context) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var initialState, modal, dashboardCfg, dashboardMO, ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (context.context === ViewContext.Device) {
                            initialState = {
                                deviceType: context.contextData.type
                            };
                        }
                        modal = this.bsModal.show(DashboardDetailComponent, {
                            class: 'modal-lg',
                            initialState: initialState,
                            ignoreBackdropClick: true
                        }).content;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 5, , 6]);
                        return [4 /*yield*/, modal.result];
                    case 2:
                        dashboardCfg = _a.sent();
                        return [4 /*yield*/, this.contextDashboardService.create(dashboardCfg, context)];
                    case 3:
                        dashboardMO = _a.sent();
                        return [4 /*yield*/, this.contextDashboardService.navigateToDashboard(dashboardMO)];
                    case 4:
                        _a.sent();
                        modal.close();
                        return [3 /*break*/, 6];
                    case 5:
                        ex_1 = _a.sent();
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    DashboardActionsFactory.prototype.isEnabled = function (context) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(context.contextData && context.contextData.id)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.permissions.canEdit(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE'], context.contextData)];
                    case 1: return [2 /*return*/, ((_a.sent()) && [ViewContext.Device, ViewContext.Group].includes(context.context))];
                    case 2: return [2 /*return*/, (this.permissions.hasAnyRole(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE']) &&
                            [ViewContext.Device, ViewContext.Group].includes(context.context))];
                }
            });
        });
    };
    DashboardActionsFactory.ctorParameters = function () { return [
        { type: Router },
        { type: ContextDashboardService },
        { type: BsModalService },
        { type: Permissions }
    ]; };
    DashboardActionsFactory = tslib_1.__decorate([
        Injectable()
    ], DashboardActionsFactory);
    return DashboardActionsFactory;
}());
export { DashboardActionsFactory };
