// @c8y/ngx-components/esm5/upgrade/auth-bridge.service.js
import { AppStateService, } from '@c8y/ngx-components';
import { BasicAuth, FetchClient } from '@c8y/client';
var AuthBridgeService = /** @class */ (function () {
    function AuthBridgeService(injector, basicAuth, fetchClient, appState) {
        this.injector = injector;
        this.basicAuth = basicAuth;
        this.fetchClient = fetchClient;
        this.appState = appState;
        this.hookAuth();
    }
    AuthBridgeService.prototype.updateBasicAuth = function (credentials) {
        var headers = this.fetchClient.getFetchOptions({}).headers;
        if (headers.Authorization) {
            var token = headers.Authorization.match(/basic\s(.*)$/i)[1];
            if (token) {
                this.basicAuth.updateCredentials(credentials);
                this.fetchClient.setAuth(this.basicAuth);
            }
        }
    };
    AuthBridgeService.prototype.hookAuth = function () {
        var _this = this;
        this.appState.currentUser.subscribe(function (u) {
            if (u) {
                _this.injector.get('c8yAuth').headers = function () { return _this.fetchClient.getFetchOptions({}).headers; };
                var headers = _this.fetchClient.getFetchOptions({}).headers;
                if (headers.Authorization) {
                    var token = headers.Authorization.match(/basic\s(.*)$/i)[1];
                    if (token) {
                        _this.setToken(token, headers.tfatoken);
                    }
                }
                else {
                    _this.setToken(undefined, headers.tfatoken, 'Oauth');
                }
                _this.injector.get('$rootScope').$emit('authStateChange', { hasAuth: true });
            }
            else {
                _this.injector.get('$rootScope').$emit('authStateChange', { hasAuth: false });
            }
        });
    };
    AuthBridgeService.prototype.setToken = function (token, tfa, type) {
        if (type === void 0) { type = 'Basic'; }
        var c8yAuth = this.injector.get('c8yAuth');
        if (type === 'Basic') {
            c8yAuth.onSetToken({ token: token, type: type });
            if (tfa) {
                c8yAuth.setTFAToken(tfa);
            }
        }
        else {
            c8yAuth.authReady();
        }
    };
    return AuthBridgeService;
}());
export { AuthBridgeService };
export function authBridgeServiceFactory(injector, basicAuth, fetchClient, appState) {
    return new AuthBridgeService(injector, basicAuth, fetchClient, appState);
}
export var authBridgeServiceProvider = {
    provide: AuthBridgeService,
    useFactory: authBridgeServiceFactory,
    deps: ['$injector', BasicAuth, FetchClient, AppStateService]
};
