// @c8y/ngx-components/esm5/core/device-status/device-availability.js
var _a, _b;
import { gettext } from '../i18n/gettext';
import { formatDate } from '@angular/common';
// TODO remove  when c8y-icon will works
export var SendStatus;
(function (SendStatus) {
    SendStatus["AVAILABLE"] = "AVAILABLE";
    SendStatus["UNAVAILABLE"] = "UNAVAILABLE";
    SendStatus["UNKNOWN"] = "UNKNOWN";
    SendStatus["MAINTENANCE"] = "MAINTENANCE";
    SendStatus["NOT_MONITORED"] = "NOT_MONITORED";
})(SendStatus || (SendStatus = {}));
export var PushStatus;
(function (PushStatus) {
    PushStatus["CONNECTED"] = "CONNECTED";
    PushStatus["DISCONNECTED"] = "DISCONNECTED";
    PushStatus["UNKNOWN"] = "UNKNOWN";
    PushStatus["MAINTENANCE"] = "MAINTENANCE";
})(PushStatus || (PushStatus = {}));
export var SendStatusLabels;
(function (SendStatusLabels) {
    SendStatusLabels[SendStatusLabels["AVAILABLE"] = gettext('Send connection: online')] = "AVAILABLE";
    SendStatusLabels[SendStatusLabels["NOT_MONITORED"] = gettext('Send connection: not monitored')] = "NOT_MONITORED";
    SendStatusLabels[SendStatusLabels["UNAVAILABLE"] = gettext('Send connection: offline')] = "UNAVAILABLE";
    SendStatusLabels[SendStatusLabels["UNKNOWN"] = gettext('Send connection status unknown')] = "UNKNOWN";
    SendStatusLabels[SendStatusLabels["MAINTENANCE"] = gettext('Device is under maintenance')] = "MAINTENANCE";
})(SendStatusLabels || (SendStatusLabels = {}));
export var PushStatusLabels;
(function (PushStatusLabels) {
    PushStatusLabels[PushStatusLabels["CONNECTED"] = gettext('Push connection: active')] = "CONNECTED";
    PushStatusLabels[PushStatusLabels["DISCONNECTED"] = gettext('Push connection: inactive')] = "DISCONNECTED";
    PushStatusLabels[PushStatusLabels["UNKNOWN"] = gettext('Push connection: inactive')] = "UNKNOWN";
    PushStatusLabels[PushStatusLabels["MAINTENANCE"] = gettext('Device is under maintenance')] = "MAINTENANCE";
})(PushStatusLabels || (PushStatusLabels = {}));
export var deviceAvailabilityIconMap = {
    sendData: (_a = {},
        _a[SendStatus.AVAILABLE] = {
            icon: 'long-arrow-right',
            class: 'statusOk'
        },
        _a[SendStatus.UNAVAILABLE] = {
            icon: 'long-arrow-right',
            class: 'statusNok'
        },
        _a[SendStatus.UNKNOWN] = {
            icon: 'long-arrow-right',
            class: 'statusUnknown'
        },
        _a[SendStatus.NOT_MONITORED] = {
            icon: 'long-arrow-right',
            class: 'statusUnknown'
        },
        _a),
    push: (_b = {},
        _b[PushStatus.CONNECTED] = {
            icon: 'long-arrow-left',
            class: 'statusOk'
        },
        _b[PushStatus.DISCONNECTED] = {
            icon: 'long-arrow-left',
            class: 'statusUnknown'
        },
        _b[PushStatus.UNKNOWN] = {
            icon: 'long-arrow-left',
            class: 'statusUnknown'
        },
        _b[PushStatus.MAINTENANCE] = {
            icon: 'wrench',
            class: 'statusAlert'
        },
        _b),
    device: {
        MAINTENANCE: {
            icon: 'wrench',
            class: 'statusAlert'
        }
    }
};
var ɵ0 = function (lastMessage, translateService) {
    var date = formatDate(new Date(lastMessage), 'medium', translateService.currentLang);
    return translateService.instant(gettext('Device last request:\n{{date}}'), { date: date });
};
export var tooltips = {
    sendStatusConnected: ɵ0,
    sendStatusDisconnected: gettext('Connection not monitored'),
    pushStatusConnected: gettext('Connection to device established'),
    pushStatusDisconnected: gettext('Connection to device not established (not supported by device or no connectivity)'),
    maintenance: gettext('Device is under maintenance')
};
export { ɵ0 };
