// @c8y/ngx-components/esm5/core/router/router-tabs.resolver.js
import * as tslib_1 from "tslib";
import { Injector, Injectable } from '@angular/core';
import { ActivatedRouteSnapshot, CanActivate, Router } from '@angular/router';
import { from, isObservable, of } from 'rxjs';
import { every, filter, map, mergeMap, toArray, tap } from 'rxjs/operators';
var RouterTabsResolver = /** @class */ (function () {
    function RouterTabsResolver(injector, router) {
        this.injector = injector;
        this.router = router;
    }
    Object.defineProperty(RouterTabsResolver.prototype, "routerStateSnapshot", {
        get: function () {
            return this.router.routerState.snapshot;
        },
        enumerable: true,
        configurable: true
    });
    RouterTabsResolver.prototype.resolve = function (route) {
        var _this = this;
        var latestChildren = this.router.config.find(function (r) { return route.routeConfig.path === r.path; }).children;
        return from(latestChildren || []).pipe(mergeMap(function (child) {
            return from(_this.getGuards(child)).pipe(map(function (guard) { return guard.canActivate(route, _this.routerStateSnapshot); }), mergeMap(function (result) {
                if (isObservable(result)) {
                    return result;
                }
                if (result && typeof result.then === 'function') {
                    return from(result);
                }
                return of(result);
            }), tap(function (result) {
                if (Array.isArray(result)) {
                    child.tabs = result;
                }
            }), every(function (result) { return !!result; }), filter(function (result) { return result; }), mergeMap(function () { return (Array.isArray(child.tabs) ? from(child.tabs) : of(child)); }), map(function (result) {
                return result.label || result.path
                    ? {
                        label: result.label,
                        icon: result.icon,
                        path: decodeURIComponent(route.url.join('/')) + "/" + result.path,
                        priority: result.priority,
                        hide: result.hide
                    }
                    : undefined;
            }));
        }), filter(function (tabs) { return !!tabs; }), toArray());
    };
    RouterTabsResolver.prototype.getGuards = function (child) {
        var _this = this;
        var guards = (child.canActivate || []).map(function (guard) {
            return _this.injector.get(guard);
        });
        if (child.data && child.data.showIf) {
            guards.push({ canActivate: child.data.showIf });
        }
        return guards;
    };
    RouterTabsResolver.ctorParameters = function () { return [
        { type: Injector },
        { type: Router }
    ]; };
    RouterTabsResolver = tslib_1.__decorate([
        Injectable()
    ], RouterTabsResolver);
    return RouterTabsResolver;
}());
export { RouterTabsResolver };
