// @c8y/ngx-components/esm5/core/action-bar/action-bar-item.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { ActionBarService } from './action-bar.service';
/**
 * A <c8y-action-bar-item> can be added with content projection.
 * It allows to show items in three positions (left, right
 * or more).
 *
 * @example
 * ```html
 * <c8y-action-bar-item [placement]="'right'">
 *   <button class="btn btn-link" (click)="addRandom()"><i c8yIcon="plus-circle"></i> Add random</button>
 * </c8y-action-bar-item>
 * ```
 */
var ActionBarItemComponent = /** @class */ (function () {
    /**
     * @ignore
     */
    function ActionBarItemComponent(actionBarService) {
        this.actionBarService = actionBarService;
        /**
         * Where the item should be shown. Use "more" to place
         * it in a dropdown on the right.
         */
        this.placement = 'left';
        /**
         * Priority for sorting the list. Higher priority means
         * an earlier position in the list.
         */
        this.priority = 0;
        /**
         * Adds additional classes to the item.
         */
        this.itemClass = '';
    }
    /**
     * @ignore
     */
    ActionBarItemComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.viewInitTimeout = setTimeout(function () { return _this.actionBarService.add(_this); });
    };
    /**
     * @ignore
     */
    ActionBarItemComponent.prototype.ngOnDestroy = function () {
        clearTimeout(this.viewInitTimeout);
        this.actionBarService.remove(this);
    };
    ActionBarItemComponent.ctorParameters = function () { return [
        { type: ActionBarService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], ActionBarItemComponent.prototype, "placement", void 0);
    tslib_1.__decorate([
        Input()
    ], ActionBarItemComponent.prototype, "priority", void 0);
    tslib_1.__decorate([
        Input()
    ], ActionBarItemComponent.prototype, "itemClass", void 0);
    tslib_1.__decorate([
        ViewChild('template', { static: false })
    ], ActionBarItemComponent.prototype, "template", void 0);
    ActionBarItemComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-action-bar-item',
            template: "<ng-template #template>\n  <li class=\"{{itemClass}}\">\n    <ng-content></ng-content>\n  </li>\n</ng-template>"
        })
    ], ActionBarItemComponent);
    return ActionBarItemComponent;
}());
export { ActionBarItemComponent };
