// @c8y/ngx-components/esm5/core/action-bar/action-bar.service.js
import * as tslib_1 from "tslib";
import { Injectable, InjectionToken, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject, Subject } from 'rxjs';
import { map, shareReplay } from 'rxjs/operators';
import { fromTrigger, sortByPriority, stateToFactory } from '../common/extension-hooks';
import { StateService } from '../common/state-service.abstract';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * A hook to add ActionBarItems using the multi provider extension concept.
 *
 * @example
 * ```typescript
 * providers: [
 *   {
 *     provide: HOOK_ACTION_BAR,
 *     useValue: [{ template: SomeComponent, priority: 10, placement: 'left' } as ActionBarItem],
 *     multi: true
 *   }
 * ]
 * ```
 *
 */
export var HOOK_ACTION_BAR = new InjectionToken('ActionBarFactory');
/**
 * A service which defines action-bar items via the multi provider concept.
 *
 * @example
 * ```typescript
 * // preferred way, multi provider concept:
 * providers: [
 *   {
 *     provide: HOOK_ACTION_BAR,
 *     useValue: [{ template: SomeComponent, priority: 10, placement: 'left' } as ActionBarItem],
 *     multi: true
 *   }
 * ]
 *
 * // use services:
 * this.actionBarService.add({ template: SomeComponent, priority: 10, placement: 'left' });
 * ```
 */
var ActionBarService = /** @class */ (function (_super) {
    tslib_1.__extends(ActionBarService, _super);
    /**
     * @ignore
     */
    function ActionBarService(injector, router) {
        var _this = _super.call(this) || this;
        _this.injector = injector;
        _this.refreshTrigger = new Subject();
        _this.state$ = new BehaviorSubject(new Set());
        _this.items$ = fromTrigger(router, _this.refreshTrigger, [
            function () { return _this.injector.get(HOOK_ACTION_BAR, []); },
            stateToFactory(_this.state$)
        ]).pipe(map(function (items) { return sortByPriority(items); }), shareReplay(1));
        return _this;
    }
    Object.defineProperty(ActionBarService.prototype, "state", {
        /**
         * Returns the current state.
         * @readonly
         * @returns The current set of actions.
         */
        get: function () {
            return this.state$.value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Refreshes the current store.
     */
    ActionBarService.prototype.refresh = function () {
        this.refreshTrigger.next();
    };
    /**
     * Adds a new item to the action bar in the header and emits a state change.
     * @param item The item to add.
     */
    ActionBarService.prototype.add = function (item) {
        this.state.add(item);
        this.emitNewState();
    };
    /**
     * Removes an action bar item from the header and emits a state change.
     * @param item The item to remove.
     */
    ActionBarService.prototype.remove = function (item) {
        this.state.delete(item);
        this.emitNewState();
    };
    ActionBarService.ctorParameters = function () { return [
        { type: Injector },
        { type: Router }
    ]; };
    ActionBarService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function ActionBarService_Factory() { return new ActionBarService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Router)); }, token: ActionBarService, providedIn: "root" });
    ActionBarService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        })
    ], ActionBarService);
    return ActionBarService;
}(StateService));
export { ActionBarService };
