// @c8y/ngx-components/esm5/protocol-opcua/opcua-auto-apply-settings.component.js
import * as tslib_1 from "tslib";
import { Component, Input, EventEmitter } from '@angular/core';
import { IManagedObject, InventoryService } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
import { reject } from 'lodash-es';
var OpcuaAutoApplySettingsComponent = /** @class */ (function () {
    function OpcuaAutoApplySettingsComponent(inventoryService) {
        this.inventoryService = inventoryService;
        this.opcuaServers = [];
        this.selectedItems = [];
        this.filteredList = [];
        this.checked = {};
        this.sizeToShowFilter = 5;
        this.constraints = {
            browsePathMatchesRegex: '',
            matchesNodeIds: [],
            serverObjectHasFragment: '',
            matchesServerIds: []
        };
        this.placeholderSelectServerIds = gettext('Select server IDs from list');
        this.updateSelectedItem = new EventEmitter();
    }
    OpcuaAutoApplySettingsComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data, matchesServerIds;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.inventoryService.list({
                            pageSize: 1000,
                            withTotalPages: true,
                            type: 'c8y_OpcuaServer'
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        this.opcuaServers = data;
                        this.filteredList = data;
                        this.selectedItems = [];
                        matchesServerIds = this.constraints.matchesServerIds;
                        data.forEach(function (server) {
                            if (matchesServerIds &&
                                matchesServerIds.length > 0 &&
                                matchesServerIds.find(function (itemId) { return itemId === server.id; })) {
                                _this.selectedItems.push(server);
                                _this.checked[server.id] = true;
                                _this.showServerIds = true;
                            }
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(OpcuaAutoApplySettingsComponent.prototype, "model", {
        get: function () {
            return this._model;
        },
        set: function (model) {
            if (model && model.applyConstraints) {
                this.constraints = model.applyConstraints;
            }
            this._model = model;
        },
        enumerable: true,
        configurable: true
    });
    OpcuaAutoApplySettingsComponent.prototype.serverIdsSelected = function (items) {
        if (this.constraints) {
            this.constraints.matchesServerIds = items.map(function (item) { return item.id; });
        }
        this.selectedItems = items;
    };
    OpcuaAutoApplySettingsComponent.prototype.onChangeNodeId = function (event) {
        if (event.target.checked) {
            this.showRootNodes = true;
            this.add();
        }
        else {
            this.showRootNodes = false;
            this.constraints.matchesNodeIds = [];
        }
    };
    OpcuaAutoApplySettingsComponent.prototype.onChangeShowServerIds = function (event) {
        if (!event.target.checked) {
            this.constraints.matchesServerIds = [];
            this.showServerIds = false;
            this.selectedItems = [];
            this.checked = {};
        }
        else {
            this.showServerIds = true;
        }
    };
    OpcuaAutoApplySettingsComponent.prototype.onChangeShowBrowsePath = function (event) {
        if (!event.target.checked) {
            this.constraints.browsePathMatchesRegex = '';
            this.showBrowsePath = false;
        }
        else {
            this.showBrowsePath = true;
        }
    };
    OpcuaAutoApplySettingsComponent.prototype.onChangeShowServerFragment = function (event) {
        if (!event.target.checked) {
            this.constraints.serverObjectHasFragment = '';
            this.showServerFragment = false;
        }
        else {
            this.showServerFragment = true;
        }
    };
    OpcuaAutoApplySettingsComponent.prototype.add = function () {
        this.constraints.matchesNodeIds.push('');
    };
    OpcuaAutoApplySettingsComponent.prototype.remove = function (index) {
        this.constraints.matchesNodeIds.splice(index, 1);
    };
    OpcuaAutoApplySettingsComponent.prototype.trackByFn = function (index, item) {
        return index;
    };
    OpcuaAutoApplySettingsComponent.prototype.updateConstraints = function (items) {
        if (this.constraints) {
            this.constraints.matchesServerIds = items.map(function (item) { return item.id; });
        }
    };
    OpcuaAutoApplySettingsComponent.prototype.filterItems = function (filterText) {
        if (filterText.length !== 0) {
            var search_1 = new RegExp(filterText, 'i');
            this.filteredList = this.opcuaServers.filter(function (_a) {
                var name = _a.name, id = _a.id;
                return search_1.test(name) || search_1.test(id);
            });
        }
        else {
            this.filteredList = this.opcuaServers;
        }
    };
    OpcuaAutoApplySettingsComponent.prototype.isChecked = function (item) {
        return this.checked[item.id];
    };
    OpcuaAutoApplySettingsComponent.prototype.onSelect = function (selected, item) {
        if (!selected) {
            this.selectedItems = reject(this.selectedItems, { id: item.id });
            delete this.checked[item.id];
        }
        else {
            this.selectedItems.push(item);
            this.checked[item.id] = selected;
        }
        this.updateSelectedItem.emit(true);
        this.updateConstraints(this.selectedItems);
    };
    OpcuaAutoApplySettingsComponent.ctorParameters = function () { return [
        { type: InventoryService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], OpcuaAutoApplySettingsComponent.prototype, "model", null);
    OpcuaAutoApplySettingsComponent = tslib_1.__decorate([
        Component({
            selector: 'opcua-auto-apply',
            template: "<div class=\"row\">\n  <div class=\"col-md-4\">\n    <p translate>\n      Specifying auto-apply constraints allows you to limit the scope where the device protocols are\n      applied, for example by specifying a set of possible servers or node IDs.\n    </p>\n    <p translate>\n      If no constraints are set, device protocols are applied at any fitting location on the OPC UA\n      server.\n    </p>\n  </div>\n\n  <div class=\"col-md-6\">\n    <ul class=\"list-group\" style=\"box-shadow: none;\">\n      <!-- Limit device type to a set of servers -->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol to a set of servers' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"\n              constraints.matchesServerIds !== null && constraints.matchesServerIds.length > 0\n            \"\n            (change)=\"onChangeShowServerIds($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol to a set of servers' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [collapse]=\"\n            (!showServerIds &&\n              (constraints.matchesServerIds !== null && constraints.matchesServerIds.length < 1)) ||\n            (!showServerIds && constraints.matchesServerIds === null)\n          \"\n        >\n          <c8y-form-group>\n            <c8y-select\n              [items]=\"opcuaServers\"\n              [selected]=\"selectedItems\"\n              [placeholder]=\"placeholderSelectServerIds\"\n              (onChange)=\"serverIdsSelected($event)\"\n              [updateItems]=\"updateSelectedItem\"\n            >\n              <c8y-li *ngIf=\"opcuaServers.length > sizeToShowFilter\">\n                <div class=\"p-l-8 p-r-8\">\n                 <c8y-filter [icon]=\"'search'\"  (onSearch)=\"filterItems($event)\"></c8y-filter>\n                </div>\n              </c8y-li>\n              <c8y-li *ngIf=\"filteredList && filteredList.length === 0\">\n                <c8y-li-body> \n                  <div class=\"c8y-empty-state text-left\">\n                  <p>{{ 'No items to display.' | translate }}</p>\n                </div>\n                </c8y-li-body>\n              </c8y-li>\n              <c8y-li *ngFor=\"let item of filteredList\">\n                <c8y-li-checkbox\n                  [selected]=\"isChecked(item)\"\n                  (onSelect)=\"onSelect($event, item)\"\n                ></c8y-li-checkbox>\n                <c8y-li-body>{{ item.name | translate }}<br>\n                  <span class=\"text-label-small m-r-4\" translate>ID</span> <small>{{ item.id }}</small>\n                </c8y-li-body>\n              </c8y-li>\n            </c8y-select>\n          </c8y-form-group>\n        </div>\n      </li>\n      <!-- Limit device type scope in the address space -->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol scope in the address space' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"\n              constraints.browsePathMatchesRegex !== null &&\n              constraints.browsePathMatchesRegex.length > 0\n            \"\n            (change)=\"onChangeShowBrowsePath($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol scope in the address space' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [isAnimated]=\"true\"\n          [collapse]=\"\n            (!showBrowsePath &&\n              constraints.browsePathMatchesRegex !== null &&\n              constraints.browsePathMatchesRegex.length < 1) ||\n            (!showBrowsePath && constraints.browsePathMatchesRegex === null)\n          \"\n        >\n          <c8y-form-group class=\"m-t-8 m-b-8\">\n            <input\n              name=\"browsePath\"\n              type=\"text\"\n              class=\"form-control\"\n              placeholder=\"{{ 'e.g.' | translate }} /objects/devices/.*\"\n              [(ngModel)]=\"constraints.browsePathMatchesRegex\"\n              ngDefaultControl\n            />\n          </c8y-form-group>\n        </div>\n      </li>\n      <!-- Limit device type to servers with a certain fragment-->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol to servers with a certain fragment' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"constraints.serverObjectHasFragment !== null && constraints.serverObjectHasFragment.length > 0\"\n            (change)=\"onChangeShowServerFragment($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol to servers with a certain fragment' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [isAnimated]=\"true\"\n          [collapse]=\"(!showServerFragment && constraints.serverObjectHasFragment !== null && constraints.serverObjectHasFragment.length < 1) || (!showServerFragment && constraints.serverObjectHasFragment === null)\"\n        >\n          <c8y-form-group class=\"m-t-8 m-b-8\">\n            <input\n              name=\"serverFragment\"\n              type=\"text\"\n              class=\"form-control\"\n              placeholder=\"{{ 'e.g.' | translate }} c8y_SomeServerMarker\"\n              [(ngModel)]=\"constraints.serverObjectHasFragment\"\n              ngDefaultControl\n            />\n          </c8y-form-group>\n        </div>\n      </li>\n      <!-- Limit device type to a specific root node ID -->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol to specific root nodes ID' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"constraints.matchesNodeIds !== null && constraints.matchesNodeIds.length > 0\"\n            (change)=\"onChangeNodeId($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol to specific root nodes ID' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [isAnimated]=\"true\"\n          [collapse]=\"(!showRootNodes && constraints.matchesNodeIds !== null && constraints.matchesNodeIds.length < 1) || ( !showRootNodes && constraints.matchesNodeIds === null)\"\n        >\n            <ul c8yInputGroupListContainer class=\"list-unstyled p-t-16\">\n              <li\n                class=\"m-b-8\"\n                *ngFor=\"let item of constraints.matchesNodeIds; let i = index; trackBy: trackByFn\"\n              >\n                <c8y-input-group-list [index]=\"i\" (onAdd)=\"add()\" (onRemove)=\"remove($event)\">\n                  <c8y-form-group class=\"form-group--tooltip-validation\">\n                    <input\n                      type=\"text\"\n                      class=\"form-control\"\n                      placeholder=\"{{ 'e.g.' | translate }} nodeId\"\n                      [(ngModel)]=\"constraints.matchesNodeIds[i]\"\n                      [required]=\"true\"\n                    />\n                  </c8y-form-group>\n                </c8y-input-group-list>\n              </li>\n            </ul> \n        </div>\n      </li>\n    </ul>\n  </div>\n</div>\n"
        })
    ], OpcuaAutoApplySettingsComponent);
    return OpcuaAutoApplySettingsComponent;
}());
export { OpcuaAutoApplySettingsComponent };
