// @c8y/ngx-components/esm2015/operations/bulk-operations-stepper-container/bulk-operations-stepper-container.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationsStepperContainerComponent } from './bulk-operations-stepper-container.component';
export const bulkOperatioStepperRoutes = [
    {
        path: 'devicecontrol/bulk/creation/:type',
        component: BulkOperationsStepperContainerComponent
    }
];
/**
 * Container module for the new bulk operation steppers. Use this module in your application
 * to provide a step by step guided process for scheduling of bulk operations.
 * The module reads the bulk operation type from the URL and displays the corresponding stepper.
 * You can provide stepper for your custom bulk operation type by registering it with the
 * HOOK_LIST_BULK_TYPE injection token.
 */
let BulkOperationsStepperContainerModule = class BulkOperationsStepperContainerModule {
};
BulkOperationsStepperContainerModule = tslib_1.__decorate([
    NgModule({
        declarations: [BulkOperationsStepperContainerComponent],
        imports: [CoreModule, RouterModule.forChild(bulkOperatioStepperRoutes)],
        entryComponents: [BulkOperationsStepperContainerComponent]
    })
], BulkOperationsStepperContainerModule);
export { BulkOperationsStepperContainerModule };
