// @c8y/ngx-components/bundles/c8y-ngx-components-operations-bulk-operation-scheduler.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@c8y/ngx-components'), require('ngx-bootstrap/datepicker'), require('ngx-bootstrap/timepicker'), require('lodash-es'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/operations/bulk-operation-scheduler', ['exports', '@angular/core', '@angular/forms', '@c8y/ngx-components', 'ngx-bootstrap/datepicker', 'ngx-bootstrap/timepicker', 'lodash-es', 'rxjs/operators'], factory) :
    (global = global || self, factory((global.c8y = global.c8y || {}, global.c8y['ngx-components'] = global.c8y['ngx-components'] || {}, global.c8y['ngx-components'].operations = global.c8y['ngx-components'].operations || {}, global.c8y['ngx-components'].operations['bulk-operation-scheduler'] = {}), global.ng.core, global.ng.forms, global.c8y['ngx-components'], global.datepicker, global.timepicker, global.lodashEs, global.rxjs.operators));
}(this, (function (exports, core, forms, ngxComponents, datepicker, timepicker, lodashEs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var OperationSchedulerComponent = /** @class */ (function () {
        function OperationSchedulerComponent(formBuilder) {
            this.formBuilder = formBuilder;
            this.placeholder = ngxComponents.gettext('Start date');
            this.delayErrors = null;
            this.pickerErrors = null;
            this.DELAY_SECONDS_DEFAULT = 1;
            this.DELAY_MILLISECONDS_DEFAULT = 1;
            this.MINUTES_AHEAD_DEFAULT = 5;
            this.delaySeconds = this.DELAY_SECONDS_DEFAULT;
            this.delayMilliseconds = this.DELAY_MILLISECONDS_DEFAULT;
            this.minutesAhead = this.MINUTES_AHEAD_DEFAULT;
            this.currentUnit = 'seconds';
        }
        OperationSchedulerComponent_1 = OperationSchedulerComponent;
        Object.defineProperty(OperationSchedulerComponent.prototype, "_minutesAhead", {
            set: function (minutes) {
                if (minutes && minutes > this.MINUTES_AHEAD_DEFAULT) {
                    this.minutesAhead = minutes;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(OperationSchedulerComponent.prototype, "_delayConfig", {
            set: function (config) {
                if (config) {
                    if (config.seconds > this.DELAY_SECONDS_DEFAULT) {
                        this.delaySeconds = config.seconds;
                    }
                    if (config.milliseconds > this.DELAY_MILLISECONDS_DEFAULT) {
                        this.delayMilliseconds = config.milliseconds;
                    }
                }
            },
            enumerable: true,
            configurable: true
        });
        OperationSchedulerComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.minDate = new Date();
            this.initialDate = new Date(this.minDate.setMinutes(this.minDate.getMinutes() + this.minutesAhead));
            this.minDelay = this.delaySeconds;
            this.fgOperationScheduler = this.formBuilder.group({
                picker: ['', [forms.Validators.required, this.dateValidation]],
                time: ['', [forms.Validators.required, this.timeValidation]],
                delay: ['', [forms.Validators.required, forms.Validators.min(this.minDelay)]],
                unit: ['seconds']
            });
            this.fgOperationScheduler.patchValue({
                picker: this.initialDate,
                time: this.initialDate,
                delay: this.minDelay
            });
            // Due to the validation of picker and time it could be possible that value changes
            // are emitted more than once. Therefore we throttle the emits.
            var valueChanges$ = this.fgOperationScheduler.valueChanges.pipe(operators.throttleTime(100));
            this.subscription = valueChanges$.subscribe(function (data) {
                _this.delayErrors = _this.fgOperationScheduler.controls.delay.errors;
                _this.pickerErrors = _this.fgOperationScheduler.controls.picker.errors;
                _this.convertDelayHandler(data.unit);
                _this.emitData(data);
            });
        };
        OperationSchedulerComponent.prototype.ngOnDestroy = function () {
            if (this.subscription && !this.subscription.closed) {
                this.subscription.unsubscribe();
            }
        };
        OperationSchedulerComponent.prototype.writeValue = function (value) {
            if (value) {
                this.fgOperationScheduler.patchValue({
                    picker: value.scheduledDate,
                    time: value.scheduledDate,
                    delay: value.delayInSeconds > 1 ? value.delayInSeconds : value.delayInSeconds * 1000,
                    unit: value.delayInSeconds > 1 ? 'seconds' : 'milliseconds'
                });
            }
        };
        OperationSchedulerComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        OperationSchedulerComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        OperationSchedulerComponent.prototype.setDisabledState = function (isDisabled) {
            isDisabled ? this.fgOperationScheduler.disable() : this.fgOperationScheduler.enable();
        };
        OperationSchedulerComponent.prototype.validate = function () {
            if (this.fgOperationScheduler.invalid) {
                return __assign({}, this.fgOperationScheduler.controls.picker.errors, this.fgOperationScheduler.controls.time.errors, this.fgOperationScheduler.controls.delay.errors);
            }
        };
        OperationSchedulerComponent.prototype.registerOnValidatorChange = function (fn) {
            this.onValidatorChanged = fn;
        };
        OperationSchedulerComponent.prototype.markAsTouched = function () {
            if (this.onTouched) {
                this.onTouched();
            }
        };
        OperationSchedulerComponent.prototype.convertDelayHandler = function (unit) {
            if (this.currentUnit === unit) {
                return;
            }
            this.currentUnit = unit;
            this.convertDelay(this.currentUnit);
            // update validator on delay control to make sure that
            // switching from minutes to seconds or vice versa does not harm validation.
            this.fgOperationScheduler.controls.delay.setValidators([forms.Validators.required]);
            this.fgOperationScheduler.controls.delay.updateValueAndValidity();
        };
        OperationSchedulerComponent.prototype.emitData = function (data) {
            if (this.onValidatorChanged) {
                this.onValidatorChanged();
            }
            if (data.picker && data.time) {
                data.picker = this.combineDateAndTime(data.picker, data.time);
            }
            this.convertDelay(this.currentUnit);
            data.delayInSeconds = this.delayInSeconds;
            if (this.onChange) {
                this.onChange({
                    delayInSeconds: data.delayInSeconds,
                    scheduledDate: data.picker
                });
            }
        };
        OperationSchedulerComponent.prototype.convertDelay = function (unit) {
            if (unit && this.fgOperationScheduler.controls.delay.value) {
                this.delayMilliseconds = this.fgOperationScheduler.controls.delay.value;
                if (unit === 'milliseconds') {
                    this.minDelay =
                        this.delayMilliseconds > this.DELAY_MILLISECONDS_DEFAULT
                            ? this.delayMilliseconds
                            : this.DELAY_MILLISECONDS_DEFAULT;
                    this.delayInSeconds = this.fgOperationScheduler.controls.delay.value / 1000;
                }
                else {
                    this.delaySeconds = this.fgOperationScheduler.controls.delay.value;
                    this.minDelay =
                        this.delaySeconds > this.DELAY_SECONDS_DEFAULT
                            ? this.delaySeconds
                            : this.DELAY_SECONDS_DEFAULT;
                    this.delayInSeconds = this.fgOperationScheduler.controls.delay.value;
                }
            }
        };
        OperationSchedulerComponent.prototype.combineDateAndTime = function (date, time) {
            return new Date(date.getFullYear(), date.getMonth(), date.getDate(), time.getHours(), time.getMinutes());
        };
        OperationSchedulerComponent.prototype.dateValidation = function (fControl) {
            if (fControl.value) {
                var date = fControl.value;
                fControl.parent.get('time').setValue(date);
                return date >= new Date()
                    ? null
                    : {
                        dateValidation: true
                    };
            }
            return { dateValidation: true };
        };
        OperationSchedulerComponent.prototype.timeValidation = function (fControl) {
            if (fControl.value) {
                var date = fControl.value;
                var result = date >= new Date()
                    ? null
                    : {
                        dateValidation: true
                    };
                var picker = fControl.parent.get('picker');
                if (result) {
                    picker.setErrors(result);
                    picker.markAsTouched();
                    return result;
                }
                if (picker && picker.errors && picker.errors.dateValidation) {
                    delete picker.errors.dateValidation;
                    if (lodashEs.isEmpty(picker.errors)) {
                        picker.setErrors(null);
                        return result;
                    }
                    picker.setErrors(picker.errors);
                }
                return result;
            }
            return { dateValidation: true };
        };
        var OperationSchedulerComponent_1;
        OperationSchedulerComponent.ctorParameters = function () { return [
            { type: forms.FormBuilder }
        ]; };
        __decorate([
            core.Input('minutesAhead')
        ], OperationSchedulerComponent.prototype, "_minutesAhead", null);
        __decorate([
            core.Input('delayConfig')
        ], OperationSchedulerComponent.prototype, "_delayConfig", null);
        OperationSchedulerComponent = OperationSchedulerComponent_1 = __decorate([
            core.Component({
                selector: 'c8y-operation-scheduler',
                template: "<div [formGroup]=\"fgOperationScheduler\">\n  <div class=\"form-group m-0\">\n    <label translate>Start date</label>\n    <div class=\"datetime-picker\">\n      <c8y-form-group class=\"datepicker\">\n        <input\n          formControlName=\"picker\"\n          class=\"form-control\"\n          placeholder=\"{{ placeholder | translate }}\"\n          [bsConfig]=\"{ customTodayClass: 'today' }\"\n          [minDate]=\"minDate\"\n          bsDatepicker\n          required\n          (blur)=\"markAsTouched()\"\n        />\n        <c8y-messages *ngIf=\"pickerErrors\">\n          <c8y-message *ngIf=\"pickerErrors.required\" translate>\n            This field is required.\n          </c8y-message>\n          <c8y-message *ngIf=\"pickerErrors.dateValidation && !pickerErrors.required\" translate>\n            Select time in the future.\n          </c8y-message>\n        </c8y-messages>\n      </c8y-form-group>\n      <timepicker\n        class=\"form-group\"\n        [showSpinners]=\"false\"\n        [showMeridian]=\"false\"\n        formControlName=\"time\"\n        (blur)=\"markAsTouched()\"\n      ></timepicker>\n    </div>\n  </div>\n  <div class=\"form-group m-0\">\n    <c8y-form-group [hasError]=\"delayErrors\">\n      <label translate>Delay</label>\n      <div class=\"input-group\">\n        <input\n          formControlName=\"delay\"\n          type=\"number\"\n          class=\"form-control\"\n          placeholder=\"{{ 'e.g.' | translate }} 15\"\n          required\n          (blur)=\"markAsTouched()\"\n        />\n        <div class=\"input-group-btn\">\n          <div class=\"c8y-select-wrapper\">\n            <select formControlName=\"unit\" class=\"form-control\" (blur)=\"markAsTouched()\">\n              <option value=\"seconds\" translate>Seconds</option>\n              <option value=\"milliseconds\" translate>Milliseconds</option>\n            </select>\n            <span></span>\n          </div>\n        </div>\n      </div>\n      <c8y-messages *ngIf=\"delayErrors\">\n        <c8y-message *ngIf=\"delayErrors.required\" translate>\n          This field is required.\n        </c8y-message>\n        <c8y-message\n          *ngIf=\"delayErrors.min && !delayErrors.required\"\n          translate\n          ngNonBindable\n          [translateParams]=\"{ delay: minDelay }\"\n        >\n          Minimum value is {{ delay }}.\n        </c8y-message>\n      </c8y-messages>\n    </c8y-form-group>\n  </div>\n</div>\n",
                providers: [
                    {
                        provide: forms.NG_VALUE_ACCESSOR,
                        multi: true,
                        useExisting: core.forwardRef(function () { return OperationSchedulerComponent_1; })
                    },
                    {
                        provide: forms.NG_VALIDATORS,
                        multi: true,
                        useExisting: core.forwardRef(function () { return OperationSchedulerComponent_1; })
                    }
                ]
            })
        ], OperationSchedulerComponent);
        return OperationSchedulerComponent;
    }());

    /**
     * This module provides components for scheduling bulk operations.
     */
    var BulkOperationSchedulerModule = /** @class */ (function () {
        function BulkOperationSchedulerModule() {
        }
        BulkOperationSchedulerModule = __decorate([
            core.NgModule({
                imports: [
                    ngxComponents.CoreModule,
                    forms.ReactiveFormsModule,
                    datepicker.BsDatepickerModule.forRoot(),
                    timepicker.TimepickerModule.forRoot()
                ],
                providers: [],
                declarations: [OperationSchedulerComponent],
                entryComponents: [OperationSchedulerComponent],
                exports: [OperationSchedulerComponent]
            })
        ], BulkOperationSchedulerModule);
        return BulkOperationSchedulerModule;
    }());

    exports.BulkOperationSchedulerModule = BulkOperationSchedulerModule;
    exports.OperationSchedulerComponent = OperationSchedulerComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
