// @c8y/ngx-components/esm5/sub-assets/sub-assets.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CoreModule, gettext, HOOK_ONCE_ROUTE, ViewContext } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { AddGroupModule } from './add-group/add-group.module';
import { AssetTypeCellRendererComponent } from './columns/asset-type.cell-renderer.component';
import { DeleteAssetsModalComponent } from './delete-assets-modal/delete-assets-modal.component';
import { GroupInfoComponent } from './group-info.component';
import { GroupsComponent } from './groups.component';
import { SubAssetsGridComponent } from './sub-assets-grid.component';
import { SubAssetsComponent } from './sub-assets.component';
import { SubAssetsService } from './sub-assets.service';
import { UnassignModalComponent } from './unassign-assets-modal/unassign-modal.component';
import { AssignDevicesComponent } from './assign-devices/assign-devices.component';
import { AssignChildDevicesComponent } from './assign-devices/assign-child-devices.component';
import { AssignDevicesService } from './assign-devices/assign-devices.service';
var ɵ0 = [
    {
        context: ViewContext.Group,
        path: 'subassets',
        priority: 1000,
        icon: 'c8y-group-open',
        label: gettext('Subassets'),
        component: SubAssetsComponent
    },
    {
        path: 'group',
        component: GroupsComponent
    }
];
var SubAssetsModule = /** @class */ (function () {
    function SubAssetsModule() {
    }
    SubAssetsModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                SubAssetsComponent,
                SubAssetsGridComponent,
                AssetTypeCellRendererComponent,
                GroupsComponent,
                GroupInfoComponent,
                DeleteAssetsModalComponent,
                UnassignModalComponent,
                AssignDevicesComponent,
                AssignChildDevicesComponent
            ],
            imports: [CoreModule, DeviceGridModule, AddGroupModule, PopoverModule],
            entryComponents: [
                SubAssetsComponent,
                AssetTypeCellRendererComponent,
                GroupsComponent,
                DeleteAssetsModalComponent,
                UnassignModalComponent
            ],
            exports: [],
            providers: [
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: ɵ0,
                    multi: true
                },
                SubAssetsService,
                AssignDevicesService
            ]
        })
    ], SubAssetsModule);
    return SubAssetsModule;
}());
export { SubAssetsModule };
export { ɵ0 };
