// @c8y/ngx-components/esm5/sub-assets/sub-assets.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { AppStateService, DisplayOptions } from '@c8y/ngx-components';
import { DeviceGroupService } from '@c8y/ngx-components/assets-navigator';
import { SubAssetsService } from './sub-assets.service';
var SubAssetsComponent = /** @class */ (function () {
    function SubAssetsComponent(route, appState, deviceGroupService, subAssetsService) {
        var _this = this;
        this.route = route;
        this.appState = appState;
        this.deviceGroupService = deviceGroupService;
        this.subAssetsService = subAssetsService;
        this.toggle = false;
        this.showAddGroup = false;
        this.showAssignDevices = false;
        this.showGroupInfo = true;
        this.showAssignChildDevices = false;
        this.refresh = new EventEmitter();
        this.filterable = true;
        this.dataSub = this.route.parent.data.subscribe(function (_a) {
            var contextData = _a.contextData;
            _this.title = contextData.name;
            _this.currentGroupId = contextData.id;
            _this.group = contextData;
            _this.filterable = !_this.deviceGroupService.isDynamicGroup(contextData);
        });
    }
    SubAssetsComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.appState.state$.subscribe(function (state) {
            _this.showGroupInfo = state.app.name === 'cockpit' || state.app.name === 'devicemanagement';
        });
        this.displayOptions = {
            striped: true,
            bordered: false,
            gridHeader: true,
            filter: !this.isSmartGroup()
        };
    };
    SubAssetsComponent.prototype.groupChange = function (group) {
        /**
         * Group is cached in the context-route guard. To avoid loading old data (when switching tabs) we need to update
         * this object in the cache.
         */
        Object.assign(this.group, group);
        this.title = group.name;
        this.refresh.emit();
    };
    SubAssetsComponent.prototype.isSmartGroup = function () {
        return this.subAssetsService.isSmartGroup(this.group);
    };
    SubAssetsComponent.prototype.ngOnDestroy = function () {
        if (this.dataSub) {
            this.dataSub.unsubscribe();
        }
    };
    SubAssetsComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: AppStateService },
        { type: DeviceGroupService },
        { type: SubAssetsService }
    ]; };
    SubAssetsComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-sub-assets',
            template: "<c8y-title>\n  {{ title }}\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add group' | translate }}\"\n    (click)=\"showAddGroup = !showAddGroup\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Add group</span>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Assign devices' | translate }}\"\n    (click)=\"showAssignDevices = !showAssignDevices\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Assign devices</span>\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"card content-fullpage d-flex d-col\">\n  <c8y-group-info\n    *ngIf=\"showGroupInfo\"\n    [group]=\"group\"\n    (onGroupChange)=\"groupChange($event)\"\n  ></c8y-group-info>\n  <c8y-sub-assets-grid\n    [refresh]=\"refresh\"\n    [parent-group]=\"group\"\n    [filterable]=\"filterable\"\n    [displayOptions]=\"displayOptions\"\n    class=\"d-contents\"\n  ></c8y-sub-assets-grid>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAddGroup }\">\n  <div class=\"bottom-drawer\">\n    <c8y-add-group\n      [currentGroupId]=\"currentGroupId\"\n      [refresh]=\"refresh\"\n      (onCancel)=\"showAddGroup = false\"\n      *ngIf=\"showAddGroup\"\n    ></c8y-add-group>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignDevices }\">\n  <div class=\"bottom-drawer\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-devices\n        class=\"d-contents\"\n        (onCancel)=\"showAssignDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        (onShowChildDevices)=\"showAssignChildDevices=$event\"\n        (selectedDevice)=\"showChildrenForDevice=$event\"\n        *ngIf=\"showAssignDevices\"\n      ></c8y-assign-devices>\n    </div>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignChildDevices }\" *ngIf=\"showAssignChildDevices\">\n  <div class=\"bottom-drawer m-t-40\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-child-devices\n        class=\"d-contents\"\n        (onCancel)=\"showAssignChildDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        [parentDevice]=\"showChildrenForDevice\"\n      ></c8y-assign-child-devices>\n    </div>\n  </div>\n</div>\n"
        })
    ], SubAssetsComponent);
    return SubAssetsComponent;
}());
export { SubAssetsComponent };
