// @c8y/ngx-components/esm5/core/list-group/list-item-radio.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * A component to display a radio button in a c8y-li. Use it with content projection. It
 * can be used standalone or with an ngModel.
 *
 * @example
 * ```html
 * <c8y-li>
 *   <c8y-li-radio></c8y-li-radio>
 * </c8y-li>
 * ```
 */
var ListItemRadioComponent = /** @class */ (function () {
    function ListItemRadioComponent() {
        /**
         * Indicates if the radio is selected.
         */
        this.selected = false;
        /**
         * The name of the radio (use different names for different radio groups).
         */
        this.name = 'list_item_radio';
        /**
         * Indicates if the radio is disabled.
         */
        this.disabled = false;
        /**
         * An event emitted as soon as the radio is selected.
         */
        this.onSelect = new EventEmitter();
    }
    ListItemRadioComponent_1 = ListItemRadioComponent;
    /**
     * Resets the native radio element.
     */
    ListItemRadioComponent.prototype.reset = function () {
        this.radio.nativeElement.checked = false;
    };
    /**
     * @ignore
     */
    ListItemRadioComponent.prototype.ngOnChanges = function (changes) {
        if (changes.selected && !changes.selected.firstChange) {
            this.onSelect.emit(changes.selected.currentValue);
        }
    };
    /**
     * Triggered when the radio is changed.
     * @param checked Should it be checked or not?
     */
    ListItemRadioComponent.prototype.change = function (checked) {
        this.onSelect.emit(checked);
        this.model = this.value;
        if (this.onChange) {
            this.onChange(this.model);
        }
        if (this.onTouched) {
            this.onTouched();
        }
    };
    /**
     * Registers an onChange event.
     */
    ListItemRadioComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    /**
     * Registers an onTouch event.
     */
    ListItemRadioComponent.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    /**
     * @ignore
     */
    ListItemRadioComponent.prototype.writeValue = function (value) {
        this.model = value;
    };
    var ListItemRadioComponent_1;
    tslib_1.__decorate([
        ViewChild('radio', { static: true })
    ], ListItemRadioComponent.prototype, "radio", void 0);
    tslib_1.__decorate([
        Input()
    ], ListItemRadioComponent.prototype, "selected", void 0);
    tslib_1.__decorate([
        Input()
    ], ListItemRadioComponent.prototype, "name", void 0);
    tslib_1.__decorate([
        Input()
    ], ListItemRadioComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        Output()
    ], ListItemRadioComponent.prototype, "onSelect", void 0);
    tslib_1.__decorate([
        Input()
    ], ListItemRadioComponent.prototype, "value", void 0);
    ListItemRadioComponent = ListItemRadioComponent_1 = tslib_1.__decorate([
        Component({
            selector: 'c8y-list-item-radio, c8y-li-radio',
            template: "<label class=\"c8y-radio\">\n  <input\n    #radio\n    type=\"radio\"\n    [name]=\"name\"\n    [checked]=\"selected\"\n    (change)=\"change($event.target.checked); (false)\"\n    [disabled]=\"disabled\"\n  />\n  <span></span>\n</label>\n<ng-content></ng-content>\n",
            host: {
                class: 'c8y-list__item__radio'
            },
            providers: [
                {
                    provide: NG_VALUE_ACCESSOR,
                    multi: true,
                    useExisting: forwardRef(function () { return ListItemRadioComponent_1; })
                }
            ]
        })
    ], ListItemRadioComponent);
    return ListItemRadioComponent;
}());
export { ListItemRadioComponent };
