// @c8y/ngx-components/esm5/core/list-group/list-group.component.js
import * as tslib_1 from "tslib";
import { Component, HostBinding, ContentChild } from '@angular/core';
import { ListItemTimelineComponent } from './list-item-timeline.component';
/**
 * List groups are flexible and powerful components for displaying
 * both simple lists of elements and complex ones with custom content.
 *
 * @example
 * ```html
 * <c8y-list-group>
 *   <c8y-li>
 *     Plain <strong>Text</strong>
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-icon [icon]="'rocket'"></c8y-li-icon>
 *     Icon
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-radio></c8y-li-radio>
 *     Radio
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-checkbox></c8y-li-checkbox>
 *     Checkbox
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-footer>
 *       Footer
 *     </c8y-li-footer>
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-action icon="rocket">
 *       Launch
 *     </c8y-li-action>
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-collapse>
 *       Collapsed content
 *     </c8y-li-collapse>
 *   </c8y-li>
 * </c8y-list-group>
 * ```
 */
var ListGroupComponent = /** @class */ (function () {
    function ListGroupComponent() {
        /**
         * @ignore
         */
        this.timeline = false;
    }
    /**
     * @ignore
     */
    ListGroupComponent.prototype.ngAfterContentChecked = function () {
        this.timeline = !!this.timelineComponent;
    };
    tslib_1.__decorate([
        HostBinding('class.c8y-list--timeline')
    ], ListGroupComponent.prototype, "timeline", void 0);
    tslib_1.__decorate([
        ContentChild(ListItemTimelineComponent, { static: false })
    ], ListGroupComponent.prototype, "timelineComponent", void 0);
    ListGroupComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-list-group',
            template: "<ng-content></ng-content>\n\n",
            host: {
                class: 'c8y-list__group'
            }
        })
    ], ListGroupComponent);
    return ListGroupComponent;
}());
export { ListGroupComponent };
