// @c8y/ngx-components/esm5/repository/software/add-software-modal.component.js
import * as tslib_1 from "tslib";
import { Component, Output, EventEmitter, ViewChild } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { RepositoryService } from '../repository.service';
import { RepositoryType } from '../repository.model';
import { AlertService, gettext } from '@c8y/ngx-components';
import { isUndefined, get, assign } from 'lodash-es';
import { BehaviorSubject, from } from 'rxjs';
import { debounceTime, distinctUntilChanged, tap, switchMap } from 'rxjs/operators';
var AddSoftwareModalComponent = /** @class */ (function () {
    function AddSoftwareModalComponent(modal, repositoryService, alert) {
        this.modal = modal;
        this.repositoryService = repositoryService;
        this.alert = alert;
        this.saved = new EventEmitter();
        this.onInput = new BehaviorSubject('');
        this.model = {
            selected: undefined,
            version: undefined,
            description: undefined,
            deviceType: undefined,
            binary: {
                file: undefined,
                url: undefined
            }
        };
        this.saving = false;
        this.softwarePreselected = false;
    }
    AddSoftwareModalComponent.prototype.ngOnInit = function () {
        this.setInitialState();
        this.loadSoftwares();
    };
    AddSoftwareModalComponent.prototype.setInitialState = function () {
        if (this.model.selected) {
            this.softwarePreselected = true;
        }
    };
    AddSoftwareModalComponent.prototype.loadSoftwares = function () {
        var _this = this;
        this.inputSubscription$ = this.onInput
            .pipe(tap(function () {
            if (!_this.softwarePreselected) {
                _this.model.description = null;
                if (_this.form) {
                    _this.form.form.get('description').reset();
                }
            }
        }), debounceTime(300), distinctUntilChanged(), switchMap(function (searchStr) { return _this.getSoftwareResult(searchStr); }))
            .subscribe(function (result) {
            _this.softwaresResult = result;
        });
    };
    AddSoftwareModalComponent.prototype.getSoftwareResult = function (searchStr) {
        return from(this.repositoryService.listRepositoryEntries(RepositoryType.SOFTWARE, {
            partialName: searchStr,
            skipLegacy: true
        }));
    };
    AddSoftwareModalComponent.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                this.saving = true;
                this.repositoryService
                    .create(this.model, RepositoryType.SOFTWARE)
                    .then(function (savedSoftware) {
                    _this.successMsg();
                    _this.saving = false;
                    _this.saved.next(savedSoftware);
                    _this.cancel();
                })
                    .catch(function (e) {
                    _this.saving = false;
                    _this.saved.error(e);
                    _this.cancel();
                });
                return [2 /*return*/];
            });
        });
    };
    AddSoftwareModalComponent.prototype.successMsg = function () {
        var msg = gettext('Software added.');
        this.alert.success(msg);
    };
    AddSoftwareModalComponent.prototype.cancel = function () {
        this.modal.hide();
        this.saved.complete();
    };
    AddSoftwareModalComponent.prototype.ngOnDestroy = function () {
        this.inputSubscription$.unsubscribe();
    };
    AddSoftwareModalComponent.prototype.onFile = function (dropped) {
        if (!isUndefined(dropped.url)) {
            this.model.binary = {
                url: dropped.url
            };
            return;
        }
        else if (!isUndefined(dropped.droppedFiles)) {
            this.model.binary = {
                file: dropped.droppedFiles[0].file
            };
            return;
        }
        else {
            this.model.binary = {
                file: undefined,
                url: undefined
            };
        }
    };
    AddSoftwareModalComponent.prototype.onSoftwareSelect = function (software) {
        assign(this.model, {
            selected: software,
            description: software.description,
            deviceType: get(software, 'c8y_Filter.type')
        });
    };
    AddSoftwareModalComponent.ctorParameters = function () { return [
        { type: BsModalRef },
        { type: RepositoryService },
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        ViewChild('softwareForm', { static: false })
    ], AddSoftwareModalComponent.prototype, "form", void 0);
    tslib_1.__decorate([
        Output()
    ], AddSoftwareModalComponent.prototype, "saved", void 0);
    AddSoftwareModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-add-software-software-modal',
            template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i [c8yIcon]=\"'c8y-tools'\"></i>\n    <h4 translate>\n      Add software\n    </h4>\n  </div>\n  <div class=\"p-16 text-center separator-bottom\" *ngIf=\"!softwarePreselected\">\n    <p class=\"lead m-0\" translate>\n      Select or create new software\n    </p>\n  </div>\n  <form\n    class=\"d-contents\"\n    autocomplete=\"off\"\n    #softwareForm=\"ngForm\"\n    (ngSubmit)=\"softwareForm.form.valid && save()\"\n  >\n    <div class=\"modal-inner-scroll\">\n      <div class=\"modal-body\">\n        <div [hidden]=\"softwarePreselected\">\n          <c8y-form-group>\n            <label for=\"softwareName\" translate>Software</label>\n            <c8y-typeahead\n              [(ngModel)]=\"model.selected\"\n              name=\"softwareName\"\n              placeholder=\"{{ 'Select or enter' | translate }}\"\n              (onSearch)=\"onInput.next($event)\"\n              [required]=\"true\"\n            >\n              <c8y-li\n                *c8yFor=\"\n                  let software of softwaresResult;\n                  loadMore: 'auto';\n                  notFound: notFoundTemplate\n                \"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"onSoftwareSelect(software)\"\n                [active]=\"model.selected === software\"\n              >\n                <c8y-highlight\n                  [text]=\"software.name || '--'\"\n                  [pattern]=\"onInput | async\"\n                ></c8y-highlight>\n              </c8y-li>\n              <ng-template #notFoundTemplate>\n                <c8y-li class=\"bg-gray-lighter p-8\" *ngIf=\"(onInput | async)?.length > 0\">\n                  <span translate>No match found, add new`software`? </span>\n                  <button\n                    title=\"{{ 'Create new`software`' | translate }}\"\n                    type=\"button\"\n                    class=\"btn btn-primary btn-xs m-l-16\"\n                    translate\n                  >\n                    Create new`software`\n                  </button>\n                </c8y-li>\n              </ng-template>\n            </c8y-typeahead>\n          </c8y-form-group>\n\n          <c8y-form-group>\n            <label for=\"softwareDescription\" translate>Description</label>\n            <input\n              id=\"softwareDescription\"\n              class=\"form-control\"\n              autocomplete=\"off\"\n              name=\"description\"\n              [(ngModel)]=\"model.description\"\n              placeholder=\"{{ 'e.g. Cloud connectivity software' | translate }}\"\n              [disabled]=\"model.selected?.id\"\n              [required]=\"true\"\n            />\n          </c8y-form-group>\n\n          <c8y-form-group>\n            <label for=\"softwareDeviceTypeFilter\" translate>Device type filter</label>\n            <input\n              id=\"softwareDeviceTypeFilter\"\n              class=\"form-control\"\n              name=\"softwareDeviceTypeFilter\"\n              popover=\"{{\n                'If the filter is set, the firmware will show up for installation only for devices of that type. If no filter is set, it will be available for all devices.'\n                  | translate\n              }}\"\n              [(ngModel)]=\"model.deviceType\"\n              placeholder=\"{{ 'e.g.' | translate }} c8y_Linux\"\n              [disabled]=\"model.selected?.id\"\n            />\n          </c8y-form-group>\n        </div>\n\n        <c8y-form-group>\n          <label for=\"softwareVersion\" translate>Version</label>\n          <input\n            id=\"softwareVersion\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            name=\"version\"\n            [(ngModel)]=\"model.version\"\n            placeholder=\"{{ 'e.g.' | translate }} 1.0.0\"\n            [required]=\"true\"\n          />\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <div class=\"legend form-block m-t-40\" translate>Software file</div>\n          <c8y-file-picker [maxAllowedFiles]=\"1\" (onFilesPicked)=\"onFile($event)\">\n          </c8y-file-picker>\n        </c8y-form-group>\n      </div>\n    </div>\n    <div class=\"modal-footer\">\n      <button\n        title=\"{{ 'Cancel' | translate }}\"\n        class=\"btn btn-default\"\n        type=\"button\"\n        (click)=\"cancel()\"\n        [disabled]=\"saving\"\n        translate\n      >\n        Cancel\n      </button>\n\n      <button\n        title=\"{{ 'Add software' | translate }}\"\n        class=\"btn btn-primary\"\n        type=\"submit\"\n        [ngClass]=\"{ 'btn-pending': saving }\"\n        [disabled]=\"\n          !softwareForm.form.valid ||\n          softwareForm.form.pristine ||\n          saving ||\n          (!model.binary?.url && !model.binary?.file)\n        \"\n        translate\n      >\n        Add software\n      </button>\n    </div>\n  </form>\n</div>\n"
        })
    ], AddSoftwareModalComponent);
    return AddSoftwareModalComponent;
}());
export { AddSoftwareModalComponent };
