// @c8y/ngx-components/esm5/operations/bulk-operations-list/bulk-operation-details-tabs.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
var BulkOperationDetailsTabsComponent = /** @class */ (function () {
    function BulkOperationDetailsTabsComponent() {
        this.readOnly = false;
        this.onRetryFailedOperations = new EventEmitter();
        this.showFailedOperation = new EventEmitter();
        this.tabs = [];
    }
    BulkOperationDetailsTabsComponent.prototype.ngOnInit = function () {
        this.template = this.detailsTemplate;
        this.tabs = [
            {
                label: gettext('Details'),
                path: null,
                icon: 'asterisk',
                template: this.detailsTemplate
            },
            {
                label: gettext('History of changes'),
                path: null,
                icon: 'archive',
                template: this.auditTemplate
            }
        ];
    };
    tslib_1.__decorate([
        Input()
    ], BulkOperationDetailsTabsComponent.prototype, "bulkOperation", void 0);
    tslib_1.__decorate([
        Input()
    ], BulkOperationDetailsTabsComponent.prototype, "readOnly", void 0);
    tslib_1.__decorate([
        Output()
    ], BulkOperationDetailsTabsComponent.prototype, "onRetryFailedOperations", void 0);
    tslib_1.__decorate([
        Output()
    ], BulkOperationDetailsTabsComponent.prototype, "showFailedOperation", void 0);
    tslib_1.__decorate([
        ViewChild('details', { static: true })
    ], BulkOperationDetailsTabsComponent.prototype, "detailsTemplate", void 0);
    tslib_1.__decorate([
        ViewChild('audit', { static: true })
    ], BulkOperationDetailsTabsComponent.prototype, "auditTemplate", void 0);
    BulkOperationDetailsTabsComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-bulk-operation-details-tabs',
            template: "<div #container class=\"tabContainer\">\n  <ul class=\"nav nav-tabs nav-tabsc8y\">\n    <li [ngClass]=\"{ active: tab.template === template }\" *ngFor=\"let tab of tabs\">\n      <button [title]=\"tab.label | translate\" class=\"btn-clean\" (click)=\"template = tab.template\">\n        <i [c8yIcon]=\"tab.icon\"></i>\n        <span class=\"txt\">{{ tab.label | translate }}</span>\n      </button>\n    </li>\n  </ul>\n</div>\n\n<ng-container *ngTemplateOutlet=\"template\"></ng-container>\n\n<ng-template #details>\n  <c8y-bulk-operation-details\n    [bulkOperation]=\"bulkOperation\"\n    [readOnly]=\"readOnly\"\n    (showFailedOperation)=\"showFailedOperation.emit($event)\"\n  >\n  </c8y-bulk-operation-details>\n</ng-template>\n\n<ng-template #audit>\n  <c8y-audit-log [source]=\"bulkOperation.id\"></c8y-audit-log>\n</ng-template>\n\n<c8y-single-operations-list\n  [bulkOperation]=\"bulkOperation\"\n  [readOnly]=\"readOnly\"\n  (onRetryFailedOperations)=\"onRetryFailedOperations.emit($event)\"\n  class=\"d-flex d-col\"\n></c8y-single-operations-list>\n"
        })
    ], BulkOperationDetailsTabsComponent);
    return BulkOperationDetailsTabsComponent;
}());
export { BulkOperationDetailsTabsComponent };
