// @c8y/ngx-components/esm5/operations/bulk-operations-list/bulk-operation-details.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
import { BulkOperationWrapperComponent } from './bulk-operation-wrapper.component';
var BulkOperationDetailsService = /** @class */ (function () {
    function BulkOperationDetailsService(bulkOperationsService, modalService) {
        this.bulkOperationsService = bulkOperationsService;
        this.modalService = modalService;
    }
    BulkOperationDetailsService.prototype.openDetails = function (bulkOperationId) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var bulkOperation, initialState;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.bulkOperationsService.getBulkOperationById(bulkOperationId)];
                    case 1:
                        bulkOperation = (_a.sent()).data;
                        initialState = { bulkOperation: bulkOperation };
                        this.modalService.show(BulkOperationWrapperComponent, {
                            initialState: initialState,
                            class: 'modal-lg'
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    BulkOperationDetailsService.ctorParameters = function () { return [
        { type: BulkOperationsService },
        { type: BsModalService }
    ]; };
    BulkOperationDetailsService = tslib_1.__decorate([
        Injectable()
    ], BulkOperationDetailsService);
    return BulkOperationDetailsService;
}());
export { BulkOperationDetailsService };
