// @c8y/ngx-components/esm5/ecosystem/ecosystem.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, HOOK_NAVIGATOR_NODES } from '@c8y/ngx-components';
import { EcosystemNavigationFactory } from './ecosystem-navigation-factory';
import { ApplicationListComponent } from './application-list.component';
import { EcosystemGuard } from './ecosystem.guard';
import { MicroserviceListComponent } from './microservice-list.component';
var ecosystemRoutes = [
    {
        path: 'application-list',
        component: ApplicationListComponent,
        canActivate: [EcosystemGuard]
    },
    {
        path: 'microservices',
        component: MicroserviceListComponent,
        canActivate: [EcosystemGuard]
    }
];
var EcosystemModule = /** @class */ (function () {
    function EcosystemModule() {
    }
    EcosystemModule = tslib_1.__decorate([
        NgModule({
            declarations: [ApplicationListComponent, MicroserviceListComponent],
            imports: [CoreModule, RouterModule.forRoot(ecosystemRoutes, { useHash: true })],
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: EcosystemNavigationFactory,
                    multi: true
                },
                EcosystemGuard
            ]
        })
    ], EcosystemModule);
    return EcosystemModule;
}());
export { EcosystemModule };
