// @c8y/ngx-components/esm5/ecosystem/ecosystem-navigation-factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { NavigatorNode, gettext } from '@c8y/ngx-components';
import { EcosystemGuard } from './ecosystem.guard';
var EcosystemNavigationFactory = /** @class */ (function () {
    function EcosystemNavigationFactory(ecosystemGuard) {
        this.ecosystemGuard = ecosystemGuard;
        this.navs = [];
    }
    EcosystemNavigationFactory.prototype.get = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var applicationsNode, microservicesNode;
            return tslib_1.__generator(this, function (_a) {
                // Check for the beta flag. To be removed when the feature is production ready.
                if (!this.ecosystemGuard.isFeatureEnabled()) {
                    return [2 /*return*/, []];
                }
                if (this.navs.length === 0) {
                    applicationsNode = new NavigatorNode({
                        label: gettext('Applications'),
                        icon: 'c8y-modules',
                        path: '/application-list'
                    });
                    microservicesNode = new NavigatorNode({
                        label: gettext('Microservices'),
                        icon: 'microchip',
                        path: '/microservices'
                    });
                    this.navs.push(new NavigatorNode({
                        label: gettext('Ecosystem'),
                        icon: 'c8y-atom',
                        priority: 3200,
                        children: [applicationsNode, microservicesNode]
                    }));
                }
                return [2 /*return*/, this.navs];
            });
        });
    };
    EcosystemNavigationFactory.ctorParameters = function () { return [
        { type: EcosystemGuard }
    ]; };
    EcosystemNavigationFactory = tslib_1.__decorate([
        Injectable()
    ], EcosystemNavigationFactory);
    return EcosystemNavigationFactory;
}());
export { EcosystemNavigationFactory };
