// @c8y/ngx-components/esm2015/connectivity/impact/impact-tab.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import { from, isObservable, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { ImpactGuard } from './impact.guard';
let ImpactTabFactory = class ImpactTabFactory {
    constructor(router, guard) {
        this.router = router;
        this.guard = guard;
        this.config = {
            label: gettext('IMPACT provider settings'),
            path: '/connectivitySettings/impact_provider_settings',
            icon: 'plug'
        };
    }
    get() {
        return this.router.url.match(/connectivitySettings/g) ? this.tab() : [];
    }
    tab() {
        return this.wrapIntoObservable(this.guard.canActivate(this.router.routerState.snapshot.root, this.router.routerState.snapshot)).pipe(map(canActivate => (canActivate ? this.config : null)));
    }
    isPromise(obj) {
        return !!obj && typeof obj.then === 'function';
    }
    wrapIntoObservable(value) {
        if (isObservable(value)) {
            return value;
        }
        if (this.isPromise(value)) {
            return from(value);
        }
        return of(value);
    }
};
ImpactTabFactory.ctorParameters = () => [
    { type: Router },
    { type: ImpactGuard }
];
ImpactTabFactory = tslib_1.__decorate([
    Injectable()
], ImpactTabFactory);
export { ImpactTabFactory };
