// @c8y/ngx-components/esm5/sub-assets/columns/asset-type.cell-renderer.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { CellRendererContext, gettext } from '@c8y/ngx-components';
import { DeviceGroupService } from '@c8y/ngx-components/assets-navigator';
var AssetTypeCellRendererComponent = /** @class */ (function () {
    function AssetTypeCellRendererComponent(context, deviceGroupService) {
        this.context = context;
        this.deviceGroupService = deviceGroupService;
    }
    AssetTypeCellRendererComponent.prototype.ngOnInit = function () {
        this.isAnyGroup = this.deviceGroupService.isAnyGroup(this.context.item);
        this.isDevice = this.deviceGroupService.isDevice(this.context.item);
        this.groupIcon = this.deviceGroupService.icon(this.context.item);
        this.iconTitle = this.getIconTitle(this.groupIcon);
    };
    AssetTypeCellRendererComponent.prototype.getIconTitle = function (icon) {
        switch (icon) {
            case 'c8y-group':
                return gettext('Group');
            case 'c8y-group-smart':
                return gettext('Smart group');
            case 'c8y-group-remote-inactive':
                return 'Inactive remote group';
            case 'c8y-group-remote':
                return gettext('Remote group');
            default:
                return gettext('Child device');
        }
    };
    AssetTypeCellRendererComponent.ctorParameters = function () { return [
        { type: CellRendererContext },
        { type: DeviceGroupService }
    ]; };
    AssetTypeCellRendererComponent = tslib_1.__decorate([
        Component({
            template: "<device-status class=\"m-l-8\" [mo]=\"context.item\" *ngIf=\"isDevice\"></device-status>\n<i\n  [c8yIcon]=\"groupIcon\"\n  title=\"{{ iconTitle | translate }}\"\n  class=\"m-l-8 c8y-icon-duocolor\"\n  *ngIf=\"isAnyGroup || !isDevice\"\n></i>\n"
        })
    ], AssetTypeCellRendererComponent);
    return AssetTypeCellRendererComponent;
}());
export { AssetTypeCellRendererComponent };
