// @c8y/ngx-components/esm5/operations/single-operation-details/single-operation-list-item.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { Router } from '@angular/router';
import { assign, get } from 'lodash-es';
import { IOperation, OperationService, OperationStatus } from '@c8y/client';
import { AlertService, ListItemComponent, Tab, gettext, operationStatusClasses, operationStatusIcons } from '@c8y/ngx-components';
var SingleOperationListItemComponent = /** @class */ (function () {
    function SingleOperationListItemComponent(router, operationService, alertService) {
        this.router = router;
        this.operationService = operationService;
        this.alertService = alertService;
        this.collapsed = true;
        this.readOnly = false;
        this.tabs = [];
        this.statusIcons = operationStatusIcons;
        this.statusClasses = operationStatusClasses;
        this.OperationStatus = OperationStatus;
    }
    Object.defineProperty(SingleOperationListItemComponent.prototype, "displayedDescription", {
        get: function () {
            var commandText = get(this.operation, 'c8y_Command.text');
            var opDescription = get(this.operation, 'description');
            return commandText || opDescription || gettext('(no command text or description available)');
        },
        enumerable: true,
        configurable: true
    });
    SingleOperationListItemComponent.prototype.ngOnInit = function () {
        this.statusIcons = operationStatusIcons;
        this.statusClasses = operationStatusClasses;
        this.tabs = [
            {
                label: gettext('Details'),
                icon: 'asterisk',
                template: this.detailsTabTemplate
            },
            {
                label: gettext('History of changes'),
                icon: 'archive',
                template: this.historyOfChangesTabTemplate
            }
        ];
        this.selectedTab = this.tabs[0];
    };
    SingleOperationListItemComponent.prototype.scheduleAsBulkOperation = function () {
        this.router.navigateByUrl("/devicecontrol/single/create-bulk/" + this.operation.id);
    };
    SingleOperationListItemComponent.prototype.cancel = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var operationAfterUpdate, ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.operationService.update({
                                id: this.operation.id,
                                status: OperationStatus.FAILED,
                                failureReason: gettext('Operation cancelled by user.')
                            })];
                    case 1:
                        operationAfterUpdate = (_a.sent()).data;
                        assign(this.operation, operationAfterUpdate);
                        this.alertService.success(gettext('Operation cancelled.'));
                        return [3 /*break*/, 3];
                    case 2:
                        ex_1 = _a.sent();
                        this.alertService.addServerFailure(ex_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    SingleOperationListItemComponent.ctorParameters = function () { return [
        { type: Router },
        { type: OperationService },
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], SingleOperationListItemComponent.prototype, "operation", void 0);
    tslib_1.__decorate([
        Input()
    ], SingleOperationListItemComponent.prototype, "collapsed", void 0);
    tslib_1.__decorate([
        Input()
    ], SingleOperationListItemComponent.prototype, "readOnly", void 0);
    tslib_1.__decorate([
        ViewChild('listItem', { static: true })
    ], SingleOperationListItemComponent.prototype, "listItem", void 0);
    tslib_1.__decorate([
        ViewChild('details', { static: true })
    ], SingleOperationListItemComponent.prototype, "detailsTabTemplate", void 0);
    tslib_1.__decorate([
        ViewChild('historyOfChanges', { static: true })
    ], SingleOperationListItemComponent.prototype, "historyOfChangesTabTemplate", void 0);
    SingleOperationListItemComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-single-operation-list-item',
            template: "<c8y-li\n  class=\"c8y-list__item--double-actions\"\n  [ngClass]=\"{ 'c8y-list__item--no-expand': !collapsed }\"\n  [collapsed]=\"collapsed\"\n  #listItem\n  id=\"single-operation-{{ operation.id }}\"\n>\n  <c8y-li-icon>\n    <i [c8yIcon]=\"statusIcons[operation.status]\" [ngClass]=\"statusClasses[operation.status]\"></i>\n  </c8y-li-icon>\n  <div [ngClass]=\"{ 'content-flex-80': !readOnly, 'content-flex-50': readOnly }\">\n    <div class=\"col-6\">\n      <span class=\"text-truncate\" title=\"{{ displayedDescription | translate }}\">\n        {{ displayedDescription | translate }}\n      </span>\n    </div>\n    <div class=\"flex-grow\"></div>\n    <div class=\"col-3\">\n      <div class=\"m-r-16\">\n        <small class=\"icon-flex\">\n          <i c8yIcon=\"calendar\" class=\"m-r-4\"></i>\n          <span>\n            {{ operation.creationTime | c8yDate }}\n          </span>\n        </small>\n      </div>\n    </div>\n  </div>\n  <div class=\"c8y-list__item__footer\">\n    <div class=\"m-r-16\">\n      <small class=\"icon-flex\">\n        <i c8yIcon=\"exchange\" class=\"m-r-4\"></i>\n        <span>\n          <a href=\"#/device/{{ operation.deviceId }}\">{{ operation.deviceName }}</a>\n        </span>\n      </small>\n    </div>\n  </div>\n  <ng-container>\n    <c8y-li-action\n      label=\"{{ 'Schedule as bulk operation' | translate }}\"\n      (click)=\"scheduleAsBulkOperation()\"\n      icon=\"c8y-energy\"\n    >\n    </c8y-li-action>\n    <c8y-li-action\n      *ngIf=\"!readOnly && operation.status === OperationStatus.PENDING\"\n      label=\"{{ 'Cancel`operation`' | translate }}\"\n      (click)=\"cancel()\"\n      icon=\"times\"\n    >\n    </c8y-li-action>\n  </ng-container>\n  <c8y-li-collapse class=\"m-b-16\">\n    <div class=\"tabContainer\">\n      <ul class=\"nav nav-tabs nav-tabsc8y\">\n        <li [ngClass]=\"{ active: selectedTab === tab }\" *ngFor=\"let tab of tabs\">\n          <button [title]=\"tab.label | translate\" (click)=\"selectedTab = tab\" class=\"btn-clean\">\n            <i [c8yIcon]=\"tab.icon\"></i>\n            <span class=\"txt\">{{ tab.label | translate }}</span>\n          </button>\n        </li>\n      </ul>\n    </div>\n\n    <ng-container *ngTemplateOutlet=\"selectedTab.template\"></ng-container>\n\n    <ng-template #details>\n      <c8y-single-operation-details [operation]=\"operation\" [readOnly]=\"readOnly\">\n      </c8y-single-operation-details>\n    </ng-template>\n\n    <ng-template #historyOfChanges>\n      <c8y-audit-log [source]=\"operation.id\"></c8y-audit-log>\n    </ng-template>\n  </c8y-li-collapse>\n</c8y-li>\n"
        })
    ], SingleOperationListItemComponent);
    return SingleOperationListItemComponent;
}());
export { SingleOperationListItemComponent };
