// @c8y/ngx-components/esm5/operations/single-operation-details/single-operation-details.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { BulkOperationDetailsService } from '@c8y/ngx-components/operations/bulk-operations-list';
var SingleOperationDetailsComponent = /** @class */ (function () {
    function SingleOperationDetailsComponent(bulkOperationDetailsService) {
        this.bulkOperationDetailsService = bulkOperationDetailsService;
        this.readOnly = false;
    }
    SingleOperationDetailsComponent.prototype.openBulkOperationDetails = function () {
        this.bulkOperationDetailsService.openDetails(this.operation.bulkOperationId);
    };
    SingleOperationDetailsComponent.ctorParameters = function () { return [
        { type: BulkOperationDetailsService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], SingleOperationDetailsComponent.prototype, "operation", void 0);
    tslib_1.__decorate([
        Input()
    ], SingleOperationDetailsComponent.prototype, "readOnly", void 0);
    SingleOperationDetailsComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-single-operation-details',
            template: "<div class=\"row\">\n  <div class=\"col-sm-3\">\n    <div class=\"legend form-block\" translate>\n      Details\n    </div>\n    <ul class=\"list-unstyled small\">\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\" *ngIf=\"operation.bulkOperationId\">\n        <label class=\"small m-0 m-r-4\" translate>Bulk operation</label>\n        <span class=\"flex-item-right\">\n          <a class=\"pointer\" role=\"button\" (click)=\"openBulkOperationDetails()\" translate\n            >See details</a\n          >\n        </span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\">\n        <label class=\"small m-0 m-r-4\" translate>Status</label>\n        <span class=\"flex-item-right\">{{ operation.status.toString() | translate }}</span>\n      </li>\n      <li *ngIf=\"operation.failureReason\" class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\">\n        <label class=\"small m-0 m-r-4\" translate>Failure reason</label>\n        <span class=\"flex-item-right\">{{ operation.failureReason | translate }}</span>\n      </li>\n      <li *ngIf=\"operation.c8y_Command\" class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\">\n        <label class=\"small flex-item-top m-r-4\" translate>Result</label>\n        <span class=\"flex-item-right\">{{ operation.c8y_Command.result }}</span>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-sm-9\">\n    <div class=\"legend form-block\" translate>\n      Operation\n    </div>\n    <pre class=\"p-16\">{{ operation | json }}</pre>\n  </div>\n</div>\n"
        })
    ], SingleOperationDetailsComponent);
    return SingleOperationDetailsComponent;
}());
export { SingleOperationDetailsComponent };
