// @c8y/ngx-components/esm5/operations/single-operation-details/single-operation-details.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { IOperation, OperationService } from '@c8y/client';
import { SingleOperationModalComponent } from './single-operation-modal.component';
var SingleOperationDetailsService = /** @class */ (function () {
    function SingleOperationDetailsService(operationService, modalService) {
        this.operationService = operationService;
        this.modalService = modalService;
    }
    /**
     * Opens a modal window with the details of given single operation.
     * @param operationOrOperationId Operation's object or id.
     * @param options Additional options:
     * - `collapsed`: boolean - whether the details pane should be initially collapsed
     * - `readOnly`: boolean - whether the details should be displayed in read-only mode
     */
    SingleOperationDetailsService.prototype.openDetails = function (operationOrOperationId, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var operation, _a, initialState;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!operationOrOperationId.id) return [3 /*break*/, 1];
                        _a = operationOrOperationId;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, this.operationService.detail(operationOrOperationId)];
                    case 2:
                        _a = (_b.sent()).data;
                        _b.label = 3;
                    case 3:
                        operation = _a;
                        initialState = tslib_1.__assign({ operation: operation }, options);
                        this.modalService.show(SingleOperationModalComponent, {
                            initialState: initialState,
                            class: 'modal-lg'
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    SingleOperationDetailsService.ctorParameters = function () { return [
        { type: OperationService },
        { type: BsModalService }
    ]; };
    SingleOperationDetailsService = tslib_1.__decorate([
        Injectable()
    ], SingleOperationDetailsService);
    return SingleOperationDetailsService;
}());
export { SingleOperationDetailsService };
