// @c8y/ngx-components/esm2015/core/provider-configuration/service/provider-configuration.service.js
var ProviderConfigurationService_1;
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { FetchClient, IIdentified, IResult, Service } from '@c8y/client';
import { get } from 'lodash-es';
let ProviderConfigurationService = ProviderConfigurationService_1 = class ProviderConfigurationService extends Service {
    constructor(client, activatedRoute) {
        super(client);
        this.baseUrl = get(activatedRoute, 'snapshot.data.endpoint.configurationEndpoint.baseUrl');
        this.listUrl = get(activatedRoute, 'snapshot.data.endpoint.configurationEndpoint.listUrl');
    }
    detail() {
        const _super = Object.create(null, {
            detail: { get: () => super.detail }
        });
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return _super.detail.call(this, ProviderConfigurationService_1.NO_DETAIL_ID);
        });
    }
    update(entity) {
        const _super = Object.create(null, {
            update: { get: () => super.update }
        });
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return _super.update.call(this, entity);
        });
    }
    delete() {
        const _super = Object.create(null, {
            delete: { get: () => super.delete }
        });
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return _super.delete.call(this, ProviderConfigurationService_1.NO_DETAIL_ID);
        });
    }
    /** There is a single provider configuration per tenant and detail URL is not needed */
    getDetailUrl(_) {
        return this.listUrl;
    }
};
ProviderConfigurationService.NO_DETAIL_ID = null;
ProviderConfigurationService.ctorParameters = () => [
    { type: FetchClient },
    { type: ActivatedRoute }
];
ProviderConfigurationService = ProviderConfigurationService_1 = tslib_1.__decorate([
    Injectable()
], ProviderConfigurationService);
export { ProviderConfigurationService };
