// @c8y/ngx-components/esm5/sms-gateway/sms-gateway.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { gettext, HOOK_PATTERN_MESSAGES, ProviderConfigurationModule } from '@c8y/ngx-components';
import { MESSAGES } from './pattern-messages.data';
import { SmsGatewayGuard } from './sms-gateway.guard';
var ɵ0 = MESSAGES;
var SmsGatewayModule = /** @class */ (function () {
    function SmsGatewayModule() {
    }
    SmsGatewayModule = tslib_1.__decorate([
        NgModule({
            imports: [
                ProviderConfigurationModule.config([
                    {
                        navigation: {
                            label: gettext('SMS provider'),
                            path: 'smsgateway',
                            icon: 'envelope-o',
                            parent: gettext('Settings'),
                            priority: 1000,
                            canActivate: [SmsGatewayGuard]
                        },
                        layout: {
                            pageTitle: gettext('SMS provider'),
                            cardTitle: gettext('Credentials'),
                            description: gettext('Enter your credentials from your SMS provider. This enables platform features that utilize SMS services (e.g. two-factor authentication and user notifications).'),
                            providerName: gettext('SMS provider'),
                            providerNamePlaceholder: gettext('Select or type to filter providers'),
                            providerNameNoMatchesHint: gettext('Select one of the existing providers.'),
                            deleteRoles: ['ROLE_SMS_ADMIN'],
                            deleteBtnLabel: gettext('Delete'),
                            saveRoles: ['ROLE_SMS_ADMIN'],
                            saveBtnLabel: gettext('Save'),
                            configurationUpdatedSuccessMsg: gettext('Credentials updated.'),
                            deleteConfigurationModalTitle: gettext('Delete credentials'),
                            deleteConfigurationModalBody: gettext("You are about to delete SMS provider credentials. Deleting credentials will lock out any users with SMS-based two-factor authentication and deactivate the SMS-based smart rules and device communication. Do you want to proceed?"),
                            deleteConfigurationModalOkBtnLabel: gettext('Delete'),
                            configurationDeletedSuccessMsg: gettext('Credentials deleted.')
                        },
                        endpoint: {
                            definitionsEndpoint: {
                                baseUrl: 'service/messaging/providers',
                                listUrl: 'definitions'
                            },
                            configurationEndpoint: {
                                baseUrl: 'service/messaging/providers',
                                listUrl: 'configuration'
                            }
                        }
                    }
                ])
            ],
            providers: [SmsGatewayGuard, { provide: HOOK_PATTERN_MESSAGES, useValue: ɵ0, multi: true }]
        })
    ], SmsGatewayModule);
    return SmsGatewayModule;
}());
export { SmsGatewayModule };
export { ɵ0 };
