// @c8y/ngx-components/esm5/operations/bulk-operation-from-single/bulk-operation-from-single.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { BulkOperationFromSingleComponent } from './bulk-operation-from-single.component';
export var bulkOperationCreateSingleRoutes = [
    {
        path: 'devicecontrol/single/create-bulk/:operationId',
        component: BulkOperationFromSingleComponent
    }
];
var BulkOperationFromSingleModule = /** @class */ (function () {
    function BulkOperationFromSingleModule() {
    }
    BulkOperationFromSingleModule = tslib_1.__decorate([
        NgModule({
            declarations: [BulkOperationFromSingleComponent],
            imports: [
                CoreModule,
                BulkOperationStepperModule,
                RouterModule.forChild(bulkOperationCreateSingleRoutes)
            ]
        })
    ], BulkOperationFromSingleModule);
    return BulkOperationFromSingleModule;
}());
export { BulkOperationFromSingleModule };
