// @c8y/ngx-components/esm5/sub-assets/assign-devices/assign-child-devices.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, HostListener } from '@angular/core';
import { AlertService, gettext } from '@c8y/ngx-components';
import { InventoryService } from '@c8y/client';
import { SubAssetsService } from '../sub-assets.service';
import { AssignDevicesService } from './assign-devices.service';
var AssignChildDevicesComponent = /** @class */ (function () {
    function AssignChildDevicesComponent(alert, subAssetsService, inventoryService, assignDevicesService) {
        this.alert = alert;
        this.subAssetsService = subAssetsService;
        this.inventoryService = inventoryService;
        this.assignDevicesService = assignDevicesService;
        this.onCancel = new EventEmitter();
        this.refresh = new EventEmitter();
        this.selected = [];
        this.pendingStatus = false;
    }
    AssignChildDevicesComponent.prototype.onEnterKeyDown = function (event) {
        if (this.selected.length > 0) {
            this.assignDevices();
        }
    };
    AssignChildDevicesComponent.prototype.onEscapeKeyDown = function (event) {
        this.onCancel.emit();
    };
    AssignChildDevicesComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.setNotIncludedInGroupQuery();
                return [2 /*return*/];
            });
        });
    };
    AssignChildDevicesComponent.prototype.setNotIncludedInGroupQuery = function () {
        var notIncludedInGroupQuery = { __not: { __bygroupid: this.currentGroupId } };
        this.baseQuery = notIncludedInGroupQuery;
    };
    AssignChildDevicesComponent.prototype.assignDevices = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var error_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.pendingStatus = true;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.assignDevicesService.assignDevices(this.currentGroupId, this.selected)];
                    case 2:
                        _a.sent();
                        this.refresh.emit();
                        this.alert.success(gettext('Child devices assigned to the group.'));
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _a.sent();
                        this.alert.danger(gettext('Could not assign child devices to the group'), error_1);
                        return [3 /*break*/, 4];
                    case 4:
                        this.pendingStatus = false;
                        this.selected = [];
                        this.onCancel.emit();
                        return [2 /*return*/];
                }
            });
        });
    };
    AssignChildDevicesComponent.prototype.onSelected = function (selectedDevicesIDs) {
        this.selected = selectedDevicesIDs;
    };
    AssignChildDevicesComponent.ctorParameters = function () { return [
        { type: AlertService },
        { type: SubAssetsService },
        { type: InventoryService },
        { type: AssignDevicesService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], AssignChildDevicesComponent.prototype, "currentGroupId", void 0);
    tslib_1.__decorate([
        Input()
    ], AssignChildDevicesComponent.prototype, "parentDevice", void 0);
    tslib_1.__decorate([
        Output()
    ], AssignChildDevicesComponent.prototype, "onCancel", void 0);
    tslib_1.__decorate([
        Input()
    ], AssignChildDevicesComponent.prototype, "refresh", void 0);
    tslib_1.__decorate([
        HostListener('document:keydown.enter', ['$event'])
    ], AssignChildDevicesComponent.prototype, "onEnterKeyDown", null);
    tslib_1.__decorate([
        HostListener('document:keydown.escape', ['$event'])
    ], AssignChildDevicesComponent.prototype, "onEscapeKeyDown", null);
    AssignChildDevicesComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-assign-child-devices',
            template: "<div class=\"card-block flex-no-shrink separator-bottom col-xs-12 large-padding p-t-24 p-b-24\">\n  <div class=\"row\">\n    <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n      <h4 class=\"text-center text-medium\">\n        {{ 'Assign child devices' | translate }}\n      </h4>\n    </div>\n  </div>\n</div>\n\n<div class=\"d-flex d-col no-align-items fit-h\">\n  <c8y-sub-assets-grid\n    [title]=\"'Select target child devices' | translate\"\n    [emptyStateText]=\"'All child devices are already assigned to the group' | translate\"\n    [refresh]=\"refresh\"\n    [actionControls]=\"[]\"\n    [selectable]=\"true\"\n    [parent-group]=\"parentDevice\"\n    [baseQuery]=\"baseQuery\"\n    (itemsSelect)=\"onSelected($event)\"\n    class=\"d-contents\"\n  >\n  </c8y-sub-assets-grid>\n</div>\n\n<div class=\"text-center card-footer p-24 separator\">\n  <button\n    (click)=\"onCancel.emit()\"\n    type=\"button\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Cancel' | translate }}\"\n  >\n    <span>{{ 'Cancel' | translate }}</span>\n  </button>\n  <button\n    (click)=\"assignDevices()\"\n    type=\"button\"\n    class=\"btn btn-primary\"\n    [ngClass]=\"{ 'btn-pending': pendingStatus }\"\n    title=\"{{ 'Assign' | translate }}\"\n    [disabled]=\"selected.length === 0\"\n  >\n    <span>{{ 'Assign' | translate }}</span>\n  </button>\n</div>\n"
        })
    ], AssignChildDevicesComponent);
    return AssignChildDevicesComponent;
}());
export { AssignChildDevicesComponent };
