// @c8y/ngx-components/esm2015/connectivity/abstract-connectivity-node.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { isObservable, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { AbstractConnectivityTabFactory } from './abstract-connectivity-tab.factory';
let AbstractConnectivityNodeFactory = class AbstractConnectivityNodeFactory {
    constructor(tabFactory) {
        this.tabFactory = tabFactory;
        this.connectivityNode = new NavigatorNode({
            parent: gettext('Settings'),
            label: gettext('Connectivity'),
            path: 'connectivitySettings',
            icon: 'plug',
            preventDuplicates: true
        });
    }
    get() {
        return this.wrapIntoObservable(this.tabFactory.tab()).pipe(map(tab => {
            return !!tab ? [this.connectivityNode] : [];
        }));
    }
    wrapIntoObservable(value) {
        if (isObservable(value)) {
            return value;
        }
        return of(value);
    }
};
AbstractConnectivityNodeFactory.ctorParameters = () => [
    { type: AbstractConnectivityTabFactory }
];
AbstractConnectivityNodeFactory = tslib_1.__decorate([
    Injectable()
], AbstractConnectivityNodeFactory);
export { AbstractConnectivityNodeFactory };
