// @c8y/ngx-components/esm5/protocol-lora/lora-set-device-protocol.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { LoraSetDeviceProtocolService } from './lora-set-device-protocol.service';
import { AlertService, gettext } from '@c8y/ngx-components';
import { IManagedObject, InventoryService, IResultList } from '@c8y/client';
import { Router } from '@angular/router';
import { pipe } from 'rxjs';
import { map } from 'rxjs/operators';
var LoraAssignDeviceProtocolComponent = /** @class */ (function () {
    function LoraAssignDeviceProtocolComponent(loraService, alertService, router, inventoryService) {
        var _this = this;
        this.loraService = loraService;
        this.alertService = alertService;
        this.router = router;
        this.inventoryService = inventoryService;
        this.filterCurrentProtocol = pipe(map(function (protocols) {
            return protocols.filter(function (protocol) { return !_this.currentProtocol || _this.currentProtocol.id !== protocol.id; });
        }));
    }
    LoraAssignDeviceProtocolComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.reload()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    LoraAssignDeviceProtocolComponent.prototype.reload = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, _b, ex_1;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        this.loading = true;
                        this.newProtocol = null;
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 5, 6, 7]);
                        return [4 /*yield*/, this.loadDevice()];
                    case 2:
                        _c.sent();
                        _a = this;
                        return [4 /*yield*/, this.loraService.getAvailableProtocols()];
                    case 3:
                        _a.availableProtocols = _c.sent();
                        _b = this;
                        return [4 /*yield*/, this.loraService.getCurrentProtocol(this.device)];
                    case 4:
                        _b.currentProtocol = _c.sent();
                        return [3 /*break*/, 7];
                    case 5:
                        ex_1 = _c.sent();
                        this.alertService.addServerFailure(ex_1);
                        return [3 /*break*/, 7];
                    case 6:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    LoraAssignDeviceProtocolComponent.prototype.loadDevice = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var deviceId, data;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        deviceId = this.router.routerState.snapshot.url.match(/\d+/)[0];
                        return [4 /*yield*/, this.inventoryService.detail(deviceId)];
                    case 1:
                        data = (_a.sent()).data;
                        this.device = data;
                        return [2 /*return*/];
                }
            });
        });
    };
    LoraAssignDeviceProtocolComponent.prototype.apply = function (selectedProtocol) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var ex_2;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.loraService.applyProtocol(this.device, selectedProtocol)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.reload()];
                    case 2:
                        _a.sent();
                        this.alertService.success(gettext('Device protocol set.'));
                        this.loraSetDeviceProtocolForm.reset('dirty');
                        return [3 /*break*/, 4];
                    case 3:
                        ex_2 = _a.sent();
                        this.alertService.danger(gettext('Could not set device protocol.'));
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    LoraAssignDeviceProtocolComponent.ctorParameters = function () { return [
        { type: LoraSetDeviceProtocolService },
        { type: AlertService },
        { type: Router },
        { type: InventoryService }
    ]; };
    tslib_1.__decorate([
        ViewChild('loraSetDeviceProtocolForm', { static: false })
    ], LoraAssignDeviceProtocolComponent.prototype, "loraSetDeviceProtocolForm", void 0);
    LoraAssignDeviceProtocolComponent = tslib_1.__decorate([
        Component({
            selector: 'set-device-protocol',
            template: "<form #loraSetDeviceProtocolForm=\"ngForm\" class=\"row\">\n  <div class=\"col-md-6\">\n    <div class=\"card\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\">\n          {{ 'LoRa device protocol' | translate }}\n        </h4>\n      </div>\n      <div *ngIf=\"loading\" class=\"card-block\">\n        <div class=\"p-relative\" style=\"min-height: 40px; min-width: 55px;\">\n          <div class=\"spinner\">\n            <div class=\"rect1\"></div>\n            <div class=\"rect2\"></div>\n            <div class=\"rect3\"></div>\n            <div class=\"rect4\"></div>\n          </div>\n        </div>\n        <span translate>Loading</span>\n      </div>\n      <div *ngIf=\"!loading\">\n        <div class=\"card-block\">\n          <div class=\"form-group\">\n            <label translate>Current device protocol</label>\n            <p class=\"form-control-static\" *ngIf=\"!currentProtocol\">\n              {{ 'c8y_LoriotLora' }}\n            </p>\n            <p class=\"form-control-static\" *ngIf=\"currentProtocol\">\n              {{ currentProtocol.name }}\n            </p>\n          </div>\n          <div class=\"form-group\">\n            <label translate>New device protocol</label>\n            <div class=\"c8y-select-wrapper\">\n              <select class=\"form-control\" [(ngModel)]=\"newProtocol\" name=\"newProtocol\" required>\n                <option *ngIf=\"!newProtocol\" [ngValue]=\"null\">\n                  {{ 'Select device protocol' | translate }}\n                </option>\n                <option\n                  *c8yFor=\"let protocol of availableProtocols; pipe: filterCurrentProtocol\"\n                  [ngValue]=\"protocol\"\n                >\n                  {{ protocol.name }}\n                </option>\n              </select>\n            </div>\n          </div>\n        </div>\n        <div class=\"card-footer separator\">\n          <button\n            title=\"{{ 'Apply' | translate }}\"\n            class=\"btn btn-primary\"\n            (click)=\"apply(newProtocol)\"\n            [disabled]=\"!loraSetDeviceProtocolForm.dirty\"\n          >\n            {{ 'Apply' | translate }}\n          </button>\n        </div>\n      </div>\n    </div>\n  </div>\n</form>\n\u200C\n"
        })
    ], LoraAssignDeviceProtocolComponent);
    return LoraAssignDeviceProtocolComponent;
}());
export { LoraAssignDeviceProtocolComponent };
