// @c8y/ngx-components/esm5/protocol-lora/lora-set-device-protocol.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { IManagedObject, InventoryService, IResultList } from '@c8y/client';
var LoraSetDeviceProtocolService = /** @class */ (function () {
    function LoraSetDeviceProtocolService(inventoryService) {
        this.inventoryService = inventoryService;
    }
    LoraSetDeviceProtocolService.prototype.getCurrentProtocol = function (device) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var protocolId, data;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        protocolId = device.c8y_LpwanDevice.type
                            ? device.c8y_LpwanDevice.type.split('/')[2]
                            : null;
                        if (!protocolId) {
                            return [2 /*return*/, null];
                        }
                        return [4 /*yield*/, this.inventoryService.detail(protocolId)];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    LoraSetDeviceProtocolService.prototype.applyProtocol = function (device, selectedProtocol) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                device.c8y_LpwanDevice.type = 'inventory/managedObjects/' + selectedProtocol.id;
                device.type = selectedProtocol.name;
                return [2 /*return*/, this.inventoryService.update(device)];
            });
        });
    };
    LoraSetDeviceProtocolService.prototype.getAvailableProtocols = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query;
            return tslib_1.__generator(this, function (_a) {
                query = {
                    type: {
                        __in: ['c8y_ActilityDeviceType', 'c8y_LoraDeviceType']
                    }
                };
                return [2 /*return*/, this.inventoryService.listQuery(query)];
            });
        });
    };
    LoraSetDeviceProtocolService.ctorParameters = function () { return [
        { type: InventoryService }
    ]; };
    LoraSetDeviceProtocolService = tslib_1.__decorate([
        Injectable()
    ], LoraSetDeviceProtocolService);
    return LoraSetDeviceProtocolService;
}());
export { LoraSetDeviceProtocolService };
