// @c8y/ngx-components/esm5/protocol-lora/lora-protocol.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CoreModule, FormsModule, gettext, HOOK_ONCE_ROUTE, ViewContext } from '@c8y/ngx-components';
import { LoraAssignDeviceProtocolComponent } from './lora-set-device-protocol.component';
import { LoraAgentGuard } from './lora-agent.guard';
import { LoraSetDeviceProtocolService } from './lora-set-device-protocol.service';
var routes = [
    {
        context: ViewContext.Device,
        path: 'assign-protocol',
        component: LoraAssignDeviceProtocolComponent,
        label: gettext('LPWAN'),
        icon: 'c8y-device-protocols',
        canActivate: [LoraAgentGuard]
    }
];
var ɵ0 = routes;
var LoraProtocolModule = /** @class */ (function () {
    function LoraProtocolModule() {
    }
    LoraProtocolModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                LoraAssignDeviceProtocolComponent
            ],
            imports: [
                CoreModule,
                FormsModule
            ],
            entryComponents: [LoraAssignDeviceProtocolComponent],
            providers: [
                LoraAgentGuard,
                LoraSetDeviceProtocolService,
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: ɵ0,
                    multi: true
                }
            ]
        })
    ], LoraProtocolModule);
    return LoraProtocolModule;
}());
export { LoraProtocolModule };
export { ɵ0 };
