// @c8y/ngx-components/esm2015/sub-assets/delete-assets-modal/delete-assets-modal.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { gettext, Status } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { Subject } from 'rxjs';
let DeleteAssetsModalComponent = class DeleteAssetsModalComponent {
    constructor(translateService) {
        this.translateService = translateService;
        this.showWithCascadeCheckbox = true;
        this.showWithDeviceUserCheckbox = false;
        this.closeSubject = new Subject();
        this.labels = { ok: gettext('Delete'), cancel: gettext('Cancel') };
        this.title = gettext('Delete asset');
        this.status = Status.DANGER;
        this.config = {
            cascade: false,
            withDeviceUser: false
        };
    }
    ngOnInit() {
        this.message = this.translateService.instant(gettext('You are about to delete asset: "{{name}}". This operation is irreversible. Do you want to proceed?'), { name: this.asset.name });
    }
    ngAfterViewInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modalRef.result;
                this.onClose();
            }
            catch (error) {
                this.onDismiss();
            }
        });
    }
    onClose() {
        this.closeSubject.next(this.config);
        this.closeSubject.complete();
    }
    onDismiss() {
        this.closeSubject.complete();
    }
};
DeleteAssetsModalComponent.ctorParameters = () => [
    { type: TranslateService }
];
tslib_1.__decorate([
    Input()
], DeleteAssetsModalComponent.prototype, "showWithCascadeCheckbox", void 0);
tslib_1.__decorate([
    Input()
], DeleteAssetsModalComponent.prototype, "showWithDeviceUserCheckbox", void 0);
tslib_1.__decorate([
    Input()
], DeleteAssetsModalComponent.prototype, "asset", void 0);
tslib_1.__decorate([
    ViewChild('modalRef', { static: false })
], DeleteAssetsModalComponent.prototype, "modalRef", void 0);
DeleteAssetsModalComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-delete-assets-modal',
        template: "<c8y-confirm-modal [title]=\"title\" [status]=\"status\" [labels]=\"labels\" #modalRef>\n  <form #assetsForm=\"ngForm\">\n    <p class=\"text-wrap m-b-16\">\n      {{ message | translate }}\n    </p>\n    <c8y-form-group *ngIf=\"showWithCascadeCheckbox\" class=\"m-b-0\">\n      <label title=\"{{ 'Delete devices' | translate }}\" class=\"c8y-checkbox\">\n        <input type=\"checkbox\" name=\"cascade\" [(ngModel)]=\"config.cascade\" />\n        <span></span>\n        <span>\n          {{ 'Also delete all devices inside selected asset and its subassets.' | translate }}\n        </span>\n      </label>\n    </c8y-form-group>\n    <c8y-form-group *ngIf=\"showWithDeviceUserCheckbox\" class=\"m-b-0\">\n      <label title=\"{{ 'Delete associated device owner' | translate }}\" class=\"c8y-checkbox\">\n        <input type=\"checkbox\" name=\"withDeviceUser\" [(ngModel)]=\"config.withDeviceUser\" />\n        <span></span>\n        <span>\n          {{ 'Also delete associated device owner.' | translate }}\n        </span>\n      </label>\n    </c8y-form-group>\n  </form>\n</c8y-confirm-modal>\n"
    })
], DeleteAssetsModalComponent);
export { DeleteAssetsModalComponent };
