// @c8y/ngx-components/esm5/core/forms/emails-validator.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input } from '@angular/core';
import { FormControl, NG_VALIDATORS, Validators } from '@angular/forms';
/**
 * A directive that adds the `emails` validator to controls marked with the
 * `emails` attribute. The directive is provided with the `NG_VALIDATORS` multi-provider list.
 *
 * **Example**
 *
 * ```html
 * <input type="text" name="emails" ngModel emails>
 * <input type="text" name="emails" ngModel emails="true">
 * <input type="text" name="emails" ngModel [emails]="true">
 * ```
 */
var EmailsValidatorDirective = /** @class */ (function () {
    function EmailsValidatorDirective() {
        this._enabled = false;
        this.separator = ',';
    }
    EmailsValidatorDirective_1 = EmailsValidatorDirective;
    Object.defineProperty(EmailsValidatorDirective.prototype, "emails", {
        /**
         * Tracks changes to the emails attribute bound to this directive.
         */
        set: function (value) {
            this._enabled = value === '' || value === true || value === 'true';
            if (this._onChange) {
                this._onChange();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Method that validates whether email addresses are valid.
     * Returns the validation result if enabled, otherwise null.
     */
    EmailsValidatorDirective.prototype.validate = function (control) {
        if (!this._enabled) {
            return null;
        }
        var value = control.value || '';
        var emails = value.split(this.separator).map(function (email) { return email.trim(); });
        var invalid = emails.some(function (email) { return Validators.email(new FormControl(email)); });
        return invalid ? { emails: true } : null;
    };
    /**
     * Registers a callback function to call when the validator inputs change.
     */
    EmailsValidatorDirective.prototype.registerOnValidatorChange = function (fn) {
        this._onChange = fn;
    };
    var EmailsValidatorDirective_1;
    tslib_1.__decorate([
        Input()
    ], EmailsValidatorDirective.prototype, "emails", null);
    EmailsValidatorDirective = EmailsValidatorDirective_1 = tslib_1.__decorate([
        Directive({
            selector: '[emails][formControlName],[emails][formControl],[emails][ngModel]',
            providers: [{ provide: NG_VALIDATORS, useExisting: EmailsValidatorDirective_1, multi: true }]
        })
    ], EmailsValidatorDirective);
    return EmailsValidatorDirective;
}());
export { EmailsValidatorDirective };
