// @c8y/ngx-components/fesm5/c8y-ngx-components-operations-stepper-bulk-type-device-profile.js
import { __decorate, __read, __extends } from 'tslib';
import { Input, Component, EventEmitter, Output, NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { gettext, CoreModule, FormsModule } from '@c8y/ngx-components';
import { DeviceProfileService, DeviceProfileModule } from '@c8y/ngx-components/device-profile';
import { BaseStepperComponent, BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { BulkOperationType, baseUrl, HOOK_LIST_BULK_TYPE } from '@c8y/ngx-components/operations/bulk-operations-service';
import { get, uniqWith, isEqual } from 'lodash-es';
import { TranslateService } from '@ngx-translate/core';
import { BehaviorSubject, combineLatest, from } from 'rxjs';
import { switchMap, shareReplay, debounceTime, distinctUntilChanged } from 'rxjs/operators';

var ConfirmDeviceProfileSelectionStepComponent = /** @class */ (function () {
    function ConfirmDeviceProfileSelectionStepComponent(translate) {
        this.translate = translate;
        this.DEVICE_TYPE_NOT_DEFINED = gettext('Device type not defined');
    }
    Object.defineProperty(ConfirmDeviceProfileSelectionStepComponent.prototype, "softwares", {
        get: function () {
            return get(this.selectedDeviceProfile, ['c8y_DeviceProfile', 'software'], []);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConfirmDeviceProfileSelectionStepComponent.prototype, "configurations", {
        get: function () {
            return get(this.selectedDeviceProfile, ['c8y_DeviceProfile', 'configuration'], []);
        },
        enumerable: true,
        configurable: true
    });
    ConfirmDeviceProfileSelectionStepComponent.prototype.getDeviceTypeTitle = function (deviceProfile) {
        return get(deviceProfile, 'c8y_Filter.type', this.translate.instant(this.DEVICE_TYPE_NOT_DEFINED));
    };
    ConfirmDeviceProfileSelectionStepComponent.ctorParameters = function () { return [
        { type: TranslateService }
    ]; };
    __decorate([
        Input()
    ], ConfirmDeviceProfileSelectionStepComponent.prototype, "selectedDeviceProfile", void 0);
    ConfirmDeviceProfileSelectionStepComponent = __decorate([
        Component({
            selector: 'c8y-confirm-device-profile-selection-step',
            template: "<div class=\"card-block p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n  <div class=\"row p-b-16\">\n    <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n      <h4 class=\"text-center text-gray\">{{ 'Confirm selection' | translate }}</h4>\n    </div>\n  </div>\n</div>\n<div class=\"col-xs-12 flex-grow no-gutter\">\n  <div class=\"card-inner-scroll fit-h\">\n    <div class=\"card-block\">\n      <div class=\"row p-b-16\">\n        <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n          <div class=\"text-truncate\" title=\"{{ 'Device type' | translate }}: {{ getDeviceTypeTitle(selectedDeviceProfile) }}\">\n            <span class=\"legend form-block\" translate>Device type</span>\n            <ng-container *ngIf=\"selectedDeviceProfile?.c8y_Filter?.type; else noType\">\n              <span>{{ selectedDeviceProfile.c8y_Filter.type }}</span>\n            </ng-container>\n            <ng-template #noType>\n              ---\n            </ng-template>\n          </div>\n        </div>\n      </div>\n      <div class=\"row p-b-16\" *ngIf=\"selectedDeviceProfile?.c8y_DeviceProfile.firmware\">\n        <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n          <span class=\"legend form-block\" translate>Firmware</span>\n          <c8y-list-group>\n            <c8y-li>\n              <c8y-li-icon>\n                <i c8yIcon=\"c8y-firmware\"></i>\n              </c8y-li-icon>\n              <c8y-li-body class=\"content-flex-500\">\n                <div class=\"col-5\">\n                  {{ selectedDeviceProfile.c8y_DeviceProfile.firmware.name }}\n                </div>\n                <div class=\"col-5\">\n                  <span class=\"text-label-small m-r-8\" translate>\n                    Version\n                  </span>\n                  <span>\n                    {{ selectedDeviceProfile.c8y_DeviceProfile.firmware.version }}\n                  </span>\n                </div>\n              </c8y-li-body>\n            </c8y-li>\n          </c8y-list-group>\n        </div>\n      </div>\n      <div class=\"row p-b-16\" *ngIf=\"softwares.length\">\n        <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n          <span class=\"legend form-block\" translate>Software</span>\n          <c8y-list-group>\n            <c8y-li *ngFor=\"let software of softwares\">\n              <c8y-li-icon>\n                <i c8yIcon=\"c8y-tools\"></i>\n              </c8y-li-icon>\n              <c8y-li-body class=\"content-flex-500\">\n                <div class=\"col-5\">\n                  {{ software.name }}\n                </div>\n                <div class=\"col-5\">\n                  <span class=\"text-label-small m-r-8\" translate>\n                    Version\n                  </span>\n                  <span>\n                    {{ software.version }}\n                  </span>\n                </div>\n              </c8y-li-body>\n            </c8y-li>\n          </c8y-list-group>\n        </div>\n      </div>\n      <div class=\"row p-b-16\" *ngIf=\"configurations.length\">\n        <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n          <span class=\"legend form-block\" translate>Configuration</span>\n          <c8y-list-group>\n            <c8y-li *ngFor=\"let configuration of configurations\">\n              <c8y-li-icon>\n                <i c8yIcon=\"c8y-tools\"></i>\n              </c8y-li-icon>\n              <c8y-li-body class=\"content-flex-500\">\n                <div class=\"col-5\">\n                  {{ configuration.name }}\n                </div>\n                <div class=\"col-5\">\n                  <span class=\"label label-info\" *ngIf=\"configuration.type\">\n                    {{ configuration.type }}\n                  </span>\n                </div>\n              </c8y-li-body>\n            </c8y-li>\n          </c8y-list-group>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n"
        })
    ], ConfirmDeviceProfileSelectionStepComponent);
    return ConfirmDeviceProfileSelectionStepComponent;
}());

var SelectDeviceProfileStepComponent = /** @class */ (function () {
    function SelectDeviceProfileStepComponent(deviceProfileService, translate) {
        var _this = this;
        this.deviceProfileService = deviceProfileService;
        this.translate = translate;
        this.deviceProfile = new EventEmitter();
        this.textFilter$ = new BehaviorSubject('');
        this.deviceType$ = new BehaviorSubject('');
        this.DEVICE_TYPE_NOT_DEFINED = gettext('Device type not defined');
        this.deviceTypes = [];
        this.selectedDeviceType = { name: '' };
        this.deviceProfile$ = combineLatest(this.textFilter$, this.deviceType$).pipe(switchMap(function (_a) {
            var _b = __read(_a, 2), name = _b[0], deviceType = _b[1];
            return _this.getDeviceProfiles(name, deviceType);
        }), shareReplay(1));
        this.loadDeviceTypes();
    }
    SelectDeviceProfileStepComponent.prototype.loadDeviceTypes = function () {
        var _this = this;
        this.deviceTypeSubscription = this.deviceType$
            .pipe(debounceTime(300), distinctUntilChanged(), switchMap(function (searchStr) {
            var query = { 'c8y_Filter.type': "*" + searchStr + "*" };
            return from(_this.deviceProfileService.getDeviceProfiles(query));
        }))
            .subscribe(function (_a) {
            var data = _a.data;
            _this.deviceTypes = uniqWith(data.map(function (val) { return ({ name: val.c8y_Filter.type }); }).filter(function (val) { return val.name; }), isEqual);
        });
    };
    SelectDeviceProfileStepComponent.prototype.ngOnDestroy = function () {
        this.deviceTypeSubscription.unsubscribe();
    };
    SelectDeviceProfileStepComponent.prototype.selectDeviceProfile = function (deviceProfile) {
        this.deviceProfile.emit(deviceProfile);
    };
    SelectDeviceProfileStepComponent.prototype.getDeviceTypeTitle = function (deviceProfile) {
        return get(deviceProfile, 'c8y_Filter.type', this.translate.instant(this.DEVICE_TYPE_NOT_DEFINED));
    };
    SelectDeviceProfileStepComponent.prototype.getDeviceProfiles = function (name, deviceType) {
        var query = deviceType ? { 'c8y_Filter.type': deviceType } : {};
        if (name) {
            query.name = "*" + name + "*";
        }
        return this.deviceProfileService.getDeviceProfiles(query);
    };
    SelectDeviceProfileStepComponent.ctorParameters = function () { return [
        { type: DeviceProfileService },
        { type: TranslateService }
    ]; };
    __decorate([
        Output()
    ], SelectDeviceProfileStepComponent.prototype, "deviceProfile", void 0);
    SelectDeviceProfileStepComponent = __decorate([
        Component({
            selector: 'c8y-select-device-profile-step',
            template: "<div class=\"card-block p-t-0 overflow-visible flex-no-shrink separator-bottom col-xs-12\">\n  <div class=\"row p-b-16\">\n    <div class=\"col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3\">\n      <h4 class=\"text-center m-b-16\">{{ 'Select device profile' | translate }}</h4>\n      <div class=\"row\">\n        <div class=\"col-md-6\">\n          <div class=\"input-group input-group-search\">\n            <input\n              type=\"search\"\n              class=\"form-control\"\n              title=\"{{ 'Filter device profiles\u2026' | translate }}\"\n              placeholder=\"{{ 'Filter device profiles\u2026' | translate }}\"\n              [ngModel]=\"textFilter$ | async\"\n              (ngModelChange)=\"textFilter$.next($event)\"\n            />\n            <span class=\"input-group-addon\">\n              <i c8yIcon=\"filter\" *ngIf=\"(textFilter$ | async).length === 0\"></i>\n              <i\n                c8yIcon=\"times\"\n                class=\"text-muted\"\n                *ngIf=\"(textFilter$ | async).length\"\n                (click)=\"textFilter$.next('')\"\n              ></i>\n            </span>\n          </div>\n        </div>\n        <div class=\"col-md-6\">\n          <c8y-form-group class=\"m-0\">\n            <c8y-typeahead\n              name=\"deviceType\"\n              [(ngModel)]=\"selectedDeviceType\"\n              placeholder=\"{{ 'Type to filter device types\u2026' | translate }}\"\n              (onSearch)=\"deviceType$.next($event)\"\n              [allowFreeEntries]=\"false\"\n            >\n              <c8y-li\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"selectedDeviceType = {name: ''}; deviceType$.next('')\"\n              >\n                <span>{{'All device types' | translate }}</span>\n              </c8y-li>\n              <c8y-li\n                *ngFor=\"let deviceType of deviceTypes\"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"selectedDeviceType = deviceType; deviceType$.next(deviceType.name)\"\n                [active]=\"selectedDeviceType === deviceType\"\n              >\n                <c8y-highlight\n                  [text]=\"deviceType.name\"\n                  [pattern]=\"deviceType$ | async\"\n                ></c8y-highlight>\n              </c8y-li>\n            </c8y-typeahead>\n            <c8y-messages\n            ><c8y-message\n              name=\"notExisting\"\n              [text]=\"'Select one of the existing device types.' | translate\"\n            ></c8y-message>\n            </c8y-messages>\n          </c8y-form-group>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n<div class=\"col-xs-12 flex-grow no-gutter\">\n  <div class=\"card-inner-scroll fit-h\">\n    <div class=\"card-block p-t-0 p-b-0\">\n      <c8y-list-group>\n        <c8y-li *c8yFor=\"let deviceProfile of deviceProfile$ | async; loadMore: 'auto'\">\n          <c8y-li-radio (onSelect)=\"selectDeviceProfile(deviceProfile)\"></c8y-li-radio>\n          <c8y-li-icon>\n            <i c8yIcon=\"c8y-device-profile\"></i>\n          </c8y-li-icon>\n          <c8y-li-body class=\"content-flex-60\">\n            <div class=\"col-5\">\n              <div class=\"text-truncate\" title=\"{{ deviceProfile.name }}\">\n                {{ deviceProfile.name }}\n              </div>\n            </div>\n            <div class=\"col-3\">\n              <div class=\"text-truncate\" title=\"{{ 'Device type' | translate }}: {{ getDeviceTypeTitle(deviceProfile) }}\">\n                <span class=\"text-label-small m-r-8\" translate>\n                  Device type\n                </span>\n                <span *ngIf=\"deviceProfile.c8y_Filter?.type; else noType\">\n                  {{ deviceProfile.c8y_Filter?.type }}\n                </span>\n                <ng-template #noType>\n                  <small><em class=\"text-muted\" translate>Undefined`device type`</em></small>\n                </ng-template>\n              </div>\n            </div>\n          </c8y-li-body>\n        </c8y-li>\n      </c8y-list-group>\n    </div>\n  </div>\n</div>\n"
        })
    ], SelectDeviceProfileStepComponent);
    return SelectDeviceProfileStepComponent;
}());

var StepperBulkTypeDeviceProfileComponent = /** @class */ (function (_super) {
    __extends(StepperBulkTypeDeviceProfileComponent, _super);
    function StepperBulkTypeDeviceProfileComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StepperBulkTypeDeviceProfileComponent.prototype.goToConfirmSelection = function ($event) {
        this.deviceTypes = get(this.selectedDeviceProfile, 'c8y_Filter.type');
        $event.stepper.next();
    };
    StepperBulkTypeDeviceProfileComponent.prototype.onDeviceProfileSelected = function (selectedItem) {
        this.selectedDeviceProfile = selectedItem;
    };
    StepperBulkTypeDeviceProfileComponent.prototype.retrieveOperationPrototype = function () {
        return {
            name: gettext('Apply device profile'),
            description: get(this.selectedDeviceProfile, 'name'),
            prototype: {
                description: "Apply device profile: " + this.selectedDeviceProfile.name,
                profileId: this.selectedDeviceProfile.id,
                profileName: this.selectedDeviceProfile.name,
                c8y_DeviceProfile: this.selectedDeviceProfile.c8y_DeviceProfile
            }
        };
    };
    StepperBulkTypeDeviceProfileComponent = __decorate([
        Component({
            selector: 'c8y-stepper-bulk-type-device-profile',
            template: "<c8y-bulk-operation-stepper>\n  <ng-container\n    *customStep=\"\n      'Select device profile' | translate; \n      completed: !!selectedDeviceProfile;\n      buttonsDisabled: !selectedDeviceProfile; \n      onNext: goToConfirmSelection.bind(this)\"\n  >\n    <c8y-select-device-profile-step\n      (deviceProfile)=\"onDeviceProfileSelected($event)\"\n      class=\"d-contents\"\n    ></c8y-select-device-profile-step>\n  </ng-container>\n  <ng-container *customStep=\"'Confirm selected device profile' | translate\">\n    <c8y-confirm-device-profile-selection-step\n      class=\"d-contents\"\n      [selectedDeviceProfile]=\"selectedDeviceProfile\"\n    ></c8y-confirm-device-profile-selection-step>\n  </ng-container>\n</c8y-bulk-operation-stepper>\n"
        })
    ], StepperBulkTypeDeviceProfileComponent);
    return StepperBulkTypeDeviceProfileComponent;
}(BaseStepperComponent));

var ɵ0 = {
    type: BulkOperationType.DEVICE_PROFILE,
    c8yIcon: 'c8y-device-profile',
    name: gettext('Apply device profile'),
    path: baseUrl + "device-profile",
    component: StepperBulkTypeDeviceProfileComponent,
    fragments: ['c8y_DeviceProfile'],
    selected: false
};
/** Module for the 'Apply device profile' operation type stepper */
var StepperBulkTypeDeviceProfileModule = /** @class */ (function () {
    function StepperBulkTypeDeviceProfileModule() {
    }
    StepperBulkTypeDeviceProfileModule = __decorate([
        NgModule({
            declarations: [
                StepperBulkTypeDeviceProfileComponent,
                SelectDeviceProfileStepComponent,
                ConfirmDeviceProfileSelectionStepComponent
            ],
            imports: [
                CoreModule,
                FormsModule,
                ReactiveFormsModule,
                BulkOperationStepperModule,
                DeviceProfileModule
            ],
            providers: [
                {
                    provide: HOOK_LIST_BULK_TYPE,
                    useValue: ɵ0,
                    multi: true
                }
            ],
            entryComponents: [StepperBulkTypeDeviceProfileComponent],
            exports: [StepperBulkTypeDeviceProfileComponent]
        })
    ], StepperBulkTypeDeviceProfileModule);
    return StepperBulkTypeDeviceProfileModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { ConfirmDeviceProfileSelectionStepComponent, SelectDeviceProfileStepComponent, StepperBulkTypeDeviceProfileComponent, StepperBulkTypeDeviceProfileModule, ɵ0 };
