// @c8y/ngx-components/fesm5/c8y-ngx-components-context-dashboard.js
import { __awaiter, __generator, __assign, __decorate, __param, __rest, __read } from 'tslib';
import { InjectionToken, Inject, ɵɵdefineInjectable, ɵɵinject, Injectable, ViewChild, Component, Input, EventEmitter, Output, Renderer2, HostBinding, NgModule, Host } from '@angular/core';
import { cloneDeep, sortBy, assign, pick, some, keys, keyBy, has, get, forEach, clone, omit, escapeRegExp, findIndex, values, kebabCase } from 'lodash-es';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { of, from, Subject, iif, timer, BehaviorSubject, combineLatest } from 'rxjs';
import { tap, map, catchError, throwIfEmpty, filter, toArray, mergeAll, mergeMap, switchMap, debounceTime, distinctUntilChanged, shareReplay, take } from 'rxjs/operators';
import { gettext, DynamicComponentService, Status, NavigatorNode, getActivatedRoute, ViewContext, TabsService, ModalService, AppStateService, NavigatorService, Permissions, ICON_LIST, DashboardChildChange, AlertService, GainsightService, HOOK_ONCE_ROUTE, HOOK_ACTION, CoreModule, FilterInputComponent, DocsService, DocsModule, QuickLinkModule, HOOK_NAVIGATOR_NODES, HOOK_COMPONENTS } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { Router, ActivatedRoute, RouterModule } from '@angular/router';
import { InventoryService, UserService, QueriesUtil, TenantService, ApplicationService } from '@c8y/client';
import { ControlContainer, NgForm } from '@angular/forms';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';

var CONTEXT_DASHBOARD_CONFIG = new InjectionToken('ContextDashboardConfig');
var ContextDashboardType;
(function (ContextDashboardType) {
    ContextDashboardType["Device"] = "device";
    ContextDashboardType["DeviceType"] = "type";
    ContextDashboardType["Group"] = "group";
    ContextDashboardType["Named"] = "name";
    ContextDashboardType["Report"] = "report";
})(ContextDashboardType || (ContextDashboardType = {}));
var WIDGET_HEADER_CLASSES = [
    {
        label: gettext('Regular`style`'),
        class: 'panel-title-regular',
        description: gettext('The widget has no border between header and content.')
    },
    {
        label: gettext('Border`style`'),
        class: 'panel-title-border',
        description: gettext('The widget has a small separation border between header and content.')
    },
    {
        label: gettext('Overlay`style`'),
        class: 'panel-title-overlay',
        description: gettext('The widget content overlays the header.')
    },
    {
        label: gettext('Hidden`style`'),
        class: 'panel-title-hidden',
        description: gettext('The widget header is not shown.')
    }
];
var WIDGET_CONTENT_CLASSES = [
    {
        label: gettext('Light`style`'),
        class: 'panel-content-light',
        description: gettext('The widget is styled with a light background and has low contrast.')
    },
    {
        label: gettext('Dark`style`'),
        class: 'panel-content-dark',
        description: gettext('The widget is styled with a dark background and has high contrast.')
    },
    {
        label: gettext('Transparent`style`'),
        class: 'panel-content-transparent',
        description: gettext('The widget is styled with a transparent background and therefore is border-less.')
    },
    {
        label: gettext('Branded`style`'),
        class: 'panel-content-branded',
        description: gettext('The widget is styled with the main brand color.')
    }
];
var DASHBOARD_THEME_CLASSES = [
    {
        label: gettext('Light`theme`'),
        class: 'dashboard-theme-light',
        description: gettext('The dashboard is styled with a light background and has low contrast.')
    },
    {
        label: gettext('Dark`theme`'),
        class: 'dashboard-theme-dark',
        description: gettext('The dashboard is styled with a dark background and has high contrast.')
    },
    {
        label: gettext('Transparent`theme`'),
        class: 'dashboard-theme-transparent',
        description: gettext('The dashboard is styled with a transparent background and therefore is border-less.')
    },
    {
        label: gettext('Branded`theme`'),
        class: 'dashboard-theme-branded',
        description: gettext('The dashboard is styled with the main brand color.')
    }
];

var WidgetService = /** @class */ (function () {
    function WidgetService(dynamicComponentService, translateService, moduleConfig) {
        this.dynamicComponentService = dynamicComponentService;
        this.translateService = translateService;
        this.moduleConfig = moduleConfig;
    }
    WidgetService.prototype.getWidgetDefinition = function (componentId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.dynamicComponentService.getById$(componentId).toPromise()];
            });
        });
    };
    WidgetService.prototype.getWidgetDefinitions = function () {
        var _this = this;
        var translatedComponents = Array.from(this.dynamicComponentService.state).map(function (cmp) { return (__assign({}, cmp, { label: _this.translateService.instant(cmp.label) })); });
        return cloneDeep(sortBy(translatedComponents, 'label').filter(this.moduleConfig.widgetFilter));
    };
    WidgetService.ctorParameters = function () { return [
        { type: DynamicComponentService },
        { type: TranslateService },
        { type: undefined, decorators: [{ type: Inject, args: [CONTEXT_DASHBOARD_CONFIG,] }] }
    ]; };
    WidgetService.ngInjectableDef = ɵɵdefineInjectable({ factory: function WidgetService_Factory() { return new WidgetService(ɵɵinject(DynamicComponentService), ɵɵinject(TranslateService), ɵɵinject(CONTEXT_DASHBOARD_CONFIG)); }, token: WidgetService, providedIn: "root" });
    WidgetService = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __param(2, Inject(CONTEXT_DASHBOARD_CONFIG))
    ], WidgetService);
    return WidgetService;
}());

var ContextDashboardService = /** @class */ (function () {
    function ContextDashboardService(inventory, tabs, modal, translateService, router, user, appState, navigator, permissions) {
        this.inventory = inventory;
        this.tabs = tabs;
        this.modal = modal;
        this.translateService = translateService;
        this.router = router;
        this.user = user;
        this.appState = appState;
        this.navigator = navigator;
        this.permissions = permissions;
        this.REPORT_PARTIAL_NAME = 'report_';
        this.cache = new Map();
        this.DEFAULT_PAGESIZE = 1000;
        this.FRAGMENT_NAME = 'c8y_Dashboard';
        this.DASHBOARD_ROUTE_PATH = 'dashboard';
        this.INDEX_SPLIT = '!';
        this._formDisabled = true;
    }
    Object.defineProperty(ContextDashboardService.prototype, "formDisabled", {
        get: function () {
            return this._formDisabled;
        },
        set: function (value) {
            this._formDisabled = value;
        },
        enumerable: true,
        configurable: true
    });
    ContextDashboardService.prototype.create = function (dashboardCfg, contextOrName) {
        return __awaiter(this, void 0, void 0, function () {
            var id, dashboardType, dashboard, value, fragmentKey, data, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (typeof contextOrName === 'string') {
                            id = contextOrName;
                            dashboardType = ContextDashboardType.Named;
                        }
                        else {
                            id = contextOrName.contextData.id;
                            dashboardType = this.getDashboardTypeFromViewContext(dashboardCfg, contextOrName);
                        }
                        dashboard = {};
                        assign(dashboard, { c8y_Dashboard: dashboardCfg });
                        value = dashboardType === ContextDashboardType.DeviceType ? dashboardCfg.deviceTypeValue : id;
                        fragmentKey = this.createFragmentKey(dashboardType, value);
                        dashboard[fragmentKey] = {};
                        if (this.shouldSetGlobal(dashboard)) {
                            assign(dashboard, { c8y_Global: {} });
                        }
                        dashboard.name = dashboard.c8y_Dashboard.name;
                        if (!(dashboardType === ContextDashboardType.Group || dashboardType === ContextDashboardType.Device)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.inventory.childAdditionsCreate(dashboard, id)];
                    case 1:
                        _a = _b.sent();
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, this.inventory.create(dashboard)];
                    case 3:
                        _a = _b.sent();
                        _b.label = 4;
                    case 4:
                        data = (_a).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    ContextDashboardService.prototype.detail = function (dashboardMO) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.inventory.detail(dashboardMO)];
                    case 1:
                        data = (_a.sent()).data;
                        this.cache.set(dashboardMO.id, data);
                        return [2 /*return*/, data];
                }
            });
        });
    };
    ContextDashboardService.prototype.update = function (dashboard) {
        return __awaiter(this, void 0, void 0, function () {
            var keepFragments, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dashboard.name = dashboard.c8y_Dashboard.name;
                        keepFragments = this.clean(pick(dashboard, [this.FRAGMENT_NAME, 'id', 'name']));
                        keepFragments.c8y_Global = this.shouldSetGlobal(dashboard);
                        return [4 /*yield*/, this.inventory.update(keepFragments)];
                    case 1:
                        data = (_a.sent()).data;
                        this.cache.set(dashboard.id, data);
                        return [2 /*return*/, data];
                }
            });
        });
    };
    ContextDashboardService.prototype.delete = function (dashboard, withConfirmation) {
        if (withConfirmation === void 0) { withConfirmation = true; }
        return __awaiter(this, void 0, void 0, function () {
            var msg, tabToRemove, ex_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        if (!withConfirmation) return [3 /*break*/, 2];
                        msg = gettext("You are about to delete the dashboard \"{{ dashboardName }}\". Do you want to proceed?");
                        if (this.isDeviceType(dashboard)) {
                            msg = gettext("You are about to delete the dashboard \"{{ dashboardName }}\" from all devices of the type \"{{ deviceType }}\".\n           Do you want to proceed?");
                        }
                        return [4 /*yield*/, this.modal.confirm(gettext('Delete dashboard'), this.translateService.instant(msg, {
                                dashboardName: dashboard.c8y_Dashboard.name,
                                deviceType: dashboard.c8y_Dashboard.deviceTypeValue
                            }), Status.DANGER, {
                                ok: gettext('Delete'),
                                cancel: gettext('Cancel')
                            })];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [4 /*yield*/, this.inventory.delete(dashboard)];
                    case 3:
                        _a.sent();
                        tabToRemove = Array.from(this.tabs.state).find(function (tab) {
                            return tab.path.endsWith(_this.DASHBOARD_ROUTE_PATH + "/" + dashboard.id);
                        });
                        this.tabs.remove(tabToRemove);
                        this.tabs.refresh();
                        return [3 /*break*/, 5];
                    case 4:
                        ex_1 = _a.sent();
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    ContextDashboardService.prototype.activateDashboards = function (route, types) {
        var dashboardId = route.params.dashboardId;
        if (dashboardId) {
            return this.getDashboard$(dashboardId, types, route.parent.data.contextData).pipe(tap(function (dashboard) {
                route.data = { dashboard: dashboard };
            }), map(function () { return true; }), catchError(function () {
                return of(false);
            }));
        }
        return this.getTabs$(route.data.contextData, types);
    };
    ContextDashboardService.prototype.getNamedDashboardOrCreate = function (name, defaultWidgets) {
        var _this = this;
        var children = this.mapWidgets(defaultWidgets);
        return this.getDashboard$(name, [ContextDashboardType.Named]).pipe(throwIfEmpty(), catchError(function () {
            return from(_this.create({
                children: children,
                widgetClasses: { 'dashboard-theme-light': true, 'panel-title-regular': true }
            }, name));
        }));
    };
    ContextDashboardService.prototype.refreshTabs = function (dashboardMO) {
        return __awaiter(this, void 0, void 0, function () {
            var tabToUpdate, data, _a, icon, priority, name_1;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!!this.isNamed(dashboardMO)) return [3 /*break*/, 2];
                        tabToUpdate = Array.from(this.tabs.state).find(function (tab) {
                            return tab.path.endsWith(_this.DASHBOARD_ROUTE_PATH + "/" + dashboardMO.id);
                        });
                        return [4 /*yield*/, this.detail(dashboardMO)];
                    case 1:
                        data = _b.sent();
                        if (tabToUpdate) {
                            _a = data.c8y_Dashboard, icon = _a.icon, priority = _a.priority, name_1 = _a.name;
                            tabToUpdate.icon = icon;
                            tabToUpdate.priority = priority;
                            tabToUpdate.label = name_1;
                        }
                        this.tabs.refresh();
                        _b.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    ContextDashboardService.prototype.updateNavigatorItem = function (mo) {
        var _this = this;
        this.navigator.state.forEach(function (node) {
            if (node.path === "reports/" + mo.id) {
                _this.navigator.remove(node);
            }
        });
        if (mo.c8y_IsNavigatorNode) {
            var nodeToAdd = new NavigatorNode({
                label: mo.name,
                path: "reports/" + mo.id,
                icon: mo.icon,
                priority: mo.priority
            });
            this.navigator.add(nodeToAdd);
        }
    };
    ContextDashboardService.prototype.navigateToDashboard = function (dashboardMO) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (/dashboard/.test(this.router.url)) {
                    this.router.navigate(['..', dashboardMO.id], {
                        relativeTo: getActivatedRoute(this.router)
                    });
                }
                else {
                    this.router.navigate(['..', this.DASHBOARD_ROUTE_PATH, dashboardMO.id], {
                        relativeTo: getActivatedRoute(this.router)
                    });
                }
                return [2 /*return*/];
            });
        });
    };
    ContextDashboardService.prototype.canEditDashboard = function (mo) {
        return this.permissions.canEdit(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE'], mo);
    };
    ContextDashboardService.prototype.isNamed = function (dashboard) {
        var _this = this;
        return some(keys(dashboard), function (prop) {
            return new RegExp("^" + _this.FRAGMENT_NAME + _this.INDEX_SPLIT + ContextDashboardType.Named + _this.INDEX_SPLIT).test(prop);
        });
    };
    ContextDashboardService.prototype.isReport = function (dashboard) {
        var _this = this;
        return some(keys(dashboard), function (prop) {
            return new RegExp("^" + _this.FRAGMENT_NAME + _this.INDEX_SPLIT + ContextDashboardType.Named + _this.INDEX_SPLIT + _this.REPORT_PARTIAL_NAME).test(prop);
        });
    };
    ContextDashboardService.prototype.isDeviceType = function (dashboard) {
        var _this = this;
        return some(keys(dashboard), function (prop) {
            return new RegExp("^" + _this.FRAGMENT_NAME + _this.INDEX_SPLIT + ContextDashboardType.DeviceType + _this.INDEX_SPLIT).test(prop);
        });
    };
    ContextDashboardService.prototype.getStyling = function (styleList, styleName, defaultValue) {
        var styling = styleList.find(function (style) { return style && new RegExp(styleName, 'i').test(style.class); });
        return styling ? styling.class : defaultValue;
    };
    ContextDashboardService.prototype.mapWidgets = function (widgets) {
        return keyBy(widgets.map(function (widget) {
            widget.id = String(Math.random()).substr(2);
            return widget;
        }), 'id');
    };
    ContextDashboardService.prototype.getDashboard$ = function (dashboardIdOrName, dashboardType, mo) {
        var _this = this;
        var cache = this.cache.get(dashboardIdOrName);
        var dashboards = mo
            ? this.getContextDashboards(mo, dashboardType)
            : [this.getNamedDashboard(dashboardIdOrName)];
        var cacheRefresh = this.getContextDashboards$(dashboards).pipe(tap(function (dashboard) { return _this.cacheDashboard(dashboard); }), filter(function (dashboard) {
            return dashboard.id === dashboardIdOrName ||
                has(dashboard, "" + _this.FRAGMENT_NAME + _this.INDEX_SPLIT + ContextDashboardType.Named + _this.INDEX_SPLIT + dashboardIdOrName);
        }));
        return cache ? of(cache) : cacheRefresh;
    };
    ContextDashboardService.prototype.getTabs$ = function (mo, dashboardType) {
        var _this = this;
        var dashboards = this.getContextDashboards(mo, dashboardType);
        this.setBaseContextRoute(mo, dashboardType);
        return this.getContextDashboards$(dashboards).pipe(map(function (dashboard) { return _this.removeDashboardMoProperty(dashboard); }), tap(function (dashboard) { return _this.cacheDashboard(dashboard); }), map(function (dashboard) { return _this.createDashboardTab(dashboard); }), toArray());
    };
    ContextDashboardService.prototype.getContextDashboards$ = function (requests) {
        return from(requests).pipe(mergeAll(), mergeMap(function (response) { return response.data; }));
    };
    ContextDashboardService.prototype.setBaseContextRoute = function (mo, dashboardType) {
        var type = dashboardType.includes(ContextDashboardType.Device)
            ? ContextDashboardType.Device
            : ContextDashboardType.Group;
        this.currentContextRoute = type + "/" + mo.id;
    };
    /**
     * Cleans already corrupted dashboards from dashboardMo property.
     * Added to fix dashboards on the cloud instance (eu-latest).
     * @deprecated This is going to be removed after 1007.7.0.
     */
    ContextDashboardService.prototype.removeDashboardMoProperty = function (dashboard) {
        var dashboardCopy = cloneDeep(dashboard);
        var children = get(dashboardCopy, 'c8y_Dashboard.children');
        var updateDashboard = false;
        forEach(children, function (child) {
            if (get(child, 'componentTransformConfigWithContext')) {
                delete child.componentTransformConfigWithContext;
                updateDashboard = true;
            }
            if (get(child, 'config.dashboardMo')) {
                delete child.config.dashboardMo;
                updateDashboard = true;
            }
        });
        if (updateDashboard) {
            this.update(dashboardCopy);
        }
        return dashboardCopy;
    };
    ContextDashboardService.prototype.cacheDashboard = function (dashboard) {
        this.cache.set(dashboard.id, dashboard);
    };
    ContextDashboardService.prototype.createDashboardTab = function (dashboard) {
        var _dashboard = dashboard.c8y_Dashboard, id = dashboard.id;
        return {
            icon: _dashboard.icon,
            path: this.DASHBOARD_ROUTE_PATH + "/" + id,
            label: _dashboard.name,
            priority: _dashboard.priority,
            hide: this.isReport(dashboard)
        };
    };
    ContextDashboardService.prototype.clean = function (dashboard) {
        var jsonString = JSON.stringify(dashboard, function (key, value) {
            if (key === '$$hashKey' || key === 'klasses') {
                return undefined;
            }
            return value;
        });
        return JSON.parse(jsonString);
    };
    ContextDashboardService.prototype.getNamedDashboard = function (name) {
        return this.inventory.list({
            fragmentType: "" + this.FRAGMENT_NAME + this.INDEX_SPLIT + ContextDashboardType.Named + this.INDEX_SPLIT + name,
            pageSize: 1
        });
    };
    ContextDashboardService.prototype.getContextDashboards = function (mo, dashboardType) {
        var _this = this;
        return dashboardType.map(function (type) {
            return _this.inventory.list({
                fragmentType: _this.createDashboardFragment(mo, type),
                pageSize: _this.DEFAULT_PAGESIZE
            });
        });
    };
    ContextDashboardService.prototype.createDashboardFragment = function (mo, type) {
        var value;
        if (mo.c8y_Report) {
            value = "" + this.REPORT_PARTIAL_NAME + mo.id;
        }
        else {
            value = type === ContextDashboardType.DeviceType ? mo.type : mo.id;
        }
        return "" + this.FRAGMENT_NAME + this.INDEX_SPLIT + type + this.INDEX_SPLIT + value;
    };
    ContextDashboardService.prototype.getDashboardTypeFromViewContext = function (dashboardCfg, context) {
        var dashboardType;
        if (context.context === ViewContext.Device) {
            dashboardType = dashboardCfg.deviceType
                ? ContextDashboardType.DeviceType
                : ContextDashboardType.Device;
        }
        if (context.context === ViewContext.Group) {
            dashboardType = ContextDashboardType.Group;
        }
        return dashboardType;
    };
    ContextDashboardService.prototype.createFragmentKey = function (contextDashboardType, value) {
        return [this.FRAGMENT_NAME, contextDashboardType, value].join(this.INDEX_SPLIT);
    };
    ContextDashboardService.prototype.shouldSetGlobal = function (dashboard) {
        if ((this.isNamed(dashboard) && !this.isReport(dashboard)) || this.isDeviceType(dashboard)) {
            return {};
        }
        return null;
    };
    ContextDashboardService.ctorParameters = function () { return [
        { type: InventoryService },
        { type: TabsService },
        { type: ModalService },
        { type: TranslateService },
        { type: Router },
        { type: UserService },
        { type: AppStateService },
        { type: NavigatorService },
        { type: Permissions }
    ]; };
    ContextDashboardService = __decorate([
        Injectable()
    ], ContextDashboardService);
    return ContextDashboardService;
}());

var WidgetConfigComponent = /** @class */ (function () {
    function WidgetConfigComponent(widgetService, modal, contextDashboardService) {
        var _this = this;
        this.widgetService = widgetService;
        this.modal = modal;
        this.contextDashboardService = contextDashboardService;
        this.mode = 'select';
        this.searchChange$ = new Subject();
        this.searchTerm = '';
        this.styling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.defaultStyling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.possibleStyling = { WIDGET_HEADER_CLASSES: WIDGET_HEADER_CLASSES, WIDGET_CONTENT_CLASSES: WIDGET_CONTENT_CLASSES };
        this.isUpgrade = false;
        this.result = new Promise(function (resolve, reject) {
            _this._save = resolve;
            _this._cancel = reject;
        });
    }
    Object.defineProperty(WidgetConfigComponent.prototype, "isEdit", {
        get: function () {
            return !!this.current;
        },
        enumerable: true,
        configurable: true
    });
    WidgetConfigComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.components = this.widgetService.getWidgetDefinitions();
        if (this.selected) {
            this.current = clone(this.selected);
            this.select(this.selected, this.isEdit ? 'config' : 'select');
        }
        this.searchSub = this.searchChange$
            .pipe(switchMap(function (event) { return iif(function () { return event.which !== 13; }, timer(200), timer(10)); }))
            .subscribe(function () {
            _this.search();
        });
        if (this.mo.c8y_Dashboard.classes) {
            this.styling = this.setDefaultStyle(__assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses, (this.isEdit ? this.current.data.classes : {})));
            this.defaultStyling = this.setDefaultStyle(__assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses));
        }
    };
    WidgetConfigComponent.prototype.save = function () {
        return __awaiter(this, void 0, void 0, function () {
            var hookSuccess, _a, _x, _y, _width, _height, _b, id, name_1, widget;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.dynamicComponent.callLifeCycleHooks().toPromise()];
                    case 1:
                        hookSuccess = _c.sent();
                        if (!hookSuccess) {
                            return [2 /*return*/];
                        }
                        _a = this.selected.data, _x = _a._x, _y = _a._y, _width = _a._width, _height = _a._height;
                        if (this.widgetConfig && this.widgetConfig.device) {
                            _b = this.widgetConfig.device, id = _b.id, name_1 = _b.name;
                            this.widgetConfig.device = { id: id, name: name_1 };
                        }
                        widget = __assign({ _x: _x,
                            _y: _y,
                            _width: _width,
                            _height: _height, config: omit(this.widgetConfig, 'settings'), title: this.selected.data.title, componentId: this.selected.id, id: this.isEdit ? this.current.data.id : String(Math.random()).substr(2), classes: this.getStyle() }, (!this.isEdit ? this.widgetConfig.settings.widgetDefaults : {}));
                        this._save(widget);
                        return [2 /*return*/];
                }
            });
        });
    };
    WidgetConfigComponent.prototype.select = function (cmp, mode) {
        if (mode === void 0) { mode = 'config'; }
        cmp.data = cmp.data || {};
        this.selected = cmp;
        this.isUpgrade = !!get(cmp, 'data.settings.upgrade');
        this.contextDashboardService.formDisabled = this.isUpgrade;
        if (this.isEdit) {
            var _a = this.current.data, _x = _a._x, _y = _a._y, _width = _a._width, _height = _a._height, classes = _a.classes, title = _a.title;
            this.selected.data = __assign({}, this.selected.data, { _x: _x, _y: _y, _width: _width, _height: _height, classes: classes, title: title });
        }
        this.widgetConfig = cloneDeep(this.composeWidgetConfig(this.selected, this.context));
        this.selected.data.title = this.selected.data.title || cmp.label;
        this.componentLabel = cmp.label;
        this.mode = mode;
    };
    WidgetConfigComponent.prototype.search = function () {
        var _this = this;
        if (this.searchTerm.length > 0) {
            this.searchResult = this.components.filter(function (cmp) {
                return new RegExp(escapeRegExp(_this.searchTerm.trim()), 'i').test(cmp.label);
            });
        }
        else {
            this.resetSearch();
        }
    };
    WidgetConfigComponent.prototype.resetSearch = function () {
        this.searchTerm = '';
        this.searchResult = undefined;
    };
    WidgetConfigComponent.prototype.changeMode = function (mode) {
        this.mode = mode;
    };
    WidgetConfigComponent.prototype.close = function () {
        this._cancel();
        this.modal.hide();
    };
    WidgetConfigComponent.prototype.getStyle = function (isPreview) {
        if (isPreview === void 0) { isPreview = false; }
        var cssClasses = {};
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.headerClass)) {
            cssClasses[this.styling.headerClass] = true;
        }
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.contentClass)) {
            cssClasses[this.styling.contentClass] = true;
        }
        if (isPreview) {
            cssClasses["dashboard-theme-" + this.defaultStyling.contentClass.split('-').pop()] = true;
        }
        return cssClasses;
    };
    WidgetConfigComponent.prototype.ngOnDestroy = function () {
        this.contextDashboardService.formDisabled = true;
        if (this.searchSub) {
            this.searchSub.unsubscribe();
        }
    };
    WidgetConfigComponent.prototype.isDashboardDefaultStyle = function (className) {
        var allClasses = __assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses);
        var styles = Object.keys(allClasses).map(function (cssClass) { return ({ class: cssClass }); });
        var style = this.contextDashboardService.getStyling(styles, className.split('-').pop(), undefined);
        return !!style;
    };
    WidgetConfigComponent.prototype.setDefaultStyle = function (setClasses) {
        var _this = this;
        var contentClass = this.styling.contentClass;
        var headerClass = this.styling.headerClass;
        var styles = Object.keys(setClasses).map(function (c) { return c.split('-').pop(); });
        styles.forEach(function (styleName) {
            contentClass = _this.contextDashboardService.getStyling(WIDGET_CONTENT_CLASSES, styleName, contentClass);
            headerClass = _this.contextDashboardService.getStyling(WIDGET_HEADER_CLASSES, styleName, headerClass);
        });
        return { headerClass: headerClass, contentClass: contentClass };
    };
    WidgetConfigComponent.prototype.composeWidgetConfig = function (selectedComponent, context) {
        if (context === void 0) { context = {}; }
        var setting = __assign({ settings: __assign({}, selectedComponent.data.settings, get(selectedComponent.data.settings, 'ng1.options'), get(selectedComponent.data, 'ng1.options'), { context: context, dashboardMo: this.mo.c8y_Dashboard }) }, selectedComponent.data.config);
        return this.applyTargetIfDeviceDashboard(setting);
    };
    WidgetConfigComponent.prototype.applyTargetIfDeviceDashboard = function (widgetConfig) {
        var isDeviceType = this.contextDashboardService.isDeviceType(this.mo);
        if (isDeviceType) {
            widgetConfig.settings.hideTarget = isDeviceType;
            var noDeviceTarget = widgetConfig.settings.ng1
                ? widgetConfig.settings.ng1.options.noDeviceTarget
                : widgetConfig.settings.noDeviceTarget;
            if (!noDeviceTarget) {
                widgetConfig.device = {
                    id: this.context.id,
                    name: this.context.name
                };
            }
        }
        return widgetConfig;
    };
    WidgetConfigComponent.ctorParameters = function () { return [
        { type: WidgetService },
        { type: BsModalRef },
        { type: ContextDashboardService }
    ]; };
    __decorate([
        ViewChild('config', { static: false })
    ], WidgetConfigComponent.prototype, "dynamicComponent", void 0);
    WidgetConfigComponent = __decorate([
        Component({
            selector: 'c8y-widget-config',
            template: "<div class=\"modal-header separator\">\n  <h3 *ngIf=\"!current\" translate>Add widget</h3>\n  <h3 *ngIf=\"current\" translate>Edit widget</h3>\n</div>\n<form #configForm=\"ngForm\" name=\"form\">\n  <div class=\"c8y-modal-tabs\">\n    <div class=\"tabContainer\">\n      <ul class=\"nav nav-tabs nav-tabsc8y p-l-24\">\n        <li [class.active]=\"mode === 'select'\">\n          <button type=\"button\" class=\"btn\" (click)=\"changeMode('select'); (false)\">\n            <i c8yIcon=\"th-large\"></i> <span class=\"txt\" translate>Select widget</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'config'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('config'); (false)\"\n          >\n            <i c8yIcon=\"cog\"></i> <span class=\"txt\" translate>Configuration</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'style'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('style'); (false)\"\n          >\n            <i c8yIcon=\"paint-brush\"></i> <span class=\"txt\" translate>Appearance</span>\n          </button>\n        </li>\n      </ul>\n    </div>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <div\n      *ngIf=\"mode === 'select'\"\n      class=\"bg-white p-l-24 p-r-24 p-t-8 p-b-8 sticky-header-top-0\"\n      style=\"z-index: 2;\"\n    >\n      <div class=\"row\">\n        <div class=\"col-sm-6\">\n          <div class=\"input-group input-group-search\">\n            <input\n              class=\"form-control\"\n              placeholder=\"{{ 'Search\u2026' | translate }}\"\n              type=\"text\"\n              [(ngModel)]=\"searchTerm\"\n              [ngModelOptions]=\"{ standalone: true }\"\n              (keydown)=\"searchChange$.next($event)\"\n            />\n            <span class=\"input-group-btn\">\n              <button class=\"btn btn-clean\" (click)=\"resetSearch()\" type=\"button\">\n                <i [c8yIcon]=\"searchTerm.length === 0 ? 'search' : 'close'\"></i>\n              </button>\n            </span>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div class=\"modal-body bg-gray-lighter\" *ngIf=\"mode === 'select'\">\n      <div class=\"card-group card-select m-b-0\" >\n        <div\n          class=\"col-md-3 col-sm-4 col-xs-6\"\n          *ngFor=\"let cmp of searchResult || components\"\n          title=\"{{ cmp.description | translate }}\"\n        >\n          <div class=\"card p-8\" [class.active]=\"selected === cmp\" (click)=\"select(cmp)\">\n            <div\n              class=\"text-center p-8 m-b-8 flex-col flex-center\"\n              style=\"min-height: 170px; background-color: var(--body-background-color, #f2f3f4);\"\n            >\n              <ng-container *ngIf=\"!cmp.previewImage; else previewImage\">\n                <h1><i c8yIcon=\"file-image-o\"></i></h1>\n                <small translate>Preview not available</small>\n              </ng-container>\n              <ng-template #previewImage>\n                <img class=\"img-responsive\" [src]=\"cmp.previewImage\" />\n              </ng-template>\n            </div>\n            <p class=\"card-title text-truncate\">\n              <c8y-highlight\n                text=\"{{ cmp.label | translate }}\"\n                [pattern]=\"searchTerm\"\n              ></c8y-highlight>\n            </p>\n          </div>\n        </div>\n\n        <div class=\"c8y-empty-state text-center\" *ngIf=\"searchResult && searchResult.length === 0\">\n          <h1 c8yIcon=\"search\"></h1>\n          <h3 translate>No widgets found.</h3>\n          <div class=\"d-flex\">\n            <p translate class=\"m-r-8\">Rephrase your search term.</p>\n            <button class=\"btn btn-primary\" (click)=\"resetSearch()\" translate>\n              Reset search\n            </button>\n          </div>\n        </div>\n      </div>\n    </div>\n\n    <!-- The following is intentional set to hidden to allow the ViewChild ref in the controller -->\n    <div\n      class=\"modal-body\"\n      *ngIf=\"selected\"\n      [hidden]=\"mode !== 'config'\"\n      style=\"min-height: calc(100vh - 290px);\"\n    >\n      <h4 class=\"text-left\">\n        <strong>\n          {{ selected.label | translate }}\n        </strong>\n      </h4>\n      <p class=\"m-b-24\">\n        {{ selected.description | translate }}\n      </p>\n      <div class=\"row\">\n        <!-- change to col-sm-6 when preview is available -->\n        <div class=\"col-sm-12\">\n          <div class=\"legend form-block\" translate>Configuration</div>\n          <c8y-form-group>\n            <label for=\"widgetTitle\" translate>Title</label>\n            <input\n              id=\"widgetTitle\"\n              [(ngModel)]=\"selected.data.title\"\n              type=\"text\"\n              name=\"title\"\n              class=\"form-control\"\n              placeholder=\"{{ 'e.g.' | translate }} {{ componentLabel | translate }}\"\n              required\n            />\n          </c8y-form-group>\n\n          <!-- This is an upgraded component for the device selector and still needs to be migrated -->\n          <c8y-dynamic-component\n            componentId=\"device.selector.legacy\"\n            [config]=\"widgetConfig\"\n            [notFoundError]=\"false\"\n            *ngIf=\"!isUpgrade\"\n          ></c8y-dynamic-component>\n\n          <c8y-dynamic-component\n            [componentId]=\"selected.id\"\n            mode=\"config\"\n            [config]=\"widgetConfig\"\n            [notFoundError]=\"false\"\n            #config\n          ></c8y-dynamic-component>\n        </div>\n        <!-- markup for the preview\n\n        <div class=\"col-sm-6 sticky-header-top-0\" >\n          <div class=\"legend form-block\">\n            Preview\n          </div>\n          <div class=\"bg-gray-lighter p-16\">\n            <div class=\"card card-dashboard m-b-0\">\n              <div class=\"card-header-actions\">\n                <div class=\"card-title text-uppercase\">{{ selected.data.title }}</div>\n                <div class=\"header-actions\">\n                  <div class=\"optionsBtn dropdow\" class=\"btnIcon dropdown-toggle c8y-dropdown\">\n                    <i c8yIcon=\"cog\"></i>\n                  </div>\n                </div>\n              </div>\n              <div class=\"card-inner-scroll\">\n                <div class=\"card-block\" style=\"min-height: 240px;\">\n                  include here the widget\n                </div>\n              </div>\n            </div>\n          </div>\n        </div>\n      --></div>\n    </div>\n\n    <div *ngIf=\"mode === 'style'\" class=\"modal-body p-t-0\" style=\"min-height: calc(100vh - 290px);\">\n      <div class=\"row\">\n        <div class=\"col-xs-6\">\n          <c8y-appearance-settings\n            [(themeClass)]=\"styling.contentClass\"\n            [(headerClass)]=\"styling.headerClass\"\n            [possibleStylingTheme]=\"possibleStyling.WIDGET_CONTENT_CLASSES\"\n            [possibleStylingHeader]=\"possibleStyling.WIDGET_HEADER_CLASSES\"\n            [defaultThemeClass]=\"defaultStyling.contentClass\"\n            [defaultHeaderClass]=\"defaultStyling.headerClass\"\n          >\n          </c8y-appearance-settings>\n        </div>\n        <div class=\"col-xs-6 sticky-header-top-0\">\n          <c8y-widget-preview [previewClasses]=\"getStyle(true)\"></c8y-widget-preview>\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      type=\"button\"\n      title=\"{{ 'Cancel' | translate }}\"\n      class=\"btn btn-default\"\n      (click)=\"close()\"\n      translate\n    >\n      Cancel\n    </button>\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      translate\n      [disabled]=\"contextDashboardService.formDisabled || configForm.invalid\"\n      c8yProductExperience\n      [actionName]=\"current ? 'editWidget' : 'createWidget'\"\n      [actionData]=\"{ widgetName: selected && selected.id }\"\n    >\n      Save\n    </button>\n  </div>\n</form>\n"
        })
    ], WidgetConfigComponent);
    return WidgetConfigComponent;
}());

var DeviceDashboardGuard = /** @class */ (function () {
    function DeviceDashboardGuard(contextDashboardService) {
        this.contextDashboardService = contextDashboardService;
    }
    DeviceDashboardGuard.prototype.canActivate = function (route) {
        return this.contextDashboardService.activateDashboards(route, [
            ContextDashboardType.Device,
            ContextDashboardType.DeviceType
        ]);
    };
    DeviceDashboardGuard.ctorParameters = function () { return [
        { type: ContextDashboardService }
    ]; };
    DeviceDashboardGuard = __decorate([
        Injectable()
    ], DeviceDashboardGuard);
    return DeviceDashboardGuard;
}());

var GroupDashboardGuard = /** @class */ (function () {
    function GroupDashboardGuard(contextDashboardService) {
        this.contextDashboardService = contextDashboardService;
    }
    GroupDashboardGuard.prototype.canActivate = function (route) {
        return this.contextDashboardService.activateDashboards(route, [
            ContextDashboardType.Group
        ]);
    };
    GroupDashboardGuard.ctorParameters = function () { return [
        { type: ContextDashboardService }
    ]; };
    GroupDashboardGuard = __decorate([
        Injectable()
    ], GroupDashboardGuard);
    return GroupDashboardGuard;
}());

var WidgetPreviewComponent = /** @class */ (function () {
    function WidgetPreviewComponent() {
        this.previewClasses = {
            'dashboard-theme-light': true,
            'panel-title-regular': true
        };
    }
    __decorate([
        Input()
    ], WidgetPreviewComponent.prototype, "previewClasses", void 0);
    __decorate([
        Input()
    ], WidgetPreviewComponent.prototype, "tab", void 0);
    WidgetPreviewComponent = __decorate([
        Component({
            selector: 'c8y-widget-preview',
            template: "<div class=\"legend form-block\" translate>Preview</div>\n<label *ngIf=\"tab\" translate>Tab and layout</label>\n<label *ngIf=\"!tab\" translate>Layout</label>\n<div class=\"dashboard-preview-slot\" [ngClass]=\"previewClasses\">\n  <div\n    class=\"page-tabs page-tabs-horizontal navigator-open p-absolute overflow-hidden\"\n    style=\"left: 0; margin: 0 15px;\"\n    *ngIf=\"tab\"\n  >\n    <div class=\"tabContainer hidden-xs\">\n      <ul class=\"nav nav-tabs nav-tabsc8y\">\n        <li class=\"active\" style=\"left: 16px;\">\n          <a routerlinkactive=\"\" title=\"{{ tab.name }}\">\n            <i class=\"text-primary\" [c8yIcon]=\"tab.icon\"></i>\n            <span class=\"txt\">{{ tab.name }}</span></a\n          >\n        </li>\n      </ul>\n    </div>\n    <div class=\"visible-xs mobile-tabs\"></div>\n  </div>\n  <div [ngClass]=\"{ 'm-t-64': !!tab }\">\n    <div class=\"card card-dashboard\" style=\"height: 360px;\">\n      <div class=\"card-header-actions\" style=\"z-index: 8;\">\n        <div class=\"card-title\">\n          <span translate>Widget title</span>\n        </div>\n        <div class=\"header-actions\">\n          <div class=\"optionsBtn dropdown\">\n            <a href=\"\" (click)=\"(false)\" class=\"dropdown-toggle c8y-dropdown\">\n              <i c8yIcon=\"cog\"></i>\n            </a>\n          </div>\n        </div>\n      </div>\n      <div class=\"card-inner-scroll p-t-16\">\n        <p translate>Widget example content.</p>\n      </div>\n    </div>\n  </div>\n</div>\n"
        })
    ], WidgetPreviewComponent);
    return WidgetPreviewComponent;
}());

var DashboardDetailComponent = /** @class */ (function () {
    function DashboardDetailComponent(modal, iconList, contextDashboardService, navigatorService) {
        var _this = this;
        this.modal = modal;
        this.contextDashboardService = contextDashboardService;
        this.navigatorService = navigatorService;
        this.styling = {
            themeClass: 'dashboard-theme-light',
            headerClass: 'panel-title-regular'
        };
        this.possibleStyling = { DASHBOARD_THEME_CLASSES: DASHBOARD_THEME_CLASSES, WIDGET_HEADER_CLASSES: WIDGET_HEADER_CLASSES };
        this.result = new Promise(function (resolve, reject) {
            _this._save = resolve;
            _this._cancel = reject;
        });
        this.DEFAULT_DASHBOARD_MARGIN = 12;
        this.DEFAULT_DASHBOARD_ICON = 'th';
        this.DEFAULT_DASHBOARD_PRIORITY = 10000;
        this.icons = iconList;
        this.filteredIcons = iconList;
    }
    DashboardDetailComponent.prototype.ngAfterContentInit = function () {
        var defaultDashboardCfg = {
            name: this.isReport ? 'Report' : 'Dashboard',
            priority: this.DEFAULT_DASHBOARD_PRIORITY,
            icon: this.DEFAULT_DASHBOARD_ICON,
            deviceTypeValue: this.deviceType
        };
        if (this.dashboard) {
            this.current = clone(this.dashboard);
            this.setDashboardStyle();
        }
        else {
            this.dashboard = defaultDashboardCfg;
            this.dashboardDetailForm.form.markAsDirty();
        }
        this.setTitle();
        this.navigatorNodes$ = this.navigatorService.items$;
        this.namePlaceholder = this.isReport ? gettext('e.g. My report') : gettext('e.g. My dashboard');
    };
    DashboardDetailComponent.prototype.setTitle = function () {
        this.titleName = this.isReport ? gettext('report') : gettext('dashboard');
        this.titleAction = this.current ? gettext('Edit') : gettext('Add');
    };
    DashboardDetailComponent.prototype.save = function () {
        var _a, _b;
        this.dashboard.classes = (_a = {}, _a[this.styling.themeClass] = true, _a);
        this.dashboard.widgetClasses = (_b = {}, _b[this.styling.headerClass] = true, _b);
        this.dashboard.c8y_IsNavigatorNode = this.dashboard.c8y_IsNavigatorNode
            ? {}
            : this.current
                ? null
                : undefined;
        this._save(this.dashboard);
    };
    DashboardDetailComponent.prototype.close = function () {
        this._cancel();
        this.modal.hide();
    };
    DashboardDetailComponent.prototype.getDashboardPreviewStyle = function () {
        var cssClasses = {};
        cssClasses[this.styling.headerClass] = true;
        cssClasses[this.styling.themeClass] = true;
        return cssClasses;
    };
    DashboardDetailComponent.prototype.selectIcon = function (icon) {
        this.dashboard.icon = icon;
        this.dashboardDetailForm.form.markAsDirty();
    };
    DashboardDetailComponent.prototype.updateFiltered = function (term) {
        if (term) {
            var search_1 = new RegExp(term, 'i');
            this.filteredIcons = this.icons.filter(function (val) { return search_1.test(val); });
        }
        else {
            this.filteredIcons = this.icons;
        }
    };
    DashboardDetailComponent.prototype.setDashboardStyle = function () {
        var _this = this;
        var allClasses = __assign({}, this.dashboard.classes, this.dashboard.widgetClasses);
        var styles = Object.keys(allClasses).map(function (c) { return c.split('-').pop(); });
        styles.forEach(function (styleName) {
            _this.styling.themeClass = _this.contextDashboardService.getStyling(DASHBOARD_THEME_CLASSES, styleName, _this.styling.themeClass);
            _this.styling.headerClass = _this.contextDashboardService.getStyling(WIDGET_HEADER_CLASSES, styleName, _this.styling.headerClass);
        });
    };
    DashboardDetailComponent.ctorParameters = function () { return [
        { type: BsModalRef },
        { type: Array, decorators: [{ type: Inject, args: [ICON_LIST,] }] },
        { type: ContextDashboardService },
        { type: NavigatorService }
    ]; };
    __decorate([
        ViewChild('dashboardDetailForm', { static: true })
    ], DashboardDetailComponent.prototype, "dashboardDetailForm", void 0);
    DashboardDetailComponent = __decorate([
        Component({
            selector: 'c8y-dashboard-detail',
            template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header separator-bottom\">\n    <h3>{{ titleAction | translate }}&nbsp;{{ titleName | translate }}</h3>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <div class=\"p-l-24 p-r-24\">\n      <form #dashboardDetailForm=\"ngForm\" class=\"d-contents\">\n        <div class=\"row\">\n          <div class=\"col-sm-6\">\n            <div *ngIf=\"!isNamedDashboard || isReport\">\n              <h6 class=\"legend form-block\">\n                <span>{{ 'General' | translate }}</span>\n              </h6>\n              <div class=\"d-flex\">\n                <c8y-form-group>\n                  <label class=\"d-block\">{{ 'Icon' | translate }}</label>\n                  <div dropdown class=\"dropdown\">\n                    <button\n                      title=\"{{ 'Icon' | translate }}\"\n                      class=\"btn-default btn btn-gray\"\n                      dropdownToggle\n                    >\n                      <i c8yIcon=\"{{ dashboard.icon }}\"></i>\n                      <span class=\"caret\"></span>\n                    </button>\n                    <ul\n                      *dropdownMenu\n                      class=\"dropdown-menu modal-inner-scroll dropdown-menu-grid-4 m-l-0\"\n                      style=\"max-height: 250px;\"\n                    >\n                      <ng-container *ngFor=\"let icon of filteredIcons\">\n                        <li (click)=\"selectIcon(icon)\">\n                          <a\n                            class=\"interact\"\n                            title=\"{{ icon }}\"\n                            [ngClass]=\"{ active: dashboard.icon === icon }\"\n                          >\n                            <i class=\"icon\" [c8yIcon]=\"icon\"></i>\n                          </a>\n                        </li>\n                      </ng-container>\n                    </ul>\n                  </div>\n                </c8y-form-group>\n                <c8y-form-group class=\"flex-grow\">\n                  <label>\n                    <span class=\"m-r-4\">{{ 'Menu label' | translate }}</span>\n                    <button\n                      class=\"btn btn-clean\"\n                      popover=\"{{\n                        'Menu label to display in submenu when dashboard is attached' | translate\n                      }}\"\n                      triggers=\"focus\"\n                      placement=\"right\"\n                      container=\"body\"\n                    >\n                      <i [c8yIcon]=\"'question-circle-o'\" class=\"text-primary\"></i>\n                    </button>\n                  </label>\n                  <input\n                    title=\"{{ 'Menu label' | translate }}\"\n                    class=\"form-control\"\n                    name=\"name\"\n                    [(ngModel)]=\"dashboard.name\"\n                    placeholder=\"{{ namePlaceholder | translate }}\"\n                    maxlength=\"512\"\n                    required\n                  />\n                </c8y-form-group>\n              </div>\n              <c8y-form-group *ngIf=\"isReport\">\n                <label translate>Description</label>\n                <textarea\n                  class=\"form-control\"\n                  rows=\"2\"\n                  name=\"description\"\n                  [(ngModel)]=\"dashboard.description\"\n                ></textarea>\n              </c8y-form-group>\n              <div class=\"row\">\n                <div class=\"col-sm-6\" *ngIf=\"!isReport\">\n                  <c8y-form-group>\n                    <label>\n                      <span class=\"m-r-4\">{{ 'Position in navigation' | translate }}</span>\n                      <button\n                        class=\"btn btn-clean\"\n                        popover=\"{{\n                          'Position in navigation menu (10000 first, -10000 last)' | translate\n                        }}\"\n                        triggers=\"focus\"\n                        placement=\"right\"\n                        container=\"body\"\n                      >\n                        <i [c8yIcon]=\"'question-circle-o'\" class=\"text-primary\"></i>\n                      </button>\n                    </label>\n                    <input\n                      title=\"{{ 'Position in navigation' | translate }}\"\n                      type=\"number\"\n                      class=\"form-control\"\n                      name=\"priority\"\n                      [(ngModel)]=\"dashboard.priority\"\n                      min=\"-10000\"\n                      max=\"10000\"\n                      placeholder=\"{{ 'e.g.' | translate }} 500\"\n                      required\n                    />\n                  </c8y-form-group>\n                </div>\n\n                <div class=\"col-sm-6\" *ngIf=\"isReport\">\n                  <label translate>Navigator menu item</label>\n                  <c8y-form-group>\n                    <label title=\"{{ 'Show in navigator' | translate }}\" class=\"c8y-checkbox\">\n                      <input\n                        type=\"checkbox\"\n                        name=\"isNavigatorNode\"\n                        [(ngModel)]=\"!!dashboard.c8y_IsNavigatorNode\"\n                      /><span></span>\n                      <span>{{ 'Show in navigator' | translate }}</span>\n                    </label>\n                  </c8y-form-group>\n                </div>\n                <div class=\"col-sm-6\" *ngIf=\"isReport\">\n                  <c8y-form-group>\n                    <label>\n                      <span class=\"m-r-4\">{{ 'Position in navigator' | translate }}</span>\n                      <ng-template #positionInNavPop>\n                        <span>\n                          {{\n                            'Position in navigator (10001 first, -10000 last).' | translate\n                          }}&nbsp;\n                          {{ 'Existing nodes:' | translate }}\n                        </span>\n                        <ul class=\"list-unstyled m-t-16\">\n                          <li *ngFor=\"let node of navigatorNodes$ | async\">\n                            <i [c8yIcon]=\"node.icon\"></i>\n                            <span class=\"word-break m-l-4 m-r-16\">\n                              {{\n                                node.label.length > 15\n                                  ? (node.label | slice: 0:15) + '...'\n                                  : node.label\n                              }}\n                            </span>\n                            <span class=\"pull-right\"> {{ node.priority }} </span>\n                          </li>\n                        </ul>\n                      </ng-template>\n                      <button\n                        class=\"btn btn-clean\"\n                        [popover]=\"positionInNavPop\"\n                        triggers=\"focus\"\n                        placement=\"right\"\n                        container=\"body\"\n                      >\n                        <i [c8yIcon]=\"'question-circle-o'\" class=\"text-primary\"></i>\n                      </button>\n                    </label>\n                    <input\n                      title=\"{{ 'Position in navigation' | translate }}\"\n                      type=\"number\"\n                      class=\"form-control\"\n                      name=\"priority\"\n                      [(ngModel)]=\"dashboard.priority\"\n                      min=\"-10000\"\n                      max=\"20000\"\n                      placeholder=\"{{ 'e.g.' | translate }} 500\"\n                    />\n                  </c8y-form-group>\n                </div>\n              </div>\n\n              <div *ngIf=\"!current && deviceType\">\n                <div class=\"form-group\">\n                  <label\n                    title=\"{{ 'Apply dashboard to all devices of type' | translate }}\"\n                    class=\"c8y-checkbox\"\n                  >\n                    <input type=\"checkbox\" name=\"deviceType\" [(ngModel)]=\"dashboard.deviceType\" />\n                    <span></span>\n                    <span class=\"m-r-4\">\n                      {{ 'Apply dashboard to all devices of type' | translate }}\n                      <i>{{ dashboard.deviceTypeValue }}</i>\n                    </span>\n                  </label>\n                </div>\n\n                <div class=\"alert alert-info  m-b-24\" *ngIf=\"isDeviceType\">\n                  <i c8y-icon=\"info\"></i>\n                  {{ 'This dashboard is shared between all devices of the type ' | translate }}\n                  <i>{{ dashboard.deviceTypeValue }}</i>\n                </div>\n              </div>\n            </div>\n            <c8y-appearance-settings\n              [(themeClass)]=\"styling.themeClass\"\n              [(headerClass)]=\"styling.headerClass\"\n            >\n            </c8y-appearance-settings>\n            <div class=\"row\">\n              <div class=\"col-sm-6\">\n                <c8y-form-group class=\"p-b-24 m-b-0\">\n                  <label>{{ 'Widget margin' | translate }}</label>\n                  <div class=\"input-group\">\n                    <input\n                      title=\"{{ 'Widget margin' | translate }}\"\n                      id=\"margin\"\n                      name=\"margin\"\n                      type=\"number\"\n                      class=\"form-control\"\n                      [(ngModel)]=\"dashboard.widgetMargin\"\n                      min=\"0\"\n                      max=\"50\"\n                      placeholder=\"{{ DEFAULT_DASHBOARD_MARGIN }}\"\n                    />\n                    <span class=\"input-group-addon\">px</span>\n                  </div>\n                </c8y-form-group>\n              </div>\n              <div class=\"col-sm-6\">\n                <c8y-form-group class=\"p-b-24 m-b-0\">\n                  <label translate>Widget titles</label>\n                  <label title=\"{{ 'Translate if possible' | translate }}\" class=\"c8y-checkbox\">\n                    <input\n                      type=\"checkbox\"\n                      name=\"translateWidgetTitle\"\n                      [(ngModel)]=\"dashboard.translateWidgetTitle\"\n                    /><span></span>\n                    <span>{{ 'Translate if possible' | translate }}</span>\n                  </label>\n                </c8y-form-group>\n              </div>\n            </div>\n          </div>\n\n          <div class=\"col-sm-6\">\n            <c8y-widget-preview\n              [tab]=\"!isNamedDashboard ? dashboard : undefined\"\n              [previewClasses]=\"getDashboardPreviewStyle()\"\n            ></c8y-widget-preview>\n          </div>\n        </div>\n      </form>\n    </div>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\">\n      {{ 'Cancel' | translate }}\n    </button>\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      [disabled]=\"dashboardDetailForm.form.invalid || dashboardDetailForm.form.pristine\"\n    >\n      {{ 'Save' | translate }}\n    </button>\n  </div>\n</div>\n"
        }),
        __param(1, Inject(ICON_LIST))
    ], DashboardDetailComponent);
    return DashboardDetailComponent;
}());

var AppearanceSettingsComponent = /** @class */ (function () {
    function AppearanceSettingsComponent() {
        this.themeClass = 'dashboard-theme-light';
        this.headerClass = 'panel-title-regular';
        this.themeClassChange = new EventEmitter();
        this.headerClassChange = new EventEmitter();
        this.possibleStylingTheme = DASHBOARD_THEME_CLASSES;
        this.possibleStylingHeader = WIDGET_HEADER_CLASSES;
    }
    AppearanceSettingsComponent.prototype.themeClassClick = function (value) {
        value = this.resetToDefault(value, this.themeClass, this.defaultThemeClass);
        this.themeClass = value;
        this.themeClassChange.emit(value);
    };
    AppearanceSettingsComponent.prototype.headerClassClick = function (value) {
        value = this.resetToDefault(value, this.headerClass, this.defaultHeaderClass);
        this.headerClass = value;
        this.headerClassChange.emit(value);
    };
    AppearanceSettingsComponent.prototype.resetToDefault = function (value, compareTo, defaultValue) {
        if (defaultValue && value === compareTo) {
            value = defaultValue;
        }
        return value;
    };
    __decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "themeClass", void 0);
    __decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "headerClass", void 0);
    __decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "defaultThemeClass", void 0);
    __decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "defaultHeaderClass", void 0);
    __decorate([
        Output()
    ], AppearanceSettingsComponent.prototype, "themeClassChange", void 0);
    __decorate([
        Output()
    ], AppearanceSettingsComponent.prototype, "headerClassChange", void 0);
    __decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "possibleStylingTheme", void 0);
    __decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "possibleStylingHeader", void 0);
    AppearanceSettingsComponent = __decorate([
        Component({
            selector: 'c8y-appearance-settings',
            template: "<h6 class=\"legend form-block\"><span translate>Layout</span></h6>\n<div class=\"row \">\n  <div class=\"col-sm-6 m-b-32\">\n    <label translate>Theme</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row\"\n        style=\"align-items: center; box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let themeClassItem of possibleStylingTheme; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradiocontentclass' + i\"\n              name=\"content\"\n              [value]=\"themeClassItem.class\"\n              [ngModel]=\"themeClass\"\n              (click)=\"themeClassClick(themeClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradiocontentclass' + i\">\n          <span>{{ themeClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"themeClassItem.class === defaultThemeClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button\n          class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ themeClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\"\n        >\n          <i [c8yIcon]=\"'question-circle-o'\" class=\"text-primary\"></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-sm-6\">\n    <label translate>Widget header style</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row a-i-center\"\n        style=\"box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let headerClassItem of possibleStylingHeader; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradioheaderclass' + i\"\n              name=\"header\"\n              [value]=\"headerClassItem.class\"\n              [ngModel]=\"headerClass\"\n              (click)=\"headerClassClick(headerClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradioheaderclass' + i\">\n          <span>{{ headerClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"headerClassItem.class === defaultHeaderClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button\n          class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ headerClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\"\n        >\n          <i [c8yIcon]=\"'question-circle-o'\" class=\"text-primary\"></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n</div>\n",
            viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
        })
    ], AppearanceSettingsComponent);
    return AppearanceSettingsComponent;
}());

/**
 * The context dashboard is a dashboard which resolves it data from the current context (device or group)
 * it is displayed on. It usually uses the route.data for it, but you can pass
 * a different managedObject to the [mo] input parameter to change that behavior.
 */
var ContextDashboardComponent = /** @class */ (function () {
    function ContextDashboardComponent(route, router, contextDashboardService, alert, renderer, moduleConfig, widgetService, bsModal, inventory, gainsightService) {
        this.route = route;
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.alert = alert;
        this.renderer = renderer;
        this.moduleConfig = moduleConfig;
        this.widgetService = widgetService;
        this.bsModal = bsModal;
        this.inventory = inventory;
        this.gainsightService = gainsightService;
        this.childrenClasses = '';
        this.setTitle = false;
        this.disabled = false;
        this.defaultWidgets = [];
        this.canDelete = true;
        this.isLoading = true;
        this.class = '';
        this.widgets = [];
    }
    ContextDashboardComponent.prototype.ngOnInit = function () {
        if (!this.name) {
            this.loadContextDashboard();
            return;
        }
        this.loadNamedDashboard();
    };
    /**
     * Applies the current context to the widget
     * @param widget The widget to apply the context to.
     */
    ContextDashboardComponent.prototype.applyDeviceTarget = function (widget) {
        if (widget.config.device) {
            widget.config.device = { id: this.context.id, name: this.context.name };
        }
    };
    /**
     * Removes the route listener.
     */
    ContextDashboardComponent.prototype.ngOnDestroy = function () {
        if (this.dataSub) {
            this.dataSub.unsubscribe();
        }
    };
    /**
     * Restores the dashboard widgets to the default widgets.
     */
    ContextDashboardComponent.prototype.restore = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.isLoading = true;
                        this.mo.c8y_Dashboard.children = this.contextDashboardService.mapWidgets(this.defaultWidgets);
                        return [4 /*yield*/, this.contextDashboardService.update(this.mo)];
                    case 1:
                        _a.sent();
                        this.onLoad();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Updates all dashboards children's. Useful for position changes on the dashboard.
     * @param child The child to change.
     */
    ContextDashboardComponent.prototype.updateDashboardChildren = function (child) {
        return __awaiter(this, void 0, void 0, function () {
            var children, dashboardMO, mappedChildren;
            var _this = this;
            return __generator(this, function (_a) {
                children = child.children;
                dashboardMO = this.mo;
                mappedChildren = keyBy(children.map(function (c) { return _this.componentToWidget(c); }), 'id');
                dashboardMO.c8y_Dashboard.children = mappedChildren;
                return [2 /*return*/, this.contextDashboardService.update(dashboardMO)];
            });
        });
    };
    /**
     * Remove the complete dashboard and navigate away.
     */
    ContextDashboardComponent.prototype.deleteDashboard = function () {
        return __awaiter(this, void 0, void 0, function () {
            var route;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contextDashboardService.delete(this.mo)];
                    case 1:
                        _a.sent();
                        if (this.route.parent) {
                            route = this.route.parent.snapshot.url.map(function (segment) { return segment.path; }).join('/');
                            this.router.navigateByUrl(route);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Edits the current dashboard
     */
    ContextDashboardComponent.prototype.editDashboard = function () {
        return __awaiter(this, void 0, void 0, function () {
            var isReport, _a, name_1, icon, priority, c8y_IsNavigatorNode, description, initialState, modal, dashboardMO, cfg, name_2, icon, c8y_IsNavigatorNode, priority, description, dashboardCfg, ex_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        isReport = this.contextDashboardService.isReport(this.mo);
                        if (isReport) {
                            _a = this.context, name_1 = _a.name, icon = _a.icon, priority = _a.priority, c8y_IsNavigatorNode = _a.c8y_IsNavigatorNode, description = _a.description;
                            Object.assign(this.dashboard, { name: name_1, icon: icon, priority: priority, c8y_IsNavigatorNode: c8y_IsNavigatorNode, description: description });
                        }
                        initialState = {
                            dashboard: this.dashboard,
                            deviceType: this.context.type,
                            isDeviceType: this.contextDashboardService.isDeviceType(this.mo),
                            isNamedDashboard: this.contextDashboardService.isNamed(this.mo),
                            isReport: isReport
                        };
                        modal = this.bsModal.show(DashboardDetailComponent, {
                            class: 'modal-lg',
                            initialState: initialState,
                            ignoreBackdropClick: true
                        }).content;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 5, , 6]);
                        dashboardMO = cloneDeep(this.mo);
                        return [4 /*yield*/, modal.result];
                    case 2:
                        cfg = _b.sent();
                        if (isReport) {
                            name_2 = cfg.name, icon = cfg.icon, c8y_IsNavigatorNode = cfg.c8y_IsNavigatorNode, priority = cfg.priority, description = cfg.description, dashboardCfg = __rest(cfg, ["name", "icon", "c8y_IsNavigatorNode", "priority", "description"]);
                            dashboardMO.c8y_Dashboard = dashboardCfg;
                            this.updateReport({ id: this.context.id, name: name_2, icon: icon, c8y_IsNavigatorNode: c8y_IsNavigatorNode, priority: priority, description: description });
                        }
                        else {
                            dashboardMO.c8y_Dashboard = cfg;
                        }
                        return [4 /*yield*/, this.contextDashboardService.update(dashboardMO)];
                    case 3:
                        _b.sent();
                        return [4 /*yield*/, this.contextDashboardService.refreshTabs(dashboardMO)];
                    case 4:
                        _b.sent();
                        this.onLoad();
                        modal.close();
                        return [3 /*break*/, 6];
                    case 5:
                        ex_1 = _b.sent();
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Edits a widget on the dashboard.
     * @param change The widget change event.
     */
    ContextDashboardComponent.prototype.editWidget = function (change) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, x, y, width, height, component;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = change.source, x = _a.x, y = _a.y, width = _a.width, height = _a.height;
                        return [4 /*yield*/, this.widgetService.getWidgetDefinition(change.widget.name || change.widget.componentId)];
                    case 1:
                        component = _b.sent();
                        if (!component) {
                            this.addWidget();
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.addWidget(__assign({}, component, { data: __assign({}, component.data, change.widget, { _x: x, _y: y, _width: width, _height: height }) }))];
                    case 2:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Adds a widget to the dashboard.
     * @param selected Define a selected component to switch to edit mode directly.
     */
    ContextDashboardComponent.prototype.addWidget = function (selected) {
        return __awaiter(this, void 0, void 0, function () {
            var initialState, modal, newWidget, ex_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        initialState = {
                            mo: this.mo,
                            context: this.context.c8y_Report ? {} : this.context,
                            selected: cloneDeep(selected)
                        };
                        modal = this.bsModal.show(WidgetConfigComponent, {
                            class: 'modal-lg',
                            initialState: initialState,
                            ignoreBackdropClick: true
                        }).content;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 4, , 5]);
                        return [4 /*yield*/, modal.result];
                    case 2:
                        newWidget = _a.sent();
                        if (!this.mo.c8y_Dashboard.children) {
                            this.mo.c8y_Dashboard.children = {};
                        }
                        this.mo.c8y_Dashboard.children[newWidget.id] = newWidget;
                        this.contextDashboardService.update(this.mo);
                        newWidget.classes = this.mergeWidgetClasses(newWidget);
                        return [4 /*yield*/, this.updateWidget(newWidget)];
                    case 3:
                        _a.sent();
                        modal.close();
                        return [3 /*break*/, 5];
                    case 4:
                        ex_2 = _a.sent();
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Updates a widget or adds a new one if it doesn't exist on
     * the dashboard.
     * @param widget The new widget
     */
    ContextDashboardComponent.prototype.updateWidget = function (widget) {
        return __awaiter(this, void 0, void 0, function () {
            var index, isNew, mappedWidget;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        index = findIndex(this.widgets, { id: widget.id });
                        isNew = index === -1;
                        return [4 /*yield*/, this.mapLegacy(widget)];
                    case 1:
                        mappedWidget = _a.sent();
                        if (isNew) {
                            this.widgets.push(mappedWidget);
                        }
                        else {
                            this.widgets.splice(index, 1, mappedWidget);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Removes a widget and rearranges the remaining ones
     * if necessary.
     * @param change The change event.
     */
    ContextDashboardComponent.prototype.deleteWidget = function (change) {
        var _this = this;
        var widget = change.widget, source = change.source;
        delete this.mo.c8y_Dashboard.children[widget.id];
        var removed = this.widgets.find(function (_a) {
            var id = _a.id;
            return id === widget.id;
        });
        this.widgets.splice(this.widgets.indexOf(removed), 1);
        // using setTimeout to give the component the chance to remove it.
        setTimeout(function () {
            var child = new DashboardChildChange(source);
            child.collapseUpAll();
            _this.updateDashboardChildren(child);
        });
    };
    /**
     * This is a workaround to ensure that the dragged-element
     * (which is attached to the body) has the right styling.
     */
    ContextDashboardComponent.prototype.addDashboardClassToBody = function () {
        var _this = this;
        this.class.split(' ').forEach(function (cssClass) {
            _this.renderer.addClass(document.body, cssClass);
        });
    };
    /**
     * This is a workaround to ensure that the dragged-element
     * (which is attached to the body) has the right styling.
     */
    ContextDashboardComponent.prototype.removeDashboardClassFromBody = function () {
        var _this = this;
        this.class.split(' ').forEach(function (cssClass) {
            _this.renderer.removeClass(document.body, cssClass);
        });
    };
    /**
     * Changes the dashboard settings to frozen or vice versa.
     * @param settings The current settings of the dashboard.
     */
    ContextDashboardComponent.prototype.toggleFreeze = function (settings) {
        return __awaiter(this, void 0, void 0, function () {
            var ex_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.toggleIsFrozenFlag(settings);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.contextDashboardService.update(this.mo)];
                    case 2:
                        _a.sent();
                        if (this.dashboard.isFrozen) {
                            this.alert.success(gettext('Your dashboard is locked now.'));
                        }
                        else {
                            this.alert.success(gettext('Your dashboard is unlocked now.'));
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        ex_3 = _a.sent();
                        this.alert.addServerFailure(ex_3);
                        this.toggleIsFrozenFlag(settings);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ContextDashboardComponent.prototype.updateReport = function (mo) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.inventory.update(mo)];
                    case 1:
                        res = _a.sent();
                        this.context = res.data;
                        if (this.route.parent) {
                            this.route.parent.snapshot.data.contextData = this.context;
                        }
                        this.contextDashboardService.updateNavigatorItem(res.data);
                        return [2 /*return*/];
                }
            });
        });
    };
    ContextDashboardComponent.prototype.toggleIsFrozenFlag = function (settings) {
        settings.isFrozen = !settings.isFrozen;
        this.dashboard.isFrozen = settings.isFrozen;
    };
    ContextDashboardComponent.prototype.loadContextDashboard = function () {
        var _this = this;
        this.dataSub = this.route.data.subscribe(function (_a) {
            var dashboard = _a.dashboard;
            _this.context = _this.route.parent.snapshot.data.contextData;
            _this.mo = dashboard;
            _this.dashboard = _this.mo.c8y_Dashboard;
            _this.onLoad(true);
        });
    };
    ContextDashboardComponent.prototype.loadNamedDashboard = function () {
        var _this = this;
        this.dataSub = this.contextDashboardService
            .getNamedDashboardOrCreate(this.name, this.defaultWidgets)
            .subscribe(function (mo) {
            _this.context = _this.context || {};
            _this.mo = mo;
            _this.dashboard = _this.mo.c8y_Dashboard;
            _this.onLoad(true);
        });
    };
    ContextDashboardComponent.prototype.onLoad = function (trackExperience) {
        return __awaiter(this, void 0, void 0, function () {
            var canEditDashboard, dashboardChildren, isDeviceType, isReport, dashboardClasses, _a;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.contextDashboardService.canEditDashboard(this.mo)];
                    case 1:
                        canEditDashboard = _b.sent();
                        this.disabled = !canEditDashboard;
                        dashboardChildren = cloneDeep(this.mo.c8y_Dashboard.children);
                        isDeviceType = this.contextDashboardService.isDeviceType(this.mo);
                        isReport = this.contextDashboardService.isReport(this.mo);
                        dashboardClasses = __assign({ 'c8y-grid-dashboard': true, dashboard: true }, this.dashboard.classes);
                        _a = this;
                        return [4 /*yield*/, Promise.all(values(dashboardChildren).map(function (widget) {
                                widget.classes = _this.mergeWidgetClasses(widget);
                                if (isDeviceType) {
                                    _this.applyDeviceTarget(widget);
                                }
                                if (trackExperience) {
                                    _this.gainsightService.triggerEvent('loadWidget', {
                                        widgetName: widget.componentId || widget.name
                                    });
                                }
                                return _this.mapLegacy(widget);
                            }))];
                    case 2:
                        _a.widgets = _b.sent();
                        this.class = Object.keys(dashboardClasses).join(' ');
                        if (isReport) {
                            this.addReportDashboardSettings();
                        }
                        this.isLoading = false;
                        return [2 /*return*/];
                }
            });
        });
    };
    ContextDashboardComponent.prototype.mergeWidgetClasses = function (widget) {
        var _a;
        var hasHeaderClass = WIDGET_HEADER_CLASSES.find(function (el) { return widget.classes && widget.classes[el.class]; });
        var widgetClasses = hasHeaderClass
            ? __assign({}, widget.classes) : __assign({}, this.dashboard.widgetClasses, widget.classes);
        return __assign((_a = { card: true, 'card-dashboard': true }, _a[kebabCase(widget.componentId || widget.name)] = true, _a), widgetClasses);
    };
    ContextDashboardComponent.prototype.componentToWidget = function (child) {
        return __assign({}, omit(child.data, ['componentTransformConfigWithContext', 'transformConfigWithContext']), {
            _x: child.x,
            _y: child.y,
            _width: child.width,
            _height: child.height
        });
    };
    ContextDashboardComponent.prototype.mapLegacy = function (widget) {
        return __awaiter(this, void 0, void 0, function () {
            var cmp;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.widgetService.getWidgetDefinition(widget.componentId || widget.name)];
                    case 1:
                        cmp = _a.sent();
                        if (get(cmp, 'data.settings.upgrade')) {
                            widget.widgetComponent = cmp.data.settings.widgetComponent;
                            widget.configComponent = cmp.data.settings.configComponent;
                            widget.templateUrl = cmp.data.settings.templateUrl;
                            widget.configTemplateUrl = cmp.data.settings.configTemplateUrl;
                            widget.transformConfigWithContext =
                                cmp.data.settings.componentTransformConfigWithContext ||
                                    cmp.data.settings.transformConfigWithContext;
                        }
                        else {
                            delete widget.templateUrl;
                            delete widget.configTemplateUrl;
                        }
                        return [2 /*return*/, widget];
                }
            });
        });
    };
    ContextDashboardComponent.prototype.addReportDashboardSettings = function () {
        this.setTitle = true;
        this.title = this.context.name;
        this.breadcrumbSettings = {
            icon: 'th',
            label: 'Reports',
            path: 'reports'
        };
        this.canDelete = false;
    };
    ContextDashboardComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: Router },
        { type: ContextDashboardService },
        { type: AlertService },
        { type: Renderer2 },
        { type: undefined, decorators: [{ type: Inject, args: [CONTEXT_DASHBOARD_CONFIG,] }] },
        { type: WidgetService },
        { type: BsModalService },
        { type: InventoryService },
        { type: GainsightService }
    ]; };
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "name", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "childrenClasses", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "context", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "setTitle", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "defaultWidgets", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "canDelete", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "isLoading", void 0);
    __decorate([
        Input()
    ], ContextDashboardComponent.prototype, "breadcrumbSettings", void 0);
    __decorate([
        HostBinding('class')
    ], ContextDashboardComponent.prototype, "class", void 0);
    ContextDashboardComponent = __decorate([
        Component({
            selector: 'c8y-context-dashboard',
            template: "<c8y-action-bar-item [placement]=\"'more'\" *ngIf=\"defaultWidgets.length > 0\">\n  <button (click)=\"restore()\" [disabled]=\"dashboard?.isFrozen || disabled\">\n    <i c8yIcon=\"undo\"></i>&nbsp;<span translate>Restore dashboard</span>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-widgets-dashboard\n  [context]=\"context\"\n  [contextDashboard]=\"dashboard\"\n  [widgets]=\"widgets\"\n  [settings]=\"{\n    isLoading: isLoading,\n    isFrozen: dashboard?.isFrozen,\n    isDisabled: disabled,\n    canDelete: canDelete,\n    translateWidgetTitle: dashboard?.translateWidgetTitle,\n    allowFullscreen: moduleConfig.allowFullscreen,\n    title: setTitle ? dashboard.name || title : undefined,\n    widgetMargin: dashboard?.widgetMargin\n  }\"\n  [breadcrumb]=\"breadcrumbSettings\"\n  (onFreeze)=\"toggleFreeze($event)\"\n  (onChangeDashboard)=\"updateDashboardChildren($event)\"\n  (onAddWidget)=\"addWidget()\"\n  (onEditWidget)=\"editWidget($event)\"\n  (onDeleteWidget)=\"deleteWidget($event)\"\n  (onChangeStart)=\"addDashboardClassToBody()\"\n  (onChangeEnd)=\"removeDashboardClassFromBody()\"\n  (onEditDashboard)=\"editDashboard()\"\n  (onDeleteDashboard)=\"deleteDashboard()\"\n>\n</c8y-widgets-dashboard>\n",
            host: {
                style: "\n      display: block;\n    ",
                class: 'dashboard c8y-grid-dashboard'
            }
        }),
        __param(5, Inject(CONTEXT_DASHBOARD_CONFIG))
    ], ContextDashboardComponent);
    return ContextDashboardComponent;
}());

var DashboardActionsFactory = /** @class */ (function () {
    function DashboardActionsFactory(router, contextDashboardService, bsModal, permissions) {
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.permissions = permissions;
    }
    DashboardActionsFactory.prototype.get = function (activatedRoute) {
        return __awaiter(this, void 0, void 0, function () {
            var context, canEditDashboard;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        context = !activatedRoute.parent || activatedRoute.snapshot.data.context
                            ? activatedRoute.snapshot.data
                            : activatedRoute.parent.snapshot.data;
                        return [4 /*yield*/, this.isEnabled(context)];
                    case 1:
                        canEditDashboard = _a.sent();
                        return [2 /*return*/, [
                                {
                                    label: gettext('Add dashboard'),
                                    priority: 2500,
                                    action: function () { return _this.addDashboard(context); },
                                    disabled: !canEditDashboard
                                }
                            ]];
                }
            });
        });
    };
    DashboardActionsFactory.prototype.addDashboard = function (context) {
        return __awaiter(this, void 0, void 0, function () {
            var initialState, modal, dashboardCfg, dashboardMO, ex_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (context.context === ViewContext.Device) {
                            initialState = {
                                deviceType: context.contextData.type
                            };
                        }
                        modal = this.bsModal.show(DashboardDetailComponent, {
                            class: 'modal-lg',
                            initialState: initialState,
                            ignoreBackdropClick: true
                        }).content;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 5, , 6]);
                        return [4 /*yield*/, modal.result];
                    case 2:
                        dashboardCfg = _a.sent();
                        return [4 /*yield*/, this.contextDashboardService.create(dashboardCfg, context)];
                    case 3:
                        dashboardMO = _a.sent();
                        return [4 /*yield*/, this.contextDashboardService.navigateToDashboard(dashboardMO)];
                    case 4:
                        _a.sent();
                        modal.close();
                        return [3 /*break*/, 6];
                    case 5:
                        ex_1 = _a.sent();
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    DashboardActionsFactory.prototype.isEnabled = function (context) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(context.contextData && context.contextData.id)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.permissions.canEdit(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE'], context.contextData)];
                    case 1: return [2 /*return*/, ((_a.sent()) && [ViewContext.Device, ViewContext.Group].includes(context.context))];
                    case 2: return [2 /*return*/, (this.permissions.hasAnyRole(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE']) &&
                            [ViewContext.Device, ViewContext.Group].includes(context.context))];
                }
            });
        });
    };
    DashboardActionsFactory.ctorParameters = function () { return [
        { type: Router },
        { type: ContextDashboardService },
        { type: BsModalService },
        { type: Permissions }
    ]; };
    DashboardActionsFactory = __decorate([
        Injectable()
    ], DashboardActionsFactory);
    return DashboardActionsFactory;
}());

var defaultConfig = {
    allowFullscreen: true,
    widgetFilter: function (_a) {
        var data = _a.data;
        return !(data && data.settings && data.settings.noNewWidgets);
    },
    routes: [
        {
            path: 'dashboard/:dashboardId',
            component: ContextDashboardComponent,
            canActivate: [DeviceDashboardGuard],
            context: ViewContext.Device
        },
        {
            path: 'dashboard/:dashboardId',
            component: ContextDashboardComponent,
            canActivate: [GroupDashboardGuard],
            context: ViewContext.Group
        }
    ]
};
var ContextDashboardModule = /** @class */ (function () {
    function ContextDashboardModule() {
    }
    ContextDashboardModule_1 = ContextDashboardModule;
    ContextDashboardModule.config = function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: ContextDashboardModule_1,
            providers: [
                {
                    provide: CONTEXT_DASHBOARD_CONFIG,
                    useValue: __assign({}, defaultConfig, config)
                },
                ContextDashboardService,
                DeviceDashboardGuard,
                GroupDashboardGuard,
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: __assign({}, defaultConfig, config).routes,
                    multi: true
                },
                {
                    provide: HOOK_ACTION,
                    useClass: DashboardActionsFactory,
                    multi: true
                }
            ]
        };
    };
    var ContextDashboardModule_1;
    ContextDashboardModule = ContextDashboardModule_1 = __decorate([
        NgModule({
            declarations: [
                AppearanceSettingsComponent,
                ContextDashboardComponent,
                WidgetConfigComponent,
                WidgetPreviewComponent,
                DashboardDetailComponent
            ],
            imports: [CoreModule, TooltipModule, PopoverModule, BsDropdownModule.forRoot()],
            exports: [ContextDashboardComponent, WidgetConfigComponent],
            entryComponents: [ContextDashboardComponent, WidgetConfigComponent, DashboardDetailComponent]
        })
    ], ContextDashboardModule);
    return ContextDashboardModule;
}());

var ReportDashboardService = /** @class */ (function () {
    function ReportDashboardService(inventoryService, navigatorService) {
        this.inventoryService = inventoryService;
        this.navigatorService = navigatorService;
        this.queriesUtil = new QueriesUtil();
    }
    ReportDashboardService.prototype.listReports = function (options) {
        if (options === void 0) { options = {}; }
        var query = { __has: 'c8y_Report' };
        if (options.filter) {
            query = this.queriesUtil.addAndFilter(query, options.filter);
        }
        var filters = {
            query: this.queriesUtil.buildQuery(query),
            pageSize: options.pageSize || 50,
            withTotalPages: true
        };
        return this.inventoryService.list(filters);
    };
    ReportDashboardService.prototype.createReport = function (reportCfg) {
        var report = {};
        Object.assign(report, reportCfg);
        Object.assign(report, { c8y_Report: {} });
        return this.inventoryService.create(report);
    };
    ReportDashboardService.prototype.addReportNavigatorNode = function (report) {
        var node = new NavigatorNode({
            label: report.name,
            path: "reports/" + report.id,
            icon: report.icon,
            priority: report.priority
        });
        this.navigatorService.add(node);
    };
    ReportDashboardService.prototype.removeNavigatorNode = function (report) {
        var _this = this;
        this.navigatorService.state.forEach(function (node) {
            if (node.path === "reports/" + report.id) {
                _this.navigatorService.remove(node);
                return;
            }
        });
    };
    ReportDashboardService.ctorParameters = function () { return [
        { type: InventoryService },
        { type: NavigatorService }
    ]; };
    ReportDashboardService = __decorate([
        Injectable()
    ], ReportDashboardService);
    return ReportDashboardService;
}());

var ReportDashboardListComponent = /** @class */ (function () {
    function ReportDashboardListComponent(inventoryService, contextDashboardService, bsModal, translateService, modal, alertService, reportDashboardService) {
        var _this = this;
        this.inventoryService = inventoryService;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.translateService = translateService;
        this.modal = modal;
        this.alertService = alertService;
        this.reportDashboardService = reportDashboardService;
        this.textFilter$ = new BehaviorSubject('');
        this.reload$ = new BehaviorSubject(null);
        this.reloading = false;
        this.reports$ = combineLatest(this.textFilter$.pipe(debounceTime(400), distinctUntilChanged()), this.reload$).pipe(tap(function () {
            _this.reloading = true;
        }), switchMap(function (_a) {
            var _b = __read(_a, 1), text = _b[0];
            return _this.loadReports(text);
        }), tap(function () {
            _this.reloading = false;
        }), shareReplay(1));
        this.DELETED_SUCCESS_MSG = gettext('Report deleted.');
    }
    ReportDashboardListComponent.prototype.loadReports = function (partialName) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, partialName
                        ? this.reportDashboardService.listReports({ filter: { name: "*" + partialName + "*" } })
                        : this.reportDashboardService.listReports()];
            });
        });
    };
    ReportDashboardListComponent.prototype.add = function () {
        return __awaiter(this, void 0, void 0, function () {
            var initialState, modal, cfg, name_1, icon, c8y_IsNavigatorNode, priority, description, dashboardCfg, report, ex_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        initialState = {
                            isReport: true,
                            isNamedDashboard: true
                        };
                        modal = this.bsModal.show(DashboardDetailComponent, {
                            class: 'modal-lg',
                            ignoreBackdropClick: true,
                            initialState: initialState
                        }).content;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 5, , 6]);
                        return [4 /*yield*/, modal.result];
                    case 2:
                        cfg = _a.sent();
                        name_1 = cfg.name, icon = cfg.icon, c8y_IsNavigatorNode = cfg.c8y_IsNavigatorNode, priority = cfg.priority, description = cfg.description, dashboardCfg = __rest(cfg, ["name", "icon", "c8y_IsNavigatorNode", "priority", "description"]);
                        return [4 /*yield*/, this.reportDashboardService.createReport({
                                name: name_1,
                                icon: icon,
                                c8y_IsNavigatorNode: c8y_IsNavigatorNode,
                                priority: priority,
                                description: description
                            })];
                    case 3:
                        report = (_a.sent()).data;
                        return [4 /*yield*/, this.contextDashboardService.create(dashboardCfg, "" + this.contextDashboardService.REPORT_PARTIAL_NAME + report.id)];
                    case 4:
                        _a.sent();
                        if (report.c8y_IsNavigatorNode) {
                            this.reportDashboardService.addReportNavigatorNode(report);
                        }
                        this.reload$.next();
                        modal.close();
                        return [3 /*break*/, 6];
                    case 5:
                        ex_1 = _a.sent();
                        if (ex_1) {
                            throw new Error("Something went wrong: " + ex_1);
                        }
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    ReportDashboardListComponent.prototype.delete = function (report) {
        return __awaiter(this, void 0, void 0, function () {
            var msg, ex_2;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        msg = gettext("You are about to delete the report \"{{ reportName }}\". Do you want to proceed?");
                        return [4 /*yield*/, this.modal.confirm(gettext('Delete report'), this.translateService.instant(msg, {
                                reportName: report.name
                            }), Status.DANGER, {
                                ok: gettext('Delete'),
                                cancel: gettext('Cancel')
                            })];
                    case 1:
                        _a.sent();
                        this.contextDashboardService
                            .getDashboard$("report_" + report.id, [ContextDashboardType.Named])
                            .pipe(take(1))
                            .subscribe(function (dashboard) { return _this.contextDashboardService.delete(dashboard, false); });
                        return [4 /*yield*/, this.inventoryService.delete(report.id)];
                    case 2:
                        _a.sent();
                        this.alertService.success(this.DELETED_SUCCESS_MSG);
                        if (report.c8y_IsNavigatorNode) {
                            this.reportDashboardService.removeNavigatorNode(report);
                        }
                        this.reload$.next();
                        return [3 /*break*/, 4];
                    case 3:
                        ex_2 = _a.sent();
                        if (ex_2) {
                            throw new Error("Something went wrong: " + ex_2);
                        }
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ReportDashboardListComponent.prototype.update = function (report) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        report.c8y_IsNavigatorNode = report.c8y_IsNavigatorNode ? {} : null;
                        return [4 /*yield*/, this.inventoryService.update(report)];
                    case 1:
                        _a.sent();
                        report.c8y_IsNavigatorNode
                            ? this.reportDashboardService.addReportNavigatorNode(report)
                            : this.reportDashboardService.removeNavigatorNode(report);
                        return [2 /*return*/];
                }
            });
        });
    };
    ReportDashboardListComponent.ctorParameters = function () { return [
        { type: InventoryService },
        { type: ContextDashboardService },
        { type: BsModalService },
        { type: TranslateService },
        { type: ModalService },
        { type: AlertService },
        { type: ReportDashboardService }
    ]; };
    __decorate([
        ViewChild(FilterInputComponent, { static: false })
    ], ReportDashboardListComponent.prototype, "filter", void 0);
    ReportDashboardListComponent = __decorate([
        Component({
            selector: 'c8y-report-dashboard-list',
            template: "<c8y-title>\n  <span translate>\n    Reports\n  </span>&nbsp;\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form\">\n  <div class=\"input-group input-group-search\">\n    <input class=\"form-control\"\n      type=\"search\"\n      title=\"{{ 'Filter\u2026' | translate }}\"\n      placeholder=\"{{ 'Filter\u2026' | translate }}\"\n      [ngModel]=\"textFilter$ | async\"\n      (ngModelChange)=\"textFilter$.next($event)\"\n    />\n    <span class=\"input-group-addon\">\n      <i c8yIcon=\"search\"\n        *ngIf=\"(textFilter$ | async).length === 0\"\n      ></i>\n      <i class=\"text-muted\"\n        c8yIcon=\"times\"\n        *ngIf=\"(textFilter$ | async).length > 0\"\n        (click)=\"textFilter$.next('')\"\n      ></i>\n    </span>\n  </div>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\"\n    title=\"{{ 'Add report' | translate }}\"\n    (click)=\"add()\"\n    >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add report' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\"\n    title=\"{{ 'Reload' | translate }}\"\n    (click)=\"loadReports()\"\n  >\n    <i [ngClass]=\"{ 'icon-spin': reloading }\"\n      c8yIcon=\"refresh\"\n    ></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"p-b-32\">\n  <c8y-list-group>\n    <c8y-li class=\"page-sticky-header hidden-xs hidden-sm\"\n      *ngIf=\"(reports$ | async)?.data.length > 0\"\n    >\n      <c8y-li-icon>\n        <i class=\"fa\"></i>\n      </c8y-li-icon>\n      <c8y-li-body class=\"content-flex-80\">\n        <div class=\"col-3\">\n          {{ 'Report' | translate }}\n        </div>\n        <div class=\"col-6\">\n          {{ 'Description' | translate }}\n        </div>\n        <div class=\"col-2\">\n          {{ 'Show in navigator' | translate }}\n        </div>\n      </c8y-li-body>\n    </c8y-li>\n\n    <c8y-li *c8yFor=\"let report of reports$; let i = index; loadMore: 'auto'\">\n      <c8y-li-icon [icon]=\"report.icon\"></c8y-li-icon>\n      <c8y-li-body class=\"content-flex-70\">\n        <div class=\"col-3\">\n          <button class=\"btn-clean\"\n            title=\"{{ report.name }}\"\n            routerLink=\"/reports/{{ report.id }}\"\n          >\n            <span class=\"text-truncate\">\n              {{ report.name }}\n            </span>\n          </button>\n        </div>\n        <div class=\"col-6\">\n          <small class=\"text-truncate-wrap\">\n            <em class=\"text-muted\"\n              *ngIf=\"!report.description; else showDescription\"\n            >\n              {{ 'No description available' | translate }}\n            </em>\n            <ng-template #showDescription>\n              {{ report.description }}\n            </ng-template>\n          </small>\n        </div>\n        <div class=\"col-2\">\n          <span class=\"m-t-8 visible-xs\"></span>\n          <label class=\"c8y-switch c8y-switch--inline\"\n            title=\"{{ 'Show in navigator' | translate }}\"\n          >\n            <input\n              [(ngModel)]=\"!!report.c8y_IsNavigatorNode\"\n              type=\"checkbox\"\n              (change)=\"update(report)\"\n            />\n            <span></span>\n            <span class=\"visible-xs\">\n              {{ 'Show in navigator' | translate }}\n            </span>\n          </label>\n        </div>\n        <div class=\"col-1 text-right hidden-xs\">\n          <button class=\"btn-dot showOnHover pull-right\"\n            (click)=\"delete(report)\"\n            title=\"{{ 'Remove report' | translate }}\"\n          >\n            <i class=\"text-danger\"\n            c8yIcon=\"minus-circle\"\n            ></i>\n          </button>\n        </div>\n        <div class=\"visible-xs p-t-8 text-right\">\n          <button class=\"btn-danger btn btn-xs\"\n            (click)=\"delete(report)\"\n            title=\"{{ 'Remove report' | translate }}\"\n          >\n            <i c8yIcon=\"trash\"></i>\n            Delete\n          </button>\n        </div>\n      </c8y-li-body>\n    </c8y-li>\n  </c8y-list-group>\n</div>\n\n<div class=\"c8y-empty-state text-center m-t-40\"\n  *ngIf=\"(reports$ | async)?.data.length === 0\"\n  >\n  <h1 c8yIcon=\"c8y-reports\" class=\"c8y-icon-duocolor\"></h1>\n  <h3 translate>\n    There are no reports defined\n  </h3>\n  <p translate>\n    Add a report first.\n  </p>\n  <div>\n    <button class=\"btn btn-primary\"\n      (click)=\"add()\"\n      translate\n    >\n      Add report\n    </button>\n  </div>\n  <p c8y-guide-docs>\n    <small translate ngNonBindable\n      >Find out more in the\n      <a c8y-guide-href=\"users-guide/cockpit/#reports\">\n        User guide`KEEP_ORIGINAL`\n      </a>.\n      </small>\n  </p>\n</div>\n"
        })
    ], ReportDashboardListComponent);
    return ReportDashboardListComponent;
}());

var CockpitDashboardComponent = /** @class */ (function () {
    function CockpitDashboardComponent() {
        this.DEFAULT_HOME_WIDGETS = [
            {
                name: 'cockpit.welcome.widget',
                title: gettext(''),
                _x: 0,
                _y: 0,
                _height: 5,
                _width: 12,
                classes: { 'panel-title-hidden': true }
            },
            {
                name: 'Asset Alarms',
                title: gettext('Active, critical alarms'),
                _x: 0,
                _y: 5,
                _height: 4,
                _width: 6
            },
            {
                name: 'Recent Alarms',
                title: gettext('Recent alarms'),
                _x: 0,
                _y: 9,
                _height: 4,
                _width: 6
            },
            {
                name: 'Map',
                title: gettext('Map'),
                _x: 6,
                _y: 5,
                _height: 8,
                _width: 6
            }
        ];
    }
    CockpitDashboardComponent = __decorate([
        Component({
            selector: 'c8y-cockpit-dashboard',
            template: "<c8y-title>\n  {{'Home' | translate}}\n</c8y-title>\n\n<c8y-context-dashboard\n  name=\"home-cockpit1\"\n  [defaultWidgets]=\"DEFAULT_HOME_WIDGETS\"\n  [canDelete]=\"false\"\n></c8y-context-dashboard>\n"
        })
    ], CockpitDashboardComponent);
    return CockpitDashboardComponent;
}());

var WelcomeToCockpit = /** @class */ (function () {
    function WelcomeToCockpit(tenantService, docs, router, navigator, applicationService) {
        this.tenantService = tenantService;
        this.docs = docs;
        this.router = router;
        this.navigator = navigator;
        this.applicationService = applicationService;
        this.quickLinks = [];
        this.CONFIGURATION_NODE = 'Configuration';
        this.TRIAL = 'TRIAL';
    }
    WelcomeToCockpit.prototype.ngOnInit = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.tenantService.currentTenantType()];
                    case 1:
                        _a.tenantType = _b.sent();
                        this.setMessage();
                        this.navSubscription = this.navigator.items$.subscribe(function (nodes) {
                            _this.navNodes = nodes;
                            _this.configurationNode = nodes.find(function (node) { return node.label === _this.CONFIGURATION_NODE; });
                        });
                        this.docsSubscription = this.docs.items$.subscribe(function (links) {
                            _this.links = links;
                        });
                        // <---TRIAL & REGULAR TENANT --->
                        this.createConnectSmartphoneQuickLink();
                        return [4 /*yield*/, this.createQuicklinkRegisterDevice()];
                    case 2:
                        _b.sent();
                        // <--- TRIAL TENANT --->
                        if (this.tenantType === this.TRIAL) {
                            this.createQuicklinkUserGuide();
                            return [2 /*return*/];
                        }
                        // <--- REGULAR TENANT --->
                        this.createQuicklinkAddGroup();
                        this.createQuickLinkReports();
                        this.createQuickLinkExports();
                        this.createQuicklinkSmartRules();
                        return [2 /*return*/];
                }
            });
        });
    };
    WelcomeToCockpit.prototype.ngOnDestroy = function () {
        if (this.docsSubscription && !this.docsSubscription.closed) {
            this.docsSubscription.unsubscribe();
        }
        if (this.navSubscription && !this.navSubscription.closed) {
            this.navSubscription.unsubscribe();
        }
    };
    WelcomeToCockpit.prototype.setMessage = function () {
        if (this.tenantType === this.TRIAL) {
            this.welcomeMessage = gettext("\n        The Cockpit application allows you to build IoT applications in minutes.\n        To get started, connect any device to the platform.\n        If you do not have an IoT device to hand, you can start by connecting your smartphone.\n        Click below to be guided through the process.\n      ");
        }
        else {
            this.welcomeMessage = gettext("\n        The Cockpit application provides you with options to manage\n        and monitor Internet of Things assets and data from business perspective.\n      ");
        }
    };
    WelcomeToCockpit.prototype.createQuicklinkAddGroup = function () {
        // comes from angularJS factory c8yQuickLinks
        var addGroup = this.links.find(function (link) { return link.label === 'Add group'; });
        if (addGroup) {
            this.quickLinks.push(addGroup);
        }
    };
    WelcomeToCockpit.prototype.createConnectSmartphoneQuickLink = function () {
        // Provider in SensorPhoneModule defines the
        // 'Connect smartphone' quicklink.
        var connectSmartphone = this.links.find(function (link) { return link.label === 'Connect smartphone'; });
        if (connectSmartphone) {
            this.quickLinks.push(connectSmartphone);
        }
    };
    WelcomeToCockpit.prototype.createQuickLinkReports = function () {
        var _this = this;
        var label = gettext('Reports');
        var reports = {
            icon: 'th',
            label: label,
            url: '/reports'
        };
        var reportsNode = this.findNavigatorNode(label, this.navNodes);
        if (reportsNode) {
            reports.click = function () {
                reportsNode.open = true;
                _this.router.navigateByUrl(reports.url);
            };
            this.quickLinks.push(reports);
        }
    };
    WelcomeToCockpit.prototype.createQuickLinkExports = function () {
        var _this = this;
        var label = gettext('Exports');
        var exports = {
            icon: 'c8y-reports',
            label: label,
            url: '/export'
        };
        if (this.isConfigChildNodeShown(label)) {
            exports.click = function () {
                _this.configurationNode.open = true;
                _this.router.navigateByUrl(exports.url);
            };
            this.quickLinks.push(exports);
        }
    };
    WelcomeToCockpit.prototype.createQuicklinkSmartRules = function () {
        var _this = this;
        var label = gettext('Smart rules');
        var smartRules = {
            icon: 'c8y-smart-rules',
            label: label,
            url: '/rules'
        };
        if (this.isConfigChildNodeShown('Global smart rules')) {
            smartRules.click = function () {
                _this.configurationNode.open = true;
                _this.router.navigateByUrl(smartRules.url);
            };
            this.quickLinks.push(smartRules);
        }
    };
    WelcomeToCockpit.prototype.createQuicklinkRegisterDevice = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data, deviceManagement, deviceMgmtUrl_1, registerDevice;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.applicationService.listByUser()];
                    case 1:
                        data = (_a.sent()).data;
                        if (data) {
                            deviceManagement = data.find(function (app) { return app.contextPath === 'devicemanagement'; });
                            if (deviceManagement) {
                                deviceMgmtUrl_1 = this.applicationService.getHref(deviceManagement);
                                registerDevice = {
                                    icon: 'c8y-device-connect',
                                    label: gettext('Register device'),
                                    click: function () { return window.open(deviceMgmtUrl_1 + "/#/deviceregistration", '_self'); }
                                };
                                this.quickLinks.push(registerDevice);
                            }
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    WelcomeToCockpit.prototype.createQuicklinkUserGuide = function () {
        var _this = this;
        var userGuide = {
            icon: 'c8y-user',
            label: gettext('User guide'),
            url: '/users-guide/getting-started',
            click: function () {
                var userGuideURL = _this.docs.getUserGuideLink(userGuide.url);
                window.open(userGuideURL);
            }
        };
        this.quickLinks.push(userGuide);
    };
    WelcomeToCockpit.prototype.isConfigChildNodeShown = function (nodeName) {
        if (this.configurationNode && this.configurationNode.show) {
            var navNode = this.findNavigatorNode(nodeName, this.configurationNode.children);
            return !!navNode && navNode.show;
        }
        return false;
    };
    WelcomeToCockpit.prototype.findNavigatorNode = function (nodeName, navNodes) {
        if (navNodes && navNodes.length > 0) {
            return navNodes.find(function (node) { return node.label === nodeName; });
        }
        return undefined;
    };
    WelcomeToCockpit.ctorParameters = function () { return [
        { type: TenantService },
        { type: DocsService },
        { type: Router },
        { type: NavigatorService },
        { type: ApplicationService }
    ]; };
    __decorate([
        Input()
    ], WelcomeToCockpit.prototype, "config", void 0);
    WelcomeToCockpit = __decorate([
        Component({
            selector: 'c8y-welcome-to-cockpit',
            template: "<div class=\"welcome-widget welcome-cockpit\">\n  <div class=\"flex-row\">\n    <div class=\"col-xs-12 col-md-5 flex-item-v-stretch p-24\">\n      <h2 class=\"text-light\">{{ 'Welcome to Cockpit' | translate }}</h2>\n      <p class=\"text-16 text-light p-t-16 p-b-24\">{{ welcomeMessage | translate }}</p>\n      <div class=\"card-group interact-grid tight-grid\">\n        <div *ngFor=\"let link of quickLinks\" class=\"col-sm-4 col-xs-6\">\n          <c8y-quick-link\n            (click)=\"link.click ? link.click() : false\"\n            [icon]=\"link.icon\"\n            [label]=\"link.label\"\n            class=\"card\"\n            c8yProductExperience\n            [actionName]=\"'welcomeWidgetClicked'\"\n            [actionData]=\"{ link: link.label }\"\n          >\n          </c8y-quick-link>\n        </div>\n      </div>\n    </div>\n    <!-- <div class=\"col-sm-6 welcome-illustration flex-item-v-stretch\"></div> -->\n  </div>\n</div>\n"
        })
    ], WelcomeToCockpit);
    return WelcomeToCockpit;
}());

var LegacyWelcomeComponent = /** @class */ (function () {
    function LegacyWelcomeComponent(dashboard) {
        this.dashboard = dashboard;
    }
    LegacyWelcomeComponent.prototype.restore = function () {
        this.dashboard.restore();
    };
    LegacyWelcomeComponent.ctorParameters = function () { return [
        { type: ContextDashboardComponent, decorators: [{ type: Host }] }
    ]; };
    LegacyWelcomeComponent = __decorate([
        Component({
            selector: 'c8y-legacy-welcome-widget',
            template: "<div class=\"card-block\">\n  <p class=\"lead\" c8y-guide-docs>\n    <span>{{ 'Welcome to Cockpit' | translate }}.</span>\n    <span title=\"{{ 'Browse the manual' | translate }}\" class=\"cockpit-welcome-manual\" translate ngNonBindable\n      >Browse the manual <a c8y-guide-href=\"users-guide/cockpit/\">here</a>.</span\n    >\n  </p>\n  <p class=\"m-t-16\">\n    <em>{{ 'New default dashboard available.' | translate }} </em>\n  </p>\n  <button class=\"btn btn-primary btn-sm m-t-16\" (click)=\"restore(); (false)\" translate>\n    Get new dashboard\n  </button>\n</div>\n"
        }),
        __param(0, Host())
    ], LegacyWelcomeComponent);
    return LegacyWelcomeComponent;
}());

// import for the preview img to get into the build.
var previewImage = '';
try {
    // tslint:disable-next-line: no-var-requires
    previewImage = require('@c8y/style/img/welcome-widget-pr.png');
}
catch (ex) {
    // intended empty
}
var ɵ0 = [{ label: gettext('Home'), icon: 'home', priority: 10000, path: '/' }], ɵ1 = [
    {
        id: 'cockpit.welcome.widget',
        label: gettext('Welcome to Cockpit'),
        description: gettext('Displays a welcome message for Cockpit'),
        component: WelcomeToCockpit,
        previewImage: previewImage,
        data: {
            settings: {
                noNewWidgets: false,
                widgetDefaults: {
                    _width: 12,
                    _height: 6
                },
                ng1: {
                    options: {
                        noDeviceTarget: true,
                        groupsSelectable: false
                    }
                }
            }
        }
    }
], ɵ2 = [
    {
        id: 'Cockpit Welcome',
        label: gettext('Welcome to Cockpit'),
        description: gettext('Displays a welcome message for Cockpit'),
        component: LegacyWelcomeComponent,
        data: {
            settings: {
                noNewWidgets: true,
                ng1: {
                    options: {
                        noDeviceTarget: true,
                        groupsSelectable: false
                    }
                }
            }
        }
    }
];
var CockpitDashboardModule = /** @class */ (function () {
    function CockpitDashboardModule() {
    }
    CockpitDashboardModule = __decorate([
        NgModule({
            entryComponents: [CockpitDashboardComponent, WelcomeToCockpit, LegacyWelcomeComponent],
            imports: [
                DocsModule,
                CoreModule,
                ContextDashboardModule.config({
                    routes: [
                        {
                            path: 'dashboard/:dashboardId',
                            component: ContextDashboardComponent,
                            canActivate: [DeviceDashboardGuard],
                            context: ViewContext.Device
                        },
                        {
                            path: 'dashboard/:dashboardId',
                            component: ContextDashboardComponent,
                            canActivate: [GroupDashboardGuard],
                            context: ViewContext.Group
                        },
                        {
                            path: '',
                            component: CockpitDashboardComponent
                        }
                    ]
                }),
                QuickLinkModule
            ],
            declarations: [CockpitDashboardComponent, WelcomeToCockpit, LegacyWelcomeComponent],
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useValue: ɵ0,
                    multi: true
                },
                {
                    provide: HOOK_COMPONENTS,
                    multi: true,
                    useValue: ɵ1
                },
                {
                    provide: HOOK_COMPONENTS,
                    multi: true,
                    useValue: ɵ2
                }
            ]
        })
    ], CockpitDashboardModule);
    return CockpitDashboardModule;
}());

var ReportDashboardGuard = /** @class */ (function () {
    function ReportDashboardGuard(contextDashboardService) {
        this.contextDashboardService = contextDashboardService;
    }
    ReportDashboardGuard.prototype.canActivate = function (route) {
        var _this = this;
        var reportId = route.data.contextData && route.data.contextData.id;
        return from(this.contextDashboardService.activateDashboards(route, [ContextDashboardType.Named])).pipe(tap(function (res) {
            return Array.isArray(res) && res.length === 0
                ? _this.createDashboardIfNotExist(reportId)
                : of(true);
        }));
    };
    ReportDashboardGuard.prototype.createDashboardIfNotExist = function (reportId) {
        var dashboardCfg = {};
        return this.contextDashboardService.create(dashboardCfg, "report_" + reportId).then(function () { return true; });
    };
    ReportDashboardGuard.ctorParameters = function () { return [
        { type: ContextDashboardService }
    ]; };
    ReportDashboardGuard = __decorate([
        Injectable()
    ], ReportDashboardGuard);
    return ReportDashboardGuard;
}());

var ReportDashboardNavigationFactory = /** @class */ (function () {
    function ReportDashboardNavigationFactory(navigatorService, inventoryService, reportDashboardService) {
        this.navigatorService = navigatorService;
        this.inventoryService = inventoryService;
        this.reportDashboardService = reportDashboardService;
        this.loaded = false;
    }
    ReportDashboardNavigationFactory.prototype.get = function () {
        return __awaiter(this, void 0, void 0, function () {
            var reports;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.loaded) return [3 /*break*/, 2];
                        this.loaded = true;
                        this.navigatorService.add(new NavigatorNode({
                            label: gettext('Reports'),
                            path: 'reports',
                            icon: 'c8y-reports',
                            priority: 5
                        }));
                        return [4 /*yield*/, this.reportDashboardService.listReports({
                                filter: { __has: 'c8y_IsNavigatorNode' },
                                pageSize: 2000
                            })];
                    case 1:
                        reports = (_a.sent()).data;
                        reports.forEach(function (report) {
                            var node = new NavigatorNode({
                                label: report.name,
                                path: "reports/" + report.id,
                                icon: report.icon,
                                priority: report.priority
                            });
                            _this.navigatorService.add(node);
                        });
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    ReportDashboardNavigationFactory.ctorParameters = function () { return [
        { type: NavigatorService },
        { type: InventoryService },
        { type: ReportDashboardService }
    ]; };
    ReportDashboardNavigationFactory = __decorate([
        Injectable()
    ], ReportDashboardNavigationFactory);
    return ReportDashboardNavigationFactory;
}());

var ɵ0$1 = [
    {
        path: 'dashboard/:dashboardId',
        component: ContextDashboardComponent,
        canActivate: [ReportDashboardGuard],
        context: ViewContext.Report
    }
];
var ReportDashboardModule = /** @class */ (function () {
    function ReportDashboardModule() {
    }
    ReportDashboardModule = __decorate([
        NgModule({
            declarations: [ReportDashboardListComponent],
            entryComponents: [],
            imports: [
                CoreModule,
                ContextDashboardModule,
                RouterModule.forChild([
                    {
                        path: 'reports',
                        component: ReportDashboardListComponent
                    }
                ])
            ],
            providers: [
                ReportDashboardGuard,
                ReportDashboardNavigationFactory,
                ReportDashboardService,
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: ɵ0$1,
                    multi: true
                },
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: ReportDashboardNavigationFactory,
                    multi: true
                }
            ]
        })
    ], ReportDashboardModule);
    return ReportDashboardModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { AppearanceSettingsComponent, CONTEXT_DASHBOARD_CONFIG, CockpitDashboardComponent, CockpitDashboardModule, ContextDashboardComponent, ContextDashboardModule, ContextDashboardService, ContextDashboardType, DASHBOARD_THEME_CLASSES, DashboardActionsFactory, DashboardDetailComponent, DeviceDashboardGuard, GroupDashboardGuard, LegacyWelcomeComponent, ReportDashboardGuard, ReportDashboardListComponent, ReportDashboardModule, ReportDashboardNavigationFactory, ReportDashboardService, WIDGET_CONTENT_CLASSES, WIDGET_HEADER_CLASSES, WelcomeToCockpit, WidgetConfigComponent, WidgetPreviewComponent, WidgetService, ɵ0, ɵ1, ɵ2 };
