// @c8y/ngx-components/esm5/sub-assets/add-group/add-group.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { IIdentified, IManagedObject, InventoryService } from '@c8y/client';
import { GroupFragment } from '@c8y/ngx-components/assets-navigator';
var AddGroupService = /** @class */ (function () {
    function AddGroupService(inventoryService) {
        this.inventoryService = inventoryService;
        this.GROUP_FRAGMENT_TYPE = 'c8y_IsDeviceGroup';
    }
    AddGroupService.prototype.createGroupAndAssignDevices = function (groupForm, groupContextId, selectedDevices) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var group, name, description, newGroupMO;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        name = groupForm.name, description = groupForm.description;
                        newGroupMO = this.getGroupMO(name, description, groupContextId);
                        if (!groupContextId) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.inventoryService.childAssetsCreate(newGroupMO, groupContextId)];
                    case 1:
                        group = (_a.sent()).data;
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, this.inventoryService.create(newGroupMO)];
                    case 3:
                        group = (_a.sent()).data;
                        _a.label = 4;
                    case 4:
                        if (!(selectedDevices.length > 0)) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.assignDevices(group.id, selectedDevices)];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6: return [2 /*return*/, group];
                }
            });
        });
    };
    AddGroupService.prototype.getGroupMO = function (name, description, groupContextId) {
        var _a;
        if (description === void 0) { description = ''; }
        var group = (_a = {
                type: this.getGroupType(groupContextId)
            },
            _a[this.GROUP_FRAGMENT_TYPE] = {},
            _a.name = name,
            _a.description = description,
            _a);
        return group;
    };
    AddGroupService.prototype.getGroupType = function (groupContextId) {
        return groupContextId ? GroupFragment.subGroupType : GroupFragment.groupType;
    };
    AddGroupService.prototype.assignDevices = function (id, selectedDevices) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var promises;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promises = [];
                        selectedDevices.forEach(function (moId) {
                            promises.push(_this.inventoryService.childAssetsAdd(moId, id));
                        });
                        return [4 /*yield*/, Promise.all(promises)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AddGroupService.ctorParameters = function () { return [
        { type: InventoryService }
    ]; };
    AddGroupService = tslib_1.__decorate([
        Injectable()
    ], AddGroupService);
    return AddGroupService;
}());
export { AddGroupService };
