// @c8y/ngx-components/esm5/connectivity/abstract-connectivity-tab.factory.js
import { from, isObservable, of } from 'rxjs';
import { map } from 'rxjs/operators';
var AbstractConnectivityTabFactory = /** @class */ (function () {
    function AbstractConnectivityTabFactory(router, guard) {
        this.router = router;
        this.guard = guard;
    }
    AbstractConnectivityTabFactory.prototype.get = function () {
        return this.router.url.match(/connectivitySettings/g) ? this.tab() : [];
    };
    AbstractConnectivityTabFactory.prototype.tab = function () {
        var _this = this;
        return this.wrapIntoObservable(this.guard.canActivate(this.router.routerState.snapshot.root, this.router.routerState.snapshot)).pipe(map(function (canActivate) { return (canActivate ? _this.config() : null); }));
    };
    AbstractConnectivityTabFactory.prototype.isPromise = function (obj) {
        return !!obj && typeof obj.then === 'function';
    };
    AbstractConnectivityTabFactory.prototype.wrapIntoObservable = function (value) {
        if (isObservable(value)) {
            return value;
        }
        if (this.isPromise(value)) {
            return from(value);
        }
        return of(value);
    };
    return AbstractConnectivityTabFactory;
}());
export { AbstractConnectivityTabFactory };
