// @c8y/ngx-components/esm5/device-profile/device-tab-profile/device-tab-profile-detail.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var DeviceTabProfileDetailComponent = /** @class */ (function () {
    function DeviceTabProfileDetailComponent() {
        this.emptyStateText = '';
        this.emptyStateDetails = '';
        this.showTextLabel = true;
    }
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "sectionTitle", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "sectionIcon", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "emptyStateText", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "emptyStateDetails", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "isProfileSelected", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "isEmpty", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceTabProfileDetailComponent.prototype, "showTextLabel", void 0);
    DeviceTabProfileDetailComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-tab-profile-detail',
            template: "<div class=\"card--grid grid__col--6-6--md\">\n  <div class=\"bg-white card-block sticky-top\">\n    <h5 class=\"legend form-block\">{{ sectionTitle | translate }}</h5>\n  </div>\n  <div class=\"bg-gray-white  card-block sticky-top hidden-xs hidden-sm\">\n    <h5 class=\"legend form-block\">{{ sectionTitle | translate }}</h5>\n  </div>\n  <div class=\"bg-white p-l-16 p-r-16\">\n    <hr class=\"m-0\" />\n  </div>\n  <div class=\"bg-gray-white p-l-16 p-r-16\">\n    <hr class=\"m-0\" />\n  </div>\n  <div class=\"d-contents\" *ngIf=\"!isProfileSelected || isEmpty\">\n    <div class=\"p-l-16 p-r-16\">\n      <div class=\"c8y-empty-state text-left\">\n        <h1 [c8yIcon]=\"sectionIcon\" class=\"c8y-icon-duocolor\"></h1>\n        <p>\n          <span>{{ emptyStateText | translate }}</span\n          ><br />\n          <small *ngIf=\"isProfileSelected; else noItems\">\n            {{ emptyStateDetails | translate }}\n          </small>\n          <ng-template #noItems>\n            <small translate>No device profile selected</small>\n          </ng-template>\n        </p>\n      </div>\n    </div>\n  </div>\n  <div class=\"bg-gray-white\" *ngIf=\"items.length === 0\"></div>\n  <c8y-device-profile-item-list\n    *ngIf=\"items.length > 0\"\n    [icon]=\"sectionIcon\"\n    [comparisonResults]=\"items\"\n    [showProfileItems]=\"isProfileSelected && !isEmpty\"\n    [showTextLabel]=\"showTextLabel\"\n    class=\"d-contents\"\n  ></c8y-device-profile-item-list>\n  <div class=\"bg-white p-t-24\" *ngIf=\"isProfileSelected && !isEmpty\"></div>\n  <div class=\"bg-gray-white p-t-24\" *ngIf=\"isProfileSelected && !isEmpty\"></div>\n</div>\n"
        })
    ], DeviceTabProfileDetailComponent);
    return DeviceTabProfileDetailComponent;
}());
export { DeviceTabProfileDetailComponent };
