// @c8y/ngx-components/esm5/core/provider-configuration/provider-configuration.module.js
import * as tslib_1 from "tslib";
import { forwardRef, NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { CoreModule } from '../core.module';
import { DynamicFormsModule } from '../dynamic-forms/dynamic-forms.module';
import { HOOK_NAVIGATOR_NODES } from '../navigator';
import { HOOK_ONCE_ROUTE } from '../router';
import { HOOK_TABS } from '../tabs';
import { HOOK_DYNAMIC_PROVIDER_CONFIG } from './provider-configuration-hook';
import { ProviderConfigurationNodeFactory } from './provider-configuration-node.factory';
import { ProviderConfigurationRouteFactory } from './provider-configuration-route.factory';
import { ProviderConfigurationTabFactory } from './provider-configuration-tab.factory';
import { ProviderConfigurationComponent } from './provider-configuration.component';
var ProviderConfigurationModule = /** @class */ (function () {
    function ProviderConfigurationModule() {
    }
    ProviderConfigurationModule_1 = ProviderConfigurationModule;
    ProviderConfigurationModule.config = function (config) {
        return {
            ngModule: ProviderConfigurationModule_1,
            providers: [{ provide: HOOK_DYNAMIC_PROVIDER_CONFIG, useValue: config, multi: true }]
        };
    };
    var ProviderConfigurationModule_1;
    ProviderConfigurationModule = ProviderConfigurationModule_1 = tslib_1.__decorate([
        NgModule({
            declarations: [ProviderConfigurationComponent],
            exports: [ProviderConfigurationComponent],
            imports: [CoreModule, CommonModule, DynamicFormsModule],
            entryComponents: [ProviderConfigurationComponent],
            providers: [
                {
                    provide: HOOK_ONCE_ROUTE,
                    useClass: forwardRef(function () { return ProviderConfigurationRouteFactory; }),
                    multi: true
                },
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: forwardRef(function () { return ProviderConfigurationNodeFactory; }),
                    multi: true
                },
                {
                    provide: HOOK_TABS,
                    useClass: forwardRef(function () { return ProviderConfigurationTabFactory; }),
                    multi: true
                }
            ]
        })
    ], ProviderConfigurationModule);
    return ProviderConfigurationModule;
}());
export { ProviderConfigurationModule };
