// @c8y/ngx-components/esm2015/core/header/title/title.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild, Input } from '@angular/core';
import { HeaderService } from '../header.service';
let TitleComponent = class TitleComponent {
    constructor(headerService) {
        this.headerService = headerService;
        this.pageTitleUpdate = true;
    }
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.headerService.changeTitle(this.template, this.pageTitleUpdate));
    }
    ngOnDestroy() {
        if (this.headerService.state.title === this.template) {
            clearTimeout(this.viewInitTimeout);
            this.headerService.changeTitle();
        }
    }
};
TitleComponent.ctorParameters = () => [
    { type: HeaderService }
];
tslib_1.__decorate([
    Input()
], TitleComponent.prototype, "pageTitleUpdate", void 0);
tslib_1.__decorate([
    ViewChild('template', { static: true })
], TitleComponent.prototype, "template", void 0);
TitleComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-title',
        template: "<ng-template #template>\n  <h1 class=\"text-truncate\">\n    <ng-content></ng-content>\n  </h1>\n</ng-template>\n\n"
    })
], TitleComponent);
export { TitleComponent };
