// @c8y/ngx-components/esm5/device-profile/device-profile-list.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { IManagedObject, InventoryService, IResultList } from '@c8y/client';
import { TranslateService } from '@ngx-translate/core';
import { AlertService, gettext, ModalService, Status } from '@c8y/ngx-components';
import { cloneDeep } from 'lodash-es';
import { BsModalService } from 'ngx-bootstrap/modal';
import { AddDeviceProfileComponent } from './add-device-profile.component';
import { DeviceProfileService } from './device-profile.service';
import { Router } from '@angular/router';
import { BehaviorSubject } from 'rxjs';
import { tap, switchMap } from 'rxjs/operators';
var DeviceProfileListComponent = /** @class */ (function () {
    function DeviceProfileListComponent(inventoryService, translateService, modalService, alertService, bsModalService, router, deviceProfileService) {
        var _this = this;
        this.inventoryService = inventoryService;
        this.translateService = translateService;
        this.modalService = modalService;
        this.alertService = alertService;
        this.bsModalService = bsModalService;
        this.router = router;
        this.deviceProfileService = deviceProfileService;
        this.reloading = false;
        this.reload = new BehaviorSubject(null);
        this.deviceProfiles$ = this.reload.pipe(tap(function () { return (_this.reloading = true); }), switchMap(function () { return _this.deviceProfileService.getDeviceProfiles(); }), tap(function () { return (_this.reloading = false); }));
    }
    DeviceProfileListComponent.prototype.ngOnInit = function () {
        this.loadDeviceProfiles();
    };
    DeviceProfileListComponent.prototype.loadDeviceProfiles = function () {
        this.reload.next();
    };
    DeviceProfileListComponent.prototype.createDeviceProfile = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var modal, profileId, ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        modal = this.bsModalService.show(AddDeviceProfileComponent, {
                            class: 'modal-sm'
                        }).content;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, modal.result];
                    case 2:
                        profileId = _a.sent();
                        modal.close();
                        this.router.navigateByUrl("/device-profiles/" + profileId);
                        return [3 /*break*/, 4];
                    case 3:
                        ex_1 = _a.sent();
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    DeviceProfileListComponent.prototype.duplicateDeviceProfile = function (deviceProfile) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var copy, mo;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        copy = cloneDeep(deviceProfile);
                        copy.id = null;
                        copy.name = 'Duplicate of ' + deviceProfile.name;
                        return [4 /*yield*/, this.deviceProfileService.createDeviceProfile(copy)];
                    case 1:
                        mo = (_a.sent()).data;
                        this.router.navigateByUrl("/device-profiles/" + mo.id);
                        return [2 /*return*/];
                }
            });
        });
    };
    DeviceProfileListComponent.prototype.deleteDeviceProfile = function (deviceProfile) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var deviceProfileName, title, confirmationText, finalQuestion, ex_2;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        deviceProfileName = deviceProfile.name;
                        title = gettext('Delete device profile');
                        confirmationText = this.translateService.instant(gettext('You are about to delete a device profile "{{ deviceProfileName }}".'), { deviceProfileName: deviceProfileName });
                        finalQuestion = this.translateService.instant(gettext('Do you want to proceed?'));
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 4, , 5]);
                        return [4 /*yield*/, this.modalService.confirm(title, confirmationText + " " + finalQuestion, Status.DANGER, {
                                ok: gettext('Delete')
                            })];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, this.delete(deviceProfile.id)];
                    case 3:
                        _a.sent();
                        this.reload.next();
                        return [3 /*break*/, 5];
                    case 4:
                        ex_2 = _a.sent();
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    DeviceProfileListComponent.prototype.delete = function (profileId) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var ex_3;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.inventoryService.delete(profileId)];
                    case 1:
                        _a.sent();
                        this.alertService.success(gettext('Device profile deleted.'));
                        return [3 /*break*/, 3];
                    case 2:
                        ex_3 = _a.sent();
                        this.alertService.addServerFailure(ex_3);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    DeviceProfileListComponent.ctorParameters = function () { return [
        { type: InventoryService },
        { type: TranslateService },
        { type: ModalService },
        { type: AlertService },
        { type: BsModalService },
        { type: Router },
        { type: DeviceProfileService }
    ]; };
    DeviceProfileListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-profile-list',
            template: "<c8y-title>{{ 'Device profiles' | translate }}</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link\"\n    (click)=\"createDeviceProfile()\"\n    title=\"{{ 'Add device profile' | translate }}\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add device profile' | translate }}\n  </button>\n</c8y-action-bar-item>\n<c8y-action-bar-item\n  [placement]=\"'right'\">\n  <button class=\"btn btn-link\"\n    (click)=\"loadDeviceProfiles()\"\n    title=\"{{ 'Reload' | translate }}\">\n    <i\n      c8yIcon=\"refresh\"\n      [ngClass]=\"{ 'icon-spin': reloading }\"\n    ></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<!-- EMPTY STATE -->\n<div class=\"c8y-empty-state m-t-40\"\n  *ngIf=\"(deviceProfiles$ | async)?.data.length === 0\"\n>\n  <h1 class=\"c8y-icon-duocolor\"\n    [c8yIcon]=\"'c8y-device-profile'\"\n  ></h1>\n  <h3 translate>\n    No device profiles available.\n  </h3>\n  <p translate>\n    Add a new device profile by clicking below.\n  </p>\n  <p>\n    <button\n      class=\"btn btn-primary\"\n      title=\"{{ 'Add device profile' | translate }}\"\n      (click)=\"createDeviceProfile()\"\n    >\n      {{ 'Add device profile' | translate }}\n    </button>\n  </p>\n</div>\n\n<!-- DEVICE PROFILES LIST -->\n<c8y-list-group\n  class=\"m-b-24\"\n  [ngClass]=\"{ 'dd-low': (deviceProfiles$ | async)?.data.length < 10 }\"\n>\n  <c8y-li\n    class=\"page-sticky-header c8y-list__item--empty-actions hidden-xs\"\n    *ngIf=\"(deviceProfiles$ | async)?.data.length > 0\"\n  >\n    <c8y-li-icon>\n      <i class=\"fa\"></i>\n    </c8y-li-icon>\n    <c8y-li-body class=\"content-flex-50\">\n      <div class=\"col-6\">\n        {{ 'Device profile' | translate }}\n      </div>\n      <div class=\"col-6\">\n        {{ 'Device type' | translate }}\n      </div>\n    </c8y-li-body>\n  </c8y-li>\n  <c8y-li *c8yFor=\"let deviceProfile of deviceProfiles$\">\n    <c8y-li-icon>\n      <i c8yIcon=\"c8y-device-profile\"></i>\n    </c8y-li-icon>\n\n    <c8y-li-body class=\"content-flex-50\">\n      <div class=\"col-6\" title=\"{{ deviceProfile.name }}\">\n        <button\n          class=\"btn-clean max-width-inherit\"\n          routerLink=\"/device-profiles/{{ deviceProfile.id }}\"\n        >\n          <span class=\"text-truncate\">\n            {{ deviceProfile.name }}\n          </span>\n        </button>\n      </div>\n      <div class=\"col-6\" title=\"{{ deviceProfile.c8y_Filter?.type || '-' }}\">\n        <span class=\"text-label-small m-r-4 visible-xs\" translate>\n          Device type\n        </span>\n        <span class=\"text-truncate\" *ngIf=\"deviceProfile.c8y_Filter?.type; else emptyText\">\n          {{ deviceProfile.c8y_Filter?.type }}\n        </span>\n      </div>\n      <ng-template #emptyText>\n        <small class=\"text-muted\">\n          <em>\n            {{ 'Undefined' | translate }}\n          </em>\n        </small>\n      </ng-template>\n    </c8y-li-body>\n\n    <c8y-li-action\n      routerLink=\"/device-profiles/{{ deviceProfile.id }}\"\n      icon=\"pencil\"\n      label=\"{{ 'Edit' | translate }}\"\n    >\n    </c8y-li-action>\n    <c8y-li-action\n      (click)=\"duplicateDeviceProfile(deviceProfile)\"\n      icon=\"copy\"\n      label=\"{{ 'Duplicate' | translate }}\"\n    >\n    </c8y-li-action>\n    <c8y-li-action\n      (click)=\"deleteDeviceProfile(deviceProfile)\"\n      icon=\"trash\"\n      label=\"{{ 'Delete' | translate }}\"\n    >\n    </c8y-li-action>\n  </c8y-li>\n</c8y-list-group>\n"
        })
    ], DeviceProfileListComponent);
    return DeviceProfileListComponent;
}());
export { DeviceProfileListComponent };
