// @c8y/ngx-components/esm2015/core/dynamic-forms/form-field/form-field.wrapper.component.js
var WrapperFormField_1;
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldWrapper } from '@ngx-formly/core';
let WrapperFormField = WrapperFormField_1 = class WrapperFormField extends FieldWrapper {
    constructor() {
        super(...arguments);
        this.maxHelpBlockLength = 64;
    }
    ngOnInit() {
        this.showDescriptionAsPopup =
            this.field.type === 'radio' ||
                (this.to.description && this.to.description.length > this.maxHelpBlockLength);
    }
};
WrapperFormField.CONFIG = {
    wrappers: [{ name: 'c8y-form-field', component: WrapperFormField_1 }]
};
WrapperFormField = WrapperFormField_1 = tslib_1.__decorate([
    Component({
        selector: 'c8y-wrapper-form-field',
        template: "<div class=\"form-group\" [class.has-error]=\"showError\">\n  <label *ngIf=\"to.label && to.hideLabel !== true\" [attr.for]=\"id\">\n    {{ to.label }}\n    <button\n      *ngIf=\"to.description && showDescriptionAsPopup\"\n      class=\"btn-clean m-l-4 flex-item-middle\"\n      [popover]=\"to.description\"\n      placement=\"auto\"\n      outsideClick=\"true\"\n      type=\"button\"\n    >\n      <i c8yIcon=\"question-circle-o\" class=\"text-primary\"></i>\n    </button>\n  </label>\n\n  <ng-template #fieldComponent></ng-template>\n\n  <div *ngIf=\"showError\" class=\"c8y-messages\">\n    <formly-validation-message\n      class=\"form-control-feedback-message\"\n      [field]=\"field\"\n    ></formly-validation-message>\n  </div>\n\n  <div class=\"help-block\" *ngIf=\"!showError && to.description && !showDescriptionAsPopup\">\n    {{ to.description }}\n  </div>\n</div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], WrapperFormField);
export { WrapperFormField };
