// @c8y/ngx-components/esm2015/ecosystem/ecosystem-navigation-factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { NavigatorNode, gettext } from '@c8y/ngx-components';
import { EcosystemGuard } from './ecosystem.guard';
let EcosystemNavigationFactory = class EcosystemNavigationFactory {
    constructor(ecosystemGuard) {
        this.ecosystemGuard = ecosystemGuard;
        this.navs = [];
    }
    get() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // Check for the beta flag. To be removed when the feature is production ready.
            if (!this.ecosystemGuard.isFeatureEnabled()) {
                return [];
            }
            if (this.navs.length === 0) {
                const applicationsNode = new NavigatorNode({
                    label: gettext('Applications'),
                    icon: 'c8y-modules',
                    path: '/application-list'
                });
                const microservicesNode = new NavigatorNode({
                    label: gettext('Microservices'),
                    icon: 'microchip',
                    path: '/microservices'
                });
                this.navs.push(new NavigatorNode({
                    label: gettext('Ecosystem'),
                    icon: 'c8y-atom',
                    priority: 3200,
                    children: [applicationsNode, microservicesNode]
                }));
            }
            return this.navs;
        });
    }
};
EcosystemNavigationFactory.ctorParameters = () => [
    { type: EcosystemGuard }
];
EcosystemNavigationFactory = tslib_1.__decorate([
    Injectable()
], EcosystemNavigationFactory);
export { EcosystemNavigationFactory };
