// @c8y/ngx-components/esm5/sensor-phone/sensor-phone.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { SensorPhoneModalComponent } from './sensor-phone-modal.component';
import { AppStateService } from '@c8y/ngx-components';
import { InventoryService } from '@c8y/client';
import { head } from 'lodash-es';
import { gettext } from '@c8y/ngx-components';
var SensorPhoneService = /** @class */ (function () {
    function SensorPhoneService(bsModalService, ui, inventoryService) {
        var _this = this;
        this.bsModalService = bsModalService;
        this.ui = ui;
        this.inventoryService = inventoryService;
        this.HTTP_STATUS_CODE = {
            OK: 200,
            CREATED: 201
        };
        this.PHONE_GROUP_FRAGMENT = 'c8y_IsSensorPhoneGroup';
        this.PHONE_GROUP_NAME = gettext('Phones');
        this.ui.currentTenant.subscribe(function (data) {
            if (data) {
                _this.domainName = data.domainName;
                _this.tenantId = data.name;
            }
        });
    }
    SensorPhoneService.prototype.createModal = function () {
        this.bsModalService.show(SensorPhoneModalComponent, {
            backdrop: 'static',
            class: 'lg'
        });
    };
    SensorPhoneService.prototype.getHost = function () {
        return this.domainName.substring(this.domainName.indexOf('.') + 1);
    };
    SensorPhoneService.prototype.getTenantName = function () {
        return this.domainName.split('.')[0];
    };
    SensorPhoneService.prototype.getTenantId = function () {
        return this.tenantId;
    };
    SensorPhoneService.prototype.checkPhoneGroupExists = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, data, res;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.inventoryService.list({
                            fragmentType: this.PHONE_GROUP_FRAGMENT
                        })];
                    case 1:
                        _a = _b.sent(), data = _a.data, res = _a.res;
                        if (data && res && res.status === this.HTTP_STATUS_CODE.OK) {
                            return [2 /*return*/, head(data)];
                        }
                        return [2 /*return*/, null];
                }
            });
        });
    };
    SensorPhoneService.prototype.addPhoneGroup = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var group;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkPhoneGroupExists()];
                    case 1:
                        group = _a.sent();
                        if (group) {
                            return [2 /*return*/, group];
                        }
                        return [4 /*yield*/, this.createPhoneGroup()];
                    case 2: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    SensorPhoneService.prototype.createPhoneGroup = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var newPhoneGroup, _a, data, res;
            var _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        newPhoneGroup = (_b = {
                                c8y_Global: {},
                                c8y_IsDeviceGroup: {},
                                type: 'c8y_DeviceGroup',
                                name: this.PHONE_GROUP_NAME
                            },
                            _b[this.PHONE_GROUP_FRAGMENT] = {},
                            _b);
                        return [4 /*yield*/, this.inventoryService.create(newPhoneGroup)];
                    case 1:
                        _a = _c.sent(), data = _a.data, res = _a.res;
                        if (data && res && res.status === this.HTTP_STATUS_CODE.CREATED) {
                            return [2 /*return*/, data];
                        }
                        return [2 /*return*/, null];
                }
            });
        });
    };
    SensorPhoneService.prototype.generateRegistrationData = function (deviceId) {
        var res = 'c8y://' + this.getTenantName() + '.' + this.getHost() + '/?deviceId=' + deviceId;
        return res;
    };
    SensorPhoneService.ctorParameters = function () { return [
        { type: BsModalService },
        { type: AppStateService },
        { type: InventoryService }
    ]; };
    SensorPhoneService = tslib_1.__decorate([
        Injectable()
    ], SensorPhoneService);
    return SensorPhoneService;
}());
export { SensorPhoneService };
