// @c8y/ngx-components/esm5/search/search.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { SubAssetsService } from '@c8y/ngx-components/sub-assets';
import { find, get, isUndefined, some } from 'lodash-es';
var SearchService = /** @class */ (function (_super) {
    tslib_1.__extends(SearchService, _super);
    function SearchService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.GRID_CONFIG_STORAGE_KEY = 'search-grid-config';
        _this.DEFAULT_PAGE_SIZE = 50;
        _this.FRAGMENTS_FOR_NO_DEVICE = [
            'c8y_Dashboard',
            'c8y_Report',
            'c8y_Kpi',
            'c8y_ExportConfiguration',
            'c8y_IsBinary',
            'c8y_NoDevice',
            'c8y_IsDeviceGroup',
            'c8y_Group',
            'com_cumulocity_model_smartrest_SmartRestTemplate',
            'com_cumulocity_model_devicesimulator_SensorTemplate',
            '_attachments',
            'c8y_IsDeviceType',
            'c8y_objectmapping_ObjectMapping'
        ];
        _this.TYPES_FOR_NO_DEVICE = [
            'c8y_ConfigurationDump',
            'c8y_Firmware',
            'c8y_SmartRule',
            'c8y_Software',
            'impact_object_mapping',
            'c8y_UserPreference',
            'c8y_TenantPolicy',
            'c8y_PrivateSmartRule',
            'c8y_SmartRest2Template',
            'c8y_JsonSchema',
            'c8y_DeviceShellTemplate',
            'c8y_DemoStatus',
            'c8y_DataBroker',
            'c8y_Application_',
            'brandingVariables',
            'c8y_DeviceSimulator',
            'c8y_CertificateMetadata',
            'lwm2m_post_registration',
            'c8y_microservice_manifest_',
            'c8y_CepAgent'
        ];
        _this.DEVICE_FRAGMENT_TYPE = 'c8y_IsDevice';
        return _this;
    }
    /**
     * Will return only valid assets (groups and devices) and filter out
     * none useful inventories (e.g. c8y_JsonSchema).
     * @param data All managed objects that should be filtered.
     */
    SearchService.prototype.filterOnlyAssets = function (data) {
        var _this = this;
        return data.filter(function (mo) { return mo.c8y_IsDeviceGroup || _this.isAnyDevice(mo); });
    };
    SearchService.prototype.buildCombinedRootQueryFilter = function (columns, pagination) {
        var rootQuery = {
            __filter: {
                __and: { __not: { __has: "c8y_IsBinary" } }
            }
        };
        var userQuery = this.getQueryObj(columns, pagination);
        var queryPart = this.queriesUtil.addOrderbys(rootQuery, userQuery.__orderby, 'append');
        var fullQuery = this.queriesUtil.addAndFilter(queryPart, userQuery.__filter);
        return this.queriesUtil.buildQuery(fullQuery);
    };
    /**
     * Returns the full-text search results.
     *
     * @param term The search term.
     */
    SearchService.prototype.search = function (term) {
        return this.inventoryService.list({
            text: term,
            withTotalPages: true,
            pageSize: this.DEFAULT_PAGE_SIZE,
            withChildren: false
        });
    };
    SearchService.prototype.isRootDevice = function (mo) {
        return !!mo[this.DEVICE_FRAGMENT_TYPE];
    };
    SearchService.prototype.isAnyDevice = function (mo) {
        var isDevice = this.isRootDevice(mo) || !this.hasFragmentOrTypeFromBlacklist(mo);
        return isDevice;
    };
    SearchService.prototype.hasFragmentOrTypeFromBlacklist = function (mo) {
        return this.hasTypeFromBlacklist(mo) || this.hasFragmentFromBlacklist(mo);
    };
    SearchService.prototype.hasTypeFromBlacklist = function (mo) {
        var moType = get(mo, 'type', '');
        return some(this.TYPES_FOR_NO_DEVICE, function (type) { return moType.includes(type); });
    };
    SearchService.prototype.hasFragmentFromBlacklist = function (mo) {
        return find(this.FRAGMENTS_FOR_NO_DEVICE, function (f) { return !isUndefined(mo[f]); });
    };
    SearchService = tslib_1.__decorate([
        Injectable()
    ], SearchService);
    return SearchService;
}(SubAssetsService));
export { SearchService };
