// @c8y/ngx-components/esm2015/core/list-group/list-item-checkbox.component.js
var ListItemCheckboxComponent_1;
import * as tslib_1 from "tslib";
import { Output, EventEmitter, Component, Input, forwardRef } from '@angular/core';
import { CheckboxControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * A component to display checkboxes in a c8y-li. Use it with content projection. It
 * can be used standalone or with an ngModel.
 *
 * @example
 * ```html
 * <c8y-li>
 *   <c8y-li-checkbox></c8y-li-checkbox>
 * </c8y-li>
 * ```
 */
let ListItemCheckboxComponent = ListItemCheckboxComponent_1 = class ListItemCheckboxComponent extends CheckboxControlValueAccessor {
    /**
     * A component to display checkboxes in a c8y-li. Use it with content projection. It
     * can be used standalone or with an ngModel.
     *
     * @example
     * ```html
     * <c8y-li>
     *   <c8y-li-checkbox></c8y-li-checkbox>
     * </c8y-li>
     * ```
     */
    constructor() {
        super(...arguments);
        /**
         * Indicates if the checkbox is selected.
         */
        this.selected = false;
        /**
         * Indicates if the checkbox is disabled.
         */
        this.disabled = false;
        /**
         * An event emitted as soon as the checkbox is selected.
         */
        this.onSelect = new EventEmitter();
    }
    /**
     * @ignore
     */
    ngOnChanges(changes) {
        if (changes.selected && !changes.selected.firstChange) {
            this.onSelect.emit(changes.selected.currentValue);
        }
    }
    /**
     * Selects the checkbox.
     * @param value The value to pass.
     */
    select(value) {
        this.onSelect.emit(value);
        this.onChange(value);
    }
};
tslib_1.__decorate([
    Input()
], ListItemCheckboxComponent.prototype, "selected", void 0);
tslib_1.__decorate([
    Input()
], ListItemCheckboxComponent.prototype, "disabled", void 0);
tslib_1.__decorate([
    Output()
], ListItemCheckboxComponent.prototype, "onSelect", void 0);
ListItemCheckboxComponent = ListItemCheckboxComponent_1 = tslib_1.__decorate([
    Component({
        selector: 'c8y-list-item-checkbox, c8y-li-checkbox',
        template: "<label class=\"c8y-checkbox\">\n  <input\n    type=\"checkbox\"\n    class=\"form-control\"\n    [disabled]=\"disabled\"\n    [checked]=\"selected\"\n    (change)=\"select($event.target.checked); (false)\"\n  />\n  <span></span>\n</label>\n<ng-content></ng-content>\n",
        host: {
            class: 'c8y-list__item__check'
        },
        providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                multi: true,
                useExisting: forwardRef(() => ListItemCheckboxComponent_1)
            }
        ]
    })
], ListItemCheckboxComponent);
export { ListItemCheckboxComponent };
