// @c8y/ngx-components/esm2015/core/action-bar/action-bar.component.js
import * as tslib_1 from "tslib";
import { Component, HostBinding, Input } from '@angular/core';
import { filter, sortBy } from 'lodash-es';
/**
 * This component is used as the outlet to show the action bars.
 * In a c8ycli app it is by default placed on the bootstrap component.
 */
let ActionBarComponent = class ActionBarComponent {
    /**
     * This component is used as the outlet to show the action bars.
     * In a c8ycli app it is by default placed on the bootstrap component.
     */
    constructor() {
        /**
         * Identifies if the navigator is opened. If yes, the action-bar
         * needs to move to the right.
         */
        this.navigatorOpen = false;
        /**
         * Identifies if the current view has tabs. If yes, the action bar needs
         * to move down.
         */
        this.hasTabs = false;
        /**
         * Identifies if the tabs are aligned horizontally. If yes, the tabs don't
         * need to move to the left.
         */
        this.isTabsHorizontal = false;
        /**
         * Identifies if in a mobile view the toolbar is expanded or not.
         */
        this.isPageToolbarExpanded = false;
        /**
         * Adds the default `c8y-ui-action-bar` class.
         */
        this.uiActionBar = true;
        /**
         * The current items to display in the ActionBar.
         */
        this.items = [];
    }
    /**
     * Returns true if no items are present and hides the action bar.
     */
    get hidden() {
        return !(this.items && this.items.length);
    }
    /**
     * Returns the items sorted by priority and grouped by the placement passed.
     * @param {placement} Which placement should be returned sorted.
     */
    getSortedItems(placement) {
        return sortBy(filter(this.items, { placement }), this.byPriority);
    }
    byPriority(item) {
        return -item.priority;
    }
};
tslib_1.__decorate([
    Input(), HostBinding('class.navigator-open')
], ActionBarComponent.prototype, "navigatorOpen", void 0);
tslib_1.__decorate([
    Input(), HostBinding('class.has-tabs')
], ActionBarComponent.prototype, "hasTabs", void 0);
tslib_1.__decorate([
    Input(), HostBinding('class.horizontal-tabs')
], ActionBarComponent.prototype, "isTabsHorizontal", void 0);
tslib_1.__decorate([
    HostBinding('class.c8y-ui-action-bar')
], ActionBarComponent.prototype, "uiActionBar", void 0);
tslib_1.__decorate([
    Input()
], ActionBarComponent.prototype, "items", void 0);
tslib_1.__decorate([
    HostBinding()
], ActionBarComponent.prototype, "hidden", null);
ActionBarComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-action-bar',
        template: "<div class=\"navbar-header\">\n  <button\n    type=\"button\"\n    class=\"navbar-toggle collapsed\"\n    (click)=\"isPageToolbarExpanded = !isPageToolbarExpanded\"\n  >\n    <span class=\"sr-only\" translate>Toggle toolbar</span>\n    <span class=\"icon-bar\"></span>\n    <span class=\"icon-bar\"></span>\n    <span class=\"icon-bar\"></span>\n  </button>\n</div>\n\n<div id=\"page-toolbar\" class=\"navbar-collapse\" [collapse]=\"!isPageToolbarExpanded\">\n  <ul class=\"nav navbar-nav navbar-left\">\n    <ng-container *ngFor=\"let item of getSortedItems('left')\">\n      <ng-container *c8yOutlet=\"item.template\"></ng-container>\n    </ng-container>\n  </ul>\n  <ul class=\"nav navbar-nav navbar-right\">\n    <ng-container *ngFor=\"let item of getSortedItems('right')\">\n      <ng-container *c8yOutlet=\"item.template\"></ng-container>\n    </ng-container>\n    <li class=\"no-remove more-wrapper\" *ngIf=\"getSortedItems('more').length > 0\">\n      <div class=\"dropdown\" dropdown>\n        <button\n          title=\"{{ 'More' | translate }}\u2026\"\n          class=\"dropdown-toggle c8y-dropdown d-flex\"\n          dropdownToggle\n          aria-haspopup=\"true\"\n          aria-expanded=\"false\"\n        >\n          <span class=\"text-truncate\" translate>More\u2026</span>\n          <i [c8yIcon]=\"'caret-down'\"></i>\n        </button>\n        <ul class=\"dropdown-menu dropdown-menu-right\" *dropdownMenu>\n          <ng-container *ngFor=\"let item of getSortedItems('more')\">\n            <ng-container *c8yOutlet=\"item.template\"></ng-container>\n          </ng-container>\n        </ul>\n        <ul class=\"dropdown-menu visible-xs\"> \n          <ng-container *ngFor=\"let item of getSortedItems('more')\">\n            <ng-container *c8yOutlet=\"item.template\"></ng-container>\n          </ng-container>\n        </ul>\n      </div>\n    </li>\n  </ul>\n</div>\n",
        host: { class: 'c8y-ui-action-bar' }
    })
], ActionBarComponent);
export { ActionBarComponent };
