// @c8y/ngx-components/esm5/operations/device-selector/device-selector.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { AlarmsDeviceGridColumn, DeviceGridComponent, GroupDeviceGridColumn, ModelDeviceGridColumn, NameDeviceGridColumn, RegistrationDateDeviceGridColumn, StatusDeviceGridColumn, TypeDeviceGridColumn } from '@c8y/ngx-components/device-grid';
import { FilteringActionType } from '@c8y/ngx-components';
import { Subject } from 'rxjs';
import { distinctUntilChanged, takeUntil } from 'rxjs/operators';
var DeviceSelectorComponent = /** @class */ (function () {
    function DeviceSelectorComponent() {
        this.typeColumn = new TypeDeviceGridColumn();
        this.onDeviceQueryStringChange = new EventEmitter();
        this.columns = [
            new StatusDeviceGridColumn(),
            new NameDeviceGridColumn(),
            this.typeColumn,
            new ModelDeviceGridColumn(),
            new GroupDeviceGridColumn(),
            new RegistrationDateDeviceGridColumn(),
            new AlarmsDeviceGridColumn()
        ];
        this.unsubscribe$ = new Subject();
    }
    DeviceSelectorComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (this.deviceTypes) {
            this.deviceTypes
                .pipe(distinctUntilChanged(), takeUntil(this.unsubscribe$))
                .subscribe(function (types) {
                _this.deviceGrid.updateFiltering([_this.typeColumn.name], {
                    type: types && types.length
                        ? FilteringActionType.ApplyFilter
                        : FilteringActionType.ResetFilter,
                    payload: types
                        ? {
                            filteringModifier: {
                                externalFilterQuery: {
                                    types: types
                                }
                            }
                        }
                        : undefined
                });
            });
        }
    };
    DeviceSelectorComponent.prototype.ngOnDestroy = function () {
        this.unsubscribe$.next(true);
        this.unsubscribe$.complete();
    };
    tslib_1.__decorate([
        ViewChild(DeviceGridComponent, { static: true })
    ], DeviceSelectorComponent.prototype, "deviceGrid", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceSelectorComponent.prototype, "deviceTypes", void 0);
    tslib_1.__decorate([
        Output()
    ], DeviceSelectorComponent.prototype, "onDeviceQueryStringChange", void 0);
    DeviceSelectorComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-selector',
            template: "<c8y-device-grid\n  [title]=\"'Select target devices' | translate\"\n  [columns]=\"columns\"\n  [actionControls]=\"[]\"\n  [infiniteScroll]=\"'auto'\"\n  (onDeviceQueryStringChange)=\"onDeviceQueryStringChange.emit($event)\"\n>\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"search\"></h1>\n    <div>\n      <p>\n        <strong>{{ 'No matching devices.' | translate }}</strong>\n      </p>\n      <small>{{ 'Refine your search terms' | translate }}</small>\n    </div>\n  </div>\n</c8y-device-grid>\n"
        })
    ], DeviceSelectorComponent);
    return DeviceSelectorComponent;
}());
export { DeviceSelectorComponent };
