// @c8y/ngx-components/esm5/operations/bulk-operation-stepper/base-stepper.component.js
import * as tslib_1 from "tslib";
import { ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { BulkOperationStepper } from './bulk-operation-stepper.component';
var BaseStepperComponent = /** @class */ (function () {
    function BaseStepperComponent() {
        /**
         * A map holding step data. The order of properties need to match the order of
         * the steps they hold data for as the index of the property is used to clear
         * step data when navigating forth after changing data at an earlier step.
         */
        this.stepData = {};
        this.endSubscriptions = new Subject();
    }
    Object.defineProperty(BaseStepperComponent.prototype, "deviceTypes", {
        set: function (deviceTypes) {
            if (this.operationStepper) {
                this.operationStepper.changeDeviceTypes(deviceTypes);
            }
        },
        enumerable: true,
        configurable: true
    });
    BaseStepperComponent.prototype.ngOnInit = function () {
        this.operationStepper.retrieveOperationDetails = this.retrieveOperationPrototype.bind(this);
        this.operationStepper.selectionChange
            .pipe(takeUntil(this.endSubscriptions))
            .subscribe(this.onSelectionChange.bind(this));
    };
    BaseStepperComponent.prototype.ngOnDestroy = function () {
        this.endSubscriptions.next();
        this.endSubscriptions.complete();
    };
    BaseStepperComponent.prototype.onSelectionChange = function (event) {
        var selectedIndex = event.selectedIndex, previouslySelectedIndex = event.previouslySelectedIndex;
        if (selectedIndex > previouslySelectedIndex &&
            selectedIndex < Object.keys(this.stepData).length) {
            // TODO clear step data only if previous step is "dirty"
            this.stepData[this.getStepDataKeyByIndex(selectedIndex)] = undefined;
        }
    };
    BaseStepperComponent.prototype.getStepDataKeyByIndex = function (index) {
        return Object.keys(this.stepData)[index];
    };
    tslib_1.__decorate([
        ViewChild(BulkOperationStepper, { static: true })
    ], BaseStepperComponent.prototype, "operationStepper", void 0);
    return BaseStepperComponent;
}());
export { BaseStepperComponent };
