// @c8y/ngx-components/esm2015/core/provider-configuration/provider-configuration-tab.factory.js
import * as tslib_1 from "tslib";
import { Inject, Injectable, Injector, Optional } from '@angular/core';
import { CanActivate, Router } from '@angular/router';
import { flatten } from 'lodash-es';
import { forkJoin, from, isObservable, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { HOOK_DYNAMIC_PROVIDER_CONFIG } from './provider-configuration-hook';
let ProviderConfigurationTabFactory = class ProviderConfigurationTabFactory {
    constructor(config, router, injector) {
        this.router = router;
        this.injector = injector;
        this.config = flatten(config);
    }
    get() {
        if (!this.config || !this.config.length) {
            return;
        }
        const configForRoute = this.config.find(c => c.tab
            ? this.router.url === '/' + this.getNodeTabPath(c.navigation.path, c.tab.path) ||
                this.router.url.startsWith('/' + c.navigation.path.replace(/^\/|\/$/g, ''))
            : false);
        const filteredRoutes = configForRoute
            ? this.config.filter(c => c.navigation.path.replace(/^\/|\/$/g, '') ===
                configForRoute.navigation.path.replace(/^\/|\/$/g, '') && c.tab)
            : [];
        const canActivate = filteredRoutes
            .map(c => c.tab.canActivate && c.tab.canActivate.length
            ? c.tab.canActivate.map(ca => this.injector.get(ca))
            : undefined)
            .map(this.checkCanActivate.bind(this));
        return canActivate.length > 0
            ? forkJoin(canActivate).pipe(map((canActivateResult) => filteredRoutes
                .map((c, index) => {
                const tab = Object.assign({}, c.tab, { path: this.getNodeTabPath(c.navigation.path, c.tab.path) });
                return canActivateResult[index] ? tab : undefined;
            })
                .filter(el => !!el)))
            : [];
    }
    checkCanActivate(ca) {
        if (!!ca && ca.length) {
            const canActivateResult = ca
                .map((canActivate) => canActivate.canActivate(undefined, undefined))
                .map(this.wrapIntoObservable.bind(this));
            return forkJoin(canActivateResult).pipe(map((caResult) => caResult.reduce((acc, curr) => acc && curr)));
        }
        return of(true);
    }
    isPromise(obj) {
        return !!obj && typeof obj.then === 'function';
    }
    wrapIntoObservable(value) {
        if (isObservable(value)) {
            return value;
        }
        if (this.isPromise(value)) {
            return from(value);
        }
        return of(value);
    }
    getNodeTabPath(nodePath, tabPath) {
        return `${nodePath.replace(/^\/|\/$/g, '')}/${tabPath.replace(/^\/|\/$/g, '')}`;
    }
};
ProviderConfigurationTabFactory.ctorParameters = () => [
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [HOOK_DYNAMIC_PROVIDER_CONFIG,] }] },
    { type: Router },
    { type: Injector }
];
ProviderConfigurationTabFactory = tslib_1.__decorate([
    Injectable(),
    tslib_1.__param(0, Optional()),
    tslib_1.__param(0, Inject(HOOK_DYNAMIC_PROVIDER_CONFIG))
], ProviderConfigurationTabFactory);
export { ProviderConfigurationTabFactory };
