// @c8y/ngx-components/esm2015/core/login/login.module.js
var LoginModule_1;
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { ChangePasswordComponent } from './change-password.component';
import { CredentialsComponent } from './credentials.component';
import { LoginComponent } from './login.component';
import { LoginService } from './login.service';
import { RecoverPasswordComponent } from './recover-password.component';
import { AuthenticationModule } from '../authentication/authentication.module';
import { AlertModule } from '../alert/alert.module';
import { TotpAuthComponent } from './totp-auth.component';
import { PasswordStrengthValidatorDirective } from './password-strength-validator.directive';
/**
 * The angular module definition for the login.
 * @exports LoginComponent
 */
let LoginModule = LoginModule_1 = class LoginModule {
    static forRoot() {
        return {
            ngModule: LoginModule_1,
            providers: LoginModule_1.providers()
        };
    }
    static providers() {
        return [LoginService];
    }
};
LoginModule = LoginModule_1 = tslib_1.__decorate([
    NgModule({
        declarations: [
            LoginComponent,
            RecoverPasswordComponent,
            ChangePasswordComponent,
            CredentialsComponent,
            TotpAuthComponent,
            PasswordStrengthValidatorDirective
        ],
        imports: [
            FormsModule,
            TooltipModule,
            CommonModule,
            AuthenticationModule,
            AlertModule,
            C8yFormsModule
        ],
        exports: [LoginComponent]
    })
], LoginModule);
export { LoginModule };
