// @c8y/ngx-components/esm2015/core/login/strength-validator-service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { PasswordService } from '../authentication';
import { PasswordStrength } from '@c8y/client';
import * as i0 from "@angular/core";
import * as i1 from "../authentication/password.service";
let StrengthValidatorService = class StrengthValidatorService {
    constructor(passwordService) {
        this.passwordService = passwordService;
    }
    isStrong(password) {
        return this.isPasswordGreen(this.passwordService.getStrengthColor(password).passwordStrength);
    }
    isPasswordGreen(strength) {
        return strength === PasswordStrength.GREEN;
    }
};
StrengthValidatorService.ctorParameters = () => [
    { type: PasswordService }
];
StrengthValidatorService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function StrengthValidatorService_Factory() { return new StrengthValidatorService(i0.ɵɵinject(i1.PasswordService)); }, token: StrengthValidatorService, providedIn: "root" });
StrengthValidatorService = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    })
], StrengthValidatorService);
export { StrengthValidatorService };
