// @c8y/ngx-components/esm5/device-grid/columns/custom.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { gettext } from '@c8y/ngx-components';
var CustomDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(CustomDeviceGridColumn, _super);
    function CustomDeviceGridColumn(initialColumnConfig) {
        var _this = _super.call(this, initialColumnConfig) || this;
        var self = _this;
        _this.name = 'custom';
        _this.header = gettext('Custom');
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    exists: {
                        type: 'boolean',
                        default: false,
                        title: gettext('Only rows where value is defined')
                    },
                    equals: {
                        type: 'array',
                        items: {
                            type: 'string'
                        }
                    }
                }
            },
            form: [
                {
                    key: 'exists'
                },
                {
                    key: 'equals',
                    title: gettext('Only rows where value equals to:'),
                    add: gettext('Or'),
                    items: [
                        {
                            key: 'equals[]',
                            notitle: true,
                            placeholder: gettext('e.g. value, you can use an asterisk (*) as wildcard character')
                        }
                    ]
                }
            ],
            getFilter: function (model) {
                var filter = {};
                if (model.exists) {
                    filter.__has = self.path;
                }
                if (model.equals) {
                    filter[self.path] = { __in: model.equals };
                }
                return filter;
            }
        };
        _this.sortable = true;
        return _this;
    }
    Object.defineProperty(CustomDeviceGridColumn.prototype, "sortingConfig", {
        get: function () {
            return {
                pathSortingConfigs: [{ path: this.path }]
            };
        },
        enumerable: true,
        configurable: true
    });
    return CustomDeviceGridColumn;
}(BaseDeviceGridColumn));
export { CustomDeviceGridColumn };
