// @c8y/ngx-components/esm5/device-grid/columns/imei.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { gettext } from '@c8y/ngx-components';
var ImeiDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(ImeiDeviceGridColumn, _super);
    function ImeiDeviceGridColumn(initialColumnConfig) {
        var _this = _super.call(this, initialColumnConfig) || this;
        var self = _this;
        _this.path = 'c8y_Mobile.imei';
        _this.name = 'imei';
        _this.header = gettext('IMEI');
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    imeis: {
                        type: 'array',
                        items: {
                            type: 'string'
                        }
                    }
                }
            },
            form: [
                {
                    key: 'imeis',
                    title: gettext('Show devices with IMEI'),
                    add: gettext('Or'),
                    items: [
                        {
                            key: 'imeis[]',
                            notitle: true,
                            placeholder: gettext('e.g. 46543432321, you can use an asterisk (*) as wildcard character')
                        }
                    ]
                }
            ],
            getFilter: function (model) {
                var filter = {};
                if (model.imeis.length) {
                    filter[self.path] = { __in: model.imeis };
                }
                return filter;
            }
        };
        _this.sortable = true;
        _this.sortingConfig = {
            pathSortingConfigs: [{ path: _this.path }]
        };
        return _this;
    }
    return ImeiDeviceGridColumn;
}(BaseDeviceGridColumn));
export { ImeiDeviceGridColumn };
