// @c8y/ngx-components/esm5/device-grid/columns/model.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { ModelCellRendererComponent } from './model.cell-renderer.component';
import { gettext } from '@c8y/ngx-components';
var ModelDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(ModelDeviceGridColumn, _super);
    function ModelDeviceGridColumn(initialColumnConfig) {
        var _this = _super.call(this, initialColumnConfig) || this;
        var hardwareModelPath = 'c8y_Hardware.model';
        var vendingModelPath = 'com_nsn_startups_vendme_fragments_VendingMachineTypeInfo.model';
        _this.name = 'model';
        _this.header = gettext('Model');
        _this.cellRendererComponent = ModelCellRendererComponent;
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    models: {
                        type: 'array',
                        items: {
                            type: 'string'
                        }
                    }
                }
            },
            form: [
                {
                    key: 'models',
                    title: gettext('Show devices with model'),
                    add: gettext('Or'),
                    items: [
                        {
                            key: 'models[]',
                            notitle: true,
                            placeholder: gettext('e.g. ModelName, you can use an asterisk (*) as wildcard character')
                        }
                    ]
                }
            ],
            getFilter: function (model) {
                var _a;
                var filter = {};
                if (model.models.length) {
                    filter.__or = (_a = {},
                        _a[hardwareModelPath] = { __in: model.models },
                        _a[vendingModelPath] = { __in: model.models },
                        _a);
                }
                return filter;
            }
        };
        _this.sortable = true;
        _this.sortingConfig = {
            pathSortingConfigs: [{ path: hardwareModelPath }, { path: vendingModelPath }]
        };
        return _this;
    }
    return ModelDeviceGridColumn;
}(BaseDeviceGridColumn));
export { ModelDeviceGridColumn };
