// @c8y/ngx-components/esm5/device-grid/columns/registration-date.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { RegistrationDateCellRendererComponent } from './registration-date.cell-renderer.component';
import { gettext } from '@c8y/ngx-components';
var RegistrationDateDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(RegistrationDateDeviceGridColumn, _super);
    function RegistrationDateDeviceGridColumn(initialColumnConfig) {
        var _this = _super.call(this, initialColumnConfig) || this;
        var self = _this;
        _this.path = 'creationTime';
        _this.name = 'registrationDate';
        _this.header = gettext('Registration date');
        _this.cellRendererComponent = RegistrationDateCellRendererComponent;
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    registrationDate: {
                        type: 'object',
                        properties: {
                            after: {
                                type: 'string',
                                format: 'datetime'
                            },
                            before: {
                                type: 'string',
                                format: 'datetime'
                            }
                        }
                    }
                }
            },
            form: [
                {
                    key: 'registrationDate.after',
                    title: gettext('Show devices registered after'),
                    maxDateKey: 'registrationDate.before'
                },
                {
                    key: 'registrationDate.before',
                    title: gettext('And before`date`'),
                    minDateKey: 'registrationDate.after'
                }
            ],
            getFilter: function (model) {
                var _a, _b;
                var filter = {};
                var dates = model && model.registrationDate;
                if (dates && (dates.after || dates.before)) {
                    filter.__and = [];
                    if (dates.after) {
                        var after = self.formatDate(dates.after);
                        filter.__and.push({
                            __or: (_a = {},
                                _a[self.path] = { __gt: after },
                                _a[self.path + ".date"] = { __gt: after },
                                _a)
                        });
                    }
                    if (dates.before) {
                        var before = self.formatDate(dates.before);
                        filter.__and.push({
                            __or: (_b = {},
                                _b[self.path] = { __lt: before },
                                _b[self.path + ".date"] = { __lt: before },
                                _b)
                        });
                    }
                }
                return filter;
            }
        };
        _this.sortable = true;
        _this.sortingConfig = {
            pathSortingConfigs: [{ path: _this.path + ".date" }, { path: _this.path }]
        };
        return _this;
    }
    RegistrationDateDeviceGridColumn.prototype.formatDate = function (dateToFormat) {
        return new Date(dateToFormat).toISOString();
    };
    return RegistrationDateDeviceGridColumn;
}(BaseDeviceGridColumn));
export { RegistrationDateDeviceGridColumn };
