// @c8y/ngx-components/esm5/device-grid/columns/serial-number.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { SerialNumberCellRendererComponent } from './serial-number.cell-renderer.component';
import { gettext } from '@c8y/ngx-components';
var SerialNumberDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(SerialNumberDeviceGridColumn, _super);
    function SerialNumberDeviceGridColumn(initialColumnConfig) {
        var _this = _super.call(this, initialColumnConfig) || this;
        var hardwareSerialNumberPath = 'c8y_Hardware.serialNumber';
        var vendingSerialNumberPath = 'com_nsn_startups_vendme_fragments_VendingMachineTypeInfo.serial';
        _this.name = 'serialNumber';
        _this.header = gettext('Serial number');
        _this.cellRendererComponent = SerialNumberCellRendererComponent;
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    serialNumbers: {
                        type: 'array',
                        items: {
                            type: 'string'
                        }
                    }
                }
            },
            form: [
                {
                    key: 'serialNumbers',
                    title: gettext('Show devices with serial number'),
                    add: gettext('Or'),
                    items: [
                        {
                            key: 'serialNumbers[]',
                            notitle: true,
                            placeholder: gettext('e.g. 54321-123, you can use an asterisk (*) as wildcard character')
                        }
                    ]
                }
            ],
            getFilter: function (model) {
                var _a;
                var filter = {};
                if (model.serialNumbers.length) {
                    filter.__or = (_a = {},
                        _a[hardwareSerialNumberPath] = { __in: model.serialNumbers },
                        _a[vendingSerialNumberPath] = { __in: model.serialNumbers },
                        _a);
                }
                return filter;
            }
        };
        _this.sortable = true;
        _this.sortingConfig = {
            pathSortingConfigs: [{ path: hardwareSerialNumberPath }, { path: vendingSerialNumberPath }]
        };
        return _this;
    }
    return SerialNumberDeviceGridColumn;
}(BaseDeviceGridColumn));
export { SerialNumberDeviceGridColumn };
