// @c8y/ngx-components/esm5/device-grid/columns/status.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { StatusCellRendererComponent } from './status.cell-renderer.component';
import { PushStatus, SendStatus, gettext } from '@c8y/ngx-components';
var StatusDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(StatusDeviceGridColumn, _super);
    function StatusDeviceGridColumn() {
        var _a, _b, _c, _d, _e, _f, _g;
        var _this = _super.call(this) || this;
        var responseIntervalPath = 'c8y_RequiredAvailability.responseInterval';
        var responseIntervalLessThanOrEqualTo0 = (_a = {}, _a[responseIntervalPath] = { __le: 0 }, _a);
        var responseIntervalNotDefined = { __not: { __has: responseIntervalPath } };
        var availabilityStatusPath = 'c8y_Availability.status';
        var availabilityStatusAvailable = (_b = {}, _b[availabilityStatusPath] = SendStatus.AVAILABLE, _b);
        var availabilityStatusUnavailable = (_c = {}, _c[availabilityStatusPath] = SendStatus.UNAVAILABLE, _c);
        var availabilityStatusMaintenance = (_d = {}, _d[availabilityStatusPath] = SendStatus.MAINTENANCE, _d);
        var availabilityStatusNotDefined = { __not: { __has: availabilityStatusPath } };
        var connectionStatusPath = 'c8y_Connection.status';
        var connectionStatusConnected = (_e = {}, _e[connectionStatusPath] = PushStatus.CONNECTED, _e);
        var connectionStatusDisconnected = (_f = {}, _f[connectionStatusPath] = PushStatus.DISCONNECTED, _f);
        var connectionStatusMaintenance = (_g = {}, _g[connectionStatusPath] = PushStatus.MAINTENANCE, _g);
        var deviceUnderMaintenance = {
            __or: [
                responseIntervalLessThanOrEqualTo0,
                availabilityStatusMaintenance,
                connectionStatusMaintenance
            ]
        };
        var deviceNotUnderMaintenance = {
            // using __and of __nots because backend does not support __not with __ors
            __and: [
                { __not: responseIntervalLessThanOrEqualTo0 },
                { __not: availabilityStatusMaintenance },
                { __not: connectionStatusMaintenance }
            ]
        };
        _this.name = 'status';
        _this.header = gettext('Status');
        _this.dataType = "icon" /* Icon */;
        _this.cellRendererComponent = StatusCellRendererComponent;
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    sendOnline: {
                        type: 'boolean'
                    },
                    sendOffline: {
                        type: 'boolean'
                    },
                    sendUnknown: {
                        type: 'boolean'
                    },
                    sendNotMonitored: {
                        type: 'boolean'
                    },
                    pushOnline: {
                        type: 'boolean'
                    },
                    pushOffline: {
                        type: 'boolean'
                    },
                    pushNotMonitored: {
                        type: 'boolean'
                    },
                    maintenance: {
                        type: 'boolean'
                    }
                }
            },
            form: [
                {
                    type: 'template',
                    template: "<label><span translate>" + gettext('Show devices with send status') + "</span>:</label>"
                },
                {
                    key: 'sendOnline',
                    title: gettext('Online')
                },
                {
                    key: 'sendOffline',
                    title: gettext('Offline')
                },
                {
                    key: 'sendUnknown',
                    title: gettext('Unknown')
                },
                {
                    key: 'sendNotMonitored',
                    title: gettext('Not monitored')
                },
                {
                    type: 'template',
                    template: "<label><span translate>" + gettext('Show devices with push status') + "</span>:</label>"
                },
                {
                    key: 'pushOnline',
                    title: gettext('Online')
                },
                {
                    key: 'pushOffline',
                    title: gettext('Offline')
                },
                {
                    key: 'pushNotMonitored',
                    title: gettext('Not monitored')
                },
                {
                    type: 'template',
                    template: "<label><span translate>" + gettext('Show devices with maintenance status') + "</span>:</label>"
                },
                {
                    key: 'maintenance',
                    title: gettext('Device is under maintenance')
                }
            ],
            getFilter: function (model) {
                var filter = {};
                var ors = [];
                if (model) {
                    if (model.sendOnline) {
                        ors.push({
                            __and: [deviceNotUnderMaintenance, availabilityStatusAvailable]
                        });
                    }
                    if (model.sendOffline) {
                        ors.push({
                            __and: [deviceNotUnderMaintenance, availabilityStatusUnavailable]
                        });
                    }
                    if (model.sendUnknown) {
                        ors.push({
                            __and: [deviceNotUnderMaintenance, availabilityStatusNotDefined]
                        });
                    }
                    if (model.sendNotMonitored || model.pushNotMonitored) {
                        ors.push(responseIntervalNotDefined);
                    }
                    if (model.pushOnline) {
                        ors.push({
                            __and: [deviceNotUnderMaintenance, connectionStatusConnected]
                        });
                    }
                    if (model.pushOffline) {
                        ors.push({
                            __and: [deviceNotUnderMaintenance, connectionStatusDisconnected]
                        });
                    }
                    if (model.maintenance) {
                        ors.push(deviceUnderMaintenance);
                    }
                    if (ors.length) {
                        filter.__or = ors;
                    }
                }
                return filter;
            }
        };
        _this.sortable = true;
        _this.sortingConfig = {
            pathSortingConfigs: [{ path: availabilityStatusPath }]
        };
        return _this;
    }
    return StatusDeviceGridColumn;
}(BaseDeviceGridColumn));
export { StatusDeviceGridColumn };
