// @c8y/ngx-components/esm5/device-grid/columns/type.device-grid-column.js
import * as tslib_1 from "tslib";
import { gettext } from '@c8y/ngx-components';
import { BaseDeviceGridColumn } from './base.device-grid-column';
var TypeDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(TypeDeviceGridColumn, _super);
    function TypeDeviceGridColumn(initialColumnConfig) {
        var _this = _super.call(this, initialColumnConfig) || this;
        _this.name = 'type';
        _this.path = 'type';
        _this.header = gettext('Type');
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    types: {
                        type: 'array',
                        items: {
                            type: 'string'
                        }
                    }
                }
            },
            form: [
                {
                    key: 'types',
                    title: gettext('Show devices with type'),
                    add: gettext('Or'),
                    items: [
                        {
                            key: 'types[]',
                            notitle: true,
                            placeholder: gettext('e.g. MyType, you can use an asterisk (*) as wildcard character')
                        }
                    ]
                }
            ],
            getFilter: function (model) {
                var filter = {};
                if (model.types.length) {
                    filter.type = { __in: model.types };
                }
                return filter;
            }
        };
        _this.sortable = true;
        _this.sortingConfig = { pathSortingConfigs: [{ path: 'type' }] };
        return _this;
    }
    return TypeDeviceGridColumn;
}(BaseDeviceGridColumn));
export { TypeDeviceGridColumn };
