// @c8y/ngx-components/esm5/operations/stepper-bulk-type-software/stepper-bulk-type-software.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { RepositoryModule } from '@c8y/ngx-components/repository';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { HOOK_LIST_BULK_TYPE, baseUrl, BulkOperationType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { ConfirmSoftwareSelectionStepComponent } from './confirm-software-selection-step.component';
import { SelectSoftwareStepComponent } from './select-software-step.component';
import { StepperBulkTypeSoftwareComponent } from './stepper-bulk-type-software.component';
var ɵ0 = {
    type: BulkOperationType.SOFTWARE,
    c8yIcon: 'c8y-tools',
    name: gettext('Software update'),
    path: baseUrl + "software",
    component: StepperBulkTypeSoftwareComponent,
    fragments: ['c8y_SoftwareList', 'c8y_SoftwareUpdate'],
    selected: false
};
/** Module for the 'Software update' operation type stepper */
var StepperBulkTypeSoftwareModule = /** @class */ (function () {
    function StepperBulkTypeSoftwareModule() {
    }
    StepperBulkTypeSoftwareModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                StepperBulkTypeSoftwareComponent,
                ConfirmSoftwareSelectionStepComponent,
                SelectSoftwareStepComponent
            ],
            imports: [
                CoreModule,
                FormsModule,
                ReactiveFormsModule,
                BulkOperationStepperModule,
                RepositoryModule
            ],
            providers: [
                {
                    provide: HOOK_LIST_BULK_TYPE,
                    useValue: ɵ0,
                    multi: true
                }
            ],
            entryComponents: [StepperBulkTypeSoftwareComponent],
            exports: [StepperBulkTypeSoftwareComponent]
        })
    ], StepperBulkTypeSoftwareModule);
    return StepperBulkTypeSoftwareModule;
}());
export { StepperBulkTypeSoftwareModule };
export { ɵ0 };
