// @c8y/ngx-components/esm2015/core/header/application/app-switcher.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ApplicationService, IApplication, ICurrentTenant, IUser } from '@c8y/client';
import { get, includes, orderBy } from 'lodash-es';
import { combineLatest } from 'rxjs';
import { first } from 'rxjs/operators';
import { AppStateService } from '../../common/ui-state.service';
let AppSwitcherComponent = class AppSwitcherComponent {
    constructor(ui, applications) {
        this.ui = ui;
        this.applications = applications;
        const { currentUser, currentTenant } = this.ui;
        combineLatest(currentUser, currentTenant)
            .pipe(first())
            .subscribe(([user, tenant]) => {
            this.tenant = tenant;
            this.listAppsForUser(user);
        });
    }
    get hideCloudApps() {
        return false; // this should be configurable from outside
    }
    isActive(app) {
        return app.contextPath === this.ui.state.app.contextPath;
    }
    getHref(app) {
        return this.applications.getHref(app);
    }
    listAppsForUser(user) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (user) {
                const { data } = yield this.applications.listByUser(user, { pageSize: 1000 });
                this.apps = this.orderApps(this.filterApps(data));
            }
        });
    }
    filterApps(apps) {
        apps = this.filterVisible(apps);
        apps = this.filterCloud(apps);
        apps = this.filterDuplicates(apps);
        return apps;
    }
    filterVisible(apps) {
        const visibleApplicationTypes = ['HOSTED', 'EXTERNAL'];
        return apps.filter(app => includes(visibleApplicationTypes, app.type) &&
            !app.noAppSwitcher &&
            !get(app, 'manifest.noAppSwitcher'));
    }
    filterCloud(apps) {
        const { hideCloudApps } = this;
        const filterFn = app => !!app.cloud;
        if (!hideCloudApps) {
            this.oneCloudApps = this.orderApps(apps.filter(filterFn));
        }
        return apps.filter(app => !filterFn(app));
    }
    filterDuplicates(apps) {
        // Filter out apps that have duplicate contextpaths and are not own owned but the current tenant;
        const filterFn = app => !apps.some(otherApp => app !== otherApp &&
            (app.contextPath && app.contextPath === otherApp.contextPath) &&
            app.owner.tenant.id !== this.tenant.name);
        return apps.filter(filterFn);
    }
    orderApps(apps) {
        return orderBy(apps, ({ name }) => name.toLowerCase());
    }
};
AppSwitcherComponent.ctorParameters = () => [
    { type: AppStateService },
    { type: ApplicationService }
];
AppSwitcherComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-app-switcher',
        template: "<div class=\"app-switcher-dropdown\" dropdown>\n  <button\n    class=\"main-header-button c8y-dropdown dropdown-toggle\"\n    id=\"appSwitcherDropdown\"\n    dropdownToggle\n  >\n    <i c8yIcon=\"th\" class=\"icon-2x\"></i>\n  </button>\n  <div\n    *dropdownMenu\n    class=\"app-switcher-dropdown-menu dropdown-menu\"\n    aria-labelledby=\"appSwitcherDropdown\"\n  >\n    <div class=\"app-switcher-sagcloud\" *ngIf=\"oneCloudApps && oneCloudApps.length > 0\">\n      <a\n        title=\"{{ app.name | humanizeAppName | async }}\"\n        class=\"appLink\"\n        *ngFor=\"let app of oneCloudApps\"\n        [href]=\"getHref(app)\"\n      >\n        <c8y-app-icon [app]=\"app\" [name]=\"app.name\" [contextPath]=\"app.contextPath\"></c8y-app-icon>\n        <span class=\"text-truncate-wrap\">{{ app | humanizeAppName | async }}</span>\n      </a>\n    </div>\n    <div class=\"app-switcher\">\n      <a\n        title=\"{{ app.name | humanizeAppName | async }}\"\n        class=\"appLink\"\n        [ngClass]=\"isActive(app) ? 'active' : ''\"\n        *ngFor=\"let app of apps\"\n        [href]=\"getHref(app)\"\n        c8yProductExperience\n        [actionName]=\"'appSwitched'\"\n        [actionData]=\"{ applicationName: app && app.name }\"\n      >\n        <c8y-app-icon [app]=\"app\" [name]=\"app.name\" [contextPath]=\"app.contextPath\"></c8y-app-icon>\n        <span class=\"text-truncate-wrap\">{{ app | humanizeAppName | async }}</span>\n      </a>\n    </div>\n  </div>\n</div>\n"
    })
], AppSwitcherComponent);
export { AppSwitcherComponent };
