// @c8y/ngx-components/esm5/repository/software-device-tab/device-software-changes.component.js
import * as tslib_1 from "tslib";
import { Component, Output, Input, EventEmitter } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map } from 'rxjs/operators';
var DeviceSoftwareChangesComponent = /** @class */ (function () {
    function DeviceSoftwareChangesComponent() {
        this.clear = new EventEmitter();
        this.drop = new EventEmitter();
        this.apply = new EventEmitter();
        this.changes$ = new BehaviorSubject([]);
        this.installations$ = this.changes$.pipe(map(function (changes) {
            return changes.filter(function (change) { return change.action === 'install'; });
        }));
        this.removals$ = this.changes$.pipe(map(function (changes) {
            return changes.filter(function (change) { return change.action === 'delete'; });
        }));
    }
    DeviceSoftwareChangesComponent.prototype.ngOnChanges = function (changes) {
        if (changes.changes) {
            this.changes$.next(changes.changes.currentValue);
        }
    };
    tslib_1.__decorate([
        Input()
    ], DeviceSoftwareChangesComponent.prototype, "changes", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceSoftwareChangesComponent.prototype, "changesInProgress", void 0);
    tslib_1.__decorate([
        Output()
    ], DeviceSoftwareChangesComponent.prototype, "clear", void 0);
    tslib_1.__decorate([
        Output()
    ], DeviceSoftwareChangesComponent.prototype, "drop", void 0);
    tslib_1.__decorate([
        Output()
    ], DeviceSoftwareChangesComponent.prototype, "apply", void 0);
    DeviceSoftwareChangesComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-software-changes',
            template: "<div class=\"inner-scroll d-flex d-col\">\n  <div class=\"card-header separator large-padding sticky-top\">\n    <h4 class=\"card-title\" translate>Software changes</h4>\n  </div>\n  <div class=\"flex-grow\">\n    <div class=\"card-block large-padding\"\n      *ngIf=\"(changes$ | async).length === 0; else changesList\"\n    >\n      <div class=\"c8y-empty-state text-center\">\n        <h1 class=\"c8y-icon c8y-icon-device-control c8y-icon-duocolor\"></h1>\n        <p>\n          <strong translate>No software changes.</strong> <br />\n          <small translate>Selected actions will be displayed here.</small>\n        </p>\n      </div>\n    </div>\n\n    <ng-template #changesList>\n      <fieldset class=\"card-block p-t-0 p-b-0\"\n        *ngIf=\"(installations$ | async).length > 0\"\n        [disabled]=\"changesInProgress\"\n        >\n        <div class=\"form-block legend\" translate>Installations`software`</div>\n        <c8y-device-software-list\n          [softwareList]=\"installations$ | async\"\n          (remove)=\"drop.emit($event)\"\n        ></c8y-device-software-list>\n      </fieldset>\n\n      <fieldset class=\"card-block p-t-0 p-b-0\"\n        *ngIf=\"(removals$ | async).length > 0\"\n        [disabled]=\"changesInProgress\"\n        >\n        <div class=\"form-block legend\" translate>Removals`software`</div>\n        <c8y-device-software-list\n          [softwareList]=\"removals$ | async\"\n          (remove)=\"drop.emit($event)\"\n        ></c8y-device-software-list>\n      </fieldset>\n    </ng-template>\n  </div>\n  <div\n    class=\"card-footer separator large-padding sticky-bottom\"\n    *ngIf=\"(changes$ | async).length > 0\"\n  >\n    <button\n      class=\"btn btn-default\"\n      (click)=\"clear.emit()\"\n      title=\"{{ 'Clear changes' | translate }}\"\n      [disabled]=\"changesInProgress\"\n      translate\n    >\n      Clear`changes`\n    </button>\n    <button\n      class=\"btn btn-primary\"\n      [ngClass]=\"{ 'btn-pending': changesInProgress }\"\n      [disabled]=\"changesInProgress\"\n      (click)=\"apply.emit()\"\n      title=\"{{ 'Apply changes' | translate }}\"\n      translate\n    >\n      Apply changes\n    </button>\n  </div>\n</div>\n"
        })
    ], DeviceSoftwareChangesComponent);
    return DeviceSoftwareChangesComponent;
}());
export { DeviceSoftwareChangesComponent };
