// @c8y/ngx-components/esm5/default-subscriptions/default-subscriptions.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ONCE_ROUTE } from '@c8y/ngx-components';
import { DefaultSubscriptionsComponent } from './default-subscriptions.component';
import { DefaultSubscriptionsService } from './default-subscriptions.service';
import { DefaultSubscriptionsGuard } from './default-subscriptions.guard';
import { DefaultSubscriptionsNavigationFactory } from './default-subscriptions-navigation.factory';
var ɵ0 = [
    {
        path: 'default-subscriptions',
        component: DefaultSubscriptionsComponent,
        canActivate: [DefaultSubscriptionsGuard]
    }
];
var DefaultSubscriptionsModule = /** @class */ (function () {
    function DefaultSubscriptionsModule() {
    }
    DefaultSubscriptionsModule = tslib_1.__decorate([
        NgModule({
            imports: [CoreModule, ReactiveFormsModule],
            declarations: [DefaultSubscriptionsComponent],
            entryComponents: [DefaultSubscriptionsComponent],
            exports: [DefaultSubscriptionsComponent],
            providers: [
                DefaultSubscriptionsService,
                DefaultSubscriptionsGuard,
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: DefaultSubscriptionsNavigationFactory,
                    multi: true
                },
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: ɵ0,
                    multi: true
                }
            ]
        })
    ], DefaultSubscriptionsModule);
    return DefaultSubscriptionsModule;
}());
export { DefaultSubscriptionsModule };
export { ɵ0 };
