// @c8y/ngx-components/esm5/connectivity/impact/impact-node.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { isObservable, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { ImpactTabFactory } from './impact-tab.factory';
var ImpactNodeFactory = /** @class */ (function () {
    function ImpactNodeFactory(tabFactory) {
        this.tabFactory = tabFactory;
        this.connectivityNode = new NavigatorNode({
            parent: gettext('Settings'),
            label: gettext('Connectivity'),
            path: 'connectivitySettings',
            icon: 'plug',
            preventDuplicates: true
        });
    }
    ImpactNodeFactory.prototype.get = function () {
        var _this = this;
        return this.wrapIntoObservable(this.tabFactory.tab()).pipe(map(function (tab) {
            return !!tab ? [_this.connectivityNode] : [];
        }));
    };
    ImpactNodeFactory.prototype.wrapIntoObservable = function (value) {
        if (isObservable(value)) {
            return value;
        }
        return of(value);
    };
    ImpactNodeFactory.ctorParameters = function () { return [
        { type: ImpactTabFactory }
    ]; };
    ImpactNodeFactory = tslib_1.__decorate([
        Injectable()
    ], ImpactNodeFactory);
    return ImpactNodeFactory;
}());
export { ImpactNodeFactory };
