// @c8y/ngx-components/esm5/connectivity/impact/impact.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, FormsModule, HOOK_NAVIGATOR_NODES, HOOK_TABS } from '@c8y/ngx-components';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { ImpactConnectivityService } from './impact-connectivity.service';
import { ImpactNodeFactory } from './impact-node.factory';
import { ImpactProviderSettingsComponent } from './impact-provider-settings.component';
import { ImpactTabFactory } from './impact-tab.factory';
import { ImpactGuard } from './impact.guard';
import { StatusDisplayComponent } from './status-display.component';
var impactRoutes = [
    {
        path: 'connectivitySettings/impact_provider_settings',
        component: ImpactProviderSettingsComponent,
        canActivate: [ImpactGuard]
    }
];
var ImpactModule = /** @class */ (function () {
    function ImpactModule() {
    }
    ImpactModule = tslib_1.__decorate([
        NgModule({
            declarations: [ImpactProviderSettingsComponent, StatusDisplayComponent],
            exports: [ImpactProviderSettingsComponent],
            imports: [
                CoreModule,
                CommonModule,
                FormsModule,
                RouterModule.forChild(impactRoutes),
                PopoverModule,
                ReactiveFormsModule
            ],
            entryComponents: [ImpactProviderSettingsComponent],
            providers: [
                ImpactConnectivityService,
                ImpactGuard,
                ImpactTabFactory,
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: ImpactNodeFactory,
                    multi: true
                },
                {
                    provide: HOOK_TABS,
                    useClass: ImpactTabFactory,
                    multi: true
                }
            ]
        })
    ], ImpactModule);
    return ImpactModule;
}());
export { ImpactModule };
