// @c8y/ngx-components/esm2015/sub-assets/groups.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter } from '@angular/core';
import { Permissions } from '@c8y/ngx-components';
let GroupsComponent = class GroupsComponent {
    constructor(permissions) {
        this.permissions = permissions;
        this.refresh = new EventEmitter();
        this.showAddGroup = false;
        this.filterable = false;
        this.sortable = true;
    }
    ngOnInit() {
        if (!this.permissions.hasRole('ROLE_INVENTORY_READ')) {
            this.sortable = false;
        }
    }
};
GroupsComponent.ctorParameters = () => [
    { type: Permissions }
];
GroupsComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-groups-name',
        template: "<c8y-title>\n  {{'Groups'|translate}}\n</c8y-title>\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add group' | translate }}\"\n    (click)=\"showAddGroup = !showAddGroup\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>&nbsp;<span translate>Add group</span>\n  </button>\n</c8y-action-bar-item>\n\n<div [ngClass]=\"{ drawerOpen: showAddGroup }\">\n  <div class=\"bottom-drawer\">\n    <c8y-add-group\n      [refresh]=\"refresh\"\n      (onCancel)=\"showAddGroup = false\"\n      *ngIf=\"showAddGroup\"\n    ></c8y-add-group>\n  </div>\n</div>\n<c8y-sub-assets-grid\n  [refresh]=\"refresh\"\n  [filterable]=\"filterable\"\n  [sortable]=\"sortable\"\n  class=\"content-fullpage d-flex d-col\"\n></c8y-sub-assets-grid>\n"
    })
], GroupsComponent);
export { GroupsComponent };
