// @c8y/ngx-components/esm5/operations/bulk-operation-scheduler/bulk-operation-scheduler.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule } from '@c8y/ngx-components';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { TimepickerModule } from 'ngx-bootstrap/timepicker';
import { OperationSchedulerComponent } from './operation-scheduler.component';
/**
 * This module provides components for scheduling bulk operations.
 */
var BulkOperationSchedulerModule = /** @class */ (function () {
    function BulkOperationSchedulerModule() {
    }
    BulkOperationSchedulerModule = tslib_1.__decorate([
        NgModule({
            imports: [
                CoreModule,
                ReactiveFormsModule,
                BsDatepickerModule.forRoot(),
                TimepickerModule.forRoot()
            ],
            providers: [],
            declarations: [OperationSchedulerComponent],
            entryComponents: [OperationSchedulerComponent],
            exports: [OperationSchedulerComponent]
        })
    ], BulkOperationSchedulerModule);
    return BulkOperationSchedulerModule;
}());
export { BulkOperationSchedulerModule };
