// @c8y/ngx-components/esm5/repository/file-download/file-download.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { RepositoryService } from './../repository.service';
import { memoize } from '@c8y/ngx-components';
import { saveAs } from 'file-saver';
import { InventoryBinaryService } from '@c8y/client';
import { LinkRenderType } from './link-render-type.enum';
var FileDownloadComponent = /** @class */ (function () {
    function FileDownloadComponent(repositoryService, inventoryBinaryService) {
        this.repositoryService = repositoryService;
        this.inventoryBinaryService = inventoryBinaryService;
        this.linkRenderType = LinkRenderType;
    }
    FileDownloadComponent.prototype.getBinaryName$ = function (binaryUrl) {
        return this.repositoryService.getBinaryName$(binaryUrl);
    };
    FileDownloadComponent.prototype.determineBehavior = function () {
        var result;
        if (this.inventoryBinaryService.getIdFromUrl(this.url)) {
            result = LinkRenderType.DOWNLOAD;
        }
        else if (this.url.match(/\/\//g)) {
            result = LinkRenderType.LINK;
        }
        else {
            result = LinkRenderType.TEXTONLY;
        }
        return result;
    };
    FileDownloadComponent.prototype.downloadFile = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var binary;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.repositoryService.getBinaryFile(this.url, {
                            allowExternal: false
                        })];
                    case 1:
                        binary = _a.sent();
                        saveAs(binary);
                        return [2 /*return*/];
                }
            });
        });
    };
    FileDownloadComponent.ctorParameters = function () { return [
        { type: RepositoryService },
        { type: InventoryBinaryService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], FileDownloadComponent.prototype, "url", void 0);
    tslib_1.__decorate([
        memoize()
    ], FileDownloadComponent.prototype, "getBinaryName$", null);
    tslib_1.__decorate([
        memoize()
    ], FileDownloadComponent.prototype, "determineBehavior", null);
    FileDownloadComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-file-download',
            template: "<a\n  *ngIf=\"determineBehavior() === linkRenderType.LINK\"\n  href=\"{{ url }}\"\n  class=\"pointer\"\n  target=\"_blank\"\n  rel=\"noopener noreferrer\"\n>\n  {{ getBinaryName$(url) | async }}\n</a>\n\n<span *ngIf=\"determineBehavior() === linkRenderType.TEXTONLY\">{{ getBinaryName$(url) | async }}</span>\n\n<a *ngIf=\"determineBehavior() === linkRenderType.DOWNLOAD\" class=\"pointer\" (click)=\"downloadFile()\">\n    {{getBinaryName$(url) | async}}\n</a>\n"
        })
    ], FileDownloadComponent);
    return FileDownloadComponent;
}());
export { FileDownloadComponent };
