// @c8y/ngx-components/esm5/core/dynamic-forms/input/input.type.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
var FieldInput = /** @class */ (function (_super) {
    tslib_1.__extends(FieldInput, _super);
    function FieldInput() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FieldInput_1 = FieldInput;
    Object.defineProperty(FieldInput.prototype, "type", {
        get: function () {
            return this.to.type || 'text';
        },
        enumerable: true,
        configurable: true
    });
    var FieldInput_1;
    FieldInput.CONFIG = {
        types: [
            {
                name: 'input',
                component: FieldInput_1,
                wrappers: ['c8y-form-field']
            },
            { name: 'string', extends: 'input' },
            {
                name: 'number',
                extends: 'input',
                defaultOptions: {
                    templateOptions: {
                        type: 'number'
                    }
                }
            },
            {
                name: 'integer',
                extends: 'input',
                defaultOptions: {
                    templateOptions: {
                        type: 'number'
                    }
                }
            }
        ]
    };
    FieldInput = FieldInput_1 = tslib_1.__decorate([
        Component({
            selector: 'c8y-field-input',
            template: "<input\n  *ngIf=\"type !== 'number'; else numberTmp\"\n  [type]=\"type\"\n  [formControl]=\"formControl\"\n  class=\"form-control\"\n  [formlyAttributes]=\"field\"\n  [required]=\"to.required\"\n  [attr.autocomplete]=\"to.autocomplete ? to.autocomplete : null\"\n  [class.is-invalid]=\"showError\"\n/>\n<ng-template #numberTmp>\n  <input\n    type=\"number\"\n    [formControl]=\"formControl\"\n    class=\"form-control\"\n    [formlyAttributes]=\"field\"\n    [required]=\"to.required\"\n    [attr.autocomplete]=\"to.autocomplete ? to.autocomplete : null\"\n    [class.is-invalid]=\"showError\"\n  />\n</ng-template>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], FieldInput);
    return FieldInput;
}(FieldType));
export { FieldInput };
