// @c8y/ngx-components/esm5/operations/bulk-operations-list/type-list/bulk-operation-type-list.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { Router } from '@angular/router';
import { BulkOperationsService, OperationType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { BulkOperationModalsService } from '../modals/bulk-operation-modals.service';
var BulkOperationTypeListComponent = /** @class */ (function () {
    function BulkOperationTypeListComponent(router, bulkOperationsService, bulkOperationModalsService) {
        this.router = router;
        this.bulkOperationsService = bulkOperationsService;
        this.bulkOperationModalsService = bulkOperationModalsService;
        this.onCustomType = new EventEmitter();
        this.bulkList = [];
    }
    BulkOperationTypeListComponent.prototype.ngOnInit = function () {
        this.bulkList = this.bulkOperationsService.getBulkTypes();
    };
    BulkOperationTypeListComponent.prototype.handleClick = function (item) {
        if (item.path) {
            this.bulkOperationModalsService.hideNewBulkOperationModal();
            this.router.navigateByUrl(item.path);
        }
        else if (item.component) {
            this.onCustomType.emit(item);
        }
    };
    BulkOperationTypeListComponent.ctorParameters = function () { return [
        { type: Router },
        { type: BulkOperationsService },
        { type: BulkOperationModalsService }
    ]; };
    tslib_1.__decorate([
        Output()
    ], BulkOperationTypeListComponent.prototype, "onCustomType", void 0);
    BulkOperationTypeListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-bulk-operation-type-list',
            template: "<div class=\"list-group list-group-links\">\n  <button\n    *ngFor=\"let item of bulkList\"\n    type=\"button\"\n    class=\"list-group-item flex-row\"\n    (click)=\"handleClick(item)\"\n  >\n    <div *ngIf=\"item.c8yIcon; else icon\" class=\"list-item-icon\">\n      <i [c8yIcon]=\"item.c8yIcon\"></i>\n    </div>\n    <ng-template #icon>\n      <div class=\"list-item-icon\">\n        <i c8yIcon=\"cogs\"></i>\n      </div>\n    </ng-template>\n    <div title=\"{{ item.name | translate }}\" class=\"list-item-body text-truncate\">\n      {{ item.name | translate }}\n    </div>\n  </button>\n</div>\n"
        })
    ], BulkOperationTypeListComponent);
    return BulkOperationTypeListComponent;
}());
export { BulkOperationTypeListComponent };
