// @c8y/ngx-components/esm5/operations/stepper-bulk-type-configuration/stepper-bulk-type-configuration.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { RepositoryModule } from '@c8y/ngx-components/repository';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { HOOK_LIST_BULK_TYPE, baseUrl, BulkOperationType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { StepperBulkTypeConfigurationComponent } from './stepper-bulk-type-configuration.component';
var ɵ0 = {
    type: BulkOperationType.CONFIGURATION,
    c8yIcon: 'cogs',
    name: gettext('Configuration update'),
    path: baseUrl + "configuration",
    component: StepperBulkTypeConfigurationComponent,
    fragments: ['c8y_DownloadConfigFile', 'c8y_Configuration'],
    selected: false
};
/** Module for the 'Configuration update' operation type stepper */
var StepperBulkTypeConfigurationModule = /** @class */ (function () {
    function StepperBulkTypeConfigurationModule() {
    }
    StepperBulkTypeConfigurationModule = tslib_1.__decorate([
        NgModule({
            declarations: [StepperBulkTypeConfigurationComponent],
            imports: [
                CoreModule,
                FormsModule,
                ReactiveFormsModule,
                BulkOperationStepperModule,
                RepositoryModule
            ],
            providers: [
                {
                    provide: HOOK_LIST_BULK_TYPE,
                    useValue: ɵ0,
                    multi: true
                }
            ],
            entryComponents: [StepperBulkTypeConfigurationComponent],
            exports: [StepperBulkTypeConfigurationComponent]
        })
    ], StepperBulkTypeConfigurationModule);
    return StepperBulkTypeConfigurationModule;
}());
export { StepperBulkTypeConfigurationModule };
export { ɵ0 };
