// @c8y/ngx-components/esm5/core/authentication/new-password.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { ControlContainer, NgForm } from '@angular/forms';
import { LoginService } from '../login/login.service';
var NewPasswordComponent = /** @class */ (function () {
    function NewPasswordComponent(loginService) {
        this.loginService = loginService;
        this.password = new EventEmitter();
        this.model = {};
        this.changePassword = false;
        this.passwordEnforced = false;
    }
    NewPasswordComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.loadPasswordStrengthSettings();
        this.loginService.getGreenMinLength().then(function (value) {
            _this.minlength = value;
        });
    };
    NewPasswordComponent.prototype.newPasswordChanged = function () {
        this.password.emit({
            password: this.model.newPassword,
            passwordStrength: this.model.strength
        });
    };
    NewPasswordComponent.prototype.loadPasswordStrengthSettings = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.loginService.getPasswordStrengthEnforced({ refresh: true })];
                    case 1:
                        _a.passwordEnforced = _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    NewPasswordComponent.prototype.showPasswordStrength = function () {
        return this.model.newPassword && this.model.newPassword.length >= this.minlength;
    };
    NewPasswordComponent.prototype.toggleChangePassword = function () {
        this.changePassword = !this.changePassword;
        if (!this.changePassword) {
            this.password.emit({});
            this.model = {};
        }
    };
    NewPasswordComponent.ctorParameters = function () { return [
        { type: LoginService }
    ]; };
    tslib_1.__decorate([
        Output()
    ], NewPasswordComponent.prototype, "password", void 0);
    NewPasswordComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-new-password',
            template: "<div class=\"form-group\">\n  <button type=\"button\" class=\"btn btn-default\" (click)=\"toggleChangePassword()\">\n    <ng-container *ngIf=\"!changePassword\">\n      {{ 'Change password' | translate }}\n    </ng-container>\n    <ng-container *ngIf=\"changePassword\">\n      {{ 'Cancel password change' | translate }}\n    </ng-container>\n  </button>\n</div>\n\n<div *ngIf=\"changePassword\" class=\"row content-flex-50\">\n  <div class=\"col-7\">\n    <c8y-form-group>\n      <label for=\"newPassword\" translate>Password</label>\n      <input\n        id=\"newPassword\"\n        name=\"newPassword\"\n        type=\"password\"\n        #newPassword=\"ngModel\"\n        class=\"form-control\"\n        [(ngModel)]=\"model.newPassword\"\n        (change)=\"newPasswordChanged()\"\n        (input)=\"newPasswordConfirm.control.updateValueAndValidity()\"\n        required\n        c8yDefaultValidation=\"password\"\n        autocomplete=\"new-password\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label for=\"newConfirmPassword\" translate>Confirm password</label>\n      <input\n        id=\"newConfirmPassword\"\n        name=\"newPasswordConfirm\"\n        type=\"password\"\n        #newPasswordConfirm=\"ngModel\"\n        class=\"form-control\"\n        [(ngModel)]=\"model.newPasswordConfirm\"\n        required\n        passwordConfirm=\"newPassword\"\n        autocomplete=\"new-password\"\n      />\n    </c8y-form-group>\n  </div>\n  <div class=\"col-5\">\n    <c8y-password-strength\n      [password]=\"model.newPassword\"\n      (strength)=\"model.strength = $event\"\n    ></c8y-password-strength>\n\n    <c8y-password-check-list\n      class=\"m-t-32\"\n      [password]=\"model.newPassword\"\n      [hidden]=\"!passwordEnforced\"\n    ></c8y-password-check-list>\n  </div>\n</div>\n",
            viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
        })
    ], NewPasswordComponent);
    return NewPasswordComponent;
}());
export { NewPasswordComponent };
