// @c8y/ngx-components/esm5/core/authentication/password-check-list.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { PasswordService } from './password.service';
import { LoginService } from '../login/login.service';
import { assign, map } from 'lodash-es';
import { gettext } from '../i18n/gettext';
var PasswordCheckListComponent = /** @class */ (function () {
    function PasswordCheckListComponent(loginService, passwordStrengthChecker, passwordService) {
        var _this = this;
        this.loginService = loginService;
        this.passwordStrengthChecker = passwordStrengthChecker;
        this.passwordService = passwordService;
        this.minGreenLength = 0;
        this.checklist = [
            {
                label: gettext('Include lowercase characters (e.g. abcdef)'),
                check: this.passwordStrengthChecker.hasLowerCase,
                icon: '',
                contextualColor: ''
            },
            {
                label: gettext('Include uppercase characters (e.g. ABCDEF)'),
                check: this.passwordStrengthChecker.hasUpperCase,
                icon: '',
                contextualColor: ''
            },
            {
                label: gettext('Include numbers (e.g. 123456)'),
                check: this.passwordStrengthChecker.hasNumbers,
                icon: '',
                contextualColor: ''
            },
            {
                label: gettext('Include symbols (e.g. !@#$%^)'),
                check: this.passwordStrengthChecker.hasSpecialChars,
                icon: '',
                contextualColor: ''
            },
            {
                label: gettext('Must have at least {{length}} characters'),
                check: function (password) { return password.length >= _this.minGreenLength; },
                icon: '',
                contextualColor: ''
            }
        ];
    }
    Object.defineProperty(PasswordCheckListComponent.prototype, "password", {
        set: function (password) {
            if (typeof password === 'string') {
                this.onPasswordChange(password);
            }
        },
        enumerable: true,
        configurable: true
    });
    PasswordCheckListComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.loginService.getGreenMinLength()];
                    case 1:
                        _a.minGreenLength = _b.sent();
                        if (!this.minGreenLength) {
                            this.minGreenLength = this.passwordService.getDefaultPasswordMinLength();
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(PasswordCheckListComponent.prototype, "translateParams", {
        get: function () {
            return {
                length: this.minGreenLength
            };
        },
        enumerable: true,
        configurable: true
    });
    PasswordCheckListComponent.prototype.onPasswordChange = function (password) {
        this.checklist = map(this.checklist, function (requirement) {
            var checked = requirement.check(password);
            assign(requirement, {
                icon: checked ? 'check-circle' : '',
                contextualColor: checked ? '' : 'text-muted'
            });
            return requirement;
        });
    };
    PasswordCheckListComponent.ctorParameters = function () { return [
        { type: LoginService },
        { type: PasswordService },
        { type: PasswordService }
    ]; };
    tslib_1.__decorate([
        Input('password')
    ], PasswordCheckListComponent.prototype, "password", null);
    PasswordCheckListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-password-check-list',
            template: "<div>\n  <span class=\"small\">{{ 'Please meet 5 out of the 5 conditions below' | translate }}:</span>\n  <ul class=\"list-unstyled\">\n    <li class=\"small\" *ngFor=\"let requirement of checklist\">\n      <i [c8yIcon]=\"requirement.icon\" class=\"text-success\"></i>\n      <span\n        class=\"small {{ requirement.contextualColor }}\"\n        [translate]=\"requirement.label\"\n        [translateParams]=\"this.translateParams\"\n      ></span>\n    </li>\n  </ul>\n</div>\n"
        })
    ], PasswordCheckListComponent);
    return PasswordCheckListComponent;
}());
export { PasswordCheckListComponent };
