// @c8y/ngx-components/esm2015/operations/single-operation-details/single-operation-modal.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
let SingleOperationModalComponent = class SingleOperationModalComponent {
    constructor(modalRef) {
        this.modalRef = modalRef;
        this.collapsed = false;
        this.readOnly = false;
    }
};
SingleOperationModalComponent.ctorParameters = () => [
    { type: BsModalRef }
];
tslib_1.__decorate([
    Input()
], SingleOperationModalComponent.prototype, "operation", void 0);
tslib_1.__decorate([
    Input()
], SingleOperationModalComponent.prototype, "collapsed", void 0);
tslib_1.__decorate([
    Input()
], SingleOperationModalComponent.prototype, "readOnly", void 0);
SingleOperationModalComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-single-operation-modal',
        template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header\">\n    <h3>{{ 'Single operation details' | translate }}</h3>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <c8y-single-operation-list-item\n      [operation]=\"operation\"\n      [collapsed]=\"collapsed\"\n      [readOnly]=\"readOnly\"\n    ></c8y-single-operation-list-item>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button title=\"{{ 'Close' | translate }}\" class=\"btn btn-default\" (click)=\"modalRef.hide()\">\n      {{ 'Close' | translate }}\n    </button>\n  </div>\n</div>\n"
    })
], SingleOperationModalComponent);
export { SingleOperationModalComponent };
