// @c8y/ngx-components/esm2015/operations/single-operation-details/single-operation-details.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationsListModule } from '@c8y/ngx-components/operations/bulk-operations-list';
import { BulkOperationsServiceModule } from '@c8y/ngx-components/operations/bulk-operations-service';
import { SingleOperationDetailsService } from './single-operation-details.service';
import { SingleOperationComponent } from './single-operation.component';
import { SingleOperationModalComponent } from './single-operation-modal.component';
import { SingleOperationListItemComponent } from './single-operation-list-item.component';
import { SingleOperationDetailsComponent } from './single-operation-details.component';
/**
 * This module allows to display details of a single operation in a modal.
 */
let SingleOperationDetailsModule = class SingleOperationDetailsModule {
};
SingleOperationDetailsModule = tslib_1.__decorate([
    NgModule({
        imports: [
            CoreModule,
            PopoverModule,
            TooltipModule,
            CollapseModule,
            BulkOperationsListModule,
            BulkOperationsServiceModule
        ],
        providers: [SingleOperationDetailsService],
        declarations: [
            SingleOperationComponent,
            SingleOperationModalComponent,
            SingleOperationListItemComponent,
            SingleOperationDetailsComponent
        ],
        entryComponents: [
            SingleOperationComponent,
            SingleOperationModalComponent,
            SingleOperationListItemComponent,
            SingleOperationDetailsComponent
        ],
        exports: [SingleOperationComponent]
    })
], SingleOperationDetailsModule);
export { SingleOperationDetailsModule };
