// @c8y/ngx-components/esm2015/device-grid/columns/group.device-grid-column.js
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { GroupCellRendererComponent } from './group.cell-renderer.component';
import { GroupFilteringFormRendererComponent } from './group.filtering-form-renderer.component';
import { gettext } from '@c8y/ngx-components';
export class GroupDeviceGridColumn extends BaseDeviceGridColumn {
    constructor(initialColumnConfig) {
        super(initialColumnConfig);
        this.name = 'group';
        this.header = gettext('Group');
        this.cellRendererComponent = GroupCellRendererComponent;
        this.filteringFormRendererComponent = GroupFilteringFormRendererComponent;
        this.filterable = true;
        this.filteringConfig = {
            getFilter(model) {
                const filter = {};
                if (model.selectedNodes) {
                    filter.__or = model.selectedNodes.map((mo) => {
                        if (mo.c8y_DeviceQueryString) {
                            return { __useFilterQueryString: mo.c8y_DeviceQueryString };
                        }
                        return { __bygroupid: mo.id };
                    });
                }
                return filter;
            }
        };
        this.sortable = false;
    }
}
