// @c8y/ngx-components/esm2015/core/authentication/provide-phone-number.component.js
import * as tslib_1 from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { LoginService } from '../login/login.service';
import { AlertService } from '../alert/alert.service';
import { LoginViews } from '../login/login.model';
import { ICredentials, UserService } from '@c8y/client';
let ProvidePhoneNumberComponent = class ProvidePhoneNumberComponent {
    constructor(loginService, alert, userService) {
        this.loginService = loginService;
        this.alert = alert;
        this.userService = userService;
        this.onCancel = new EventEmitter();
        this.onChangeView = new EventEmitter();
        this.requestInProgress = false;
        this.sendTfa = '0';
    }
    save() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                this.requestInProgress = true;
                yield this.userService.savePhoneNumber(this.phoneNumber);
                yield this.sendTFASms();
                this.onChangeView.emit({
                    view: LoginViews.SmsChallenge,
                    credentials: this.credentials
                });
            }
            catch (e) {
                this.alert.addServerFailure(e);
            }
            finally {
                this.requestInProgress = false;
            }
        });
    }
    sendTFASms() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                yield this.userService.verifyTFACode(this.sendTfa);
            }
            catch (e) {
                if (e.res.status === 403) {
                    this.loginService.cleanMessages();
                    this.loginService.addSuccessMessage('send_sms');
                }
                else {
                    throw e;
                }
            }
        });
    }
};
ProvidePhoneNumberComponent.ctorParameters = () => [
    { type: LoginService },
    { type: AlertService },
    { type: UserService }
];
tslib_1.__decorate([
    Input()
], ProvidePhoneNumberComponent.prototype, "credentials", void 0);
tslib_1.__decorate([
    Output()
], ProvidePhoneNumberComponent.prototype, "onCancel", void 0);
tslib_1.__decorate([
    Output()
], ProvidePhoneNumberComponent.prototype, "onChangeView", void 0);
ProvidePhoneNumberComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-provide-phone-number',
        template: "<form #twoFactorForm=\"ngForm\" role=\"form\" class=\"loginForm\" (ngSubmit)=\"save()\" novalidate>\n  <div class=\"legend form-block center\" translate>\n    Two-factor authentication\n  </div>\n\n  <c8y-form-group [ngClass]=\"requestInProgress || twoFactorForm.invalid ? 'p-b-8' : ''\">\n    <label translate>Provide your phone number</label>\n\n    <input\n      class=\"form-control\"\n      [(ngModel)]=\"phoneNumber\"\n      #contactPhone=\"ngModel\"\n      type=\"text\"\n      name=\"phone\"\n      autocomplete=\"off\"\n      placeholder=\"{{ 'e.g. +49 9 876 543 210`LOCALIZE`' | translate }}\"\n      c8yPhoneValidation\n      c8yDefaultValidation=\"phoneNumber\"\n      required\n    />\n  </c8y-form-group>\n\n  <button\n    title=\"{{ 'Save and continue' | translate }}\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n    [disabled]=\"requestInProgress || twoFactorForm.invalid\"\n  >\n    {{ 'Save and continue' | translate }}\n  </button>\n\n  <div class=\"d-flex m-t-8\">\n    <a\n      title=\"{{ 'Login' | translate }}\"\n      class=\"small pointer m-l-auto\"\n      href=\"#\"\n      (click)=\"onCancel.emit()\"\n    >\n      {{ 'Login' | translate }}\n    </a>\n  </div>\n</form>\n"
    })
], ProvidePhoneNumberComponent);
export { ProvidePhoneNumberComponent };
