// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-bulk-operation-from-single.js
import { __awaiter, __decorate } from 'tslib';
import { Component, NgModule } from '@angular/core';
import { ActivatedRoute, RouterModule } from '@angular/router';
import { get } from 'lodash-es';
import { BaseStepperComponent, BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
import { CoreModule } from '@c8y/ngx-components';

let BulkOperationFromSingleComponent = class BulkOperationFromSingleComponent extends BaseStepperComponent {
    constructor(activatedRoute, bulkOperationService) {
        super();
        this.activatedRoute = activatedRoute;
        this.bulkOperationService = bulkOperationService;
    }
    retrieveOperationPrototype() {
        return __awaiter(this, void 0, void 0, function* () {
            const prototype = (yield this.bulkOperationService.getOperation(this.activatedRoute.snapshot.params.operationId)).data;
            return {
                description: get(prototype, 'description'),
                prototype
            };
        });
    }
};
BulkOperationFromSingleComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: BulkOperationsService }
];
BulkOperationFromSingleComponent = __decorate([
    Component({
        selector: 'bulk-operation-from-single',
        template: "<c8y-bulk-operation-stepper> </c8y-bulk-operation-stepper>\n"
    })
], BulkOperationFromSingleComponent);

const bulkOperationCreateSingleRoutes = [
    {
        path: 'devicecontrol/single/create-bulk/:operationId',
        component: BulkOperationFromSingleComponent
    }
];
let BulkOperationFromSingleModule = class BulkOperationFromSingleModule {
};
BulkOperationFromSingleModule = __decorate([
    NgModule({
        declarations: [BulkOperationFromSingleComponent],
        imports: [
            CoreModule,
            BulkOperationStepperModule,
            RouterModule.forChild(bulkOperationCreateSingleRoutes)
        ]
    })
], BulkOperationFromSingleModule);

/**
 * Generated bundle index. Do not edit.
 */

export { BulkOperationFromSingleComponent, BulkOperationFromSingleModule, bulkOperationCreateSingleRoutes };
