// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-bulk-operations-stepper-container.js
import { __decorate } from 'tslib';
import { ComponentFactoryResolver, ChangeDetectorRef, ViewChild, ViewContainerRef, Component, NgModule } from '@angular/core';
import { ActivatedRoute, RouterModule } from '@angular/router';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';

let BulkOperationsStepperContainerComponent = class BulkOperationsStepperContainerComponent {
    constructor(activatedRoute, bulkOpsService, resolver, changeDetector) {
        this.activatedRoute = activatedRoute;
        this.bulkOpsService = bulkOpsService;
        this.resolver = resolver;
        this.changeDetector = changeDetector;
    }
    ngOnInit() {
        this.activatedRoute.params.subscribe(params => {
            if (params && params.type) {
                this.bulkOperationType = params.type;
                this.component = this.bulkOpsService
                    .getBulkTypes()
                    .find(element => element.path.includes(this.bulkOperationType)).component;
                this.changeDetector.detectChanges();
                this.createComponent(this.component);
            }
        });
    }
    createComponent(component) {
        if (component) {
            const factory = this.resolver.resolveComponentFactory(component);
            this.container.createComponent(factory);
        }
    }
};
BulkOperationsStepperContainerComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: BulkOperationsService },
    { type: ComponentFactoryResolver },
    { type: ChangeDetectorRef }
];
__decorate([
    ViewChild('container', { static: false, read: ViewContainerRef })
], BulkOperationsStepperContainerComponent.prototype, "container", void 0);
BulkOperationsStepperContainerComponent = __decorate([
    Component({
        selector: 'bulk-operations-stepper-container',
        template: "<c8y-title>{{ 'New bulk operation' | translate }}</c8y-title>\n<div class=\"card card--grid--fullpage\">\n  <div class=\"d-contents\">\n    <template #container></template>\n  </div>\n</div>\n"
    })
], BulkOperationsStepperContainerComponent);

const bulkOperatioStepperRoutes = [
    {
        path: 'devicecontrol/bulk/creation/:type',
        component: BulkOperationsStepperContainerComponent
    }
];
/**
 * Container module for the new bulk operation steppers. Use this module in your application
 * to provide a step by step guided process for scheduling of bulk operations.
 * The module reads the bulk operation type from the URL and displays the corresponding stepper.
 * You can provide stepper for your custom bulk operation type by registering it with the
 * HOOK_LIST_BULK_TYPE injection token.
 */
let BulkOperationsStepperContainerModule = class BulkOperationsStepperContainerModule {
};
BulkOperationsStepperContainerModule = __decorate([
    NgModule({
        declarations: [BulkOperationsStepperContainerComponent],
        imports: [CoreModule, RouterModule.forChild(bulkOperatioStepperRoutes)],
        entryComponents: [BulkOperationsStepperContainerComponent]
    })
], BulkOperationsStepperContainerModule);

/**
 * Generated bundle index. Do not edit.
 */

export { BulkOperationsStepperContainerComponent, BulkOperationsStepperContainerModule, bulkOperatioStepperRoutes };
