// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-operation-summary.js
import { __decorate } from 'tslib';
import { Input, Component, NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { InventoryService } from '@c8y/client';

let OperationSummaryComponent = class OperationSummaryComponent {
    constructor(inventoryService) {
        this.inventoryService = inventoryService;
        this.deviceCount = 0;
        this.displayDeviceCount = false;
    }
    set deviceQueryString(query) {
        this.displayDeviceCount = true;
        const filters = {
            q: query || '',
            pageSize: 1,
            currentPage: 1,
            withTotalPages: true
        };
        this.inventoryService
            .list(filters)
            .then((result) => (this.deviceCount = result.paging.totalPages));
    }
};
OperationSummaryComponent.ctorParameters = () => [
    { type: InventoryService }
];
__decorate([
    Input()
], OperationSummaryComponent.prototype, "name", void 0);
__decorate([
    Input()
], OperationSummaryComponent.prototype, "description", void 0);
__decorate([
    Input()
], OperationSummaryComponent.prototype, "deviceQueryString", null);
OperationSummaryComponent = __decorate([
    Component({
        selector: 'c8y-operation-summary',
        template: "<div class=\"row\">\n  <div class=\"col-md-12 text-center\">\n    <p>{{ name | translate }}</p>\n    <div class=\"row\">\n      <div class=\"col-md-4 col-md-offset-4 text-center\">\n        <h4 class=\"m-t-16 m-b-8 text-truncate\" title=\"{{ description | translate }}\">{{ description | translate }}</h4>\n      </div>\n    </div>\n    <span\n      class=\"text-muted\"\n      [ngPlural]=\"deviceCount\"\n      *ngIf=\"displayDeviceCount && deviceCount !== undefined && deviceCount !== null\"\n    >\n      <ng-template ngPluralCase=\"=0\">\n        <span translate>No devices selected</span>\n      </ng-template>\n      <ng-template ngPluralCase=\"=1\">\n        <span translate>1 device selected</span>\n      </ng-template>\n      <ng-template ngPluralCase=\"other\">\n        <span translate ngNonBindable [translateParams]=\"{ deviceCount: deviceCount }\">\n          {{ deviceCount }} devices selected</span\n        >\n      </ng-template>\n    </span>\n  </div>\n</div>\n"
    })
], OperationSummaryComponent);

/**
 * This module provides a component for displaying the summary of given bulk operation.
 */
let OperationSummaryModule = class OperationSummaryModule {
};
OperationSummaryModule = __decorate([
    NgModule({
        imports: [CoreModule],
        declarations: [OperationSummaryComponent],
        entryComponents: [OperationSummaryComponent],
        exports: [OperationSummaryComponent]
    })
], OperationSummaryModule);

/**
 * Generated bundle index. Do not edit.
 */

export { OperationSummaryComponent, OperationSummaryModule };
