// @c8y/ngx-components/esm2015/context-dashboard/report-dashboard/report-dashboard.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ONCE_ROUTE, ViewContext } from '@c8y/ngx-components';
import { ContextDashboardModule } from '../context-dashboard.module';
import { ContextDashboardComponent } from '../context-dashboard.component';
import { ReportDashboardGuard } from './report-dashboard.guard';
import { ReportDashboardNavigationFactory } from './report-dashboard-navigation.factory';
import { ReportDashboardListComponent } from './report-dashboard-list.component';
import { ReportDashboardService } from './report-dashboard.service';
const ɵ0 = [
    {
        path: 'dashboard/:dashboardId',
        component: ContextDashboardComponent,
        canActivate: [ReportDashboardGuard],
        context: ViewContext.Report
    }
];
let ReportDashboardModule = class ReportDashboardModule {
};
ReportDashboardModule = tslib_1.__decorate([
    NgModule({
        declarations: [ReportDashboardListComponent],
        entryComponents: [],
        imports: [
            CoreModule,
            ContextDashboardModule,
            RouterModule.forChild([
                {
                    path: 'reports',
                    component: ReportDashboardListComponent
                }
            ])
        ],
        providers: [
            ReportDashboardGuard,
            ReportDashboardNavigationFactory,
            ReportDashboardService,
            {
                provide: HOOK_ONCE_ROUTE,
                useValue: ɵ0,
                multi: true
            },
            {
                provide: HOOK_NAVIGATOR_NODES,
                useClass: ReportDashboardNavigationFactory,
                multi: true
            }
        ]
    })
], ReportDashboardModule);
export { ReportDashboardModule };
export { ɵ0 };
