// @c8y/ngx-components/esm2015/sub-assets/unassign-assets-modal/unassign-modal.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { gettext, Status } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { Subject } from 'rxjs';
let UnassignModalComponent = class UnassignModalComponent {
    constructor(translateService) {
        this.translateService = translateService;
        this.closeSubject = new Subject();
        this.labels = { ok: gettext('Unassign'), cancel: gettext('Cancel') };
        this.title = gettext('Unassign devices');
        this.status = Status.WARNING;
    }
    ngOnInit() {
        this.message = this.translateService.instant(gettext('You are about to unassign asset: "{{name}}" . Do you want to proceed?'), { name: this.asset.name });
    }
    ngAfterViewInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modalRef.result;
                this.onClose();
            }
            catch (error) {
                this.onDismiss();
            }
        });
    }
    onClose() {
        this.closeSubject.next(true);
        this.closeSubject.complete();
    }
    onDismiss() {
        this.closeSubject.complete();
    }
};
UnassignModalComponent.ctorParameters = () => [
    { type: TranslateService }
];
tslib_1.__decorate([
    Input()
], UnassignModalComponent.prototype, "asset", void 0);
tslib_1.__decorate([
    ViewChild('modalRef', { static: false })
], UnassignModalComponent.prototype, "modalRef", void 0);
UnassignModalComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-unassign-modal',
        template: "<c8y-confirm-modal [title]=\"title\" [status]=\"status\" [labels]=\"labels\" #modalRef>\n  <span>{{ message | translate }}</span>\n</c8y-confirm-modal>\n"
    })
], UnassignModalComponent);
export { UnassignModalComponent };
