// @c8y/ngx-components/esm5/core/alert/alert.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ContentChild } from '@angular/core';
import { AlertService } from './alert.service';
import { AlertDetailsComponent } from './alert-details.component';
/**
 * Alert component allows creating alert with complex view.
 */
var AlertComponent = /** @class */ (function () {
    /**
     * @ignore
     */
    function AlertComponent(alertService) {
        this.alertService = alertService;
    }
    /**
     * AlertComponent adds new alert to collection stored in alertService after view init.
     */
    AlertComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (!this.containDetailsComponent) {
            this.detailedData = null;
        }
        this.viewInitTimeout = setTimeout(function () {
            return _this.alertService.add(_this);
        });
    };
    /**
     * AlertComponent removes alert from collection stored in alertService on destroy component.
     */
    AlertComponent.prototype.ngOnDestroy = function () {
        clearTimeout(this.viewInitTimeout);
        this.alertService.remove(this);
    };
    AlertComponent.ctorParameters = function () { return [
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], AlertComponent.prototype, "type", void 0);
    tslib_1.__decorate([
        Input()
    ], AlertComponent.prototype, "onDetail", void 0);
    tslib_1.__decorate([
        Input()
    ], AlertComponent.prototype, "onClose", void 0);
    tslib_1.__decorate([
        ViewChild('detailedData', { static: false })
    ], AlertComponent.prototype, "detailedData", void 0);
    tslib_1.__decorate([
        ViewChild('text', { static: false })
    ], AlertComponent.prototype, "text", void 0);
    tslib_1.__decorate([
        ContentChild(AlertDetailsComponent, { static: false })
    ], AlertComponent.prototype, "containDetailsComponent", void 0);
    AlertComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-alert',
            template: "<ng-template #text>\r\n  <ng-content select=\"c8y-alert-text\"></ng-content>\r\n  <ng-content></ng-content>\r\n</ng-template>\r\n<ng-template #detailedData>\r\n  <ng-content select=\"c8y-alert-details\"></ng-content>\r\n</ng-template>"
        })
    ], AlertComponent);
    return AlertComponent;
}());
export { AlertComponent };
