// @c8y/ngx-components/esm5/core/header/application/app-switcher.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ApplicationService, IApplication, ICurrentTenant, IUser } from '@c8y/client';
import { get, includes, orderBy } from 'lodash-es';
import { combineLatest } from 'rxjs';
import { first } from 'rxjs/operators';
import { AppStateService } from '../../common/ui-state.service';
var AppSwitcherComponent = /** @class */ (function () {
    function AppSwitcherComponent(ui, applications) {
        var _this = this;
        this.ui = ui;
        this.applications = applications;
        var _a = this.ui, currentUser = _a.currentUser, currentTenant = _a.currentTenant;
        combineLatest(currentUser, currentTenant)
            .pipe(first())
            .subscribe(function (_a) {
            var _b = tslib_1.__read(_a, 2), user = _b[0], tenant = _b[1];
            _this.tenant = tenant;
            _this.listAppsForUser(user);
        });
    }
    Object.defineProperty(AppSwitcherComponent.prototype, "hideCloudApps", {
        get: function () {
            return false; // this should be configurable from outside
        },
        enumerable: true,
        configurable: true
    });
    AppSwitcherComponent.prototype.isActive = function (app) {
        return app.contextPath === this.ui.state.app.contextPath;
    };
    AppSwitcherComponent.prototype.getHref = function (app) {
        return this.applications.getHref(app);
    };
    AppSwitcherComponent.prototype.listAppsForUser = function (user) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!user) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.applications.listByUser(user, { pageSize: 1000 })];
                    case 1:
                        data = (_a.sent()).data;
                        this.apps = this.orderApps(this.filterApps(data));
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    AppSwitcherComponent.prototype.filterApps = function (apps) {
        apps = this.filterVisible(apps);
        apps = this.filterCloud(apps);
        apps = this.filterDuplicates(apps);
        return apps;
    };
    AppSwitcherComponent.prototype.filterVisible = function (apps) {
        var visibleApplicationTypes = ['HOSTED', 'EXTERNAL'];
        return apps.filter(function (app) {
            return includes(visibleApplicationTypes, app.type) &&
                !app.noAppSwitcher &&
                !get(app, 'manifest.noAppSwitcher');
        });
    };
    AppSwitcherComponent.prototype.filterCloud = function (apps) {
        var hideCloudApps = this.hideCloudApps;
        var filterFn = function (app) { return !!app.cloud; };
        if (!hideCloudApps) {
            this.oneCloudApps = this.orderApps(apps.filter(filterFn));
        }
        return apps.filter(function (app) { return !filterFn(app); });
    };
    AppSwitcherComponent.prototype.filterDuplicates = function (apps) {
        var _this = this;
        // Filter out apps that have duplicate contextpaths and are not own owned but the current tenant;
        var filterFn = function (app) {
            return !apps.some(function (otherApp) {
                return app !== otherApp &&
                    (app.contextPath && app.contextPath === otherApp.contextPath) &&
                    app.owner.tenant.id !== _this.tenant.name;
            });
        };
        return apps.filter(filterFn);
    };
    AppSwitcherComponent.prototype.orderApps = function (apps) {
        return orderBy(apps, function (_a) {
            var name = _a.name;
            return name.toLowerCase();
        });
    };
    AppSwitcherComponent.ctorParameters = function () { return [
        { type: AppStateService },
        { type: ApplicationService }
    ]; };
    AppSwitcherComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-app-switcher',
            template: "<div class=\"app-switcher-dropdown\" dropdown>\n  <button\n    class=\"main-header-button c8y-dropdown dropdown-toggle\"\n    id=\"appSwitcherDropdown\"\n    dropdownToggle\n  >\n    <i c8yIcon=\"th\" class=\"icon-2x\"></i>\n  </button>\n  <div\n    *dropdownMenu\n    class=\"app-switcher-dropdown-menu dropdown-menu\"\n    aria-labelledby=\"appSwitcherDropdown\"\n  >\n    <div class=\"app-switcher-sagcloud\" *ngIf=\"oneCloudApps && oneCloudApps.length > 0\">\n      <a\n        title=\"{{ app.name | humanizeAppName | async }}\"\n        class=\"appLink\"\n        *ngFor=\"let app of oneCloudApps\"\n        [href]=\"getHref(app)\"\n      >\n        <c8y-app-icon [app]=\"app\" [name]=\"app.name\" [contextPath]=\"app.contextPath\"></c8y-app-icon>\n        <span class=\"text-truncate-wrap\">{{ app | humanizeAppName | async }}</span>\n      </a>\n    </div>\n    <div class=\"app-switcher\">\n      <a\n        title=\"{{ app.name | humanizeAppName | async }}\"\n        class=\"appLink\"\n        [ngClass]=\"isActive(app) ? 'active' : ''\"\n        *ngFor=\"let app of apps\"\n        [href]=\"getHref(app)\"\n        c8yProductExperience\n        [actionName]=\"'appSwitched'\"\n        [actionData]=\"{ applicationName: app && app.name }\"\n      >\n        <c8y-app-icon [app]=\"app\" [name]=\"app.name\" [contextPath]=\"app.contextPath\"></c8y-app-icon>\n        <span class=\"text-truncate-wrap\">{{ app | humanizeAppName | async }}</span>\n      </a>\n    </div>\n  </div>\n</div>\n"
        })
    ], AppSwitcherComponent);
    return AppSwitcherComponent;
}());
export { AppSwitcherComponent };
