// @ngx-formly/core/bundles/ngx-formly-core-json-schema.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@ngx-formly/core')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/core/json-schema', ['exports', '@angular/core', '@angular/forms', '@ngx-formly/core'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].core = global['ngx-formly'].core || {}, global['ngx-formly'].core['json-schema'] = {}),global.ng.core,global.ng.forms,global['ngx-formly'].core));
}(this, (function (exports,i0,forms,core) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @param {?} value
     * @return {?}
     */
    function isInteger(value) {
        return Number.isInteger
            ? Number.isInteger(value)
            : typeof value === 'number' && Math.floor(value) === value;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    function isEmpty(v) {
        return v === '' || v === undefined || v === null;
    }
    /**
     * @param {?} schema
     * @return {?}
     */
    function isConst(schema) {
        return schema.hasOwnProperty('const') || (schema.enum && schema.enum.length === 1);
    }
    /**
     * @param {?} field
     * @return {?}
     */
    function totalMatchedFields(field) {
        if (!field.fieldGroup) {
            return field.key && core.ɵgetFieldInitialValue(field) !== undefined ? 1 : 0;
        }
        return field.fieldGroup.reduce(( /**
         * @param {?} s
         * @param {?} f
         * @return {?}
         */function (s, f) { return totalMatchedFields(f) + s; }), 0);
    }
    var FormlyJsonschema = /** @class */ (function () {
        function FormlyJsonschema() {
        }
        /**
         * @param {?} schema
         * @param {?=} options
         * @return {?}
         */
        FormlyJsonschema.prototype.toFieldConfig = /**
         * @param {?} schema
         * @param {?=} options
         * @return {?}
         */
            function (schema, options) {
                return this._toFieldConfig(schema, __assign({ schema: schema }, (options || {})));
            };
        /**
         * @private
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
        FormlyJsonschema.prototype._toFieldConfig = /**
         * @private
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
            function (schema, options) {
                var _this_1 = this;
                schema = this.resolveSchema(schema, options);
                /** @type {?} */
                var field = {
                    type: this.guessType(schema),
                    defaultValue: schema.default,
                    templateOptions: {
                        label: schema.title,
                        readonly: schema.readOnly,
                        description: schema.description,
                    },
                };
                if (options.autoClear) {
                    field['autoClear'] = true;
                }
                if (options.shareFormControl === false) {
                    field['shareFormControl'] = false;
                }
                if (options.ignoreDefault) {
                    delete field.defaultValue;
                }
                switch (field.type) {
                    case 'null': {
                        this.addValidator(field, 'null', ( /**
                         * @param {?} __0
                         * @return {?}
                         */function (_a) {
                            var value = _a.value;
                            return value === null;
                        }));
                        break;
                    }
                    case 'number':
                    case 'integer': {
                        field.parsers = [( /**
                                 * @param {?} v
                                 * @return {?}
                                 */function (v) { return isEmpty(v) ? null : Number(v); })];
                        if (schema.hasOwnProperty('minimum')) {
                            field.templateOptions.min = schema.minimum;
                        }
                        if (schema.hasOwnProperty('maximum')) {
                            field.templateOptions.max = schema.maximum;
                        }
                        if (schema.hasOwnProperty('exclusiveMinimum')) {
                            field.templateOptions.exclusiveMinimum = schema.exclusiveMinimum;
                            this.addValidator(field, 'exclusiveMinimum', ( /**
                             * @param {?} __0
                             * @return {?}
                             */function (_a) {
                                var value = _a.value;
                                return isEmpty(value) || (value > schema.exclusiveMinimum);
                            }));
                        }
                        if (schema.hasOwnProperty('exclusiveMaximum')) {
                            field.templateOptions.exclusiveMaximum = schema.exclusiveMaximum;
                            this.addValidator(field, 'exclusiveMaximum', ( /**
                             * @param {?} __0
                             * @return {?}
                             */function (_a) {
                                var value = _a.value;
                                return isEmpty(value) || (value < schema.exclusiveMaximum);
                            }));
                        }
                        if (schema.hasOwnProperty('multipleOf')) {
                            field.templateOptions.step = schema.multipleOf;
                            this.addValidator(field, 'multipleOf', ( /**
                             * @param {?} __0
                             * @return {?}
                             */function (_a) {
                                var value = _a.value;
                                if (isEmpty(value) || typeof value !== 'number' || value === 0 || schema.multipleOf <= 0) {
                                    return true;
                                }
                                return isInteger(value / schema.multipleOf);
                            }));
                        }
                        break;
                    }
                    case 'string': {
                        /** @type {?} */
                        var schemaType = ( /** @type {?} */(schema.type));
                        if (Array.isArray(schemaType) && (schemaType.indexOf('null') !== -1)) {
                            field.parsers = [( /**
                                     * @param {?} v
                                     * @return {?}
                                     */function (v) { return isEmpty(v) ? null : v; })];
                        }
                        ['minLength', 'maxLength', 'pattern'].forEach(( /**
                         * @param {?} prop
                         * @return {?}
                         */function (prop) {
                            if (schema.hasOwnProperty(prop)) {
                                field.templateOptions[prop] = schema[prop];
                            }
                        }));
                        break;
                    }
                    case 'object': {
                        field.fieldGroup = [];
                        var _a = __read(this.resolveDependencies(schema), 2), propDeps_1 = _a[0], schemaDeps_1 = _a[1];
                        Object.keys(schema.properties || {}).forEach(( /**
                         * @param {?} key
                         * @return {?}
                         */function (key) {
                            /** @type {?} */
                            var f = _this_1._toFieldConfig(( /** @type {?} */(schema.properties[key])), options);
                            field.fieldGroup.push(f);
                            f.key = key;
                            if (Array.isArray(schema.required) && schema.required.indexOf(key) !== -1) {
                                f.templateOptions.required = true;
                            }
                            if (f.templateOptions && !f.templateOptions.required && propDeps_1[key]) {
                                f.expressionProperties = {
                                    'templateOptions.required': ( /**
                                     * @param {?} m
                                     * @return {?}
                                     */function (m) {
                                        return m && propDeps_1[key].some(( /**
                                         * @param {?} k
                                         * @return {?}
                                         */function (k) { return !isEmpty(m[k]); }));
                                    }),
                                };
                            }
                            if (schemaDeps_1[key]) {
                                /** @type {?} */
                                var getConstValue_1 = ( /**
                                 * @param {?} s
                                 * @return {?}
                                 */function (s) {
                                    return s.hasOwnProperty('const') ? s.const : s.enum[0];
                                });
                                /** @type {?} */
                                var oneOfSchema = schemaDeps_1[key].oneOf;
                                if (oneOfSchema
                                    && oneOfSchema.every(( /**
                                     * @param {?} o
                                     * @return {?}
                                     */function (o) { return o.properties && o.properties[key] && isConst(o.properties[key]); }))) {
                                    oneOfSchema.forEach(( /**
                                     * @param {?} oneOfSchemaItem
                                     * @return {?}
                                     */function (oneOfSchemaItem) {
                                        var _a = oneOfSchemaItem.properties, _b = key, constSchema = _a[_b], properties = __rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
                                        field.fieldGroup.push(__assign({}, _this_1._toFieldConfig(__assign({}, oneOfSchemaItem, { properties: properties }), __assign({}, options, { autoClear: true })), { hideExpression: ( /**
                                                 * @param {?} m
                                                 * @return {?}
                                                 */function (m) { return !m || getConstValue_1(constSchema) !== m[key]; }) }));
                                    }));
                                }
                                else {
                                    field.fieldGroup.push(__assign({}, _this_1._toFieldConfig(schemaDeps_1[key], options), { hideExpression: ( /**
                                             * @param {?} m
                                             * @return {?}
                                             */function (m) { return !m || isEmpty(m[key]); }) }));
                                }
                            }
                        }));
                        if (schema.oneOf) {
                            field.fieldGroup.push(this.resolveMultiSchema('oneOf', ( /** @type {?} */(schema.oneOf)), __assign({}, options, { shareFormControl: false })));
                        }
                        if (schema.anyOf) {
                            field.fieldGroup.push(this.resolveMultiSchema('anyOf', ( /** @type {?} */(schema.anyOf)), options));
                        }
                        break;
                    }
                    case 'array': {
                        if (schema.hasOwnProperty('minItems')) {
                            field.templateOptions.minItems = schema.minItems;
                            this.addValidator(field, 'minItems', ( /**
                             * @param {?} __0
                             * @return {?}
                             */function (_a) {
                                var value = _a.value;
                                return isEmpty(value) || (value.length >= schema.minItems);
                            }));
                        }
                        if (schema.hasOwnProperty('maxItems')) {
                            field.templateOptions.maxItems = schema.maxItems;
                            this.addValidator(field, 'maxItems', ( /**
                             * @param {?} __0
                             * @return {?}
                             */function (_a) {
                                var value = _a.value;
                                return isEmpty(value) || (value.length <= schema.maxItems);
                            }));
                        }
                        if (schema.hasOwnProperty('uniqueItems')) {
                            field.templateOptions.uniqueItems = schema.uniqueItems;
                            this.addValidator(field, 'uniqueItems', ( /**
                             * @param {?} __0
                             * @return {?}
                             */function (_a) {
                                var value = _a.value;
                                if (isEmpty(value) || !schema.uniqueItems) {
                                    return true;
                                }
                                /** @type {?} */
                                var uniqueItems = Array.from(new Set(value.map(( /**
                                 * @param {?} v
                                 * @return {?}
                                 */function (v) { return JSON.stringify(v); }))));
                                return uniqueItems.length === value.length;
                            }));
                        }
                        // resolve items schema needed for isEnum check
                        if (schema.items && !Array.isArray(schema.items)) {
                            schema.items = this.resolveSchema(( /** @type {?} */(schema.items)), options);
                        }
                        // TODO: remove isEnum check once adding an option to skip extension
                        if (!this.isEnum(schema)) {
                            /** @type {?} */
                            var _this_2 = this;
                            Object.defineProperty(field, 'fieldArray', {
                                get: ( /**
                                 * @return {?}
                                 */function () {
                                    if (!Array.isArray(schema.items)) {
                                        // When items is a single schema, the additionalItems keyword is meaningless, and it should not be used.
                                        return _this_2._toFieldConfig(( /** @type {?} */(schema.items)), options);
                                    }
                                    /** @type {?} */
                                    var length = this.fieldGroup ? this.fieldGroup.length : 0;
                                    /** @type {?} */
                                    var itemSchema = schema.items[length]
                                        ? schema.items[length]
                                        : schema.additionalItems;
                                    return itemSchema
                                        ? _this_2._toFieldConfig(( /** @type {?} */(itemSchema)), options)
                                        : {};
                                }),
                                enumerable: true,
                                configurable: true,
                            });
                        }
                        break;
                    }
                }
                if (schema.hasOwnProperty('const')) {
                    field.templateOptions.const = schema.const;
                    this.addValidator(field, 'const', ( /**
                     * @param {?} __0
                     * @return {?}
                     */function (_a) {
                        var value = _a.value;
                        return value === schema.const;
                    }));
                    if (!field.type) {
                        field.defaultValue = schema.const;
                    }
                }
                if (this.isEnum(schema)) {
                    field.templateOptions.multiple = field.type === 'array';
                    field.type = 'enum';
                    field.templateOptions.options = this.toEnumOptions(schema);
                }
                // map in possible formlyConfig options from the widget property
                if (schema['widget'] && schema['widget'].formlyConfig) {
                    field = core.ɵreverseDeepMerge(schema['widget'].formlyConfig, field);
                }
                // if there is a map function passed in, use it to allow the user to
                // further customize how fields are being mapped
                return options.map ? options.map(field, schema) : field;
            };
        /**
         * @private
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
        FormlyJsonschema.prototype.resolveSchema = /**
         * @private
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
            function (schema, options) {
                if (schema.$ref) {
                    schema = this.resolveDefinition(schema, options);
                }
                if (schema.allOf) {
                    schema = this.resolveAllOf(schema, options);
                }
                return schema;
            };
        /**
         * @private
         * @param {?} __0
         * @param {?} options
         * @return {?}
         */
        FormlyJsonschema.prototype.resolveAllOf = /**
         * @private
         * @param {?} __0
         * @param {?} options
         * @return {?}
         */
            function (_a, options) {
                var _this_1 = this;
                var allOf = _a.allOf, baseSchema = __rest(_a, ["allOf"]);
                if (!allOf.length) {
                    throw Error("allOf array can not be empty " + allOf + ".");
                }
                return allOf.reduce(( /**
                 * @param {?} base
                 * @param {?} schema
                 * @return {?}
                 */function (base, schema) {
                    schema = _this_1.resolveSchema(schema, options);
                    if (base.required && schema.required) {
                        base.required = __spread(base.required, schema.required);
                    }
                    if (schema.uniqueItems) {
                        base.uniqueItems = schema.uniqueItems;
                    }
                    // resolve to min value
                    ['maxLength', 'maximum', 'exclusiveMaximum', 'maxItems', 'maxProperties']
                        .forEach(( /**
                 * @param {?} prop
                 * @return {?}
                 */function (prop) {
                        if (!isEmpty(base[prop]) && !isEmpty(schema[prop])) {
                            base[prop] = base[prop] < schema[prop] ? base[prop] : schema[prop];
                        }
                    }));
                    // resolve to max value
                    ['minLength', 'minimum', 'exclusiveMinimum', 'minItems', 'minProperties']
                        .forEach(( /**
                 * @param {?} prop
                 * @return {?}
                 */function (prop) {
                        if (!isEmpty(base[prop]) && !isEmpty(schema[prop])) {
                            base[prop] = base[prop] > schema[prop] ? base[prop] : schema[prop];
                        }
                    }));
                    return core.ɵreverseDeepMerge(base, schema);
                }), baseSchema);
            };
        /**
         * @private
         * @param {?} mode
         * @param {?} schemas
         * @param {?} options
         * @return {?}
         */
        FormlyJsonschema.prototype.resolveMultiSchema = /**
         * @private
         * @param {?} mode
         * @param {?} schemas
         * @param {?} options
         * @return {?}
         */
            function (mode, schemas, options) {
                var _this_1 = this;
                return {
                    type: 'multischema',
                    fieldGroup: [
                        {
                            type: 'enum',
                            defaultValue: -1,
                            templateOptions: {
                                multiple: mode === 'anyOf',
                                options: schemas
                                    .map(( /**
                             * @param {?} s
                             * @param {?} i
                             * @return {?}
                             */function (s, i) { return ({ label: s.title, value: i }); })),
                            },
                        },
                        {
                            fieldGroup: schemas.map(( /**
                             * @param {?} s
                             * @param {?} i
                             * @return {?}
                             */function (s, i) {
                                return (__assign({}, _this_1._toFieldConfig(s, __assign({}, options, { autoClear: true })), { hideExpression: ( /**
                                         * @param {?} m
                                         * @param {?} fs
                                         * @param {?} f
                                         * @return {?}
                                         */function (m, fs, f) {
                                        /** @type {?} */
                                        var control = f.parent.parent.fieldGroup[0].formControl;
                                        if (control.value === -1) {
                                            /** @type {?} */
                                            var value = f.parent.fieldGroup
                                                .map(( /**
                                         * @param {?} f
                                         * @param {?} i
                                         * @return {?}
                                         */function (f, i) { return ( /** @type {?} */([f, i])); }))
                                                .filter(( /**
                                         * @param {?} __0
                                         * @return {?}
                                         */function (_a) {
                                                var _b = __read(_a, 2), f = _b[0], i = _b[1];
                                                return _this_1.isFieldValid(f, schemas[i], options);
                                            }))
                                                .sort(( /**
                                         * @param {?} __0
                                         * @param {?} __1
                                         * @return {?}
                                         */function (_a, _b) {
                                                var _c = __read(_a, 1), f1 = _c[0];
                                                var _d = __read(_b, 1), f2 = _d[0];
                                                /** @type {?} */
                                                var matchedFields1 = totalMatchedFields(f1);
                                                /** @type {?} */
                                                var matchedFields2 = totalMatchedFields(f2);
                                                if (matchedFields1 === matchedFields2) {
                                                    return 0;
                                                }
                                                return matchedFields2 > matchedFields1 ? 1 : -1;
                                            }))
                                                .map(( /**
                                         * @param {?} __0
                                         * @return {?}
                                         */function (_a) {
                                                var _b = __read(_a, 2), i = _b[1];
                                                return i;
                                            }));
                                            /** @type {?} */
                                            var normalizedValue = [value.length === 0 ? 0 : value[0]];
                                            /** @type {?} */
                                            var formattedValue = mode === 'anyOf' ? normalizedValue : normalizedValue[0];
                                            control.setValue(formattedValue);
                                        }
                                        return Array.isArray(control.value)
                                            ? control.value.indexOf(i) === -1
                                            : control.value !== i;
                                    }) }));
                            })),
                        },
                    ],
                };
            };
        /**
         * @private
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
        FormlyJsonschema.prototype.resolveDefinition = /**
         * @private
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
            function (schema, options) {
                var _a = __read(schema.$ref.split('#/'), 2), uri = _a[0], pointer = _a[1];
                if (uri) {
                    throw Error("Remote schemas for " + schema.$ref + " not supported yet.");
                }
                /** @type {?} */
                var definition = !pointer ? null : pointer.split('/').reduce(( /**
                 * @param {?} def
                 * @param {?} path
                 * @return {?}
                 */function (def, path) { return def && def.hasOwnProperty(path) ? def[path] : null; }), options.schema);
                if (!definition) {
                    throw Error("Cannot find a definition for " + schema.$ref + ".");
                }
                if (definition.$ref) {
                    return this.resolveDefinition(definition, options);
                }
                return __assign({}, definition, ['title', 'description', 'default'].reduce(( /**
                 * @param {?} annotation
                 * @param {?} p
                 * @return {?}
                 */function (annotation, p) {
                    if (schema.hasOwnProperty(p)) {
                        annotation[p] = schema[p];
                    }
                    return annotation;
                }), {}));
            };
        /**
         * @private
         * @param {?} schema
         * @return {?}
         */
        FormlyJsonschema.prototype.resolveDependencies = /**
         * @private
         * @param {?} schema
         * @return {?}
         */
            function (schema) {
                /** @type {?} */
                var deps = {};
                /** @type {?} */
                var schemaDeps = {};
                Object.keys(schema.dependencies || {}).forEach(( /**
                 * @param {?} prop
                 * @return {?}
                 */function (prop) {
                    /** @type {?} */
                    var dependency = ( /** @type {?} */(schema.dependencies[prop]));
                    if (Array.isArray(dependency)) {
                        // Property dependencies
                        dependency.forEach(( /**
                         * @param {?} dep
                         * @return {?}
                         */function (dep) {
                            if (!deps[dep]) {
                                deps[dep] = [prop];
                            }
                            else {
                                deps[dep].push(prop);
                            }
                        }));
                    }
                    else {
                        // schema dependencies
                        schemaDeps[prop] = dependency;
                    }
                }));
                return [deps, schemaDeps];
            };
        /**
         * @private
         * @param {?} schema
         * @return {?}
         */
        FormlyJsonschema.prototype.guessType = /**
         * @private
         * @param {?} schema
         * @return {?}
         */
            function (schema) {
                /** @type {?} */
                var type = ( /** @type {?} */(schema.type));
                if (!type && schema.properties) {
                    return 'object';
                }
                if (Array.isArray(type)) {
                    if (type.length === 1) {
                        return type[0];
                    }
                    if (type.length === 2 && type.indexOf('null') !== -1) {
                        return type[type[0] === 'null' ? 1 : 0];
                    }
                }
                return type;
            };
        /**
         * @private
         * @param {?} field
         * @param {?} name
         * @param {?} validator
         * @return {?}
         */
        FormlyJsonschema.prototype.addValidator = /**
         * @private
         * @param {?} field
         * @param {?} name
         * @param {?} validator
         * @return {?}
         */
            function (field, name, validator) {
                field.validators = field.validators || {};
                field.validators[name] = validator;
            };
        /**
         * @private
         * @param {?} schema
         * @return {?}
         */
        FormlyJsonschema.prototype.isEnum = /**
         * @private
         * @param {?} schema
         * @return {?}
         */
            function (schema) {
                return schema.enum
                    || (schema.anyOf && schema.anyOf.every(isConst))
                    || (schema.oneOf && schema.oneOf.every(isConst))
                    || schema.uniqueItems && schema.items && !Array.isArray(schema.items) && this.isEnum(( /** @type {?} */(schema.items)));
            };
        /**
         * @private
         * @param {?} schema
         * @return {?}
         */
        FormlyJsonschema.prototype.toEnumOptions = /**
         * @private
         * @param {?} schema
         * @return {?}
         */
            function (schema) {
                if (schema.enum) {
                    return schema.enum.map(( /**
                     * @param {?} value
                     * @return {?}
                     */function (value) { return ({ value: value, label: value }); }));
                }
                /** @type {?} */
                var toEnum = ( /**
                 * @param {?} s
                 * @return {?}
                 */function (s) {
                    /** @type {?} */
                    var value = s.hasOwnProperty('const') ? s.const : s.enum[0];
                    return { value: value, label: s.title || value };
                });
                if (schema.anyOf) {
                    return schema.anyOf.map(toEnum);
                }
                if (schema.oneOf) {
                    return schema.oneOf.map(toEnum);
                }
                return this.toEnumOptions(( /** @type {?} */(schema.items)));
            };
        /**
         * @private
         * @param {?} field
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
        FormlyJsonschema.prototype.isFieldValid = /**
         * @private
         * @param {?} field
         * @param {?} schema
         * @param {?} options
         * @return {?}
         */
            function (field, schema, options) {
                var form = (( /** @type {?} */(field.options)))._buildField({
                    form: new forms.FormGroup({}),
                    fieldGroup: [this._toFieldConfig(schema, __assign({}, options, { autoClear: true, ignoreDefault: true }))],
                    model: field.model ? core.ɵclone(field.model) : (field.fieldArray ? [] : {}),
                }).form;
                return form.valid;
            };
        FormlyJsonschema.decorators = [
            { type: i0.Injectable, args: [{ providedIn: 'root' },] }
        ];
        /** @nocollapse */ FormlyJsonschema.ngInjectableDef = i0.defineInjectable({ factory: function FormlyJsonschema_Factory() { return new FormlyJsonschema(); }, token: FormlyJsonschema, providedIn: "root" });
        return FormlyJsonschema;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.FormlyJsonschema = FormlyJsonschema;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
