// @c8y/ngx-components/esm2015/sub-assets/add-group/add-group.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_ONCE_ROUTE } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { AddGroupComponent } from './add-group.component';
import { AddGroupService } from './add-group.service';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
const ɵ0 = [
    {
        path: 'add-group',
        component: AddGroupComponent
    }
];
let AddGroupModule = class AddGroupModule {
};
AddGroupModule = tslib_1.__decorate([
    NgModule({
        declarations: [AddGroupComponent],
        imports: [CoreModule, DeviceGridModule, FormsModule, ReactiveFormsModule],
        entryComponents: [AddGroupComponent],
        exports: [AddGroupComponent],
        providers: [
            {
                provide: HOOK_ONCE_ROUTE,
                useValue: ɵ0,
                multi: true
            },
            AddGroupService
        ]
    })
], AddGroupModule);
export { AddGroupModule };
export { ɵ0 };
